<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Population genetic analyses for hierarchical analysis of partially
    clonal populations built upon the architecture of the adegenet package. 
    Originally described in Kamvar, Tabima, and Grünwald (2014) 
    &lt;doi:10.7717/peerj.281&gt; with version 2.0 described in Kamvar, Brooks, and 
    Grünwald (2015) &lt;doi:10.3389/fgene.2015.00208&gt;. ">
<title>Genetic Analysis of Populations with Mixed Reproduction • poppr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Genetic Analysis of Populations with Mixed Reproduction">
<meta property="og:description" content="Population genetic analyses for hierarchical analysis of partially
    clonal populations built upon the architecture of the adegenet package. 
    Originally described in Kamvar, Tabima, and Grünwald (2014) 
    &lt;doi:10.7717/peerj.281&gt; with version 2.0 described in Kamvar, Brooks, and 
    Grünwald (2015) &lt;doi:10.3389/fgene.2015.00208&gt;. ">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="poppr-version-2-">Poppr version 2 <a class="anchor" aria-label="anchor" href="#poppr-version-2-"></a>
</h1>
</div>
<!-- badges: start -->

<div class="section level2">
<h2 id="what-is-poppr">What is <em>poppr</em>?<a class="anchor" aria-label="anchor" href="#what-is-poppr"></a>
</h2>
<p><em>Poppr</em> is an R package designed for analysis of populations with mixed modes of sexual and clonal reproduction. It is built around the framework of <a href="https://github.com/thibautjombart/adegenet/" class="external-link">adegenet’s</a> genind and genlight objects and offers the following implementations:</p>
<ul>
<li>clone censoring of populations at any of multiple levels of a hierarchy</li>
<li>convenient counting of multilocus genotypes and sub-setting of populations with multiple levels of hierarchy</li>
<li>define multilocus genotypes</li>
<li>calculation of indices of genotypic diversity, evenness, richness, and rarefaction</li>
<li>drawing of dendrograms with bootstrap support for genetic distances</li>
<li>drawing of minimum spanning networks for genetic distances</li>
<li>calculation of the index of association (<img src="https://latex.codecogs.com/gif.latex?I_A" alt="Index of association">) or (<img src="https://latex.codecogs.com/gif.latex?%5Cbar%7Br%7D_d" alt="Standardized index of association">)</li>
<li>batch processing on any server that has R (≥ 2.15.1) installed</li>
<li>calculation of Bruvo’s distance for microsatellite (SSR) markers (implemented in C for speed)</li>
<li>import of data from and export to <a href="https://biology-assets.anu.edu.au/GenAlEx/Welcome.html" title="GenAlEx Homepage" class="external-link">GenAlEx</a>
</li>
</ul>
<div class="section level3">
<h3 id="new-in-version-20">New in version 2.0:<a class="anchor" aria-label="anchor" href="#new-in-version-20"></a>
</h3>
<ul>
<li>handling of genomic SNP data</li>
<li>custom multilocus genotype definitions</li>
<li>collapse multilocus lineages by genetic distance</li>
<li>calculate reticulate minimum spanning networks</li>
<li>calculate index of association in a sliding window across snps</li>
<li>bootstrapping of MLG diversity statistics</li>
<li><a href="https://github.com/zkamvar/poppr_msn_shiny" class="external-link">interactive exploration of minimum spanning networks</a></li>
<li>and more!</li>
</ul>
<p>For full details, see the NEWS file or type in your R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/news.html" class="external-link">news</a></span><span class="op">(</span><span class="va">Version</span> <span class="op">&gt;=</span> <span class="st">"2.0.0"</span>, package <span class="op">=</span> <span class="st">"poppr"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use <em>poppr</em> at all, please specify the version and cite:</p>
<blockquote>
<p>Kamvar ZN, Tabima JF, Grünwald NJ. (2014) Poppr: an R package for genetic analysis of populations with clonal, partially clonal, and/or sexual reproduction. PeerJ 2:e281 <a href="https://doi.org/10.7717/peerj.281" class="external-link">https://doi.org/10.7717/peerj.281</a></p>
</blockquote>
<p>If you use <em>poppr</em> in a presentation please mention it as the <em>poppr</em> R package, specify the version, and use our logo: <a href="https://github.com/grunwaldlab/poppr/blob/main/vignettes/jalapeno_logo.png" class="external-link">(png)</a> | <a href="https://github.com/grunwaldlab/poppr/blob/main/vignettes/jalapeno_logo.svg" class="external-link">(svg)</a>.</p>
<p>Additionally, if you use any following functionalities:</p>
<ul>
<li>minimum spanning networks with reticulation</li>
<li>collapsing multilocus genotypes into multilocus lineages with <code><a href="reference/mlg.filter.html">mlg.filter()</a></code>
</li>
<li>custom multilocus genotype definitions with <code>mlg.custom()</code>
</li>
<li>index of association for genomic data with <code><a href="reference/win.ia.html">win.ia()</a></code> or <code><a href="reference/samp.ia.html">samp.ia()</a></code>
</li>
<li>bootstrapping any genetic distance with genind, genlight, or genpop objects with <code><a href="reference/aboot.html">aboot()</a></code>
</li>
</ul>
<p>Please also cite:</p>
<blockquote>
<p>Kamvar ZN, Brooks JC and Grünwald NJ (2015) Novel R tools for analysis of genome-wide population genetic data with emphasis on clonality. Front. Genet. 6:208. doi: <a href="https://www.frontiersin.org/articles/10.3389/fgene.2015.00208/abstract" class="external-link">10.3389/fgene.2015.00208</a></p>
</blockquote>
<p>You can obtain citation information in R by typing:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"poppr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="from-cran">From CRAN<a class="anchor" aria-label="anchor" href="#from-cran"></a>
</h3>
<p><a href="https://cran.r-project.org/package=poppr" class="external-link">Binary versions for mac and windows are available for R ≥ 2.15.1 <strong>here</strong></a>.</p>
<p>To install, make sure R is at least version 2.15.1 (the authors recommend ≥ 3.0), and in your console, type:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"poppr"</span><span class="op">)</span></span></code></pre></div>
<p>If you want the absolute latest version of <em>poppr</em>, see about installing unreleased versions below.</p>
<hr>
</div>
<div class="section level3">
<h3 id="stable-version">Stable version<a class="anchor" aria-label="anchor" href="#stable-version"></a>
</h3>
<p>New features are occasionally added to {poppr}, but it can take time for it to get to CRAN. If you know that you want the latest version of {poppr}, (which will contain bug fixes and new features to be included in future releases), then you can use the custom R-Universe repository, which is updated hourly: <a href="https://zkamvar.r-universe.dev/builds" class="external-link uri">https://zkamvar.r-universe.dev/builds</a></p>
<p>To install poppr from the R-Universe, you can use the following code:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://zkamvar.r-universe.dev"</span>, <span class="st">"https://cloud.r-project.org"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"poppr"</span>, repos <span class="op">=</span> <span class="va">universe</span><span class="op">)</span></span></code></pre></div>
<p>The universe repository also contain up-to-date versions of {adegenet} and {hierfstat}, which are commonly used in conjunction with {poppr} and are notoriously out of date on CRAN.</p>
</div>
<div class="section level3">
<h3 id="unstabledevelopment-versions">Unstable/Development versions<a class="anchor" aria-label="anchor" href="#unstabledevelopment-versions"></a>
</h3>
<p>All Development versions of {poppr} will be on GitHub, but need to be compiled.</p>
<p>To install this package from github, make sure you have the following:</p>
<ul>
<li>
<a href="https://developer.apple.com/xcode/" class="external-link">Xcode</a> (OSX) OR <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a> (Windows)</li>
<li>
<a href="https://github.com/r-lib/remotes" class="external-link">{remotes}</a> (to install, use: <code>install.packages("remotes")</code>)</li>
</ul>
<blockquote>
<p>For Linux users, make sure that the function <code>getOption("unzip")</code> returns <code>"unzip"</code> or <code>"internal"</code>. If it does not, then run <code>options(unzip = "internal")</code>.</p>
</blockquote>
<p>Once you have {remotes} and a C compiler installed, you can use the <code>install_github()</code> function to install the current version from github.</p>
<p>All new features in testing will be released on different branches. These features will be in various stages of development and may or may not be documented. Install with caution. The below command would install features on the branch called “devel”. Note that these branches might be out of date from the main branch. Note: if you don’t have LaTeX installed, you should set <code>build_vignettes = FALSE</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"grunwaldlab/poppr@devel"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="help--documentation">Help / Documentation<a class="anchor" aria-label="anchor" href="#help--documentation"></a>
</h2>
<div class="section level3">
<h3 id="r-documentation">R documentation<a class="anchor" aria-label="anchor" href="#r-documentation"></a>
</h3>
<p>To access a descriptive index of help files in <em>poppr</em>, type in your console:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">package</span><span class="op">?</span><span class="va">poppr</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vignettes">Vignettes<a class="anchor" aria-label="anchor" href="#vignettes"></a>
</h3>
<p>A few vignettes have been written for poppr:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Title</th>
<th align="left">Command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Algorightms and Equations</td>
<td align="left"><code>vignette("algo", "poppr")</code></td>
</tr>
<tr class="even">
<td align="left">Data import and manipulation</td>
<td align="left"><code><a href="articles/poppr_manual.html">vignette("poppr_manual", "poppr")</a></code></td>
</tr>
<tr class="odd">
<td align="left">Multilocus Genotype Analysis</td>
<td align="left"><code><a href="articles/mlg.html">vignette("mlg", "poppr")</a></code></td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="user-group">User Group<a class="anchor" aria-label="anchor" href="#user-group"></a>
</h3>
<p>Users who have any questions/comments/suggestions regarding any version of poppr (stable or development) should direct their comments to the <a href="https://groups.google.com/d/forum/poppr" class="external-link">Poppr google group</a></p>
</div>
<div class="section level3">
<h3 id="bookprimer">Book/Primer<a class="anchor" aria-label="anchor" href="#bookprimer"></a>
</h3>
<p>In Spring of 2014, Dr. Niklaus J. Grünwald, Dr. Sydney E. Everhart and Zhian N. Kamvar wrote a primer for population genetic analysis in R located at <a href="https://grunwaldlab.github.io/Population_Genetics_in_R" class="external-link uri">https://grunwaldlab.github.io/Population_Genetics_in_R</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>Please note that this project is released with a <a href="https://github.com/grunwaldlab/poppr/blob/main/CODE_OF_CONDUCT.md" class="external-link">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms. If you wish to contribute code to <em>poppr</em>, please fork the repository and create a pull request with your added feature.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=poppr" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/grunwaldlab/poppr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/grunwaldlab/poppr/issues/" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a> | <a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing poppr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Zhian N. Kamvar <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-1458-7108" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Javier F. Tabima <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3603-2691" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jonah C. Brooks <br><small class="roles"> Author </small>  </li>
<li>David Folarin <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/grunwaldlab/poppr/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/grunwaldlab/poppr/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
