<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>All news • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">

  <div class="col-md-9">
    <div class="page-header">
      <h1>Change log <small>All releases</small></h1>
    </div>

    <div class="contents">
    <div id="poppr-2-5-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-5-0" class="anchor"></a>poppr 2.5.0</h1>
<div id="algorithmic-change" class="section level2">
<h2 class="hasAnchor">
<a href="#algorithmic-change" class="anchor"></a>ALGORITHMIC CHANGE</h2>
<ul>
<li>Identified in <a href="https://github.com/grunwaldlab/poppr/issues/139" class="uri">https://github.com/grunwaldlab/poppr/issues/139</a>, Bruvo’s distance will now consider all possible combinations of ordered alleles in the calculation under the genome addition and loss models for missing data. This will affect those who have polyploid data that contain more than one missing allele at any genotype</li>
</ul>
<p>To facilitate comparison, the global option old.bruvo.model, has been created. By default it is set to FALSE, indicating that poppr should use the ordered allele combinations. If the user wants to use the method considering unorderd allele combinations, they can set options(old.bruvo.model = TRUE)</p>
<p>It must be repeated that this does not affect haploid or diploid comparisons, those that use the infinite alleles model, or those who do not have more than one missing allele at any genotype.</p>
</div>
<div id="deprecation" class="section level2">
<h2 class="hasAnchor">
<a href="#deprecation" class="anchor"></a>DEPRECATION</h2>
<ul>
<li>The warning for a short repeat length vector for Bruvo’s distance is deprecated and will become an error in the future</li>
<li>jack.ia is deprecated in favor of resample.ia for clarity.</li>
</ul>
</div>
<div id="bug-fix" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix" class="anchor"></a>BUG FIX</h2>
<ul>
<li>A bug in read.genalex() where removed samples would have incorrect strata labels was fixed. Thanks to Hernán Dario Capador-Barreto for identifying it. See <a href="https://github.com/grunwaldlab/poppr/issues/147" class="uri">https://github.com/grunwaldlab/poppr/issues/147</a>.</li>
</ul>
</div>
<div id="misc" class="section level2">
<h2 class="hasAnchor">
<a href="#misc" class="anchor"></a>MISC</h2>
<ul>
<li>The internal plotting function for mlg.table now uses tidy evaluation for dplyr versions &gt; 0.5.0</li>
<li>The package reshape2 was removed from imports and replaced with base functions (see <a href="https://github.com/grunwaldlab/poppr/issues/144" class="uri">https://github.com/grunwaldlab/poppr/issues/144</a> for details)</li>
</ul>
</div>
<div id="new-imports" class="section level2">
<h2 class="hasAnchor">
<a href="#new-imports" class="anchor"></a>NEW IMPORTS</h2>
<ul>
<li>Due to the migration to dplyr version 0.7.0, poppr now imports the “!!” operator from the rlang package</li>
</ul>
</div>
</div>
    <div id="poppr-2-4-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-4-1" class="anchor"></a>poppr 2.4.1</h1>
<div id="bug-fix-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-1" class="anchor"></a>BUG FIX</h2>
<ul>
<li>A corner case where repeat length vectors out of order would be erroneously subset with <code><a href="../reference/test_replen.html">test_replen()</a></code> and <code><a href="../reference/fix_replen.html">fix_replen()</a></code> has been fixed. See <a href="https://github.com/grunwaldlab/poppr/issues/136" class="uri">https://github.com/grunwaldlab/poppr/issues/136</a> for details.</li>
<li>All functions that perform filtering will now run serially due to a bug on Fedora machines with at least two threads. Details can be found at <a href="https://github.com/grunwaldlab/poppr/issues/138" class="uri">https://github.com/grunwaldlab/poppr/issues/138</a>.</li>
</ul>
</div>
</div>
    <div id="poppr-2-4-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-4-0" class="anchor"></a>poppr 2.4.0</h1>
<div id="new-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>
<code><a href="../reference/ia.html">jack.ia()</a></code> will randomly jackknife your sample to a specified n (default is the number of MLG), and calculate the index of association over multiple iterations, giving a distribution of possible values at a given sample size.</li>
</ul>
</div>
<div id="new-features" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>The function <code><a href="../reference/mlg.html">mlg.table()</a></code> gains new parameters, “color” and “background”. The “color” parameter will create a single barplot with colors representing populations while the “background” parameter will create a background plot showing the abundance of MLGs across populations within the facets.</li>
<li>The function <code><a href="../reference/win.ia.html">win.ia()</a></code> will now take into consideration chromosomal coordinates when constructing windows. It has additionally acquired a new parameter <code>chromosome_buffer</code>, which allows the user to specify whether or not the window should be limited to within chromosomes.</li>
</ul>
</div>
<div id="notable-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#notable-changes" class="anchor"></a>NOTABLE CHANGES</h2>
<ul>
<li>calculation of MLGs for snpclone and genlight objects will be performed via distance-based methods by default. This is in contrast to the previous behavior where individuals were assumed to have unique genotypes. see <a href="https://github.com/grunwaldlab/poppr/issues/125" class="uri">https://github.com/grunwaldlab/poppr/issues/125</a> for details.</li>
<li>An error will be thrown when attempting to use <code><a href="../reference/mlg.html">mlg.crosspop()</a></code> with an object that has &lt; 2 populations.</li>
<li>
<code><a href="../reference/genotype_curve.html">genotype_curve()</a></code> will now remove monomorphic loci before calculation by default as these loci misleadingly influence the shape of the curve. This will change the shape of the curve if you have monomorphic loci. This change IS optional via the drop and dropna parameters, but it is not recommended to change these parameters.</li>
<li>The calculation for <code><a href="../reference/psex.html">psex()</a></code> has changed to be more accurate when using method = “multiple”. It also gains the ability to use several values of G, one for each population. Documentation for <code><a href="../reference/psex.html">psex()</a></code> has also been improved. For details of the change, see <a href="https://github.com/grunwaldlab/poppr/issues/101" class="uri">https://github.com/grunwaldlab/poppr/issues/101</a>
</li>
</ul>
</div>
<div id="bug-fix-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-2" class="anchor"></a>BUG FIX</h2>
<ul>
<li>The error given when a genlight object is passed to <code><a href="../reference/poppr.html">poppr()</a></code> now correctly identifies the substitute function as <code><a href="../reference/diversity_stats.html">diversity_stats()</a></code> and not diversity table (see <a href="https://github.com/grunwaldlab/poppr/issues/123" class="uri">https://github.com/grunwaldlab/poppr/issues/123</a>).</li>
<li>GenAlEx files imported with duplicate loci will generate a warning telling the user that the duplicated loci have been renamed (usually to <locus name>_1) (see <a href="https://github.com/grunwaldlab/poppr/issues/122" class="uri">https://github.com/grunwaldlab/poppr/issues/122</a>).</locus>
</li>
<li>Haplodiploids imported in genalex files will be properly treated (see <a href="https://github.com/grunwaldlab/poppr/issues/124" class="uri">https://github.com/grunwaldlab/poppr/issues/124</a>).</li>
<li>
<code><a href="../reference/read.genalex.html">read.genalex()</a></code> will now implicitly check for the correct number of individuals in the data (see <a href="https://github.com/grunwaldlab/poppr/issues/128" class="uri">https://github.com/grunwaldlab/poppr/issues/128</a>).</li>
<li>The function <code><a href="../reference/poppr.html">poppr()</a></code> no longer throws an error if the sample &gt; 0 and the data has no population (see <a href="https://github.com/grunwaldlab/poppr/issues/130" class="uri">https://github.com/grunwaldlab/poppr/issues/130</a>).</li>
<li>A bug where round-robin allele frequencies calculated with by_pop = TRUE were inaccurate for all but the first population was fixed. For details, see <a href="https://github.com/grunwaldlab/poppr/issues/132" class="uri">https://github.com/grunwaldlab/poppr/issues/132</a>.</li>
<li>A potential integer overflow was fixed in <code>SEXP association_index_haploid</code>. This was a ghost from <a href="https://github.com/grunwaldlab/poppr/issues/100" class="uri">https://github.com/grunwaldlab/poppr/issues/100</a>.</li>
<li>PROTECT statements were placed around allocation statements. For details, see <a href="https://github.com/grunwaldlab/poppr/issues/133" class="uri">https://github.com/grunwaldlab/poppr/issues/133</a>.</li>
</ul>
</div>
<div id="misc-1" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-1" class="anchor"></a>MISC</h2>
<ul>
<li>The documentation for <code><a href="../reference/bitwise.dist.html">bitwise.dist()</a></code> clarifies the role of the <code>differences_only</code> flag (see <a href="https://github.com/grunwaldlab/poppr/issues/119" class="uri">https://github.com/grunwaldlab/poppr/issues/119</a>).</li>
<li>Interruptions in C code is now handled gracefully via <code>R_CheckUserInterrupt()</code>. The benefit is that long-running calculations are interrupted near instantly, but at the cost of a few more milliseconds of computation time. (see <a href="https://github.com/grunwaldlab/poppr/issues/86" class="uri">https://github.com/grunwaldlab/poppr/issues/86</a>)</li>
<li>Bruvo’s distance now has complete tests for recursion as of commit 4e4fa40d16</li>
</ul>
</div>
</div>
    <div id="poppr-2-3-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-3-0" class="anchor"></a>poppr 2.3.0</h1>
<div id="new-functions-1" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-1" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>The function <code><a href="../reference/coercion-methods.html">bootgen2genind()</a></code> will help users take advantage of bootstrapping distance functions from other packages that require genind objects. For details, see <a href="https://github.com/grunwaldlab/poppr/issues/112" class="uri">https://github.com/grunwaldlab/poppr/issues/112</a> and <a href="https://github.com/grunwaldlab/poppr/issues/111" class="uri">https://github.com/grunwaldlab/poppr/issues/111</a>
</li>
</ul>
</div>
<div id="new-features-1" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-1" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>There is now a <code>plot</code> parameter for the genotype curve to enable or suppress plotting.</li>
<li>Progress bars are now automatically suppressed when running non-interactively. to turn them on when running non-interactively, use <code>options(poppr.debug = TRUE)</code>.</li>
<li>The progress bar for <code><a href="../reference/ia.html">ia()</a></code> and <code><a href="../reference/poppr.html">poppr()</a></code> will now show estimated time. This is from dplyr’s <code>progress_estimated()</code>.</li>
</ul>
</div>
<div id="misc-2" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-2" class="anchor"></a>MISC</h2>
<ul>
<li>The <code>hist</code> argument in the <code><a href="../reference/ia.html">ia()</a></code> is deprecated in favor of <code>plot</code>.</li>
<li>The x axis for the <code><a href="../reference/genotype_curve.html">genotype_curve()</a></code> plot is now numeric, allowing you to fit a smoothing function over the points without having to use the hack <code>geom_smooth(aes(group = 1))</code>. This is thanks to Kara Woo for pointing this out on twitter (<a href="https://twitter.com/kara_woo/status/783336540407685120" class="uri">https://twitter.com/kara_woo/status/783336540407685120</a>).</li>
<li>The “show” method for genclone objects now delimits populations and strata by a comma, avoiding confusion with multi-word population names. Thanks to @knausb for the fix in <a href="https://github.com/grunwaldlab/poppr/pull/116" class="uri">https://github.com/grunwaldlab/poppr/pull/116</a>.<br>
</li>
<li>Documentation for <code>poppr.amova</code> now contains a note about significance testing with the ade4 function <code>randtest.amova</code>.</li>
</ul>
</div>
<div id="bug-fix-3" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-3" class="anchor"></a>BUG FIX</h2>
<ul>
<li>The subsetting methods will now properly handle mlgs when using sample names to subset genclone and snpclone objects. See <a href="https://github.com/grunwaldlab/poppr/issues/114" class="uri">https://github.com/grunwaldlab/poppr/issues/114</a> for details.</li>
<li>A plotting bug for <code><a href="../reference/mlg.html">mlg.table()</a></code> was fixed so that the plots now show the maximum value.</li>
<li>Bugs with subsetting bootgen and bruvomat objects with no loci specified were fixed. See <a href="https://github.com/grunwaldlab/poppr/issues/118" class="uri">https://github.com/grunwaldlab/poppr/issues/118</a> for details.</li>
</ul>
</div>
</div>
    <div id="poppr-2-2-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-2-1" class="anchor"></a>poppr 2.2.1</h1>
<div id="bug-fix-4" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-4" class="anchor"></a>BUG FIX</h2>
<ul>
<li>A problem exporting haploid sequence data to genalex format was fixed (see <a href="https://github.com/grunwaldlab/poppr/issues/108" class="uri">https://github.com/grunwaldlab/poppr/issues/108</a>)</li>
<li>Data without population structure no longer throws an error in <code><a href="../reference/imsn.html">imsn()</a></code>
</li>
</ul>
</div>
<div id="new-features-2" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-2" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>You can now specify layout parameters in <code><a href="../reference/imsn.html">imsn()</a></code>
</li>
<li>The *.msn functions will now take into account whether or not your data is set to collapsed MLGs. (see <a href="https://github.com/grunwaldlab/poppr/issues/107" class="uri">https://github.com/grunwaldlab/poppr/issues/107</a>)</li>
</ul>
</div>
<div id="misc-3" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-3" class="anchor"></a>MISC</h2>
<ul>
<li>Errors from mlg.filter now make more sense. (see <a href="https://github.com/grunwaldlab/poppr/issues/109" class="uri">https://github.com/grunwaldlab/poppr/issues/109</a>)</li>
<li>The vignette poppr_manual has been converted to HTML format. (see <a href="https://github.com/grunwaldlab/poppr/issues/113" class="uri">https://github.com/grunwaldlab/poppr/issues/113</a>)</li>
</ul>
</div>
</div>
    <div id="poppr-2-2-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-2-0" class="anchor"></a>poppr 2.2.0</h1>
<div id="new-functions-2" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-2" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>
<code><a href="../reference/incomp.html">incomp()</a></code> will check your data to see if there are any incomparable samples.</li>
</ul>
</div>
<div id="new-features-3" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-3" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>Threshold argument added to <code><a href="../reference/filter_stats.html">filter_stats()</a></code> (see <a href="https://github.com/grunwaldlab/poppr/issues/94" class="uri">https://github.com/grunwaldlab/poppr/issues/94</a>)</li>
<li>The pipe operator (<code>%&gt;%</code>) is now exported from magrittr to make chaining commands easier.</li>
<li>
<code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> can now return multiple statistics.</li>
<li>The user can now control the size of the labels in the index of association plots with the labsize and linesize arguments in the plot method.</li>
<li>
<code><a href="../reference/private_alleles.html">private_alleles()</a></code> gains a drop argument.</li>
<li>
<code><a href="../reference/recode_polyploids.html">recode_polyploids()</a></code> can now take haplodiploid data.</li>
</ul>
</div>
<div id="bug-fix-5" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-5" class="anchor"></a>BUG FIX</h2>
<ul>
<li>An issue that caused errors in the <code><a href="../reference/imsn.html">imsn()</a></code> code output was fixed (see <a href="https://github.com/grunwaldlab/poppr/issues/93" class="uri">https://github.com/grunwaldlab/poppr/issues/93</a>)</li>
<li>Caught bug where the <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> assignment method was using <code><a href="../reference/genetic_distance.html">nei.dist()</a></code> instead of <code><a href="../reference/diss.dist.html">diss.dist()</a></code> when no distance was specified.</li>
<li>A bug that resulted in significantly negative values from <code><a href="../reference/bitwise.ia.html">bitwise.ia()</a></code> with large sample sizes was fixed. Spotted by @knausb (see <a href="https://github.com/grunwaldlab/poppr/issues/100" class="uri">https://github.com/grunwaldlab/poppr/issues/100</a>)</li>
<li>Fixed issue where <code><a href="../reference/plot_filter_stats.html">plot_filter_stats()</a></code> wasn’t displaying the full range of MLGs</li>
<li>Color vectors are now correctly parsed when passed to msn functions (see <a href="https://github.com/grunwaldlab/poppr/issues/55" class="uri">https://github.com/grunwaldlab/poppr/issues/55</a>) for details. Long color vectors are now accepted, albiet with warning.</li>
<li>An issue where <code><a href="../reference/mll.reset-method.html">mll.reset()</a></code> did not reset non-MLG class objects in the mlg slot was fixed.</li>
</ul>
</div>
<div id="misc-4" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-4" class="anchor"></a>MISC</h2>
<ul>
<li>Documentation for <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> was clarified and updated with more examples.</li>
<li>The vignette “Migration from poppr version 1” has been removed.</li>
<li>Previously deprecated <code>*hierarchy()</code> functions have been removed.</li>
</ul>
</div>
</div>
    <div id="poppr-2-1-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-1-1" class="anchor"></a>poppr 2.1.1</h1>
<div id="new-features-4" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-4" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>
<code><a href="../reference/imsn.html">imsn()</a></code> now has collapsible side panels</li>
<li>
<code><a href="../reference/mll-method.html">nmll()</a></code> and <code><a href="../reference/mll-method.html">mll()</a></code> will now handle genind and genlight objects</li>
<li>
<code><a href="../reference/rraf.html">rraf()</a></code> now gives options for minor allele correction encompassed in the internal function <code><a href="../reference/rare_allele_correction.html">rare_allele_correction()</a></code>. This extends also to <code><a href="../reference/pgen.html">pgen()</a></code> and <code><a href="../reference/psex.html">psex()</a></code>, which must correct minor allele frequencies by default. See <a href="https://github.com/grunwaldlab/poppr/issues/81" class="uri">https://github.com/grunwaldlab/poppr/issues/81</a> for details.</li>
</ul>
</div>
<div id="misc-5" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-5" class="anchor"></a>MISC</h2>
<ul>
<li>
<code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> now defaults to using <code><a href="../reference/diss.dist.html">diss.dist()</a></code>
</li>
<li>default threshold for <code><a href="../reference/filter_stats.html">filter_stats()</a></code> is now 1e+6</li>
<li>
<code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> now returns a list instead of a pairlist</li>
<li>the “hist” argument for <code><a href="../reference/poppr.html">poppr()</a></code> is now deprecated in favor of “plot”</li>
<li>documentation improvements</li>
<li>the show method for snpclone objects now looks distinct from genlight</li>
<li>genlight objects no longer get passed to <code><a href="../reference/missingno.html">missingno()</a></code> in <code><a href="../reference/filter_stats.html">filter_stats()</a></code>
</li>
<li>
<code><a href="../reference/filter_stats.html">filter_stats()</a></code> now returns invisibly when plot = TRUE; see <a href="https://github.com/grunwaldlab/poppr/issues/87" class="uri">https://github.com/grunwaldlab/poppr/issues/87</a> for details.</li>
</ul>
</div>
<div id="bug-fix-6" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-6" class="anchor"></a>BUG FIX</h2>
<ul>
<li>When supplied an object with no strata, <code><a href="../reference/clonecorrect.html">clonecorrect()</a></code> will default to <code>strata = NA</code>.</li>
<li>
<code><a href="../reference/read.genalex.html">read.genalex()</a></code> will no longer fail if missing data is not coded as zero; see <a href="https://github.com/grunwaldlab/poppr/issues/84" class="uri">https://github.com/grunwaldlab/poppr/issues/84</a> for details</li>
<li>
<code><a href="../reference/missingno.html">missingno()</a></code> no longer removes genotypes AT specified threshold; see <a href="https://github.com/grunwaldlab/poppr/issues/90" class="uri">https://github.com/grunwaldlab/poppr/issues/90</a> for details</li>
</ul>
</div>
</div>
    <div id="poppr-2-1-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-1-0" class="anchor"></a>poppr 2.1.0</h1>
<div id="improvements" class="section level2">
<h2 class="hasAnchor">
<a href="#improvements" class="anchor"></a>IMPROVEMENTS</h2>
<ul>
<li>
<code>win.ia</code> and <code>samp.ia</code> gain a significant speedup thanks to Jonah Brooks implementing the code in C.</li>
<li>The internal code for the <code>genotype_curve</code> has been implemented in C for a 10x increase in speed.</li>
</ul>
</div>
<div id="new-features-5" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-5" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>
<code>poppr.msn</code>, <code>bruvo.msn</code>, and <code>plot_poppr_msn</code> gain the ability to take character vectors for color palettes. See issue #55 (<a href="https://github.com/grunwaldlab/poppr/issues/55" class="uri">https://github.com/grunwaldlab/poppr/issues/55</a>) for details.</li>
<li>
<code>plot_poppr_msn</code> returns the modified graph.</li>
<li>All functions related to Bruvo’s distance can now take a named vector of repeat lengths in any order. See issue #61 (<a href="https://github.com/grunwaldlab/poppr/issues/61" class="uri">https://github.com/grunwaldlab/poppr/issues/61</a>) for details.</li>
<li>
<code>aboot</code> gains the argument strata so that you can automatically convert genind to genpop.</li>
<li>
<code>genotype_curve</code> can now take in loci objects from pegas.</li>
<li>You can now specify the maximum number of loci to analyze in <code>genotype_curve</code>.</li>
<li>
<code>filter_stats</code> can now optionally plot a histogram in the background.</li>
<li>
<code>bruvo.dist</code> can now optionally return distance matrices by locus. This is addresed in issue #60 (<a href="https://github.com/grunwaldlab/poppr/issues/60" class="uri">https://github.com/grunwaldlab/poppr/issues/60</a>)</li>
<li>
<code>aboot</code> can now handle matrices as previously specified in the documentation.</li>
<li>
<code>aboot</code> can now take custom functions to calculate distance for genlight objects.</li>
<li>
<code>poppr.amova</code> can now perform amova using the pegas implementation.</li>
</ul>
</div>
<div id="new-functions-3" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-3" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>
<code>rrmlg</code> will calculate round-robin multilocus genotypes for each locus.</li>
<li>
<code>rraf</code> will calculate round-robin allele frequencies for each locus.</li>
<li>
<code>pgen</code> will calculate the probabilities of observed genotypes.</li>
<li>
<code>psex</code> will calculate the probability that an observed genotype will be observed more than once by chance.</li>
</ul>
</div>
<div id="new-traditionalists" class="section level2">
<h2 class="hasAnchor">
<a href="#new-traditionalists" class="anchor"></a>NEW TRADITIONALISTS</h2>
<ul>
<li>because we’re through being cool.</li>
</ul>
</div>
<div id="misc-6" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-6" class="anchor"></a>MISC</h2>
<ul>
<li>Documentation for genclone and snpclone classes are more coherent.</li>
<li>Accessors added for internal MLG objects (for developers).</li>
<li>The genotype accumulation curve displays the iteration and locus number instead of a progress bar.</li>
<li>The genotype accumulation plot is now scaled from 0 to the number of observed mlgs.</li>
<li>Documentation for <code>poppr.amova</code> no longer references the “hierarchy” slot.</li>
</ul>
</div>
<div id="bug-fix-7" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-7" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Single locus data sets can now be read in with <code>read.genalex</code>. This was brought up in issue <a href="https://github.com/grunwaldlab/poppr/issues/58">#58</a> (Thanks to Nick Wong for spotting it).</li>
<li>A bug in <code>informloci</code> where the MAF argument wasn’t being applied to P/A data has been fixed.</li>
<li>Printing of genclone objects with mixed ploidies previously reported erroneously due to a sorting error. This has been fixed.</li>
<li>Edge case where a missing cell in the genalex matrix was interpreted as a literal “NA” was fixed.</li>
<li>msn functions now return nodes that are correctly named. See Issue #66 (<a href="https://github.com/grunwaldlab/poppr/issues/66" class="uri">https://github.com/grunwaldlab/poppr/issues/66</a>) for details.</li>
</ul>
</div>
</div>
    <div id="poppr-2-0-2" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-0-2" class="anchor"></a>poppr 2.0.2</h1>
<div id="bug-fix-8" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-8" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Definition of Hexp was fixed. It originally was mis-calculated, inflating the metric. It is now correctly calculated and documented. More information at issue <a href="https://github.com/grunwaldlab/poppr/issues/47">#47</a>
</li>
</ul>
</div>
</div>
    <div id="poppr-2-0-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-0-1" class="anchor"></a>poppr 2.0.1</h1>
<div id="bug-fix-9" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-9" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Memory leak with Bruvo’s distance was fixed by @JonahBrooks in 90facb4 (issue #40)</li>
<li>Cutoff field now works for distances other than dissimilarity in <code>imsn</code> (issue #41)</li>
<li>Switching between data sets no longer shows an error in <code>imsn</code>
</li>
<li>
<code>read.genealex</code> can now correctly import missing data for diploids (issue #42)</li>
</ul>
</div>
<div id="new-features-6" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-6" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>Startup message now tells you if poppr was compiled with OMP support.</li>
</ul>
</div>
</div>
    <div id="poppr-2-0-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-2-0-0" class="anchor"></a>poppr 2.0.0</h1>
<div id="compatibility-update" class="section level2">
<h2 class="hasAnchor">
<a href="#compatibility-update" class="anchor"></a>COMPATIBILITY UPDATE</h2>
<ul>
<li>poppr has moved to version 2.0 due to adegenet’s recent update. The hierarchy slot introduced in version 1.1 is now being moved to adegenet and renamed <code>strata</code>. For maximum backwards compatibility, all of the hierarchy methods still exist, but they are deprecated and will print a warning with the proper function to use. If you were accessing the hierarchy slot without using the <code>*hierarchy()</code> methods, your code will fail as the hierarchy slot now should only contain a formula object.</li>
</ul>
</div>
<div id="new-dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#new-dependencies" class="anchor"></a>NEW DEPENDENCIES</h2>
<ul>
<li>poppr now imports elements from dplyr and shiny</li>
<li>As required as of 2015-06-29, poppr now explicitly imports: stats, graphics, grDevices, and utils</li>
</ul>
</div>
<div id="new-suggests" class="section level2">
<h2 class="hasAnchor">
<a href="#new-suggests" class="anchor"></a>NEW SUGGESTS</h2>
<ul>
<li>poppr now suggests the cowplot, poweRlaw, and polysat packages.</li>
</ul>
</div>
<div id="new-data" class="section level2">
<h2 class="hasAnchor">
<a href="#new-data" class="anchor"></a>NEW DATA</h2>
<ul>
<li>A data set called <code>Pram</code> containing SSR genotypes from the Sudden Oak Death pathogen <em>Phytophthora ramorum</em> (Kamvar et.al., 2015)</li>
</ul>
</div>
<div id="new-mint-flavor" class="section level2">
<h2 class="hasAnchor">
<a href="#new-mint-flavor" class="anchor"></a>NEW MINT FLAVOR</h2>
<ul>
<li>refreshing!</li>
</ul>
</div>
<div id="new-features-7" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-7" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>The default plot for the index of association will now be a single histogram. The user has the option to visualize the standardized index of association (<code>index = "rbarD"</code>, default) or the classic index of association (<code>index =   "Ia"</code>). If the user uses the function <code>ia</code> with the argument <code>valuereturn =   TRUE</code>, then the resulting object can be plotted with the plot function.</li>
<li>The function <code>poppr</code> will now plot all populations in a single faceted plot instead of one plot per population.</li>
<li>
<code>aboot</code> and <code>bruvo.boot</code> will now be able to utilize any function to generate trees (suggested in issue #18).</li>
<li>The mlg slot in the genclone object can now optionally hold an MLG class object. This object will contain different definitions of multilocus genotypes, allowing the user to switch between observed, custom, and mlgs defined given a genetic distance threshold.</li>
<li>minimum spanning network functions gain the ability to include reticulations using the option <code>include.ties</code>
</li>
<li>minimum spanning network functions gain the ability to collapse multilocus genotypes by genetic distance with the option <code>threshold</code>.</li>
<li>
<code>poppr.amova</code> gains the ability to filter multilocus genotypes before calculation.</li>
<li>
<code>informloci</code> gains the argument “MAF”, which allows the specification of a minor allele frequency cutoff in addition to the cutoff argument. Examples have been updated.</li>
<li>
<code>aboot</code> can now take genlight objects.</li>
<li>
<code>poppr.msn</code> and <code>plot_poppr_msn</code> can now take genlight objects.</li>
<li>
<code>plot_poppr_msn</code> now gives users the option to exclude the legends.</li>
<li>Default plotting for <code>mlg.table</code> will no longer produce one plot per population. It will now produce a single ggplot object for all populations. Note that the bars are no longer colored by count.</li>
<li>
<code>poppr</code> will no longer calculate “Hexp”. Instead, Simpson’s index will be calumniated, but the old index can be retrieved by using (N/(N - 1))*lambda.</li>
<li>
<code>poppr</code> can now take any statistic that can be calculated from a table of multilocus genotype counts.</li>
<li>
<code>genind2genalex</code> gains the ability to selectively write different strata.</li>
</ul>
</div>
<div id="new-functions-4" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-4" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>
<code>mlg.filter</code> will contract multilocus genotypes given a genetic distance and threshold using one of three algorithms. It can report statistics such as the multilocus genotypes returned, the number of samples within each multilocus genotype, the thresholds at which multilocus genotypes were collapsed, and the genetic distance matrix that represents the new multilocus genotypes.</li>
<li>
<code>filter_stats</code> will show you graphical output of all the algorithms in <code>mlg.filter</code>.</li>
<li>
<code>cutoff_predictor</code> will predict the cutoff threshold from <code>mlg.filter</code>.</li>
<li>
<code>bitwise.dist</code> can efficiently calculate absolute genetic distance for genlight objects.</li>
<li>
<code>mll</code> “multilocus lineages” is a new replacement for <code>mlg.vector</code> which gains the functionality of selecting the multilocus genotype definition from the mlg slot.</li>
<li>
<code>nmll</code> counts the number of multilocus lineages</li>
<li>
<code>mll.custom</code> allows the user to define custom multilocus genotypes.</li>
<li>
<code>mll.levels</code> allows the user to edit the names of custom multilocus genotypes.</li>
<li>
<code>poppr_has_parallel</code> will return <code>TRUE</code> if poppr was built with OpenMP parallel library.</li>
<li>
<code>win.ia</code> calculates windows of \bar{r}_d along genlight chromosomes.</li>
<li>
<code>samp.ia</code> calculates \bar{r}_d for genlight object by randomly sampling a user-defined number of SNPs.</li>
<li>
<code>test_replen</code> will test repeat lengths of microsatellite markers for consistency.</li>
<li>
<code>fix_replen</code> will fix inconsistent repeat lengths for microsatellite markers.</li>
<li>
<code>diversity_stats</code> returns a matrix containing diversity statistics. Defaults to 4 found in <code>poppr</code>, but can be extended to any statistic that can be calculated on a vector of MLG counts.</li>
<li>
<code>diversity_boot</code> will bootstrap a MLG matrix over the statistics specified for <code>get_stats</code>. Can also perform rarefaction bootstrap.</li>
<li>
<code>diversity_ci</code> will calculate and plot confidence intervals for bootstrap resampling of an MLG matrix. This includes rarefaction to the smallest sample size.</li>
<li>
<code>imsn</code> provides an interactive shiny interface for construction of minimum spanning networks.</li>
<li>
<code>pair.ia</code> will calculate the index of association for pairs of loci and plot heatmaps.</li>
</ul>
</div>
<div id="new-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#new-classes" class="anchor"></a>NEW CLASSES</h2>
<ul>
<li>snpclone is an extension of the genlight object that acts very much like genclone. It contains an mlg slot.</li>
<li>MLG is an internal class that lives inside the mlg slot of snpclone and genclone objects. It allows the user to easily switch between multilocus genotype definitions.</li>
</ul>
</div>
<div id="bug-fixes" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>I was going too fast to count.</li>
</ul>
</div>
</div>
    <div id="poppr-1-1-5" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-1-5" class="anchor"></a>poppr 1.1.5</h1>
<div id="bug-fix-10" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-10" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Fixed internal bug for <code>fix_negative_branch</code> when only one branch had a negative edge.</li>
<li>Fixed bug in <code>diss.dist</code> where a single locus would return an error.</li>
<li>Fixed bug in <code>poppr.amova</code> where a single locus would return an error due to <code>repool_haplotypes</code>.</li>
<li>Fixed bug from the future! mlg.table will now return a matrix all the time (Fix #25).</li>
</ul>
</div>
</div>
    <div id="poppr-1-1-4" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-1-4" class="anchor"></a>poppr 1.1.4</h1>
<div id="bug-fix-11" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-11" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Fixed an internal bug that fails only on Windows OS.</li>
</ul>
</div>
</div>
    <div id="poppr-1-1-3" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-1-3" class="anchor"></a>poppr 1.1.3</h1>
<div id="new-features-8" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-8" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>new arguments to <code>plot_poppr_msn</code> to allow for easier manipulation of node sizes and of labeling</li>
<li>read.genalex can now take read text connections as input. Addresses issue #8</li>
<li>users can now specify cutoff for missing values in <code>aboot</code>
</li>
</ul>
</div>
<div id="bug-fix-12" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-12" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Fixed issue where monomorphic loci would cause an error in <code>recode_polyploids</code>
</li>
<li>Fixed logical error that would cause The infinite alleles model of Bruvo’s distance to inflate the distance. (Found by Michael Metzger. Addresses issue #5).</li>
<li>AMOVA can now take subset genclone objects (Addresses issue #7).</li>
<li>in <code>mlg.table</code>, the mlgsub argument will now subset by name instead of index (fixed in #7).</li>
<li>Fixed issue for neighbor-joining trees where the internal function to fix negative branch lengths was accidentally shuffling the corrected branches. Addresses issue #11.</li>
<li>
<code>diss.dist</code> can now be used with <code>aboot</code>
</li>
</ul>
</div>
<div id="misc-7" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-7" class="anchor"></a>MISC</h2>
<ul>
<li>
<code>info_table</code> will print a discrete scale as opposed to colorbar when type = “ploidy”</li>
<li>attempted to make model choices for Bruvo’s distance more clear in the documentation</li>
</ul>
</div>
</div>
    <div id="poppr-1-1-2" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-1-2" class="anchor"></a>poppr 1.1.2</h1>
<div id="bug-fix-13" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-13" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Fixed memory allocation bug (Further addresses issue #2).</li>
<li>Memory allocated in C function bruvo_dist is now properly freed.</li>
</ul>
</div>
</div>
    <div id="poppr-1-1-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-1-1" class="anchor"></a>poppr 1.1.1</h1>
<div id="bug-fix-14" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-14" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Fixed bug where the loss and add options for Bruvo’s distance were switched.</li>
<li>Fixed illegal memory access error by UBSAN. Made memory management of internal C functions more sane. (Addresses issue #2).</li>
<li>Fixed directional quotes and em-dashes produced by Mavericks (Addresses issue #3).</li>
</ul>
</div>
</div>
    <div id="poppr-1-1-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-1-0" class="anchor"></a>poppr 1.1.0</h1>
<div id="new-features-9" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-9" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>Polyploids with ambiguous genotypes are now supported in poppr. See documentation for <code>recode_polyploids</code> for details.</li>
<li>Calculations of Bruvo’s distance now features correction for partial missing data utilizing genome addition and genome loss models as presented in Bruvo et al. 2004.</li>
<li>
<code>diss.dist</code> now has options to return raw distances and a matrix instead of a dist object.</li>
<li>
<code>read.genalex</code> now has the option to import as a genclone object. This is the default action.</li>
<li>
<code>poppr.all</code> will be able to analyze lists of genind or genclone objects.</li>
<li>
<code>ia</code> now has the argument valuereturn which will return the sampled data.</li>
<li>
<code>[bruvo,poppr].msn</code> functions now give the user the choice to show the graph.</li>
<li>
<code>bruvo.boot</code> has a cleaner plot style.</li>
</ul>
</div>
<div id="new-data-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#new-data-classes" class="anchor"></a>NEW DATA CLASSES</h2>
<ul>
<li>The <code>genclone</code> object is a new extension of the <code>genind</code> object from adegenet. This object contains slots containing population hierarchies and multilocus genotype definitions and will work with all analyses in adegenet and poppr.</li>
</ul>
</div>
<div id="new-functions-5" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-5" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>[get,set,name,split,add]hierarchy - functions that will manipulate the hierarchy slot in a <code>genclone</code> object utilizing hierarchical formulae as arguments for simplification.</li>
<li>
<code>setpop</code> will set the population of a <code>genclone</code> object utilizing model formulae regarding the hierarchy slot.</li>
<li>
<code>as.genclone</code> will automatically convert genind objects to genclone objects.</li>
<li>
<code>is.genclone</code> checks the validity of genclone objects.</li>
<li>
<code>poppr.amova</code> will run amova on any hierarchical level. This also includes the feature to run amova on clone censored data sets. It utilizes the ade4 version of amova.</li>
<li>
<code>info_table</code> will calculate missing data per population per locus or ploidy per individual per locus and gives the user the option to visualize this as a heatmap.</li>
<li>
<code>locus_table</code> will calculate diversity and evenness statistics over all loci in a genind or genclone object.</li>
<li>
<code>*.dist</code> functions will calculate Nei’s distance, Rogers’ Distance, Edwards’ Distance, Reynolds’ Distance, and Provestis’ Distance.</li>
<li>
<code>aboot</code> will allow the user to create bootstrapped dendrograms for ANY distance that can be calculated on genind or genpop objects.</li>
<li>
<code>plot_poppr_msn</code> will plot minimum spanning networks produced with poppr.</li>
<li>
<code>private_alleles</code> will give information about the presence of private alleles within a genind or genclone object.</li>
<li>
<code>recode_polyploids</code> will take in a polyploid genind/genclone object (with missing alleles coded as extra zero-value allele) and recode them to have frequencies relative to the observed number of alleles.</li>
<li>
<code>genotype_curve</code> will create a genotype accumulation curve for increasing number of loci.</li>
<li>
<code>mlg.id</code> will return a list indicating the samples belonging to a specific multilocus genotype.</li>
</ul>
</div>
<div id="new-data-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#new-data-sets" class="anchor"></a>NEW DATA SETS</h2>
<ul>
<li>Pinf - a data set of 86 isolates from different populations of the late blight pathogen, Phytophthora infestans. Provided by Erica Goss</li>
<li>monpop - a large data set of 694 Monilinia fructicola isolates from a single orchard over three years. Provided by Sydney E. Everhart</li>
</ul>
</div>
<div id="new-car" class="section level2">
<h2 class="hasAnchor">
<a href="#new-car" class="anchor"></a>NEW CAR</h2>
<ul>
<li>Not really.</li>
</ul>
</div>
<div id="namespace-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#namespace-changes" class="anchor"></a>NAMESPACE CHANGES</h2>
<ul>
<li>poppr no longer depends on pegas.</li>
<li>ade4 and reshape2 are now explicitly required.</li>
</ul>
</div>
<div id="improvements-1" class="section level2">
<h2 class="hasAnchor">
<a href="#improvements-1" class="anchor"></a>IMPROVEMENTS</h2>
<ul>
<li>default shuffling algorithm has been implemented in C to increase speed.</li>
<li>output of the mlg functions are now represented as integers to decrease their size in memory.</li>
<li>
<code>mlg.matrix</code> is now calculated faster utilizing R’s internal tabulating capabilities.</li>
<li>The function <code>poppr</code> will no longer return rounded results, but rather is printed with three significant digits.</li>
</ul>
</div>
<div id="misc-8" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-8" class="anchor"></a>MISC</h2>
<ul>
<li>Added unit tests.</li>
<li>The poppr user manual has been shortened to only include instructions on data manipulation.</li>
<li>A new vignette, “Algorithms and Equations” gives algorithmic details for calculations performed in poppr.</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-7" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-7" class="anchor"></a>poppr 1.0.7</h1>
<div id="update" class="section level2">
<h2 class="hasAnchor">
<a href="#update" class="anchor"></a>UPDATE</h2>
<ul>
<li>Updated README to include link to poppr google group.</li>
</ul>
</div>
<div id="bug-fix-update" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-update" class="anchor"></a>BUG FIX UPDATE</h2>
<ul>
<li>Made last bug fix more stable (corrected on ape side).</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-6" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-6" class="anchor"></a>poppr 1.0.6</h1>
<div id="bug-fix-15" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-15" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Fixed bug for users who have downloaded ape version 3.1 or higher where bruvo.boot would throw an error.</li>
</ul>
</div>
<div id="misc-9" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-9" class="anchor"></a>MISC</h2>
<ul>
<li>Updated citation information.</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-5" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-5" class="anchor"></a>poppr 1.0.5</h1>
<div id="notable-change" class="section level2">
<h2 class="hasAnchor">
<a href="#notable-change" class="anchor"></a>NOTABLE CHANGE</h2>
<ul>
<li>The default shuffling algorithm for calculating the index of association has changed from multilocus-style sampling to permutation of alleles. All of the 4 methods are available, but new assignments are as follows: Method 1: permute alleles, Method 2: parametric bootstrap, Method 3: non-parametric bootstrap, Method 4: Multilocus-style sampling. Previously, Multilocus was 1 and the rest followed in the same order. There should be no compatibility issues with this change. Functions affected: <code>ia</code>, <code>poppr</code> <code>shufflepop</code>
</li>
</ul>
</div>
<div id="bug-fix-16" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-16" class="anchor"></a>BUG FIX</h2>
<ul>
<li>Bootstrapping algorithm for <code>bruvo.boot</code> function was not shuffling the repeat lengths for each locus resulting in potentially erroneous bootstrap support values. This has been fixed by implementing an internal S4 class that will allow direct bootstrapping of the data and repeat lengths together.</li>
<li>An occasional error, “INTEGER() can only be applied to a ‘integer’, not a ‘NULL’” in <code>bruvo.boot</code> or <code>bruvo.dist</code> fixed.</li>
</ul>
</div>
<div id="improvements-2" class="section level2">
<h2 class="hasAnchor">
<a href="#improvements-2" class="anchor"></a>IMPROVEMENTS</h2>
<ul>
<li>Changes to <code>bruvo.boot</code> allow for ever so slightly faster bootstrapping.</li>
</ul>
</div>
<div id="misc-10" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-10" class="anchor"></a>MISC</h2>
<ul>
<li>Permutations for I_A and \bar{r}_d are now visualized as a progress bar as opposed to dots.</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-4" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-4" class="anchor"></a>poppr 1.0.4</h1>
<div id="bug-fix-17" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fix-17" class="anchor"></a>BUG FIX</h2>
<ul>
<li>A previous error where bootstrap values greater than 100 were reported from <code>bruvo.boot</code> on UPGMA trees has been fixed.</li>
<li>Fixed correction of negative branch lengths using Kuhner and Felsenstein</li>
</ul>
<p>(1994) normalization for NJ trees.</p>
</div>
<div id="misc-11" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-11" class="anchor"></a>MISC</h2>
<ul>
<li>github repository for poppr has changed from github.com/poppr/poppr to github.com/grunwaldlab/poppr</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-3" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-3" class="anchor"></a>poppr 1.0.3</h1>
<div id="improvements-3" class="section level2">
<h2 class="hasAnchor">
<a href="#improvements-3" class="anchor"></a>IMPROVEMENTS</h2>
<ul>
<li>Optimized internal sampling function to run up to 2x faster.</li>
<li>Utilized rmultinom function to increase speed of bootstrap sampling methods for shufflepop and ia.</li>
</ul>
</div>
<div id="new-features-10" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-10" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>Function <code>informloci</code> will remove phylogenetically uninformative loci.</li>
</ul>
</div>
<div id="namespace" class="section level2">
<h2 class="hasAnchor">
<a href="#namespace" class="anchor"></a>NAMESPACE</h2>
<ul>
<li>Now importing specific functions from igraph and ape due to dependency issues.</li>
<li>Removed igraph, ape, ggplot2, and phangorn form “Dependencies”, but keeping them in “Imports”.</li>
</ul>
</div>
<div id="bug-fixes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-1" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>
<code>read.genalex</code> will no longer insert an “X” in front of loci with numeric names.</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-2" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-2" class="anchor"></a>poppr 1.0.2</h1>
<div id="bug-fixes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-2" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>Fixed bug in diss.dist function that would return an inflated distance for haploids.</li>
</ul>
</div>
<div id="documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation" class="anchor"></a>DOCUMENTATION</h2>
<ul>
<li>Added explanation for the index of association in poppr_manual.</li>
<li>Expanded installation section to include installation instructions from github.</li>
</ul>
</div>
<div id="misc-12" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-12" class="anchor"></a>MISC</h2>
<ul>
<li>internal permutation algorithm no longer lists permutations in reverse order</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-1" class="anchor"></a>poppr 1.0.1</h1>
<div id="improvements-4" class="section level2">
<h2 class="hasAnchor">
<a href="#improvements-4" class="anchor"></a>IMPROVEMENTS</h2>
<ul>
<li>Algorithm for the index of association was updated to increase speed.</li>
</ul>
</div>
<div id="bug-fixes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-3" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>Removed unnecessary rounding factor for missing data in <code>bruvo.dist</code>.</li>
<li>Corrected handling of duplicate entries for <code>read.genind</code>.</li>
<li>Input values that are not multiples of the specified repeat length for Bruvo’s distance are now rounded (as opposed to being forced as integers).</li>
</ul>
</div>
<div id="misc-13" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-13" class="anchor"></a>MISC</h2>
<ul>
<li>Vignette updated for aesthetics and to reflect algorithmic changes.</li>
</ul>
</div>
</div>
    <div id="poppr-1-0-0" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-1-0-0" class="anchor"></a>poppr 1.0.0</h1>
<div id="misc-14" class="section level2">
<h2 class="hasAnchor">
<a href="#misc-14" class="anchor"></a>MISC</h2>
<ul>
<li>Poppr has been confirmed to work on Linux, Mac, and Windows systems with R 3.0.0.</li>
<li>Vignette <code>poppr_manual</code> now has cross-references to different sections.</li>
<li>Vignette <code>poppr_manual</code> is quicker loading.</li>
</ul>
</div>
<div id="bug-fixes-4" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-4" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>removed alpha channel from plot for resampled values of I_A and \bar{r}_d due to warnings.</li>
</ul>
</div>
</div>
    <div id="poppr-0-4-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-4-1" class="anchor"></a>poppr 0.4.1</h1>
<div id="new-features-11" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-11" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>
<code>getfile</code> has a new argument, “combine”, which will automatically add the path to the list of files, so they can be read without switching working directory.</li>
<li>information printed to screen from <code>missingno</code> and <code>mlg.crosspop</code> will now be wrapped to 80 characters.</li>
</ul>
</div>
<div id="bug-fixes-5" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-5" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>
<code>poppr</code> will now be able to correctly recognize GenAlEx files with both geographic and regional data.</li>
<li>calculation of the index of association on P/A data with missing values will no longer return an error.</li>
</ul>
</div>
</div>
    <div id="poppr-0-4" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-4" class="anchor"></a>poppr 0.4</h1>
<div id="bug-fixes-6" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-6" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>mistake in Bruvo’s distance where it did not correctly check for ploidy level was fixed.</li>
<li>
<code>read.genalex</code> will be able to correctly distinguish between SNP and AFLP data.</li>
<li>
<code>read.genalex</code> can now correctly recognize regional formatting without an extra column.</li>
</ul>
</div>
<div id="new-features-12" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-12" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>
<code>read.genalex</code> will now be able to take in a file that is formatted with both regional and geographic data.</li>
<li>
<code>genind2genalex</code> can now export xy coordinates into the GenAlEx format.</li>
<li>
<code>poppr_manual</code> vignette now contains images of example GenAlEx files.</li>
</ul>
</div>
<div id="new-files" class="section level2">
<h2 class="hasAnchor">
<a href="#new-files" class="anchor"></a>NEW FILES</h2>
<ul>
<li>
<code>rootrot2.csv</code> is an example of a GenAlEx file formatted with regional data.</li>
</ul>
</div>
<div id="other-updates" class="section level2">
<h2 class="hasAnchor">
<a href="#other-updates" class="anchor"></a>OTHER UPDATES</h2>
<ul>
<li>function for guessing repeat lengths for Bruvo’s distance moved into internal file.</li>
<li>redundancy in <code>read.genalex</code> was removed.</li>
<li>changed instructions in README</li>
</ul>
</div>
</div>
    <div id="poppr-0-3-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-3-1" class="anchor"></a>poppr 0.3.1</h1>
<div id="bug-fixes-7" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-7" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>
<code>read.genalex</code> will now give a warning whenever the input file is not comma delimited.</li>
</ul>
</div>
</div>
    <div id="poppr-0-3" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-3" class="anchor"></a>poppr 0.3</h1>
<div id="new-functions-6" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-6" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>
<code>poppr.msn</code> will draw a minimum spanning network for any distance matrix derived from your data set.</li>
</ul>
</div>
<div id="new-features-13" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-13" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>vignette now has sections describing <code>poppr.msn</code>, <code>diss.dist</code>, <code>greycurve</code>, and a section discussing how to export graphics.</li>
</ul>
</div>
<div id="bug-fixes-8" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-8" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>The graphs output by <code>poppr</code> and <code>ia</code> will now display \bar{r}_d instead of \bar{r}_D.</li>
<li>
<code>bruvo.boot</code> now has a dedicated <code>quiet</code> argument.</li>
</ul>
</div>
</div>
    <div id="poppr-0-2-2" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-2-2" class="anchor"></a>poppr 0.2.2</h1>
<div id="new-features-14" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-14" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>index of association distributions will now feature a rug plot at the bottom as a better way to visualize the distribution of the index of association from the shuffled data sets.</li>
</ul>
</div>
</div>
    <div id="poppr-0-2-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-2-1" class="anchor"></a>poppr 0.2.1</h1>
<div id="new-functions-7" class="section level2">
<h2 class="hasAnchor">
<a href="#new-functions-7" class="anchor"></a>NEW FUNCTIONS</h2>
<ul>
<li>
<code>diss.dist</code> will produce a distance matrix based on discreet distances.</li>
<li>
<code>greycurve</code> will produce a grey scale adjusted to user-supplied parameters. This will be useful for future minimum spanning network functions.</li>
</ul>
</div>
<div id="new-features-15" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-15" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>
<code>bruvo.msn</code> can now adjust the edge grey level to be weighted toward either closely or distantly weighted individuals.</li>
<li>
<code>bruvo.msn</code> will now return a list giving the user the graph with all of the color, label, and weight properties so that they can plot it themselves. The legend arguments are also returned.</li>
</ul>
</div>
<div id="bug-fixes-9" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-fixes-9" class="anchor"></a>BUG FIXES</h2>
<ul>
<li>fixed shufflepop so that it will now shuffle PA markers with a specific method</li>
<li>fixed warning message mistakes in clonecorrect function.</li>
</ul>
</div>
</div>
    <div id="poppr-0-2" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-2" class="anchor"></a>poppr 0.2</h1>
<div id="new-features-16" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features-16" class="anchor"></a>NEW FEATURES</h2>
<ul>
<li>Added NEWS file and will now be incrementing version number (3/15/2013)</li>
</ul>
</div>
</div>
    <div id="poppr-0-1" class="section level1">
<h1 class="hasAnchor">
<a href="#poppr-0-1" class="anchor"></a>poppr 0.1</h1>
<ul>
<li>First development version of poppr (2012 - 3/2013)</li>
</ul>
</div>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#poppr-2-5-0">2.5.0</a></li>
        <li><a href="#poppr-2-4-1">2.4.1</a></li>
        <li><a href="#poppr-2-4-0">2.4.0</a></li>
        <li><a href="#poppr-2-3-0">2.3.0</a></li>
        <li><a href="#poppr-2-2-1">2.2.1</a></li>
        <li><a href="#poppr-2-2-0">2.2.0</a></li>
        <li><a href="#poppr-2-1-1">2.1.1</a></li>
        <li><a href="#poppr-2-1-0">2.1.0</a></li>
        <li><a href="#poppr-2-0-2">2.0.2</a></li>
        <li><a href="#poppr-2-0-1">2.0.1</a></li>
        <li><a href="#poppr-2-0-0">2.0.0</a></li>
        <li><a href="#poppr-1-1-5">1.1.5</a></li>
        <li><a href="#poppr-1-1-4">1.1.4</a></li>
        <li><a href="#poppr-1-1-3">1.1.3</a></li>
        <li><a href="#poppr-1-1-2">1.1.2</a></li>
        <li><a href="#poppr-1-1-1">1.1.1</a></li>
        <li><a href="#poppr-1-1-0">1.1.0</a></li>
        <li><a href="#poppr-1-0-7">1.0.7</a></li>
        <li><a href="#poppr-1-0-6">1.0.6</a></li>
        <li><a href="#poppr-1-0-5">1.0.5</a></li>
        <li><a href="#poppr-1-0-4">1.0.4</a></li>
        <li><a href="#poppr-1-0-3">1.0.3</a></li>
        <li><a href="#poppr-1-0-2">1.0.2</a></li>
        <li><a href="#poppr-1-0-1">1.0.1</a></li>
        <li><a href="#poppr-1-0-0">1.0.0</a></li>
        <li><a href="#poppr-0-4-1">0.4.1</a></li>
        <li><a href="#poppr-0-4">0.4</a></li>
        <li><a href="#poppr-0-3-1">0.3.1</a></li>
        <li><a href="#poppr-0-3">0.3</a></li>
        <li><a href="#poppr-0-2-2">0.2.2</a></li>
        <li><a href="#poppr-0-2-1">0.2.1</a></li>
        <li><a href="#poppr-0-2">0.2</a></li>
        <li><a href="#poppr-0-1">0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
