<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate dissimilarity or Euclidean distance for genlight objects — bitwise.dist • poppr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calculate dissimilarity or Euclidean distance for genlight objects — bitwise.dist" />
<meta property="og:description" content="This function calculates both dissimilarity and Euclidean distances for
genlight or snpclone objects." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.7.99-41</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate dissimilarity or Euclidean distance for genlight objects</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/bitwise.r'><code>R/bitwise.r</code></a></small>
    <div class="hidden name"><code>bitwise.dist.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates both dissimilarity and Euclidean distances for
genlight or <a href='genclone-class.html'>snpclone</a> objects.</p>
    </div>

    <pre class="usage"><span class='fu'>bitwise.dist</span><span class='op'>(</span>
  <span class='va'>x</span>,
  percent <span class='op'>=</span> <span class='cn'>TRUE</span>,
  mat <span class='op'>=</span> <span class='cn'>FALSE</span>,
  missing_match <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale_missing <span class='op'>=</span> <span class='cn'>FALSE</span>,
  euclidean <span class='op'>=</span> <span class='cn'>FALSE</span>,
  differences_only <span class='op'>=</span> <span class='cn'>FALSE</span>,
  threads <span class='op'>=</span> <span class='fl'>0L</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a genlight or <a href='genclone-class.html'>snpclone</a> object.</p></td>
    </tr>
    <tr>
      <th>percent</th>
      <td><p><code>logical</code>. Should the distance be represented from 0 to
1? Default set to <code>TRUE</code>. <code>FALSE</code> will return the distance
represented as integers from 1 to n where n is the number of loci.
This option has no effect if <code>euclidean = TRUE</code></p></td>
    </tr>
    <tr>
      <th>mat</th>
      <td><p><code>logical</code>. Return a matrix object. Default set to
<code>FALSE</code>, returning a dist object. <code>TRUE</code> returns a matrix object.</p></td>
    </tr>
    <tr>
      <th>missing_match</th>
      <td><p><code>logical</code>. Determines whether two samples differing
by missing data in a location should be counted as matching at that
location. Default set to <code>TRUE</code>, which forces missing data to match
with anything. <code>FALSE</code> forces missing data to not match with any other
information, <strong>including other missing data</strong>.</p></td>
    </tr>
    <tr>
      <th>scale_missing</th>
      <td><p>A logical. If <code>TRUE</code>, comparisons with missing
data is scaled up proportionally to the number of columns used by
multiplying the value by <code>m / (m - x)</code> where m is the number of
loci and x is the number of missing sites. This option matches the behavior
of base R's <code><a href='https://rdrr.io/r/stats/dist.html'>dist()</a></code> function.
Defaults to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>euclidean</th>
      <td><p><code>logical</code>. if <code>TRUE</code>, the Euclidean distance will
be calculated.</p></td>
    </tr>
    <tr>
      <th>differences_only</th>
      <td><p><code>logical</code>. When <code>differences_only = TRUE</code>,
the output will reflect the number of different loci. The default setting,
<code>differences_only = FALSE</code>, reflects the number of different alleles.
Note: this has no effect on haploid organisms since 1 locus = 1 allele.
This option is NOT recommended.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>The maximum number of parallel threads to be used within this
function. A value of 0 (default) will attempt to use as many threads as
there are available cores/CPUs. In most cases this is ideal. A value of 1
will force the function to run serially, which may increase stability on
some systems. Other values may be specified, but should be used with
caution.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dist object containing pairwise distances between samples.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The default distance calculated here is quite simple and goes by
many names depending on its application. The most familiar name might be
the Hamming distance, or the number of differences between two strings.</p>
<p>As of poppr version 2.8.0, this function now also calculates Euclidean
distance and is considerably faster and more memory-efficient than the
standard <code><a href='https://rdrr.io/r/stats/dist.html'>dist()</a></code> function.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>This function is optimized for genlight and
<a href='genclone-class.html'>snpclone</a> objects. This does not mean that it is a
catch-all optimization for SNP data. Three assumptions must be met for this
function to work:</p><ol>
<li><p>SNPs are bi-allelic</p></li>
<li><p>Samples are haploid or diploid</p></li>
<li><p>All samples have the same ploidy</p></li>
</ol>

<p>If the user supplies a genind or
<a href='genclone-class.html'>genclone</a> object, <code><a href='nei.dist.html'>prevosti.dist()</a></code> will be used for
calculation.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='diss.dist.html'>diss.dist()</a></code>, <a href='genclone-class.html'>snpclone</a>,
genlight, <code><a href='win.ia.html'>win.ia()</a></code>,  <code><a href='samp.ia.html'>samp.ia()</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Zhian N. Kamvar, Jonah C. Brooks</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>999</span><span class='op'>)</span>
<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'>glSim</span><span class='op'>(</span>n.ind <span class='op'>=</span> <span class='fl'>10</span>, n.snp.nonstruc <span class='op'>=</span> <span class='fl'>5e2</span>, n.snp.struc <span class='op'>=</span> <span class='fl'>5e2</span>, ploidy <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='va'>x</span>
</div><div class='output co'>#&gt;  /// GENLIGHT OBJECT /////////
#&gt; 
#&gt;  // 10 genotypes,  1,000 binary SNPs, size: 20.6 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  // Basic content
#&gt;    @gen: list of 10 SNPbin
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 4-6)
#&gt;    @other: a list containing: ancestral.pops 
#&gt; </div><div class='input'><span class='co'># Assess fraction of different alleles</span>
<span class='fu'><a href='https://rdrr.io/r/base/system.time.html'>system.time</a></span><span class='op'>(</span><span class='va'>xd</span> <span class='op'>&lt;-</span> <span class='fu'>bitwise.dist</span><span class='op'>(</span><span class='va'>x</span>, threads <span class='op'>=</span> <span class='fl'>1L</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.000   0.000   0.001 </div><div class='input'><span class='va'>xd</span>
</div><div class='output co'>#&gt;         1      2      3      4      5      6      7      8      9
#&gt; 2  0.2230                                                        
#&gt; 3  0.2260 0.2280                                                 
#&gt; 4  0.2250 0.2170 0.2040                                          
#&gt; 5  0.3795 0.3835 0.3795 0.3805                                   
#&gt; 6  0.4035 0.3985 0.4055 0.3975 0.2100                            
#&gt; 7  0.4005 0.3955 0.3935 0.3885 0.2000 0.2200                     
#&gt; 8  0.3880 0.3860 0.3870 0.3960 0.2035 0.2205 0.2135              
#&gt; 9  0.3920 0.4030 0.4080 0.3970 0.2135 0.2125 0.2005 0.2150       
#&gt; 10 0.3935 0.3905 0.4015 0.3885 0.2230 0.2160 0.2120 0.2265 0.2195</div><div class='input'>
<span class='co'># Calculate Euclidean distance</span>
<span class='fu'><a href='https://rdrr.io/r/base/system.time.html'>system.time</a></span><span class='op'>(</span><span class='va'>xdt</span> <span class='op'>&lt;-</span> <span class='fu'>bitwise.dist</span><span class='op'>(</span><span class='va'>x</span>, euclidean <span class='op'>=</span> <span class='cn'>TRUE</span>, scale_missing <span class='op'>=</span> <span class='cn'>TRUE</span>, threads <span class='op'>=</span> <span class='fl'>1L</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;       0       0       0 </div><div class='input'><span class='va'>xdt</span>
</div><div class='output co'>#&gt;           1        2        3        4        5        6        7        8
#&gt; 2  23.40940                                                               
#&gt; 3  23.74868 24.04163                                                      
#&gt; 4  23.36664 23.10844 22.31591                                             
#&gt; 5  34.36568 34.82815 34.71311 34.56877                                    
#&gt; 6  35.59494 35.45420 36.09709 35.76311 22.93469                           
#&gt; 7  35.53871 35.17101 35.79106 35.00000 22.31591 23.57965                  
#&gt; 8  34.75629 34.98571 35.12834 35.49648 22.24860 23.08679 23.13007         
#&gt; 9  35.49648 35.88872 36.19392 35.41186 23.04344 23.13007 22.24860 22.93469
#&gt; 10 34.71311 34.68429 35.81899 35.05710 23.57965 23.10844 22.80351 23.81176
#&gt;           9
#&gt; 2          
#&gt; 3          
#&gt; 4          
#&gt; 5          
#&gt; 6          
#&gt; 7          
#&gt; 8          
#&gt; 9          
#&gt; 10 23.51595</div><div class='input'>
<span class='co'># \dontrun{</span>

<span class='co'># This function is more efficient in both memory and speed than [dist()] for</span>
<span class='co'># calculating Euclidean distance on genlight objects. For example, we can</span>
<span class='co'># observe a clear speed increase when we attempt a calculation on 100k SNPs</span>
<span class='co'># with 10% missing data:</span>

<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>999</span><span class='op'>)</span>
<span class='va'>mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>:</span><span class='fl'>2</span>, <span class='cn'>NA</span><span class='op'>)</span>, 
                     <span class='fl'>100000</span> <span class='op'>*</span> <span class='fl'>50</span>, 
                     replace <span class='op'>=</span> <span class='cn'>TRUE</span>, 
                     prob <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.3</span>, <span class='fl'>0.3</span>, <span class='fl'>0.3</span>, <span class='fl'>0.1</span><span class='op'>)</span><span class='op'>)</span>,
              nrow <span class='op'>=</span> <span class='fl'>50</span><span class='op'>)</span>
<span class='va'>glite</span> <span class='op'>&lt;-</span> <span class='fu'>new</span><span class='op'>(</span><span class='st'>"genlight"</span>, <span class='va'>mat</span>, ploidy <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>

<span class='co'># Default Euclidean distance </span>
<span class='fu'><a href='https://rdrr.io/r/base/system.time.html'>system.time</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/dist.html'>dist</a></span><span class='op'>(</span><span class='va'>glite</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   2.182   0.052   2.279 </div><div class='input'>
<span class='co'># Bitwise dist</span>
<span class='fu'><a href='https://rdrr.io/r/base/system.time.html'>system.time</a></span><span class='op'>(</span><span class='fu'>bitwise.dist</span><span class='op'>(</span><span class='va'>glite</span>, euclidean <span class='op'>=</span> <span class='cn'>TRUE</span>, scale_missing <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.756   0.005   0.788 </div><div class='input'>
<span class='co'># }</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


