<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="GENclone is an S4 class that extends the 
  genind object. SNPclone is an S4 class
  that extends the genlight object. They will have
  all of the same attributes as their parent classes, but they will contain
  one extra slot to store extra information about multilocus genotypes."><title>GENclone and SNPclone classes — genclone-class • poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GENclone and SNPclone classes — genclone-class"><meta property="og:description" content="GENclone is an S4 class that extends the 
  genind object. SNPclone is an S4 class
  that extends the genlight object. They will have
  all of the same attributes as their parent classes, but they will contain
  one extra slot to store extra information about multilocus genotypes."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GENclone and SNPclone classes</h1>
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/R/classes.r" class="external-link"><code>R/classes.r</code></a></small>
      <div class="d-none name"><code>genclone-class.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><strong>GENclone</strong> is an S4 class that extends the 
  <code><a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a></code> object.<br><strong>SNPclone</strong> is an S4 class
  that extends the <code><a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a></code> object.<br><br> They will have
  all of the same attributes as their parent classes, but they will contain
  one extra slot to store extra information about multilocus genotypes.</p>
    </div>


    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The genclone and snpclone classes will allow for more optimized 
  methods of clone correction.</p>  
<p>Previously for <a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a> and <a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a> objects, 
  multilocus genotypes were not retained after a data set was subset by 
  population. The new <strong><code>mlg</code></strong> slot allows us to assign the 
  multilocus genotypes and retain that information no matter how we subset 
  the data set. This new slot can either contain numeric values for 
  multilocus genotypes OR it can contain a special internal 
  <code><a href="MLG-class.html">MLG</a></code> class that allows for custom multilocus genotype 
  definitions and filtering.</p>
    </div>
    <div class="section level2">
    <h2 id="slots">Slots<a class="anchor" aria-label="anchor" href="#slots"></a></h2>
    

<dl><dt><code>mlg</code></dt>
<dd><p>a vector representing multilocus genotypes for the data set OR an 
object of class <code><a href="MLG-class.html">MLG</a></code>.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>When calculating multilocus genotypes for genclone objects, a rank 
  function is used, but calculation of multilocus genotypes for snpclone
  objects is distance-based (via <code><a href="bitwise.dist.html">bitwise.dist</a></code> and 
  <code><a href="mlg.filter.html">mlg.filter</a></code>). This means that genclone objects are sensitive
  to missing data, whereas snpclone objects are insensitive.</p>
    </div>
    <div class="section level2">
    <h2 id="extends">Extends<a class="anchor" aria-label="anchor" href="#extends"></a></h2>
    <p>The <code>genclone</code> class extends class 
  <code>"<a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a>"</code>, directly. <br> The <code>snpclone</code> class 
  extends class <code>"<a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a>"</code>, directly.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="coercion-methods.html">as.genclone</a></code> <code><a href="snpclone-coercion-methods.html">as.snpclone</a></code> 
  <code><a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a></code> <code><a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a></code> 
  <code>strata</code> <code>setPop</code> 
  <code><a href="MLG-class.html">MLG</a></code> <code><a href="mll-method.html">mll</a></code> <code><a href="mlg.filter.html">mlg.filter</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Zhian N. Kamvar</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genclone objects can be created from genind objects</span></span></span>
<span class="r-in"><span><span class="co">#</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">partial_clone</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">partial_clone</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 50 individuals; 10 loci; 35 alleles; size: 22.6 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  50 x 35 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 3-5)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 35 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: old2new_genind(object = x, donor = new(class(x)))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 12-13)</span>
<span class="r-in"><span><span class="op">(</span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">partial_clone</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    26 original multilocus genotypes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    50 diploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    10 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0 strata. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4 populations defined - 1, 2, 3, 4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># snpclone objects can be created from genlight objects</span></span></span>
<span class="r-in"><span><span class="co">#</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">999</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">gl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/glSim.html" class="external-link">glSim</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">0</span>, n.snp.struc <span class="op">=</span> <span class="fl">1e3</span>, ploidy <span class="op">=</span> <span class="fl">2</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  /// GENLIGHT OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 100 genotypes,  1,000 binary SNPs, size: 172 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0 (0 %) missing data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @gen: list of 100 SNPbin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 50-50)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: ancestral.pops </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="op">(</span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="snpclone-coercion-methods.html">as.snpclone</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">gl</span>, <span class="va">gl</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ||| SNPCLONE OBJECT |||||||||</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || 200 genotypes,  1,000 binary SNPs, size: 353.7 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0 (0 %) missing data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @gen: list of 200 SNPbin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @mlg: 100 original multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 100-100)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: elements without names </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># Use mlg.filter to create a distance threshold to define multilocus genotypes.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">sc</span>, threads <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.25</span></span></span>
<span class="r-in"><span><span class="va">sc</span> <span class="co"># 82 mlgs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ||| SNPCLONE OBJECT |||||||||</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || 200 genotypes,  1,000 binary SNPs, size: 353.7 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0 (0 %) missing data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @gen: list of 200 SNPbin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @mlg: 82 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          (0.25) [t], (bitwise.dist) [d], (farthest) [a]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 100-100)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: elements without names </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

