<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="Aeut.html">Aeut</a></code> 
        </dt>
        <dd>Oomycete root rot pathogen *Aphanomyces euteiches* AFLP data</dd>
      </dl><dl><dt>
          
          <code><a href="Pinf.html">Pinf</a></code> 
        </dt>
        <dd>Phytophthora infestans data from Mexico and South America.</dd>
      </dl><dl><dt>
          
          <code><a href="Pram.html">Pram</a></code> 
        </dt>
        <dd>Phytophthora ramorum data from OR Forests and Nurseries (OR and CA)</dd>
      </dl><dl><dt>
          
          <code><a href="aboot.html">aboot()</a></code> 
        </dt>
        <dd>Calculate a dendrogram with bootstrap support using any distance applicable
to genind or genclone objects.</dd>
      </dl><dl><dt>
          
          <code><a href="bitwise.dist.html">bitwise.dist()</a></code> 
        </dt>
        <dd>Calculate dissimilarity or Euclidean distance for genlight objects</dd>
      </dl><dl><dt>
          
          <code><a href="boot.ia.html">boot.ia()</a></code> 
        </dt>
        <dd>Bootstrap the index of association</dd>
      </dl><dl><dt>
          
          <code><a href="bruvo.boot.html">bruvo.boot()</a></code> 
        </dt>
        <dd>Create a tree using Bruvo's Distance with non-parametric bootstrapping.</dd>
      </dl><dl><dt>
          
          <code><a href="bruvo.dist.html">bruvo.dist()</a></code> <code><a href="bruvo.dist.html">bruvo.between()</a></code> 
        </dt>
        <dd>Bruvo's distance for microsatellites</dd>
      </dl><dl><dt>
          
          <code><a href="bruvo.msn.html">bruvo.msn()</a></code> 
        </dt>
        <dd>Create minimum spanning network of selected populations using Bruvo's 
distance.</dd>
      </dl><dl><dt>
          
          <code><a href="clonecorrect.html">clonecorrect()</a></code> 
        </dt>
        <dd>Remove potential bias caused by cloned genotypes in genind or genclone 
object.</dd>
      </dl><dl><dt>
          
          <code><a href="coercion-methods.html">bootgen2genind()</a></code> <code><a href="coercion-methods.html">as.genclone()</a></code> <code><a href="coercion-methods.html">genclone2genind()</a></code> <code><a href="coercion-methods.html">as.genambig()</a></code> 
        </dt>
        <dd>Switch between genind and genclone objects.</dd>
      </dl><dl><dt>
          
          <code><a href="cutoff_predictor.html">cutoff_predictor()</a></code> 
        </dt>
        <dd>Predict cutoff thresholds for use with mlg.filter</dd>
      </dl><dl><dt>
          
          <code><a href="diss.dist.html">diss.dist()</a></code> 
        </dt>
        <dd>Calculate a distance matrix based on relative dissimilarity</dd>
      </dl><dl><dt>
          
          <code><a href="diversity_boot.html">diversity_boot()</a></code> 
        </dt>
        <dd>Perform a bootstrap analysis on diversity statistics</dd>
      </dl><dl><dt>
          
          <code><a href="diversity_ci.html">diversity_ci()</a></code> 
        </dt>
        <dd>Perform bootstrap statistics, calculate, and plot confidence intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="diversity_stats.html">diversity_stats()</a></code> 
        </dt>
        <dd>Produce a table of diversity statistics</dd>
      </dl><dl><dt>
          
          <code><a href="filter_stats.html">filter_stats()</a></code> 
        </dt>
        <dd>Utilize all algorithms of mlg.filter</dd>
      </dl><dl><dt>
          
          <code><a href="fix_replen.html">fix_replen()</a></code> 
        </dt>
        <dd>Find and fix inconsistent repeat lengths</dd>
      </dl><dl><dt>
          
          <code><a href="genclone-class.html">genclone-class</a></code> <code><a href="genclone-class.html">genclone</a></code> <code><a href="genclone-class.html">snpclone-class</a></code> <code><a href="genclone-class.html">snpclone</a></code> 
        </dt>
        <dd>GENclone and SNPclone classes</dd>
      </dl><dl><dt>
          
          <code><a href="genetic_distance.html">nei.dist()</a></code> <code><a href="genetic_distance.html">edwards.dist()</a></code> <code><a href="genetic_distance.html">rogers.dist()</a></code> <code><a href="genetic_distance.html">reynolds.dist()</a></code> <code><a href="genetic_distance.html">provesti.dist()</a></code> <code><a href="genetic_distance.html">prevosti.dist</a></code> 
        </dt>
        <dd>Calculate Genetic Distance for a genind or genclone object.</dd>
      </dl><dl><dt>
          
          <code><a href="genind2genalex.html">genind2genalex()</a></code> 
        </dt>
        <dd>Export data from genind objects to genalex formatted \*.csv files.</dd>
      </dl><dl><dt>
          
          <code><a href="genotype_curve.html">genotype_curve()</a></code> 
        </dt>
        <dd>Produce a genotype accumulation curve</dd>
      </dl><dl><dt>
          
          <code><a href="getfile.html">getfile()</a></code> 
        </dt>
        <dd>Get a file name and path and store them in a list.</dd>
      </dl><dl><dt>
          
          <code><a href="greycurve.html">greycurve()</a></code> 
        </dt>
        <dd>Display a greyscale gradient adjusted to specific parameters</dd>
      </dl><dl><dt>
          
          <code><a href="ia.html">ia()</a></code> <code><a href="ia.html">pair.ia()</a></code> <code><a href="ia.html">resample.ia()</a></code> <code><a href="ia.html">jack.ia()</a></code> 
        </dt>
        <dd>Index of Association</dd>
      </dl><dl><dt>
          
          <code><a href="imsn.html">imsn()</a></code> 
        </dt>
        <dd>Create minimum spanning networks interactively</dd>
      </dl><dl><dt>
          
          <code><a href="incomp.html">incomp()</a></code> 
        </dt>
        <dd>Check for samples that are incomparable due to missing data</dd>
      </dl><dl><dt>
          
          <code><a href="info_table.html">info_table()</a></code> 
        </dt>
        <dd>Create a table summarizing missing data or ploidy information of a genind or
genclone object</dd>
      </dl><dl><dt>
          
          <code><a href="informloci.html">informloci()</a></code> 
        </dt>
        <dd>Remove all non-phylogentically informative loci</dd>
      </dl><dl><dt>
          
          <code><a href="is.clone.html">is.snpclone()</a></code> <code><a href="is.clone.html">is.clone()</a></code> <code><a href="is.clone.html">is.genclone()</a></code> 
        </dt>
        <dd>Check for validity of a genclone or snpclone object</dd>
      </dl><dl><dt>
          
          <code><a href="locus_table.html">locus_table()</a></code> 
        </dt>
        <dd>Create a table of summary statistics per locus.</dd>
      </dl><dl><dt>
          
          <code><a href="make_haplotypes-method.html">make_haplotypes()</a></code> 
        </dt>
        <dd>Split samples from a genind object into pseudo-haplotypes</dd>
      </dl><dl><dt>
          
          <code><a href="missingno.html">missingno()</a></code> 
        </dt>
        <dd>Treat missing data</dd>
      </dl><dl><dt>
          
          <code><a href="mlg.html">mlg()</a></code> <code><a href="mlg.html">mlg.table()</a></code> <code><a href="mlg.html">mlg.vector()</a></code> <code><a href="mlg.html">mlg.crosspop()</a></code> <code><a href="mlg.html">mlg.id()</a></code> 
        </dt>
        <dd>Create counts, vectors, and matrices of multilocus genotypes.</dd>
      </dl><dl><dt>
          
          <code><a href="mlg.filter.html">mlg.filter()</a></code> <code><a href="mlg.filter.html">`mlg.filter&lt;-`()</a></code> 
        </dt>
        <dd>MLG definitions based on genetic distance</dd>
      </dl><dl><dt>
          
          <code><a href="mll-method.html">mll()</a></code> <code><a href="mll-method.html">nmll()</a></code> <code><a href="mll-method.html">`mll&lt;-`()</a></code> 
        </dt>
        <dd>Access and manipulate multilocus lineages.</dd>
      </dl><dl><dt>
          
          <code><a href="mll.custom.html">mll.custom()</a></code> <code><a href="mll.custom.html">`mll.custom&lt;-`()</a></code> <code><a href="mll.custom.html">mll.levels()</a></code> <code><a href="mll.custom.html">`mll.levels&lt;-`()</a></code> 
        </dt>
        <dd>Define custom multilocus lineages</dd>
      </dl><dl><dt>
          
          <code><a href="mll.reset-method.html">mll.reset()</a></code> 
        </dt>
        <dd>Reset multilocus lineages</dd>
      </dl><dl><dt>
          
          <code><a href="monpop.html">monpop</a></code> 
        </dt>
        <dd>Peach brown rot pathogen *Monilinia fructicola*</dd>
      </dl><dl><dt>
          
          <code><a href="old2new_genclone.html">old2new_genclone()</a></code> 
        </dt>
        <dd>Convert an old genclone object to a new genclone object</dd>
      </dl><dl><dt>
          
          <code><a href="partial_clone.html">partial_clone</a></code> 
        </dt>
        <dd>Simulated data illustrating a Minimum Spanning Network based on Bruvo's
Distance</dd>
      </dl><dl><dt>
          
          <code><a href="pgen.html">pgen()</a></code> 
        </dt>
        <dd>Genotype Probability</dd>
      </dl><dl><dt>
          
          <code><a href="plot_poppr_msn.html">plot_poppr_msn()</a></code> 
        </dt>
        <dd>Plot minimum spanning networks produced in poppr.</dd>
      </dl><dl><dt>
          
          <code><a href="poppr-package.html">poppr-package</a></code> 
        </dt>
        <dd>The <span class="pkg">poppr</span> R package</dd>
      </dl><dl><dt>
          
          <code><a href="poppr.html">poppr()</a></code> 
        </dt>
        <dd>Produce a basic summary table for population genetic analyses.</dd>
      </dl><dl><dt>
          
          <code><a href="poppr.all.html">poppr.all()</a></code> 
        </dt>
        <dd>Process a list of files with poppr</dd>
      </dl><dl><dt>
          
          <code><a href="poppr.amova.html">poppr.amova()</a></code> 
        </dt>
        <dd>Perform Analysis of Molecular Variance (AMOVA) on genind or genclone objects.</dd>
      </dl><dl><dt>
          
          <code><a href="poppr.msn.html">poppr.msn()</a></code> 
        </dt>
        <dd>Create a minimum spanning network of selected populations using a distance 
matrix.</dd>
      </dl><dl><dt>
          
          <code><a href="poppr_has_parallel.html">poppr_has_parallel()</a></code> 
        </dt>
        <dd>Determines whether openMP is support on this system.</dd>
      </dl><dl><dt>
          
          <code><a href="popsub.html">popsub()</a></code> 
        </dt>
        <dd>Subset data by population</dd>
      </dl><dl><dt>
          
          <code><a href="private_alleles.html">private_alleles()</a></code> 
        </dt>
        <dd>Tabulate alleles the occur in only one population.</dd>
      </dl><dl><dt>
          
          <code><a href="psex.html">psex()</a></code> 
        </dt>
        <dd>Probability of encountering a genotype more than once by chance</dd>
      </dl><dl><dt>
          
          <code><a href="rare_allele_correction.html">rare_allele_correction</a></code> 
        </dt>
        <dd>Correcting rare allele frequencies</dd>
      </dl><dl><dt>
          
          <code><a href="read.genalex.html">read.genalex()</a></code> 
        </dt>
        <dd>Importing data from genalex formatted \*.csv files.</dd>
      </dl><dl><dt>
          
          <code><a href="recode_polyploids.html">recode_polyploids()</a></code> 
        </dt>
        <dd>Recode polyploid microsatellite data for use in frequency based statistics.</dd>
      </dl><dl><dt>
          
          <code><a href="rraf.html">rraf()</a></code> 
        </dt>
        <dd>Round Robin Allele Frequencies</dd>
      </dl><dl><dt>
          
          <code><a href="rrmlg.html">rrmlg()</a></code> 
        </dt>
        <dd>Round Robin Multilocus Genotypes</dd>
      </dl><dl><dt>
          
          <code><a href="samp.ia.html">samp.ia()</a></code> 
        </dt>
        <dd>Calculate random samples of the index of association for genlight objects.</dd>
      </dl><dl><dt>
          
          <code><a href="shufflepop.html">shufflepop()</a></code> 
        </dt>
        <dd>Shuffle individuals in a <code>genclone</code> or
<code>genind</code> object independently over each locus.</dd>
      </dl><dl><dt>
          
          <code><a href="snpclone-coercion-methods.html">as.snpclone()</a></code> 
        </dt>
        <dd>Create a snpclone object from a genlight object.</dd>
      </dl><dl><dt>
          
          <code><a href="test_replen.html">test_replen()</a></code> 
        </dt>
        <dd>Test repeat length consistency.</dd>
      </dl><dl><dt>
          
          <code><a href="upgma.html">upgma()</a></code> 
        </dt>
        <dd>UPGMA</dd>
      </dl><dl><dt>
          
          <code><a href="win.ia.html">win.ia()</a></code> 
        </dt>
        <dd>Calculate windows of the index of association for genlight objects.</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

