<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function will facilitate in removing phylogenetically uninformative loci
from a genclone or genind object. 
The user has the ability to define what uninformative means by setting a 
cutoff value for either percentage of differentiating genotypes or minor 
allele frequency."><title>Remove all non-phylogentically informative loci — informloci • poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Remove all non-phylogentically informative loci — informloci"><meta property="og:description" content="This function will facilitate in removing phylogenetically uninformative loci
from a genclone or genind object. 
The user has the ability to define what uninformative means by setting a 
cutoff value for either percentage of differentiating genotypes or minor 
allele frequency."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Remove all non-phylogentically informative loci</h1>
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/R/data_subset.r" class="external-link"><code>R/data_subset.r</code></a></small>
      <div class="d-none name"><code>informloci.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function will facilitate in removing phylogenetically uninformative loci
from a <code><a href="genclone-class.html">genclone</a></code> or <code><a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a></code> object. 
The user has the ability to define what uninformative means by setting a 
cutoff value for either percentage of differentiating genotypes or minor 
allele frequency.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">informloci</span><span class="op">(</span><span class="va">pop</span>, cutoff <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">pop</span><span class="op">)</span>, MAF <span class="op">=</span> <span class="fl">0.01</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>pop</dt>
<dd><p>a <code><a href="genclone-class.html">genclone</a></code> or <code><a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a></code> 
object.</p></dd>


<dt>cutoff</dt>
<dd><p><code>numeric</code>. A number from 0 to 1 defining the minimum 
number of differentiating samples.</p></dd>


<dt>MAF</dt>
<dd><p><code>numeric</code>. A number from 0 to 1 defining the minimum minor
allele frequency. This is passed as the <code>thresh</code> parameter of
<code><a href="https://rdrr.io/pkg/adegenet/man/isPoly.html" class="external-link">isPoly</a></code>.</p></dd>


<dt>quiet</dt>
<dd><p><code>logical</code>. When <code>quiet = TRUE</code> (default), messages 
indicating the loci removed will be printed to screen. When <code>quiet = 
FALSE</code>, nothing will be printed to screen.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>genind</code> object with user-defined informative loci.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function will remove uninformative loci using a traditional MAF
  cutoff (using <code><a href="https://rdrr.io/pkg/adegenet/man/isPoly.html" class="external-link">isPoly</a></code> from <span class="pkg">adegenet</span>) as well
  as analyzing the number of observed genotypes in a locus. This is important
  for clonal organisms that can have fixed heterozygous sites not detected by
  MAF methods.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>This will have a few side effects that affect certain analyses. First, 
  the number of multilocus genotypes might be reduced due to the reduced 
  number of markers (if you are only using a genind object). Second, if you 
  plan on using this data for analysis of the index of association, be sure 
  to use the standardized version (rbarD) that corrects for the number of 
  observed loci.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Zhian N. Kamvar</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># We will use a dummy data set to demonstrate how this detects uninformative</span></span></span>
<span class="r-in"><span><span class="co"># loci using both MAF and a cutoff.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">genos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A/A"</span>, <span class="st">"A/B"</span>, <span class="st">"A/C"</span>, <span class="st">"B/B"</span>, <span class="st">"B/C"</span>, <span class="st">"C/C"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">genos</span>, <span class="fl">100</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">genos</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fl">99</span><span class="op">)</span>, <span class="va">genos</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>           <span class="co"># found by cutoff</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">genos</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">98</span><span class="op">)</span>, <span class="va">genos</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">genos</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># found by MAF</span></span></span>
<span class="r-in"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">genos</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">99</span><span class="op">)</span>, <span class="va">genos</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>           <span class="co"># found by both</span></span></span>
<span class="r-in"><span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">genos</span>, <span class="fl">100</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html" class="external-link">df2genind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">v</span>, w <span class="op">=</span> <span class="va">w</span>, x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, z <span class="op">=</span> <span class="va">z</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">informloci</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cutoff value: 2 % ( 2 samples ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MAF         : 0.01</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Found 3 uninformative loci </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  ============================ </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  2 loci found with a cutoff of 2 samples :</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  w, y </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  2 loci found with MAF &lt; 0.01 :</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  x, y</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 100 individuals; 2 loci; 6 alleles; size: 14.6 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  100 x 6 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 3-3)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 6 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    - empty -</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Ignore MAF</span></span></span>
<span class="r-in"><span><span class="fu">informloci</span><span class="op">(</span><span class="va">dat</span>, MAF <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cutoff value: 2 % ( 2 samples ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MAF         : 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Found 2 uninformative loci </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  ============================ </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  2 loci found with a cutoff of 2 samples :</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  w, y </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  0 loci found with MAF &lt; 0  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 100 individuals; 3 loci; 9 alleles; size: 16.5 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  100 x 9 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 3-3)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 9 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    - empty -</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Ignore cutoff</span></span></span>
<span class="r-in"><span><span class="fu">informloci</span><span class="op">(</span><span class="va">dat</span>, cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cutoff value: 0 % ( 0 samples ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MAF         : 0.01</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Found 2 uninformative loci </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  ============================ </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  0 loci found with a cutoff of 0 samples   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  2 loci found with MAF &lt; 0.01 :</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  x, y</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 100 individuals; 3 loci; 9 alleles; size: 16.5 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  100 x 9 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 3-3)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 9 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    - empty -</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Real data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">informloci</span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> cutoff value: 0.105097214923805 % ( 2 samples ).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MAF         : 0.01</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Found 5 uninformative loci </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  ============================ </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  1 locus found with a cutoff of 2 samples :</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  597 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  5 loci found with MAF &lt; 0.01 :</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  42, 313, 433, 597, 915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 1,903 individuals; 120 loci; 322 alleles; size: 3.3 Mb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  1903 x 322 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 2-4)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 322 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 1-1)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: x  xy  epid </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

