<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Split samples from a genind object into pseudo-haplotypes"><title>Split samples from a genind object into pseudo-haplotypes — make_haplotypes • poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Split samples from a genind object into pseudo-haplotypes — make_haplotypes"><meta property="og:description" content="Split samples from a genind object into pseudo-haplotypes"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Split samples from a genind object into pseudo-haplotypes</h1>
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/R/methods.r" class="external-link"><code>R/methods.r</code></a></small>
      <div class="d-none name"><code>make_haplotypes-method.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Split samples from a genind object into pseudo-haplotypes</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">make_haplotypes</span><span class="op">(</span><span class="va">gid</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gid</dt>
<dd><p>a genind or <a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a> object.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a haploid genind object with an extra strata</p>


<p>column called "Individual".</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Certain analyses, such as <a href="poppr.amova.html">amova</a> work best if within-sample
variance (error) can be estimated. Practically, this is performed by
splitting the genotypes across all loci to create multiple haplotypes. This
way, the within-sample distance can be calculated and incorporated into the
model. Please note that the haplotypes generated are based on the order of
the unphased alleles in the genind object and do not represent true
haplotypes.</p>
<p>Haploid data will be returned un-touched.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The other slot will not be copied over to the new
genind object.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="poppr.amova.html">poppr.amova()</a></code> <code><a href="https://rdrr.io/pkg/pegas/man/amova.html" class="external-link">pegas::amova()</a></code> <code><a href="coercion-methods.html">as.genambig()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Diploid data is doubled -------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">nan9</span> <span class="op">&lt;-</span> <span class="va">nancycats</span><span class="op">[</span>pop <span class="op">=</span> <span class="fl">9</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">nan9hap</span> <span class="op">&lt;-</span> <span class="fu">make_haplotypes</span><span class="op">(</span><span class="va">nan9</span><span class="op">)</span> </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No strata found... creating one from the population.</span>
<span class="r-in"><span><span class="va">nan9</span>              <span class="co"># 9 individuals from population 9</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 9 individuals; 9 loci; 108 alleles; size: 28.6 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  9 x 108 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 8-18)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 108 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: .local(x = x, i = i, j = j, pop = 9, drop = drop)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 9-9)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: xy </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">nan9hap</span>           <span class="co"># 18 haplotypes</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 18 individuals; 9 loci; 43 alleles; size: 21.3 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  18 x 43 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 3-8)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 43 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 1-1)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: df2genind(X = newdf, ploidy = 1, strata = df)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @strata: a data frame with 2 columns ( pop, Individual )</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">nan9hap</span><span class="op">)</span>   <span class="co"># strata gains a new column: Individual</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    pop Individual</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 01 P09       N104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 02 P09       N104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03 P09       N105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 04 P09       N105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 05 P09       N106</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 06 P09       N106</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 07 P09       N107</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 08 P09       N107</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 09 P09       N108</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 P09       N108</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 P09       N109</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 P09       N109</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 P09       N111</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 P09       N111</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 P09       N112</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 P09       N112</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 P09       N113</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18 P09       N113</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">indNames</a></span><span class="op">(</span><span class="va">nan9hap</span><span class="op">)</span> <span class="co"># individuals are renamed sequentially</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "01" "02" "03" "04" "05" "06" "07" "08" "09" "10" "11" "12" "13" "14" "15"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16] "16" "17" "18"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Mix ploidy data can be split, but should be treated with caution --------</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># For example, the Pinf data set contains 86 tetraploid individuals, </span></span></span>
<span class="r-in"><span><span class="co"># but there appear to only be diploids and triploid genotypes. When </span></span></span>
<span class="r-in"><span><span class="co"># we convert to haplotypes, those with all missing data are dropped.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Pinf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    72 multilocus genotypes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    86 tetraploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    11 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 strata - Continent, Country</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 populations defined - South America, North America</span>
<span class="r-in"><span><span class="va">pmiss</span> <span class="op">&lt;-</span> <span class="fu"><a href="info_table.html">info_table</a></span><span class="op">(</span><span class="va">Pinf</span>, type <span class="op">=</span> <span class="st">"ploidy"</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="make_haplotypes-method-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># No samples appear to be triploid across all loci. This will cause</span></span></span>
<span class="r-in"><span><span class="co"># several haplotypes to have a lot of missing data.</span></span></span>
<span class="r-in"><span><span class="va">p_haps</span> <span class="op">&lt;-</span> <span class="fu">make_haplotypes</span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">p_haps</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> /// GENIND OBJECT /////////</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // 203 individuals; 11 loci; 95 alleles; size: 137.9 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @tab:  203 x 95 matrix of allele counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.n.all: number of alleles per locus (range: 2-25)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @loc.fac: locus factor for the 95 columns of @tab</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @all.names: list of allele names for each locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 1-1)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @type:  codom</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @call: df2genind(X = newdf, ploidy = 1, strata = df)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  // Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 2-3)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @strata: a data frame with 3 columns ( Continent, Country, Individual )</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/genind2df.html" class="external-link">genind2df</a></span><span class="op">(</span><span class="va">p_haps</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       pop Pi02  D13 Pi33 Pi04 Pi4B Pi16  G11 Pi56 Pi63 Pi70 Pi89</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  PiCO01 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  PiCO01  162  136  203  166  213  178  156  174  157  192  179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  PiCO01  162  136  203  170  217  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  PiCO02 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  PiCO02  162  132  203  166  213  178  156  174  157  192  179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  PiCO02  162  136  203  170  217  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 PiCO03 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 PiCO03  162  136  203  166  213  178  156  174  157  192  179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 PiCO03  162  136  203  170  217  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 PiCO04 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 PiCO04  162  136  203  166  213  178  156  174  157  192  179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 PiCO04  162  136  203  170  217  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19 PiCO05  160  108  203  166  213  174  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20 PiCO05  162  112  203  170  225  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23 PiEC01  162  134  203  166  213  178  156  174  157  192  179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24 PiEC01  162  136  203  170  217  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26 PiEC02 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27 PiEC02  162  136  203  166  213  178  156  174  157  192  179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28 PiEC02  162  144  203  170  217  178  156  176  157  192  181</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31 PiEC03  162  136  203  166  205  178  156  174  157  192  179</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

