<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Multilocus genotypes are initially defined by naive string matching, but this
definition does not take into account missing data or genotyping error,
casting these as unique genotypes. Defining multilocus genotypes by genetic
distance allows you to incorporate genotypes that have missing data o
genotyping error into their parent clusters."><title>MLG definitions based on genetic distance — mlg.filter • poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MLG definitions based on genetic distance — mlg.filter"><meta property="og:description" content="Multilocus genotypes are initially defined by naive string matching, but this
definition does not take into account missing data or genotyping error,
casting these as unique genotypes. Defining multilocus genotypes by genetic
distance allows you to incorporate genotypes that have missing data o
genotyping error into their parent clusters."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>MLG definitions based on genetic distance</h1>
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/R/methods.r" class="external-link"><code>R/methods.r</code></a></small>
      <div class="d-none name"><code>mlg.filter.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Multilocus genotypes are initially defined by naive string matching, but this
definition does not take into account missing data or genotyping error,
casting these as unique genotypes. Defining multilocus genotypes by genetic
distance allows you to incorporate genotypes that have missing data o
genotyping error into their parent clusters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mlg.filter</span><span class="op">(</span></span>
<span>  <span class="va">pop</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  missing <span class="op">=</span> <span class="st">"asis"</span>,</span>
<span>  memory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  algorithm <span class="op">=</span> <span class="st">"farthest_neighbor"</span>,</span>
<span>  distance <span class="op">=</span> <span class="st">"diss.dist"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  stats <span class="op">=</span> <span class="st">"MLGs"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mlg.filter</span><span class="op">(</span></span>
<span>  <span class="va">pop</span>,</span>
<span>  missing <span class="op">=</span> <span class="st">"asis"</span>,</span>
<span>  memory <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  algorithm <span class="op">=</span> <span class="st">"farthest_neighbor"</span>,</span>
<span>  distance <span class="op">=</span> <span class="st">"diss.dist"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">value</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>pop</dt>
<dd><p>a <code><a href="genclone-class.html">genclone</a></code>, <code><a href="genclone-class.html">snpclone</a></code>, or
<code><a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a></code> object.</p></dd>


<dt>threshold</dt>
<dd><p>a number indicating the minimum distance two MLGs must be
separated by to be considered different. Defaults to 0, which will reflect
the original (naive) MLG definition.</p></dd>


<dt>missing</dt>
<dd><p>any method to be used by <code><a href="missingno.html">missingno</a></code>: "mean", 
"zero", "loci", "genotype", or "asis" (default).</p></dd>


<dt>memory</dt>
<dd><p>whether this function should remember the last distance matrix 
it generated. TRUE will attempt to reuse the last distance matrix if the 
other parameters are the same. (default) FALSE will ignore any stored 
matrices and not store any it generates.</p></dd>


<dt>algorithm</dt>
<dd><p>determines the type of clustering to be done.</p><dl><dt>"farthest_neighbor"</dt>
<dd><p><em> (default) </em>merges clusters based on the 
  maximum distance between points in either cluster. This is the strictest of
  the three.</p></dd>

  <dt>"nearest_neighbor"</dt>
<dd><p>merges clusters based on the minimum distance
  between points in either cluster. This is the loosest of the three.</p></dd>

  <dt>"average_neighbor"</dt>
<dd><p>merges clusters based on the average distance
  between every pair of points between clusters.</p></dd>


</dl></dd>


<dt>distance</dt>
<dd><p>a character or function defining the distance to be applied 
to pop. Defaults to <code><a href="diss.dist.html">diss.dist</a></code> for genclone objects and
<code><a href="bitwise.dist.html">bitwise.dist</a></code> for snpclone objects. A matrix or table
containing distances between individuals (such as the output of 
<code><a href="genetic_distance.html">rogers.dist</a></code>) is also accepted for this parameter.</p></dd>


<dt>threads</dt>
<dd><p>(unused) Previously, this was the maximum number of parallel 
threads to be used within this function. Default is 1 indicating that this
function will run serially. Any other number will result in a warning.</p></dd>


<dt>stats</dt>
<dd><p>a character vector specifying which statistics should be
returned (details below). Choices are "MLG", "THRESHOLDS", "DISTANCES",
"SIZES", or "ALL". If choosing "ALL" or more than one, a named list will be
returned.</p></dd>


<dt>...</dt>
<dd><p>any parameters to be passed off to the distance method.</p></dd>


<dt>value</dt>
<dd><p>the threshold at which genotypes should be collapsed.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Default, a vector of collapsed multilocus genotypes. Otherwise, any
  combination of the following:</p><div class="section">
<h3 id="mlgs">MLGs<a class="anchor" aria-label="anchor" href="#mlgs"></a></h3>
<p>a numeric vector defining the multilocus genotype cluster of each
  individual in the dataset. Each genotype cluster is separated from every
  other genotype cluster by at least the defined threshold value, as 
  calculated by the selected algorithm.</p>
</div>


<div class="section">
<h3 id="thresholds">THRESHOLDS<a class="anchor" aria-label="anchor" href="#thresholds"></a></h3>
<p>A numeric vector representing the thresholds <strong>beyond</strong> which clusters
  of multilocus genotypes were collapsed.</p>
</div>


<div class="section">
<h3 id="distances">DISTANCES<a class="anchor" aria-label="anchor" href="#distances"></a></h3>
<p>A square matrix representing the distances between each cluster.</p>
</div>


<div class="section">
<h3 id="sizes">SIZES<a class="anchor" aria-label="anchor" href="#sizes"></a></h3>
<p>The sizes of the multilocus genotype clusters in order.</p>
</div>


    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function will take in any distance matrix or function and
collapse multilocus genotypes below a given threshold. If you use this
function as the assignment method (mlg.filter(myData, distance = myDist) &lt;-
0.5), the distance function or matrix will be remembered by the object. This
means that if you define your own distance matrix or function, you must keep
it in memory to further utilize mlg.filter.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p><code>mlg.vector</code> makes use of <code>mlg.vector</code> grouping prior to 
  applying the given threshold. Genotype numbers returned by
  <code>mlg.vector</code> represent the lowest numbered genotype (as returned by
  <code>mlg.vector</code>) in in each new multilocus genotype. Therefore
  <strong><code>mlg.filter</code> and <code>mlg.vector</code> return the same vector when
  threshold is set to 0 or less</strong>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="filter_stats.html">filter_stats</a></code>, 
  <code><a href="cutoff_predictor.html">cutoff_predictor</a></code>, 
  <code><a href="mll-method.html">mll</a></code>, 
  <code><a href="genclone-class.html">genclone</a></code>, 
  <code><a href="genclone-class.html">snpclone</a></code>, 
  <code><a href="diss.dist.html">diss.dist</a></code>, 
  <code><a href="bruvo.dist.html">bruvo.dist</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">partial_clone</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">partial_clone</span>, threads <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span> <span class="co"># convert to genclone object</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Basic Use ---------------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Show MLGs at threshold 0.05</span></span></span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">pc</span>, threshold <span class="op">=</span> <span class="fl">0.05</span>, distance <span class="op">=</span> <span class="st">"nei.dist"</span>, threads <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1]  8  7 23 24 22 21 10  3 22 11 24  7 25  4 12  2 14  1  7  7  7 26  7 13 23</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [26]  3 17 22  6 20 22 12  5 25 13 21 15 13 13 13  2 19 18 13 23 16  1 11 25  4</span>
<span class="r-in"><span><span class="va">pc</span> <span class="co"># 26 mlgs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    26 original multilocus genotypes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    50 diploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    10 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0 strata. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4 populations defined - 1, 2, 3, 4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Set MLGs at threshold 0.05</span></span></span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">pc</span>, distance <span class="op">=</span> <span class="st">"nei.dist"</span>, threads <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span></span>
<span class="r-in"><span><span class="va">pc</span> <span class="co"># 25 mlgs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    25 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (0.05) [t], (nei.dist) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    50 diploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    10 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0 strata. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4 populations defined - 1, 2, 3, 4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The distance definition is persistant</span></span></span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span></span>
<span class="r-in"><span><span class="va">pc</span> <span class="co"># 24 mlgs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    24 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (0.1) [t], (nei.dist) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    50 diploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    10 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0 strata. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4 populations defined - 1, 2, 3, 4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># But you can still change the definition</span></span></span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">pc</span>, distance <span class="op">=</span> <span class="st">"diss.dist"</span>, percent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span></span>
<span class="r-in"><span><span class="va">pc</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    25 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (0.1) [t], (diss.dist) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    50 diploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    10 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     0 strata. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4 populations defined - 1, 2, 3, 4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Choosing a threshold ----------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Thresholds for collapsing multilocus genotypes should not be arbitrary. It</span></span></span>
<span class="r-in"><span><span class="co"># is important to consider what threshold is suitable. One method of choosing</span></span></span>
<span class="r-in"><span><span class="co"># a threshold is to find a gap in the distance distribution that represents</span></span></span>
<span class="r-in"><span><span class="co"># clonal groups. You can look at this by analyzing the distribution of all</span></span></span>
<span class="r-in"><span><span class="co"># possible thresholds with the function "cutoff_predictor".</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For this example, we'll use Bruvo's distance to predict the cutoff for</span></span></span>
<span class="r-in"><span><span class="co"># P. infestans.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Pinf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    72 multilocus genotypes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    86 tetraploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    11 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 strata - Continent, Country</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 populations defined - South America, North America</span>
<span class="r-in"><span><span class="co"># Repeat lengths are necessary for Bruvo's distance</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">pinfreps</span> <span class="op">&lt;-</span> <span class="fu"><a href="fix_replen.html">fix_replen</a></span><span class="op">(</span><span class="va">Pinf</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">6</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Pi02     D13    Pi33    Pi04    Pi4B    Pi16     G11    Pi56    Pi63    Pi70 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2.00000 2.00000 6.00000 2.00000 1.99999 2.00000 2.00000 2.00000 3.00000 3.00000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Pi89 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1.99999 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Now we can collect information of the thresholds. We can set threshold = 1</span></span></span>
<span class="r-in"><span><span class="co"># because we know that this will capture the maximum possible distance:</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">thresholds</span> <span class="op">&lt;-</span> <span class="fu">mlg.filter</span><span class="op">(</span><span class="va">Pinf</span>, distance <span class="op">=</span> <span class="va">bruvo.dist</span>, stats <span class="op">=</span> <span class="st">"THRESHOLDS"</span>,</span></span>
<span class="r-in"><span>                          replen <span class="op">=</span> <span class="va">pinfreps</span>, threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 0.01262626 0.02189867 0.02272727 0.02272727 0.03535354 0.04166667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] 0.04261364 0.04545454 0.04699337 0.04758520 0.05681818 0.05681818</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] 0.05835701 0.06451231 0.06534091 0.06818182 0.07871686 0.07954545</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] 0.08772748 0.08877841 0.09375000 0.09469697 0.09722222 0.10144413</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] 0.12500000 0.13099747 0.13593750 0.13740234 0.14488636 0.15000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31] 0.15656566 0.16688366 0.18115234 0.19317072 0.21022726 0.21590909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37] 0.21874983 0.22561553 0.23115234 0.23295450 0.23437499 0.23532197</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43] 0.24147723 0.25850032 0.27201702 0.27500000 0.28338066 0.29208984</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [49] 0.30000000 0.30042336 0.30326990 0.30549961 0.30632250 0.31325684</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [55] 0.33639034 0.33709162 0.34089797 0.35170815 0.35619658 0.36079233</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [61] 0.36487924 0.39646464 0.40280346 0.40332030 0.42325106 0.43731965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [67] 0.45436164 0.47047212 0.50827603 0.51496688 0.57474140</span>
<span class="r-in"><span><span class="co"># We can use these thresholds to find an appropriate cutoff</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">pcut</span> <span class="op">&lt;-</span> <span class="fu"><a href="cutoff_predictor.html">cutoff_predictor</a></span><span class="op">(</span><span class="va">thresholds</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1132221</span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">Pinf</span>, distance <span class="op">=</span> <span class="va">bruvo.dist</span>, replen <span class="op">=</span> <span class="va">pinfreps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pcut</span></span></span>
<span class="r-in"><span><span class="va">Pinf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    48 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (0.113) [t], (bruvo.dist) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    86 tetraploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    11 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 strata - Continent, Country</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 populations defined - South America, North America</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># This can also be visualized with the "filter_stats" function.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Special case: threshold = 0 ---------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># It's important to remember that a threshold of 0 is equal to the original</span></span></span>
<span class="r-in"><span><span class="co"># MLG definition. This example will show a data set that contains genotypes</span></span></span>
<span class="r-in"><span><span class="co"># with missing data that share all alleles with other genotypes except for </span></span></span>
<span class="r-in"><span><span class="co"># the missing one.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">monpop</span> <span class="co"># 264 mlg</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    264 multilocus genotypes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    694 haploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     13 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1 stratum - Pop</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     12 populations defined - </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></span>
<span class="r-in"><span><span class="fu"><a href="mll-method.html">nmll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="co"># 264 mlg</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 264</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># In order to merge these genotypes with missing data, we should set the </span></span></span>
<span class="r-in"><span><span class="co"># threshold to be slightly higher than 0. We will use the smallest fraction </span></span></span>
<span class="r-in"><span><span class="co"># the computer can store.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span> <span class="op">^</span> <span class="fl">0.5</span></span></span>
<span class="r-in"><span><span class="fu"><a href="mll-method.html">nmll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="co"># 236 mlg</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 236</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Custom distance ---------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Custom genetic distances can be used either in functions from other</span></span></span>
<span class="r-in"><span><span class="co"># packages or user-defined functions</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Pinf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    48 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (0.113) [t], (bruvo.dist) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    86 tetraploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    11 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 strata - Continent, Country</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 populations defined - South America, North America</span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">Pinf</span>, distance <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">3</span></span></span>
<span class="r-in"><span><span class="va">Pinf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    48 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (3) [t], (function(x) dist(tab(x))) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    86 tetraploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    11 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 strata - Continent, Country</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 populations defined - South America, North America</span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="va">Pinf</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    51 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       (4) [t], (function(x) dist(tab(x))) [d], (farthest) [a] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    86 tetraploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    11 codominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 strata - Continent, Country</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2 populations defined - South America, North America</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genlight / snpclone objects ---------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">999</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gc</span> <span class="op">&lt;-</span> <span class="fu"><a href="snpclone-coercion-methods.html">as.snpclone</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/glSim.html" class="external-link">glSim</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">0</span>, n.snp.struc <span class="op">=</span> <span class="fl">1e3</span>, ploidy <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gc</span> <span class="co"># 100 mlgs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ||| SNPCLONE OBJECT |||||||||</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || 100 genotypes,  1,000 binary SNPs, size: 183.5 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0 (0 %) missing data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @gen: list of 100 SNPbin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @mlg: 100 original multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 50-50)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: ancestral.pops </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span><span class="fu">mlg.filter</span><span class="op">(</span><span class="va">gc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.25</span></span></span>
<span class="r-in"><span><span class="va">gc</span> <span class="co"># 82 mlgs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ||| SNPCLONE OBJECT |||||||||</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || 100 genotypes,  1,000 binary SNPs, size: 183.5 Kb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0 (0 %) missing data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Basic content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @gen: list of 100 SNPbin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @mlg: 82 contracted multilocus genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          (0.25) [t], (bitwise.dist) [d], (farthest) [a]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @ploidy: ploidy of each individual  (range: 2-2)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  || Optional content</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @pop: population of each individual (group size range: 50-50)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    @other: a list containing: ancestral.pops </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

