<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function simplifies the process necessary for performing AMOVA in R. It
gives user the choice of utilizing either the ade4 or the pegas
implementation of AMOVA. See ade4::amova() (ade4) and pegas::amova()
(pegas) for details on the specific implementation."><title>Perform Analysis of Molecular Variance (AMOVA) on genind or genclone objects. — poppr.amova • poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Analysis of Molecular Variance (AMOVA) on genind or genclone objects. — poppr.amova"><meta property="og:description" content="This function simplifies the process necessary for performing AMOVA in R. It
gives user the choice of utilizing either the ade4 or the pegas
implementation of AMOVA. See ade4::amova() (ade4) and pegas::amova()
(pegas) for details on the specific implementation."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Perform Analysis of Molecular Variance (AMOVA) on genind or genclone objects.</h1>
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/R/amova.r" class="external-link"><code>R/amova.r</code></a></small>
      <div class="d-none name"><code>poppr.amova.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function simplifies the process necessary for performing AMOVA in R. It
gives user the choice of utilizing either the <span class="pkg">ade4</span> or the <span class="pkg">pegas</span>
implementation of AMOVA. See <code><a href="https://rdrr.io/pkg/ade4/man/amova.html" class="external-link">ade4::amova()</a></code> (ade4) and <code><a href="https://rdrr.io/pkg/pegas/man/amova.html" class="external-link">pegas::amova()</a></code>
(pegas) for details on the specific implementation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">poppr.amova</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  hier <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clonecorrect <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  within <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  squared <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  freq <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correction <span class="op">=</span> <span class="st">"quasieuclid"</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">"_"</span>,</span>
<span>  filter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  algorithm <span class="op">=</span> <span class="st">"farthest_neighbor"</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  missing <span class="op">=</span> <span class="st">"loci"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ade4"</span>, <span class="st">"pegas"</span><span class="op">)</span>,</span>
<span>  nperm <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>a <a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a>, <a href="genclone-class.html">genclone</a>, <a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a>, or <a href="genclone-class.html">snpclone</a> object</p></dd>


<dt>hier</dt>
<dd><p>a hierarchical <a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a> that defines your population
hierarchy. (e.g.: <code>~Population/Subpopulation</code>). <strong>See Details below</strong>.</p></dd>


<dt>clonecorrect</dt>
<dd><p><code>logical</code> if <code>TRUE</code>, the data set will be clone corrected
with respect to the lowest level of the hierarchy. The default is set to
<code>FALSE</code>. See <code><a href="clonecorrect.html">clonecorrect()</a></code> for details.</p></dd>


<dt>within</dt>
<dd><p><code>logical</code>. When this is set to <code>TRUE</code> (Default), variance
within individuals are calculated as well. If this is set to <code>FALSE</code>, The
lowest level of the hierarchy will be the sample level. See Details below.</p></dd>


<dt>dist</dt>
<dd><p>an optional distance matrix calculated on your data. If this is
set to <code>NULL</code> (default), the raw pairwise distances will be calculated via
<code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist()</a></code>.</p></dd>


<dt>squared</dt>
<dd><p>if a distance matrix is supplied, this indicates whether or
not it represents squared distances.</p></dd>


<dt>freq</dt>
<dd><p><code>logical</code>. If <code>within = FALSE</code>, the parameter rho is calculated
(Ronfort et al. 1998; Meirmans and Liu 2018). By setting <code>freq = TRUE</code>,
(default) allele counts will be converted to frequencies before the
distance is calculated, otherwise, the distance will be calculated on
allele counts, which can bias results in mixed-ploidy data sets. Note that
this option has no effect for haploid or presence/absence data sets.</p></dd>


<dt>correction</dt>
<dd><p>a <code>character</code> defining the correction method for
non-euclidean distances. Options are <code><a href="https://rdrr.io/pkg/ade4/man/quasieuclid.html" class="external-link">ade4::quasieuclid()</a></code> (Default),
<code><a href="https://rdrr.io/pkg/ade4/man/lingoes.html" class="external-link">ade4::lingoes()</a></code>, and <code><a href="https://rdrr.io/pkg/ade4/man/cailliez.html" class="external-link">ade4::cailliez()</a></code>. See Details below.</p></dd>


<dt>sep</dt>
<dd><p>Deprecated. As of poppr version 2, this argument serves no
purpose.</p></dd>


<dt>filter</dt>
<dd><p><code>logical</code> When set to <code>TRUE</code>, mlg.filter will be run to
determine genotypes from the distance matrix. It defaults to <code>FALSE</code>. You
can set the parameters with <code>algorithm</code> and <code>threshold</code> arguments. Note
that this will not be performed when <code>within = TRUE</code>. Note that the
threshold should be the number of allowable substitutions if you don't
supply a distance matrix.</p></dd>


<dt>threshold</dt>
<dd><p>a number indicating the minimum distance two MLGs must be
separated by to be considered different. Defaults to 0, which will reflect
the original (naive) MLG definition.</p></dd>


<dt>algorithm</dt>
<dd><p>determines the type of clustering to be done.</p><dl><dt>"farthest_neighbor"</dt>
<dd><p><em> (default) </em>merges clusters based on the 
  maximum distance between points in either cluster. This is the strictest of
  the three.</p></dd>

  <dt>"nearest_neighbor"</dt>
<dd><p>merges clusters based on the minimum distance
  between points in either cluster. This is the loosest of the three.</p></dd>

  <dt>"average_neighbor"</dt>
<dd><p>merges clusters based on the average distance
  between every pair of points between clusters.</p></dd>


</dl></dd>


<dt>threads</dt>
<dd><p><code>integer</code> When using filtering or genlight objects, this
parameter specifies the number of parallel processes passed to
<code><a href="mlg.filter.html">mlg.filter()</a></code> and/or <code><a href="bitwise.dist.html">bitwise.dist()</a></code>.</p></dd>


<dt>missing</dt>
<dd><p>specify method of correcting for missing data utilizing
options given in the function <code><a href="missingno.html">missingno()</a></code>. Default is <code>"loci"</code>. This only
applies to genind or genclone objects.</p></dd>


<dt>cutoff</dt>
<dd><p>specify the level at which missing data should be
removed/modified. See <code><a href="missingno.html">missingno()</a></code> for details. This only applies to
genind or genclone objects.</p></dd>


<dt>quiet</dt>
<dd><p><code>logical</code> If <code>FALSE</code> (Default), messages regarding any
corrections will be printed to the screen. If <code>TRUE</code>, no messages will be
printed.</p></dd>


<dt>method</dt>
<dd><p>Which method for calculating AMOVA should be used? Choices
refer to package implementations: "ade4" (default) or "pegas". See details
for differences.</p></dd>


<dt>nperm</dt>
<dd><p>the number of permutations passed to the pegas implementation of
amova.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a list of class <code>amova</code> from the ade4 or pegas package. See
<code><a href="https://rdrr.io/pkg/ade4/man/amova.html" class="external-link">ade4::amova()</a></code> or <code><a href="https://rdrr.io/pkg/pegas/man/amova.html" class="external-link">pegas::amova()</a></code> for details.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The poppr implementation of AMOVA is a very detailed wrapper for the
ade4 implementation. The output is an <code><a href="https://rdrr.io/pkg/ade4/man/amova.html" class="external-link">ade4::amova()</a></code> class list that
contains the results in the first four elements. The inputs are contained
in the last three elements. The inputs required for the ade4 implementation
are:</p><ol><li><p>a distance matrix on all unique genotypes (haplotypes)</p></li>
<li><p>a data frame defining the hierarchy of the distance matrix</p></li>
<li><p>a genotype (haplotype) frequency table.</p></li>
</ol><p>All of this data can be constructed from a <a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a> or
<a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a> object, but can be daunting for a novice R user.
<em>This function automates the entire process</em>. Since there are many
variables regarding genetic data, some points need to be highlighted:</p>
<div class="section">
<h3 id="on-hierarchies-">On Hierarchies:<a class="anchor" aria-label="anchor" href="#on-hierarchies-"></a></h3>
<p>The hierarchy is defined by different
population strata that separate your data hierarchically. These strata are
defined in the <strong>strata</strong> slot of <a href="https://rdrr.io/pkg/adegenet/man/genind.html" class="external-link">genind</a>,
<a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a>, <a href="genclone-class.html">genclone</a>, and
<a href="genclone-class.html">snpclone</a> objects. They are useful for defining the
population factor for your data. See the function <code><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata()</a></code> for details on
how to properly define these strata.</p>
</div>

<div class="section">
<h3 id="on-within-individual-variance-">On Within Individual Variance:<a class="anchor" aria-label="anchor" href="#on-within-individual-variance-"></a></h3>
<p>Heterozygosities within
genotypes are sources of variation from within individuals and can be
quantified in AMOVA. When <code>within = TRUE</code>, poppr will split genotypes into
haplotypes with the function <code><a href="make_haplotypes-method.html">make_haplotypes()</a></code> and use those to calculate
within-individual variance. No estimation of phase is made. This acts much
like the default settings for AMOVA in the Arlequin software package.
Within individual variance will not be calculated for haploid individuals
or dominant markers as the haplotypes cannot be split further. Setting
<code>within = FALSE</code> uses the euclidean distance of the allele frequencies
within each individual. <strong>Note:</strong> <code>within = TRUE</code> is incompatible with
<code>filter = TRUE</code>. In this case, <code>within</code> will be set to <code>FALSE</code></p>
</div>

<div class="section">
<h3 id="on-euclidean-distances-">On Euclidean Distances:<a class="anchor" aria-label="anchor" href="#on-euclidean-distances-"></a></h3>
<p>With the <span class="pkg">ade4</span> implementation of
AMOVA (utilized by <span class="pkg">poppr</span>), distances must be Euclidean (due to the
nature of the calculations). Unfortunately, many genetic distance measures
are not always euclidean and must be corrected for before being analyzed.
Poppr automates this with three methods implemented in <span class="pkg">ade4</span>,
<code>quasieuclid()</code>, <code>lingoes()</code>, and <code>cailliez()</code>. The correction of these
distances should not adversely affect the outcome of the analysis.</p>
</div>

<div class="section">
<h3 id="on-filtering-">On Filtering:<a class="anchor" aria-label="anchor" href="#on-filtering-"></a></h3>
<p>Filtering multilocus genotypes is performed by
<code><a href="mlg.filter.html">mlg.filter()</a></code>. This can necessarily only be done AMOVA tests that do not
account for within-individual variance. The distance matrix used to
calculate the amova is derived from using <code><a href="mlg.filter.html">mlg.filter()</a></code> with the option
<code>stats = "distance"</code>, which reports the distance between multilocus
genotype clusters. One useful way to utilize this feature is to correct for
genotypes that have equivalent distance due to missing data. (See example
below.)</p>
</div>

<div class="section">
<h3 id="on-methods-">On Methods:<a class="anchor" aria-label="anchor" href="#on-methods-"></a></h3>
<p>Both <span class="pkg">ade4</span> and <span class="pkg">pegas</span> have
implementations of AMOVA, both of which are appropriately called "amova".
The ade4 version is faster, but there have been questions raised as to the
validity of the code utilized. The pegas version is slower, but careful
measures have been implemented as to the accuracy of the method. It must be
noted that there appears to be a bug regarding permuting analyses where
within individual variance is accounted for (<code>within = TRUE</code>) in the pegas
implementation. If you want to perform permutation analyses on the pegas
implementation, you must set <code>within = FALSE</code>. In addition, while clone
correction is implemented for both methods, filtering is only implemented
for the ade4 version.</p>
</div>

<div class="section">
<h3 id="on-polyploids-">On Polyploids:<a class="anchor" aria-label="anchor" href="#on-polyploids-"></a></h3>
<p>As of <span class="pkg">poppr</span> version 2.7.0, this
function is able to calculate phi statistics for within-individual variance
for polyploid data with <strong>full dosage information</strong>. When a data set does
not contain full dosage information for all samples, then the resulting
pseudo-haplotypes will contain missing data, which would result in an
incorrect estimate of variance.</p>
<p>Instead, the AMOVA will be performed on the distance matrix derived from
allele counts or allele frequencies, depending on the <code>freq</code> option. This
has been shown to be robust to estimates with mixed ploidy (Ronfort et al.
1998; Meirmans and Liu 2018). If you wish to brute-force your way to
estimating AMOVA using missing values, you can split your haplotypes with
the <code><a href="make_haplotypes-method.html">make_haplotypes()</a></code> function.</p>
<p>One strategy for addressing ambiguous dosage in your polyploid data set
would be to convert your data to <span class="pkg">polysat</span>'s <code>genambig</code> class with the
<code><a href="coercion-methods.html">as.genambig()</a></code>, estimate allele frequencies with <code><a href="https://rdrr.io/pkg/polysat/man/deSilvaFreq.html" class="external-link">polysat::deSilvaFreq()</a></code>,
and use these frequencies to randomly sample alleles to fill in the
ambiguous alleles.</p>
</div>

    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Excoffier, L., Smouse, P.E. and Quattro, J.M. (1992) Analysis of
molecular variance inferred from metric distances among DNA haplotypes:
application to human mitochondrial DNA restriction data. <em>Genetics</em>,
<strong>131</strong>, 479-491.</p>
<p>Ronfort, J., Jenczewski, E., Bataillon, T., and Rousset, F. (1998). Analysis
of population structure in autotetraploid species. <em>Genetics</em>, <strong>150</strong>,
921–930.</p>
<p>Meirmans, P., Liu, S. (2018) Analysis of Molecular Variance (AMOVA) for
Autopolyploids <em>Submitted</em>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/ade4/man/amova.html" class="external-link">ade4::amova()</a></code>, <code><a href="https://rdrr.io/pkg/pegas/man/amova.html" class="external-link">pegas::amova()</a></code>, <code><a href="clonecorrect.html">clonecorrect()</a></code>, <code><a href="diss.dist.html">diss.dist()</a></code>,
<code><a href="missingno.html">missingno()</a></code>, <code><a href="https://rdrr.io/pkg/ade4/man/is.euclid.html" class="external-link">ade4::is.euclid()</a></code>, <code><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata()</a></code>, <code><a href="make_haplotypes-method.html">make_haplotypes()</a></code>,
<code><a href="coercion-methods.html">as.genambig()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span><span class="op">$</span><span class="va">population_hierarchy</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">agc</span> <span class="op">&lt;-</span> <span class="fu"><a href="coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">agc</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is a genclone object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotype information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    119 original multilocus genotypes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    187 diploid individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     56 dominant loci</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Population information:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2 strata - Pop, Subpop</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2 populations defined - Athena, Mt. Vernon</span>
<span class="r-in"><span><span class="va">amova.result</span> <span class="op">&lt;-</span> <span class="fu">poppr.amova</span><span class="op">(</span><span class="va">agc</span>, <span class="op">~</span><span class="va">Pop</span><span class="op">/</span><span class="va">Subpop</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  No missing values detected.</span>
<span class="r-in"><span><span class="va">amova.result</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $call</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ade4::amova(samples = xtab, distances = xdist, structures = xstruct)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $results</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             Df    Sum Sq     Mean Sq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between Pop                  1 1051.2345 1051.234516</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between samples Within Pop  16  273.4575   17.091091</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Within samples             169  576.5059    3.411277</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total                      186 1901.1979   10.221494</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $componentsofcovariance</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                            Sigma          %</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between Pop                11.063446  70.006786</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between samples Within Pop  1.328667   8.407483</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Within samples              3.411277  21.585732</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total variations                       15.803391 100.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statphi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         Phi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-total 0.7841427</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-Pop   0.2803128</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-Pop-total     0.7000679</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">amova.test</span> <span class="op">&lt;-</span> <span class="fu">randtest</span><span class="op">(</span><span class="va">amova.result</span><span class="op">)</span> <span class="co"># Test for significance</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amova.test</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="poppr.amova-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">amova.test</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: krandtest lightkrandtest </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Monte-Carlo tests</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call: randtest.amova(xtest = amova.result)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of tests:   3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Adjustment method for multiple comparisons:   none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Permutation number:   99 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         Test       Obs   Std.Obs   Alter Pvalue</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  Variations within samples  3.411277 -28.82359    less   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Variations between samples  1.328667  20.76983 greater   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     Variations between Pop 11.063446  10.20109 greater   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># You can get the same results with the pegas implementation</span></span></span>
<span class="r-in"><span><span class="va">amova.pegas</span> <span class="op">&lt;-</span> <span class="fu">poppr.amova</span><span class="op">(</span><span class="va">agc</span>, <span class="op">~</span><span class="va">Pop</span><span class="op">/</span><span class="va">Subpop</span>, method <span class="op">=</span> <span class="st">"pegas"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  No missing values detected.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Zero distance(s)</span>
<span class="r-in"><span><span class="va">amova.pegas</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Analysis of Molecular Variance</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call: pegas::amova(formula = hier, data = hierdf, nperm = nperm, is.squared = FALSE)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              SSD         MSD  df</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pop    1051.2345 1051.234516   1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Subpop  273.4575   17.091091  16</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Error   576.5059    3.411277 169</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total  1901.1979   10.221494 186</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variance components:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Pop    Subpop     Error </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11.063446  1.328667  3.411277 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-statistics:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Pop.in.GLOBAL (Phi_CT) Subpop.in.GLOBAL (Phi_ST)    Subpop.in.Pop (Phi_SC) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 0.7000679                 0.7841427                 0.2803128 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variance coefficients:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        a        b        c </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10.29589 11.16949 93.36898 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">amova.pegas</span><span class="op">$</span><span class="va">varcomp</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">amova.pegas</span><span class="op">$</span><span class="va">varcomp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        Pop     Subpop      Error </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.70006786 0.08407483 0.21585732 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Clone correction is possible</span></span></span>
<span class="r-in"><span><span class="va">amova.cc.result</span> <span class="op">&lt;-</span> <span class="fu">poppr.amova</span><span class="op">(</span><span class="va">agc</span>, <span class="op">~</span><span class="va">Pop</span><span class="op">/</span><span class="va">Subpop</span>, clonecorrect <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  No missing values detected.</span>
<span class="r-in"><span><span class="va">amova.cc.result</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $call</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ade4::amova(samples = xtab, distances = xdist, structures = xstruct)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $results</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             Df    Sum Sq    Mean Sq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between Pop                  1  741.9872 741.987234</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between samples Within Pop  16  185.6877  11.605483</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Within samples             123  520.1123   4.228555</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total                      140 1447.7872  10.341337</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $componentsofcovariance</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                             Sigma          %</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between Pop                10.4131525  66.777680</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between samples Within Pop  0.9520545   6.105355</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Within samples              4.2285550  27.116965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total variations                       15.5937620 100.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statphi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         Phi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-total 0.7288303</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-Pop   0.1837727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-Pop-total     0.6677768</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">amova.cc.test</span> <span class="op">&lt;-</span> <span class="fu">randtest</span><span class="op">(</span><span class="va">amova.cc.result</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">amova.cc.test</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="poppr.amova-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">amova.cc.test</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: krandtest lightkrandtest </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Monte-Carlo tests</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call: randtest.amova(xtest = amova.cc.result)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of tests:   3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Adjustment method for multiple comparisons:   none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Permutation number:   99 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         Test        Obs    Std.Obs   Alter Pvalue</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  Variations within samples  4.2285550 -20.171106    less   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Variations between samples  0.9520545   9.802126 greater   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     Variations between Pop 10.4131525   9.237767 greater   0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example with filtering</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">splitStrata</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">Tree</span><span class="op">/</span><span class="va">Year</span><span class="op">/</span><span class="va">Symptom</span></span></span>
<span class="r-in"><span><span class="fu">poppr.amova</span><span class="op">(</span><span class="va">monpop</span>, <span class="op">~</span><span class="va">Symptom</span><span class="op">/</span><span class="va">Year</span><span class="op">)</span> <span class="co"># gets a warning of zero distances</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  No loci with missing values above 5% found.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Zero distance(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distance matrix is non-euclidean.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using quasieuclid correction method. See ?quasieuclid for details.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Zero distance(s)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $call</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ade4::amova(samples = xtab, distances = xdist, structures = xstruct)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $results</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Df     Sum Sq    Mean Sq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between Symptom                  1   17.62137  17.621372</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between samples Within Symptom   4  444.47746 111.119364</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Within samples                 688 4880.09229   7.093157</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total                          693 5342.19112   7.708789</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $componentsofcovariance</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                 Sigma          %</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between Symptom                -0.3380505  -4.369677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between samples Within Symptom  0.9811742  12.682763</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Within samples                  7.0931574  91.686913</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total variations                            7.7362812 100.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statphi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             Phi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-total    0.08313087</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-Symptom  0.12151770</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-Symptom-total   -0.04369677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">poppr.amova</span><span class="op">(</span><span class="va">monpop</span>, <span class="op">~</span><span class="va">Symptom</span><span class="op">/</span><span class="va">Year</span>, filter <span class="op">=</span> <span class="cn">TRUE</span>, threshold <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="co"># no warning</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Original multilocus genotypes   ... 264</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Contracted multilocus genotypes ... 236</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  No loci with missing values above 5% found.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distance matrix is non-euclidean.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using quasieuclid correction method. See ?quasieuclid for details.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $call</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ade4::amova(samples = xtab, distances = xdist, structures = xstruct)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $results</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 Df     Sum Sq    Mean Sq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between Symptom                  1   18.13215  18.132145</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Between samples Within Symptom   4  452.34660 113.086650</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Within samples                 688 5002.62587   7.271259</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total                          693 5473.10462   7.897698</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $componentsofcovariance</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                 Sigma          %</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between Symptom                -0.3430635  -4.328197</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Between samples Within Symptom  0.9980498  12.591711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Variations  Within samples                  7.2712585  91.736486</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total variations                            7.9262448 100.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statphi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             Phi</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-total    0.08263514</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-samples-Symptom  0.12069326</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phi-Symptom-total   -0.04328197</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

