<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="For the poppr package description, please see package?poppr
This function allows the user to quickly view indices of heterozygosity,
evenness, and linkage to aid in the decision of a path to further analyze a
specified dataset. It natively takes adegenet::genind and
genclone objects, but can convert any raw data formats
that adegenet can take (fstat, structure, genetix, and genpop) as well as
genalex files exported into a csv format (see read.genalex() for details)."><title>Produce a basic summary table for population genetic analyses. — poppr • poppr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Produce a basic summary table for population genetic analyses. — poppr"><meta property="og:description" content="For the poppr package description, please see package?poppr
This function allows the user to quickly view indices of heterozygosity,
evenness, and linkage to aid in the decision of a path to further analyze a
specified dataset. It natively takes adegenet::genind and
genclone objects, but can convert any raw data formats
that adegenet can take (fstat, structure, genetix, and genpop) as well as
genalex files exported into a csv format (see read.genalex() for details)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Produce a basic summary table for population genetic analyses.</h1>
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/R/Index_calculations.r" class="external-link"><code>R/Index_calculations.r</code></a></small>
      <div class="d-none name"><code>poppr.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For the <span class="pkg">poppr</span> package description, please see <code><a href="poppr-package.html">package?poppr</a></code></p>
<p>This function allows the user to quickly view indices of heterozygosity,
evenness, and linkage to aid in the decision of a path to further analyze a
specified dataset. It natively takes adegenet::genind and
<a href="genclone-class.html">genclone</a> objects, but can convert any raw data formats
that adegenet can take (fstat, structure, genetix, and genpop) as well as
genalex files exported into a csv format (see <code><a href="read.genalex.html">read.genalex()</a></code> for details).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">poppr</span><span class="op">(</span></span>
<span>  <span class="va">dat</span>,</span>
<span>  total <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sublist <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blacklist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  method <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  missing <span class="op">=</span> <span class="st">"ignore"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  clonecorrect <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  strata <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  keep <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  hist <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  index <span class="op">=</span> <span class="st">"rbarD"</span>,</span>
<span>  minsamp <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>dat</dt>
<dd><p>a adegenet::genind object OR a <a href="genclone-class.html">genclone</a>
object OR any fstat, structure, genetix, genpop, or genalex formatted
file.</p></dd>


<dt>total</dt>
<dd><p>When <code>TRUE</code> (default), indices will be calculated for the
pooled populations.</p></dd>


<dt>sublist</dt>
<dd><p>a list of character strings or integers to indicate specific
population names (accessed via <code><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">adegenet::popNames()</a></code>).
Defaults to "ALL".</p></dd>


<dt>exclude</dt>
<dd><p>a <code>vector</code> of population names or indexes that the user
wishes to discard. Default to <code>NULL</code>.</p></dd>


<dt>blacklist</dt>
<dd><p>DEPRECATED, use exclude.</p></dd>


<dt>sample</dt>
<dd><p>an integer indicating the number of permutations desired to
obtain p-values. Sampling will shuffle genotypes at each locus to simulate
a panmictic population using the observed genotypes. Calculating the
p-value includes the observed statistics, so set your sample number to one
off for a round p-value (eg. <code>sample = 999</code> will give you p = 0.001 and
<code>sample = 1000</code> will give you p = 0.000999001).</p></dd>


<dt>method</dt>
<dd><p>an integer from 1 to 4 indicating the method of sampling
desired. see <code><a href="shufflepop.html">shufflepop()</a></code> for details.</p></dd>


<dt>missing</dt>
<dd><p>how should missing data be treated? <code>"zero"</code> and
<code>"mean"</code> will set the missing values to those documented in
<code><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab()</a></code>. <code>"loci"</code> and <code>"geno"</code> will remove any loci or
genotypes with missing data, respectively (see <code><a href="missingno.html">missingno()</a></code> for
more information.</p></dd>


<dt>cutoff</dt>
<dd><p><code>numeric</code> a number from 0 to 1 indicating the percent
missing data allowed for analysis. This is to be used in conjunction with
the flag <code>missing</code> (see <code><a href="missingno.html">missingno()</a></code> for details)</p></dd>


<dt>quiet</dt>
<dd><p><code>FALSE</code> (default) will display a progress bar for each
population analyzed.</p></dd>


<dt>clonecorrect</dt>
<dd><p>default <code>FALSE</code>. must be used with the <code>strata</code>
parameter, or the user will potentially get undesired results. see
<code><a href="clonecorrect.html">clonecorrect()</a></code> for details.</p></dd>


<dt>strata</dt>
<dd><p>a <code>formula</code> indicating the hierarchical levels to be used.
The hierarchies should be present in the <code>strata</code> slot. See
<code><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata()</a></code> for details.</p></dd>


<dt>keep</dt>
<dd><p>an <code>integer</code>. This indicates which strata you wish to keep
after clone correcting your data sets. To combine strata, just set keep
from 1 to the number of straifications set in strata. see
<code><a href="clonecorrect.html">clonecorrect()</a></code> for details.</p></dd>


<dt>plot</dt>
<dd><p><code>logical</code> if <code>TRUE</code> (default) and <code>sampling &gt; 0</code>,
a histogram will be produced for each population.</p></dd>


<dt>hist</dt>
<dd><p><code>logical</code> Deprecated. Use plot.</p></dd>


<dt>index</dt>
<dd><p><code>character</code> Either "Ia" or "rbarD". If <code>hist = TRUE</code>,
this will determine the index used for the visualization.</p></dd>


<dt>minsamp</dt>
<dd><p>an <code>integer</code> indicating the minimum number of individuals
to resample for rarefaction analysis. See <code><a href="https://rdrr.io/pkg/vegan/man/rarefy.html" class="external-link">vegan::rarefy()</a></code> for
details.</p></dd>


<dt>legend</dt>
<dd><p><code>logical</code>. When this is set to <code>TRUE</code>, a legend describing the
resulting table columns will be printed. Defaults to <code>FALSE</code></p></dd>


<dt>...</dt>
<dd><p>arguments to be passed on to <code><a href="diversity_stats.html">diversity_stats()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data frame with populations in rows and the following columns:</p><ul><li><p><strong>Pop</strong>: A vector indicating the population factor</p></li>
<li><p><strong>N</strong>: An integer vector indicating the number of individuals/isolates in
the specified population.</p></li>
<li><p><strong>MLG</strong>: An integer vector indicating the number of multilocus genotypes
found in the specified population, (see: <code><a href="mlg.html">mlg()</a></code>)</p></li>
<li><p><strong>eMLG</strong>: The expected number of MLG at the lowest common sample size (set
by the parameter <code>minsamp</code>).</p></li>
<li><p><strong>SE</strong>: The standard error for the rarefaction analysis</p></li>
<li><p><strong>H</strong>: Shannon-Weiner Diversity index</p></li>
<li><p><strong>G</strong>: Stoddard and Taylor's Index</p></li>
<li><p><strong>lambda</strong>: Simpson's index</p></li>
<li><p><strong>E.5</strong>: Evenness</p></li>
<li><p><strong>Hexp</strong>: Nei's gene diversity (expected heterozygosity)</p></li>
<li><p><strong>Ia</strong>: A numeric vector giving the value of the Index of Association for
each population factor, (see <code><a href="ia.html">ia()</a></code>).</p></li>
<li><p><strong>p.Ia</strong>: A numeric vector indicating the p-value for Ia from the number
of reshufflings indicated in <code>sample</code>. Lowest value is 1/n where n is the
number of observed values.</p></li>
<li><p><strong>rbarD</strong>: A numeric vector giving the value of the Standardized Index of
Association for each population factor, (see <code><a href="ia.html">ia()</a></code>).</p></li>
<li><p><strong>p.rD</strong>: A numeric vector indicating the p-value for rbarD from the
number of reshuffles indicated in <code>sample</code>. Lowest value is 1/n where n is
the number of observed values.</p></li>
<li><p><strong>File</strong>: A vector indicating the name of the original data file.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This table is intended to be a first look into the dynamics of mutlilocus
genotype diversity. Many of the statistics (except for the the index of
association) are simply based on counts of multilocus genotypes and do not
take into account the actual allelic states. <strong>Descriptions of the
statistics can be found in the Algorithms and Equations vignette</strong>:
<code>vignette("algo", package = "poppr")</code>.</p><div class="section">
<h3 id="sampling">sampling<a class="anchor" aria-label="anchor" href="#sampling"></a></h3>


<p>The sampling procedure is explicitly for testing the index of association.
None of the other diversity statistics (H, G, lambda, E.5) are tested with
this sampling due to the differing data types. To obtain confidence
intervals for these statistics, please see <code><a href="diversity_ci.html">diversity_ci()</a></code>.</p>
</div>

<div class="section">
<h3 id="rarefaction">rarefaction<a class="anchor" aria-label="anchor" href="#rarefaction"></a></h3>


<p>Rarefaction analysis is performed on the number of multilocus genotypes
because it is relatively easy to estimate (Grünwald et al., 2003). To
obtain rarefied estimates of diversity, it is possible to use
<code><a href="diversity_ci.html">diversity_ci()</a></code> with the argument <code>rarefy = TRUE</code></p>
</div>

<div class="section">
<h3 id="graphic">graphic<a class="anchor" aria-label="anchor" href="#graphic"></a></h3>


<p>This function outputs a <span class="pkg">ggplot2</span> graphic of histograms. These can be
manipulated to be visualized in another manner by retrieving the plot with
the <code><a href="https://ggplot2.tidyverse.org/reference/last_plot.html" class="external-link">last_plot()</a></code> command from <span class="pkg">ggplot2</span>. A useful manipulation would
be to arrange the graphs into a single column so that the values of the
statistic line up: <code>p &lt;- last_plot(); p + facet_wrap(~population, ncol = 1, scales = "free_y")</code> The name for the groupings is
"population" and the name for the x axis is "value".</p>
</div>

    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The calculation of <code>Hexp</code> has changed from <span class="pkg">poppr</span> 1.x. It was
previously calculated based on the diversity of multilocus genotypes,
resulting in a value of 1 for sexual populations. This was obviously not
Nei's 1978 expected heterozygosity. We have thus changed the statistic to
be the true value of Hexp by calculating \((\frac{n}{n-1}) 1 - \sum_{i =
  1}^k{p^{2}_{i}}\) where p is the allele
frequencies at a given locus and n is the number of observed alleles (Nei,
1978) in each locus and then returning the average. Caution should be
exercised in interpreting the results of Hexp with polyploid organisms with
ambiguous ploidy. The lack of allelic dosage information will cause rare
alleles to be over-represented and artificially inflate the index. This is
especially true with small sample sizes.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Paul-Michael Agapow and Austin Burt. Indices of multilocus
linkage disequilibrium. <em>Molecular Ecology Notes</em>, 1(1-2):101-102,
2001</p>
<p>A.H.D. Brown, M.W. Feldman, and E. Nevo. Multilocus structure of natural
populations of <em>Hordeum spontaneum</em>. <em>Genetics</em>, 96(2):523-536,
1980.</p>
<p>Niklaus J. Gr\"unwald, Stephen B. Goodwin, Michael G. Milgroom, and William
E. Fry. Analysis of genotypic diversity data for populations of
microorganisms. Phytopathology, 93(6):738-46, 2003</p>
<p>Bernhard Haubold and Richard R. Hudson. Lian 3.0: detecting linkage
disequilibrium in multilocus data. Bioinformatics, 16(9):847-849, 2000.</p>
<p>Kenneth L.Jr. Heck, Gerald van Belle, and Daniel Simberloff. Explicit
calculation of the rarefaction diversity measurement and the determination
of sufficient sample size. Ecology, 56(6):pp. 1459-1461, 1975</p>
<p>Masatoshi Nei. Estimation of average heterozygosity and genetic distance
from a small number of individuals. Genetics, 89(3):583-590, 1978.</p>
<p>S H Hurlbert. The nonconcept of species diversity: a critique and
alternative parameters. Ecology, 52(4):577-586, 1971.</p>
<p>J.A. Ludwig and J.F. Reynolds. Statistical Ecology. A Primer on Methods and
Computing. New York USA: John Wiley and Sons, 1988.</p>
<p>Simpson, E. H. Measurement of diversity. Nature 163: 688, 1949
doi:10.1038/163688a0</p>
<p>Good, I. J. (1953). On the Population Frequency of Species and the
Estimation of Population Parameters. <em>Biometrika</em> 40(3/4): 237-264.</p>
<p>Lande, R. (1996). Statistics and partitioning of species diversity, and
similarity among multiple communities. <em>Oikos</em> 76: 5-13.</p>
<p>Jari Oksanen, F. Guillaume Blanchet, Roeland Kindt, Pierre Legendre, Peter
R. Minchin, R. B. O'Hara, Gavin L. Simpson, Peter Solymos, M. Henry H.
Stevens, and Helene Wagner. vegan: Community Ecology Package, 2012. R
package version 2.0-5.</p>
<p>E.C. Pielou. Ecological Diversity. Wiley, 1975.</p>
<p>Claude Elwood Shannon. A mathematical theory of communication. Bell Systems
Technical Journal, 27:379-423,623-656, 1948</p>
<p>J M Smith, N H Smith, M O'Rourke, and B G Spratt. How clonal are bacteria?
Proceedings of the National Academy of Sciences, 90(10):4384-4388, 1993.</p>
<p>J.A. Stoddart and J.F. Taylor. Genotypic diversity: estimation and
prediction in samples. Genetics, 118(4):705-11, 1988.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="clonecorrect.html">clonecorrect()</a></code>,
<code><a href="poppr.all.html">poppr.all()</a></code>,
<code><a href="ia.html">ia()</a></code>,
<code><a href="missingno.html">missingno()</a></code>,
<code><a href="mlg.html">mlg()</a></code>,
<code><a href="diversity_stats.html">diversity_stats()</a></code>,
<code><a href="diversity_ci.html">diversity_ci()</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Zhian N. Kamvar</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">poppr</span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Pop   N MLG eMLG       SE    H   G lambda E.5  Hexp      Ia   rbarD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    P01  10  10   10 0.00e+00 2.30  10  0.900   1 0.649  0.1656  0.0211</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    P02  22  22   10 0.00e+00 3.09  22  0.955   1 0.701  0.1818  0.0230</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    P03  12  12   10 0.00e+00 2.48  12  0.917   1 0.719  0.3546  0.0452</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    P04  23  23   10 5.03e-07 3.14  23  0.957   1 0.750  0.4494  0.0563</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    P05  15  15   10 2.77e-07 2.71  15  0.933   1 0.640 -0.0475 -0.0060</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    P06  11  11   10 0.00e+00 2.40  11  0.909   1 0.745  0.3337  0.0426</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7    P07  14  14   10 0.00e+00 2.64  14  0.929   1 0.667  0.2569  0.0326</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8    P08  10  10   10 0.00e+00 2.30  10  0.900   1 0.752  0.2388  0.0301</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9    P09   9   9    9 0.00e+00 2.20   9  0.889   1 0.694  2.0845  0.2636</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10   P10  11  11   10 0.00e+00 2.40  11  0.909   1 0.698  0.5955  0.0763</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11   P11  20  20   10 0.00e+00 3.00  20  0.950   1 0.783  0.2847  0.0363</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12   P12  14  14   10 0.00e+00 2.64  14  0.929   1 0.667  0.4899  0.0643</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13   P13  13  13   10 7.30e-08 2.56  13  0.923   1 0.688  0.1855  0.0237</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14   P14  17  17   10 2.31e-07 2.83  17  0.941   1 0.789  0.2210  0.0282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15   P15  11  11   10 0.00e+00 2.40  11  0.909   1 0.723  0.6933  0.0873</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16   P16  12  12   10 0.00e+00 2.48  12  0.917   1 0.700  0.2345  0.0295</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17   P17  13  13   10 7.30e-08 2.56  13  0.923   1 0.605 -0.0906 -0.0138</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18 Total 237 237   10 0.00e+00 5.47 237  0.996   1 0.774  0.1721  0.0218</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         File</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18 nancycats</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Sampling</span></span></span>
<span class="r-in"><span><span class="fu">poppr</span><span class="op">(</span><span class="va">nancycats</span>, sample <span class="op">=</span> <span class="fl">999</span>, total <span class="op">=</span> <span class="cn">FALSE</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="poppr-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Pop  N MLG eMLG       SE    H  G lambda E.5  Hexp      Ia  p.Ia   rbarD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  P01 10  10   10 0.00e+00 2.30 10  0.900   1 0.649  0.1656 0.229  0.0211</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  P02 22  22   10 0.00e+00 3.09 22  0.955   1 0.701  0.1818 0.074  0.0230</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  P03 12  12   10 0.00e+00 2.48 12  0.917   1 0.719  0.3546 0.042  0.0452</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  P04 23  23   10 5.03e-07 3.14 23  0.957   1 0.750  0.4494 0.001  0.0563</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  P05 15  15   10 2.77e-07 2.71 15  0.933   1 0.640 -0.0475 0.578 -0.0060</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  P06 11  11   10 0.00e+00 2.40 11  0.909   1 0.745  0.3337 0.060  0.0426</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  P07 14  14   10 0.00e+00 2.64 14  0.929   1 0.667  0.2569 0.098  0.0326</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  P08 10  10   10 0.00e+00 2.30 10  0.900   1 0.752  0.2388 0.138  0.0301</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  P09  9   9    9 0.00e+00 2.20  9  0.889   1 0.694  2.0845 0.001  0.2636</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 P10 11  11   10 0.00e+00 2.40 11  0.909   1 0.698  0.5955 0.012  0.0763</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 P11 20  20   10 0.00e+00 3.00 20  0.950   1 0.783  0.2847 0.307  0.0363</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 P12 14  14   10 0.00e+00 2.64 14  0.929   1 0.667  0.4899 0.014  0.0643</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 P13 13  13   10 7.30e-08 2.56 13  0.923   1 0.688  0.1855 0.146  0.0237</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 P14 17  17   10 2.31e-07 2.83 17  0.941   1 0.789  0.2210 0.068  0.0282</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 P15 11  11   10 0.00e+00 2.40 11  0.909   1 0.723  0.6933 0.008  0.0873</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 P16 12  12   10 0.00e+00 2.48 12  0.917   1 0.700  0.2345 0.112  0.0295</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 P17 13  13   10 7.30e-08 2.56 13  0.923   1 0.605 -0.0906 0.667 -0.0138</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     p.rD      File</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  0.228 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  0.074 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  0.042 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  0.001 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  0.577 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  0.060 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  0.098 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  0.140 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  0.001 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 0.011 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 0.309 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 0.013 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 0.144 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 0.068 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 0.009 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 0.112 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 0.667 nancycats</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Customizing the plot</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/last_plot.html" class="external-link">last_plot</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">population</span>, scales <span class="op">=</span> <span class="st">"free_y"</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="poppr-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Turning off diversity statistics (see get_stats)</span></span></span>
<span class="r-in"><span><span class="fu">poppr</span><span class="op">(</span><span class="va">nancycats</span>, total<span class="op">=</span><span class="cn">FALSE</span>, H <span class="op">=</span> <span class="cn">FALSE</span>, G <span class="op">=</span> <span class="cn">FALSE</span>, lambda <span class="op">=</span> <span class="cn">FALSE</span>, E5 <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Pop  N MLG eMLG       SE  Hexp      Ia   rbarD      File</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  P01 10  10   10 0.00e+00 0.649  0.1656  0.0211 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  P02 22  22   10 0.00e+00 0.701  0.1818  0.0230 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  P03 12  12   10 0.00e+00 0.719  0.3546  0.0452 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  P04 23  23   10 5.03e-07 0.750  0.4494  0.0563 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  P05 15  15   10 2.77e-07 0.640 -0.0475 -0.0060 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  P06 11  11   10 0.00e+00 0.745  0.3337  0.0426 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  P07 14  14   10 0.00e+00 0.667  0.2569  0.0326 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  P08 10  10   10 0.00e+00 0.752  0.2388  0.0301 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  P09  9   9    9 0.00e+00 0.694  2.0845  0.2636 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 P10 11  11   10 0.00e+00 0.698  0.5955  0.0763 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 P11 20  20   10 0.00e+00 0.783  0.2847  0.0363 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 P12 14  14   10 0.00e+00 0.667  0.4899  0.0643 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 P13 13  13   10 7.30e-08 0.688  0.1855  0.0237 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 P14 17  17   10 2.31e-07 0.789  0.2210  0.0282 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 P15 11  11   10 0.00e+00 0.723  0.6933  0.0873 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 P16 12  12   10 0.00e+00 0.700  0.2345  0.0295 nancycats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 P17 13  13   10 7.30e-08 0.605 -0.0906 -0.0138 nancycats</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The previous version of poppr contained a definition of Hexp, which</span></span></span>
<span class="r-in"><span><span class="co"># was calculated as (N/(N - 1))*lambda. It basically looks like an unbiased </span></span></span>
<span class="r-in"><span><span class="co"># Simpson's index. This statistic was originally included in poppr because it</span></span></span>
<span class="r-in"><span><span class="co"># was originally included in the program multilocus. It was finally figured</span></span></span>
<span class="r-in"><span><span class="co"># to be an unbiased Simpson's diversity metric (Lande, 1996; Good, 1953).</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">uSimp</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">lambda</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/diversity.html" class="external-link">diversity</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"simpson"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/drop.html" class="external-link">drop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">}</span></span></span>
<span class="r-in"><span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="op">(</span><span class="va">N</span><span class="op">/</span><span class="op">(</span><span class="va">N</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="va">lambda</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span><span class="fu">poppr</span><span class="op">(</span><span class="va">Aeut</span>, uSimp <span class="op">=</span> <span class="va">uSimp</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          Pop   N MLG eMLG   SE    H    G lambda   E.5 uSimp  Hexp    Ia  rbarD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     Athena  97  70 66.0 1.25 4.06 42.2  0.976 0.721 0.986 0.170  2.91 0.0724</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Mt. Vernon  90  50 50.0 0.00 3.67 28.7  0.965 0.726 0.976 0.158 13.30 0.2816</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      Total 187 119 68.5 2.99 4.56 69.0  0.986 0.720 0.991 0.365 14.37 0.2706</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   File</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 Aeut</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 Aeut</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Aeut</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Demonstration with viral data</span></span></span>
<span class="r-in"><span><span class="co"># Note: this is a larger data set that could take a couple of minutes to run</span></span></span>
<span class="r-in"><span><span class="co"># on slower computers. </span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/population-methods.html" class="external-link">setPop</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">country</span></span></span>
<span class="r-in"><span><span class="fu">poppr</span><span class="op">(</span><span class="va">H3N2</span>, total <span class="op">=</span> <span class="cn">FALSE</span>, sublist<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Austria"</span>, <span class="st">"China"</span>, <span class="st">"USA"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  clonecorrect <span class="op">=</span> <span class="cn">TRUE</span>, strata <span class="op">=</span> <span class="op">~</span><span class="va">country</span><span class="op">/</span><span class="va">year</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Pop   N MLG eMLG    SE    H     G lambda   E.5   Hexp    Ia  rbarD File</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     USA 275 254 42.5 0.687 5.51 238.6  0.996 0.964 0.1341 10.75 0.1167 H3N2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   China  82  79 42.2 0.757 4.36  76.4  0.987 0.980 0.0929  2.54 0.0371 H3N2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 Austria  43  41 41.0 0.000 3.70  39.3  0.975 0.975 0.1140 13.02 0.2226 H3N2</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

