<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genetic Analysis of Populations with Mixed Reproduction • poppr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><meta property="og:title" content="Genetic Analysis of Populations with Mixed Reproduction">
<meta property="og:description" content="Population genetic analyses for hierarchical analysis of partially
    clonal populations built upon the architecture of the 'adegenet' package.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="poppr-version-2" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#poppr-version-2" class="anchor"></a>Poppr version 2 <img src="articles/small_logo.png" align="right">
</h1></div>

<p><a href="http://www.r-pkg.org/pkg/poppr"><img src="https://cranlogs.r-pkg.org/badges/grand-total/poppr" alt="Downloads from Rstudio mirror"></a> <a href="http://depsy.org/package/r/poppr"><img src="http://depsy.org/api/package/cran/poppr/badge.svg" alt="Research software impact"></a></p>
<div id="what-is-poppr" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-poppr" class="anchor"></a>What is <em>poppr</em>?</h2>
<p><em>Poppr</em> is an R package designed for analysis of populations with mixed modes of sexual and clonal reproduction. It is built around the framework of <a href="http://github.com/thibautjombart/adegenet/">adegenet’s</a> genind and genlight objects and offers the following implementations:</p>
<ul>
<li>clone censoring of populations at any of multiple levels of a hierarchy</li>
<li>convenient counting of multilocus genotypes and sub-setting of populations with multiple levels of hierarchy</li>
<li>define multilocus genotypes</li>
<li>calculation of indices of genotypic diversity, evenness, richness, and rarefaction</li>
<li>drawing of dendrograms with bootstrap support for genetic distances</li>
<li>drawing of minimum spanning networks for genetic distances</li>
<li>calculation of the index of association (<img src="http://latex.codecogs.com/gif.latex?I_A" alt="Index of association">) or (<img src="http://latex.codecogs.com/gif.latex?%5Cbar%7Br%7D_d" alt="Standardized index of association">)</li>
<li>batch processing on any server that has R ( ≥ 2.15.1) installed</li>
<li>calculation of Bruvo’s distance for microsatellite (SSR) markers (implemented in C for speed)</li>
<li>import of data from and export to <a href="http://biology.anu.edu.au/GenAlEx/Welcome.html" title="GenAlEx Homepage">GenAlEx</a>
</li>
</ul>
<div id="new-in-version-2-0" class="section level3">
<h3 class="hasAnchor">
<a href="#new-in-version-2-0" class="anchor"></a>New in version 2.0:</h3>
<ul>
<li>handling of genomic SNP data</li>
<li>custom multilocus genotype definitions</li>
<li>collapse multilocus lineages by genetic distance</li>
<li>calculate reticulate minimum spanning networks</li>
<li>calculate index of association in a sliding window across snps</li>
<li>bootstrapping of MLG diversity statistics</li>
<li><a href="https://github.com/zkamvar/poppr_msn_shiny">interactive exploration of minimum spanning networks</a></li>
<li>and more!</li>
</ul>
<p>For full details, see the NEWS file or type in your R console:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">news</span>(Version <span class="op">&gt;=</span><span class="st"> "2.0.0"</span>, <span class="dt">package =</span> <span class="st">"poppr"</span>)</code></pre></div>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>If you use <em>poppr</em> at all, please specify the version and cite:</p>
<blockquote>
<p>Kamvar ZN, Tabima JF, Grünwald NJ. (2014) Poppr: an R package for genetic analysis of populations with clonal, partially clonal, and/or sexual reproduction. PeerJ 2:e281 <a href="http://dx.doi.org/10.7717/peerj.281" class="uri">http://dx.doi.org/10.7717/peerj.281</a></p>
</blockquote>
<p>If you use <em>poppr</em> in a presentation please mention it as the <em>poppr</em> R package, specify the version, and use our logo: <a href="http://github.com/grunwaldlab/poppr/tree/master/vignettes/jalapeno_logo.png">(png)</a> | <a href="http://github.com/grunwaldlab/poppr/tree/master/vignettes/jalapeno_logo.svg">(svg)</a>.</p>
<p>Additionally, if you use any following functionalities:</p>
<ul>
<li>minimum spanning networks with reticulation</li>
<li>collapsing multilocus genotypes into multilocus lineages with <code><a href="reference/mlg.filter.html">mlg.filter()</a></code>
</li>
<li>custom multilocus genotype definitions with <code>mlg.custom()</code>
</li>
<li>index of association for genomic data with <code><a href="reference/win.ia.html">win.ia()</a></code> or <code><a href="reference/samp.ia.html">samp.ia()</a></code>
</li>
<li>bootstrapping any genetic distance with genind, genlight, or genpop objects with <code><a href="reference/aboot.html">aboot()</a></code>
</li>
</ul>
<p>Please also cite:</p>
<blockquote>
<p>Kamvar ZN, Brooks JC and Grünwald NJ (2015) Novel R tools for analysis of genome-wide population genetic data with emphasis on clonality. Front. Genet. 6:208. doi: <a href="http://journal.frontiersin.org/article/10.3389/fgene.2015.00208/abstract">10.3389/fgene.2015.00208</a></p>
</blockquote>
<p>You can obtain citation information in R by typing:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citation</span>(<span class="dt">package =</span> <span class="st">"poppr"</span>)</code></pre></div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="from-cran" class="section level3">
<h3 class="hasAnchor">
<a href="#from-cran" class="anchor"></a>From CRAN</h3>
<p><a href="https://cran.r-project.org/package=poppr">Binary versions for mac and windows are available for R ≥ 2.15.1 <strong>here</strong></a>.</p>
<p>To install, make sure R is at least version 2.15.1 (the authors recommend ≥ 3.0), and in your console, type:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"poppr"</span>)</code></pre></div>
<p>If you want the absolute latest version of <em>poppr</em>, see about installing from github below.</p>
<hr>
</div>
<div id="stable-and-development-versions" class="section level3">
<h3 class="hasAnchor">
<a href="#stable-and-development-versions" class="anchor"></a>Stable and Development versions</h3>
<p>To install this package from github, make sure you have the following:</p>
<ul>
<li>
<a href="https://developer.apple.com/xcode">Xcode</a> (OSX) OR <a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a> (Windows)</li>
<li>
<a href="https://github.com/hadley/devtools">devtools</a> (to install, use: <code>install.packages("devtools")</code>)</li>
</ul>
<p>For Linux users, make sure that the function <code>getOption("unzip")</code> returns <code>"unzip"</code> or <code>"internal"</code>. If it doesn’t, then run <code>options(unzip = "internal")</code>.</p>
<p>Once you have devtools and a C compiler installed, you can use the <code>install_github()</code> function to install the current version from github.</p>
<div id="stable-version" class="section level4">
<h4 class="hasAnchor">
<a href="#stable-version" class="anchor"></a>Stable version</h4>
<p>This release will contain bug fixes and new, documented, and stable features that will be included in future releases. Note: if you don’t have LaTeX installed, you should set <code>build_vignettes = FALSE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="dt">repo =</span> <span class="st">"grunwaldlab/poppr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)
<span class="kw">library</span>(<span class="st">"poppr"</span>)</code></pre></div>
</div>
<div id="unstabledevelopment-versions" class="section level4">
<h4 class="hasAnchor">
<a href="#unstabledevelopment-versions" class="anchor"></a>Unstable/Development versions</h4>
<p>All new features in testing will be released on different branches. These features will be in various stages of development and may or may not be documented. Install with caution. The below command would install features on the the branch called “devel”. Note that these branches might be out of date from the master branch.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="dt">repo =</span> <span class="st">"grunwaldlab/poppr@devel"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)
<span class="kw">library</span>(<span class="st">"poppr"</span>)</code></pre></div>
</div>
</div>
</div>
<div id="help-documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#help-documentation" class="anchor"></a>Help / Documentation</h2>
<div id="r-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#r-documentation" class="anchor"></a>R documentation</h3>
<p>To access a descriptive index of help files in <em>poppr</em>, type in your console:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><a href="reference/poppr-package.html">package?poppr</a></code></pre></div>
</div>
<div id="vignettes" class="section level3">
<h3 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h3>
<p>A few vignettes have been written for poppr:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Title</th>
<th align="left">Command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Algorightms and Equations</td>
<td align="left"><code>vignette("algo", "poppr")</code></td>
</tr>
<tr class="even">
<td align="left">Data import and manipulation</td>
<td align="left"><code><a href="articles/poppr_manual.html">vignette("poppr_manual", "poppr")</a></code></td>
</tr>
<tr class="odd">
<td align="left">Multilocus Genotype Analysis</td>
<td align="left"><code><a href="articles/mlg.html">vignette("mlg", "poppr")</a></code></td>
</tr>
</tbody>
</table>
</div>
<div id="user-group" class="section level3">
<h3 class="hasAnchor">
<a href="#user-group" class="anchor"></a>User Group</h3>
<p>Users who have any questions/comments/suggestions regarding any version of poppr (stable or development) should direct their comments to the <a href="http://groups.google.com/group/poppr">Poppr google group</a></p>
</div>
<div id="bookprimer" class="section level3">
<h3 class="hasAnchor">
<a href="#bookprimer" class="anchor"></a>Book/Primer</h3>
<p>In Spring of 2014, Dr. Niklaus J. Grünwald, Dr. Sydney E. Everhart and Zhian N. Kamvar wrote a primer for population genetic analysis in R located at <a href="http://grunwaldlab.github.io/Population_Genetics_in_R" class="uri">http://grunwaldlab.github.io/Population_Genetics_in_R</a>.</p>
</div>
</div>
<div id="contributing" class="section level2">
<h2 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h2>
<p>Please note that this project is released with a <a href="CONDUCT.html">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms. If you wish to contribute code to <em>poppr</em>, please fork the repository and create a pull request with your added feature.</p>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cran.r-project.org/package=poppr">https://​cran.r-project.org/​package=poppr</a>
</li>
<li>Browse source code at <br><a href="https://github.com/grunwaldlab/poppr">https://​github.com/​grunwaldlab/​poppr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/grunwaldlab/poppr/issues">https://​github.com/​grunwaldlab/​poppr/​issues</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a> | <a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></p>
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing poppr</a></li>
</ul>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Zhian N. Kamvar <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-1458-7108" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid"></a> </li>
<li>Javier F. Tabima <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3603-2691" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid"></a> </li>
<li>Jonah C. Brooks <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/grunwaldlab/poppr"><img src="https://travis-ci.org/grunwaldlab/poppr.svg?branch=update-amova" alt="Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/grunwaldlab/poppr"><img src="https://ci.appveyor.com/api/projects/status/github/grunwaldlab/poppr?branch=master&amp;svg=true" alt="AppVeyor build status"></a></li>
<li><a href="https://coveralls.io/r/grunwaldlab/poppr?branch=update-amova"><img src="https://coveralls.io/repos/grunwaldlab/poppr/badge.svg?branch=update-amova" alt="Coverage Status"></a></li>
<li><a href="https://cran.r-project.org/package=poppr"><img src="http://www.r-pkg.org/badges/version/poppr" alt="CRAN version"></a></li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
