<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of Multilocus Genotypes and Lineages in poppr 2.7.1 • poppr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of Multilocus Genotypes and Lineages in poppr 2.7.1">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Analysis of Multilocus Genotypes and Lineages in poppr 2.7.1</h1>
                        <h4 class="author">Zhian N. Kamvar and Niklaus J. Grünwald</h4>
            
            <h4 class="date">2018-03-18</h4>
      
      <small>Source: <a href="https:/github.com/grunwaldlab/poppr/blob/master/vignettes/mlg.Rmd"><code>vignettes/mlg.Rmd</code></a></small>

    </div>

    
    
<div class="contents">
<hr>
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>A multilocus genotype is a unique combination of alleles across two or more loci. For organisms such as plant pathogens that often reproduce clonally, multilocus genotypes are invaluable for identifying the mode and spread of an organism. This document will describe in detail how you can define multilocus genotypes/lineages and how you can use them for your analyses of both genclone and snpclone objects.</p>
<p><br><br></p>
<center>
<img src="jalapeno_logo.png" width="300">
</center>
<hr>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
</div>
<div id="multilocus-genotype-flavors" class="section level1">
<h1 class="hasAnchor">
<a href="#multilocus-genotype-flavors" class="anchor"></a>Multilocus Genotype Flavors</h1>
<p>There are three different ways for constructing multilocus genotypes in <em>poppr</em>.</p>
<table class="table">
<thead><tr class="header">
<th>Method</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>"original"</code></td>
<td>Default MLG definition; strict matching</td>
</tr>
<tr class="even">
<td><code>"contracted"</code></td>
<td>Collapsing MLGs by genetic distance with <code>mlg.filter</code>
</td>
</tr>
<tr class="odd">
<td><code>"custom"</code></td>
<td>User-defined multilocus genotypes</td>
</tr>
</tbody>
</table>
<p>The first is a simply naïve construction where all alleles must match to make a unique multilocus genotype. New in version 2.0 is the ability to use genetic distance to collapse multilocus genotypes or define custom multilocus lineages based on other information. You can use the function <code><a href="../reference/mll-method.html">mll()</a></code> to display and switch between different multilocus genotypes/lineages.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"poppr"</span>)
<span class="kw">data</span>(monpop)
monpop</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    264 multilocus genotypes 
##    694 haploid individuals
##     13 codominant loci
## 
## Population information:
## 
##      1 stratum - Pop
##     12 populations defined - 
## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop, <span class="st">"original"</span>), <span class="dv">20</span>) <span class="co"># Showing the definitions for the first 20 samples</span></code></pre></div>
<pre><code>## Warning: 
##  The @mlg slot does not contain an MLG class object.
##  Returning the original mlgs. Please use:
## 
##  mll(monpop) &lt;- "original"
##  
##  to convert your object.</code></pre>
<pre><code>##  [1] 260 179 169 168 167 221 152 133 144  78  78  79  81  44  41  40  40
## [18]  38 119 118</code></pre>
<p>Notice how we got a warning. This is because the <code>monpop</code> data set was created before <code><a href="../reference/mll-method.html">mll()</a></code> was implemented. Luckily, this warning gives us information on how to fix it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop) &lt;-<span class="st"> "original"</span>
monpop</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    264 original multilocus genotypes 
##    694 haploid individuals
##     13 codominant loci
## 
## Population information:
## 
##      1 stratum - Pop
##     12 populations defined - 
## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</code></pre>
<p>We can see that the description says 264 <strong>original</strong> multilocus genotypes . This is how we know what the current MLG definition is for our object. Let’ see what happens when we change it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop) &lt;-<span class="st"> "custom"</span>
monpop</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    264 custom multilocus genotypes 
##    694 haploid individuals
##     13 codominant loci
## 
## Population information:
## 
##      1 stratum - Pop
##     12 populations defined - 
## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</code></pre>
<p>Now it says 264 <strong>custom</strong> multilocus genotypes . Of course, if we look at the MLGs, we will see that they appear to be the same as our original definition:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop, <span class="st">"custom"</span>), <span class="dv">20</span>) <span class="co"># Showing the definitions for the first 20 samples</span></code></pre></div>
<pre><code>##  [1] 260 179 169 168 167 221 152 133 144 78  78  79  81  44  41  40  40 
## [18] 38  119 118
## 264 Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 ... 264</code></pre>
<p>All of these functions will work for both <strong>genclone</strong> and <strong>snpclone</strong> objects. In this section, we will demonstrate how to construct each of these three multilocus genotype definitions in different data sets.</p>
<div id="naive-original" class="section level2">
<h2 class="hasAnchor">
<a href="#naive-original" class="anchor"></a>Naïve (“original”)</h2>
<p>This is the default way <em>poppr</em> calculates multilocus genotypes. You see it immediately when you load a genclone object. Let’s use the example <code>monpop</code> from <span class="citation">(Everhart and Scherm 2015)</span>. First, we need to reset the data to our original MLGs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop) &lt;-<span class="st"> "original"</span>
monpop</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    264 original multilocus genotypes 
##    694 haploid individuals
##     13 codominant loci
## 
## Population information:
## 
##      1 stratum - Pop
##     12 populations defined - 
## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</code></pre>
<p>You notice that there are 694 samples, yet only 264 multilocus genotypes in the data set over 13 loci. In this sub-section, I will demonstrate how these MLGs are constructed.</p>
<p>The naïve definition simply takes strings of alleles and compares them for equality. This method is quick and easily interpretable, but means that things like genotyping error, hyper-variable loci, and missing data all contribute to a unique genotype that might not be truly unique <span class="citation">(Kamvar, Brooks, and Grünwald 2015)</span>.</p>
<p>To give an example, let’s imagine that we have five samples with alleles typed at two loci.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grid_example &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">4</span>,
                         <span class="dv">1</span>, <span class="dv">1</span>,
                         <span class="dv">5</span>, <span class="dv">1</span>,
                         <span class="dv">9</span>, <span class="dv">1</span>,
                         <span class="dv">9</span>, <span class="dv">4</span>),
                       <span class="dt">ncol =</span> <span class="dv">2</span>,
                       <span class="dt">byrow =</span> <span class="ot">TRUE</span>)
<span class="kw">rownames</span>(grid_example) &lt;-<span class="st"> </span>LETTERS[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="kw">colnames</span>(grid_example) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)
grid_example</code></pre></div>
<pre><code>##   x y
## A 1 4
## B 1 1
## C 5 1
## D 9 1
## E 9 4</code></pre>
<p>We notice how they all have different multilocus genotypes. Indeed, if we analyze them, we see that there are five multilocus genotypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"poppr"</span>)
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coercion-methods.html">as.genclone</a></span>(<span class="kw">df2genind</span>(grid_example, <span class="dt">ploidy =</span> <span class="dv">1</span>))
<span class="kw">tab</span>(x)  <span class="co"># Look at the multilocus genotype table</span></code></pre></div>
<pre><code>##   x.1 x.5 x.9 y.4 y.1
## A   1   0   0   1   0
## B   1   0   0   0   1
## C   0   1   0   0   1
## D   0   0   1   0   1
## E   0   0   1   1   0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">nmll</a></span>(x) <span class="co"># count the number of multilocus genotypes</span></code></pre></div>
<pre><code>## [1] 5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x)  <span class="co"># show the multilocus genotype definitions</span></code></pre></div>
<pre><code>## [1] 5 4 3 1 2</code></pre>
<p>What we did above was to analyze the <strong>tab</strong> slot of the object showing the counts of alleles per sample across the two loci. We can clearly see by eye that no two samples are alike. The <code><a href="../reference/mll-method.html">nmll()</a></code> function counts the <strong>n</strong>umber of <strong>m</strong>ulti<strong>l</strong>ocus <strong>l</strong>ineages in the sample. the <code><a href="../reference/mll-method.html">mll()</a></code> function displays the assignment of the multilocus lineages in the sample.</p>
<blockquote>
<p>Note: we used <code><a href="../reference/mll-method.html">mll()</a></code> to show us our multilocus lineages. Later on, we will use this same function to switch between different definitions. See ?mll for details.</p>
</blockquote>
<p>Now let’s say we included another sample with partial missing data. Let’s say it was typed as allele “5” at the <strong>x</strong> locus, but missing the allele at the <strong>y</strong> locus.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coercion-methods.html">as.genclone</a></span>(<span class="kw">df2genind</span>(<span class="kw">rbind</span>(grid_example, <span class="dt">new =</span> <span class="kw">c</span>(<span class="dv">5</span>, <span class="ot">NA</span>)), <span class="dt">ploidy =</span> <span class="dv">1</span>))
<span class="kw">tab</span>(x)  <span class="co"># Note the missing data at locus 2.</span></code></pre></div>
<pre><code>##     x.1 x.5 x.9 y.4 y.1
## A     1   0   0   1   0
## B     1   0   0   0   1
## C     0   1   0   0   1
## D     0   0   1   0   1
## E     0   0   1   1   0
## new   0   1   0  NA  NA</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">nmll</a></span>(x)</code></pre></div>
<pre><code>## [1] 6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x)</code></pre></div>
<pre><code>## [1] 6 5 3 1 2 4</code></pre>
<p>Now we see that we have 6 multilocus genotypes even though one of them contains missing data.</p>
<p>Let’s also imagine that we included yet another sample that had a low-frequency allele at locus <strong>y</strong>, “2”, but had the allele “5” at locus <strong>x</strong>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grid_new &lt;-<span class="st"> </span><span class="kw">rbind</span>(grid_example,
                  <span class="dt">new =</span> <span class="kw">c</span>(<span class="dv">5</span>, <span class="ot">NA</span>),
                  <span class="dt">mut =</span> <span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>)
                  )
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coercion-methods.html">as.genclone</a></span>(<span class="kw">df2genind</span>(grid_new, <span class="dt">ploidy =</span> <span class="dv">1</span>))
<span class="kw">tab</span>(x)</code></pre></div>
<pre><code>##     x.1 x.5 x.9 y.4 y.1 y.2
## A     1   0   0   1   0   0
## B     1   0   0   0   1   0
## C     0   1   0   0   1   0
## D     0   0   1   0   1   0
## E     0   0   1   1   0   0
## new   0   1   0  NA  NA  NA
## mut   0   1   0   0   0   1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">nmll</a></span>(x)</code></pre></div>
<pre><code>## [1] 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x)</code></pre></div>
<pre><code>## [1] 7 6 4 1 2 5 3</code></pre>
<p>Again, we get completely different genotypes, but notice how the genotypes we added are closer to the genotype that shares the 5 allele. The reason this happens is because the naïve algorithm reports the rank of the concatenated alleles like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(xt &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">tab</span>(x), <span class="dv">1</span>, paste, <span class="dt">collapse =</span> <span class="st">""</span>))</code></pre></div>
<pre><code>##           A           B           C           D           E         new 
##    "100100"    "100010"    "010010"    "001010"    "001100" "010NANANA" 
##         mut 
##    "010001"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rank</span>(xt, <span class="dt">ties.method =</span> <span class="st">"first"</span>)</code></pre></div>
<pre><code>##   A   B   C   D   E new mut 
##   7   6   4   1   2   5   3</code></pre>
<p>Even though we have reason to believe that the samples “new” and “mut” might actually have the genotype x.5 and y.1 (the MLG of sample C), the algorithm sees them as different. If we calculate the pairwise euclidean distances between samples, we see that “new”, “mut” and, “C” are very similar to each other:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"phangorn"</span>)
<span class="kw">library</span>(<span class="st">"ape"</span>)
raw_dist &lt;-<span class="st"> </span><span class="cf">function</span>(x){
  <span class="kw">dist</span>(<span class="kw">genind2df</span>(x, <span class="dt">usepop =</span> <span class="ot">FALSE</span>))
}
(xdis &lt;-<span class="st"> </span><span class="kw">raw_dist</span>(x))</code></pre></div>
<pre><code>##            A        B        C        D        E      new
## B   3.000000                                             
## C   5.000000 4.000000                                    
## D   8.544004 8.000000 4.000000                           
## E   8.000000 8.544004 5.000000 3.000000                  
## new 5.656854 5.656854 0.000000 5.656854 5.656854         
## mut 4.472136 4.123106 1.000000 4.123106 4.472136 0.000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/plot.phylo">plot.phylo</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/phangorn/topics/upgma">upgma</a></span>(xdis))</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-12-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Therefore, we might want to consider “new”, “mut” and, “C” to be the same Multilocus Lineage (MLL). In the next section, you will see how to collapse multilocus genotypes by genetic distance.</p>
</div>
<div id="filtered-contracted" class="section level2">
<h2 class="hasAnchor">
<a href="#filtered-contracted" class="anchor"></a>Filtered (“contracted”)</h2>
<p>To remedy the issues with a naïve definition of multilocus genotypes, we can utilize genetic distance, which will allow us to collapse multilocus genotypes that are under a specific distance threshold. The main function we will use in this section is <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>. It will create a dissimiliarity distance matrix from the data and then filter based off of that matrix. You can also utilize your own supplied distance matrix or distance function.</p>
<div id="main-usage" class="section level3">
<h3 class="hasAnchor">
<a href="#main-usage" class="anchor"></a>Main usage</h3>
<p>To use <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>, you need to define a thresold. This threshold defines the minimum distance to consider two genotypes unique, so anything below that will be clustered into the same MLG. In practice, it’s used like this (using the data and the euclidean distance described above):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x <span class="co"># We have 7 MLGs before filtering</span></code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    7 original multilocus genotypes 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> xdis) &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">+</span><span class="st"> </span>.Machine<span class="op">$</span>double.eps<span class="op">^</span><span class="fl">0.5</span>
x <span class="co"># Now we have 5 MLGs</span></code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    5 contracted multilocus genotypes
##      (1) [t], (xdis) [d], (farthest) [a] 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x) &lt;-<span class="st"> "original"</span> <span class="co"># We'll reset to the naive definition</span></code></pre></div>
<p>Here, we told <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> to set the threshold to 1 plus a very tiny number (this will be explained in the next section) based off of the distance matrix <strong>xdis</strong>. This is a way to manipulate the object in place. In this section, there will be examples of retrieving the MLGs and other statistics from the function as well as how to use different distance functions and reset the MLGs.</p>
</div>
<div id="thresholds" class="section level3">
<h3 class="hasAnchor">
<a href="#thresholds" class="anchor"></a>Thresholds</h3>
<p>As we saw above, in order to collapse all three new samples into the original 5 MLGs, we had to set a threshold of <em>just above</em> 1. To show you why, we will use the function to return to us the MLGs it created and the thresholds that were passed to cluster MLLs (by using the <code>stats</code> argument).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x, <span class="st">"original"</span>)</code></pre></div>
<pre><code>## [1] 7 6 4 1 2 5 3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> xdis, <span class="dt">threshold =</span> <span class="dv">1</span>, <span class="dt">stats =</span> <span class="kw">c</span>(<span class="st">"mlg"</span>, <span class="st">"thresholds"</span>))</code></pre></div>
<pre><code>## $MLGS
## [1] 7 6 4 1 2 3 3
## 
## $THRESHOLDS
## [1] 0</code></pre>
<p>In our output, we can see that MLG 5 has collapsed into MLG 3. This occured as soon as our threshold passed 0. Take a look at what happens when we add a very small number to 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(e &lt;-<span class="st"> </span>.Machine<span class="op">$</span>double.eps<span class="op">^</span><span class="fl">0.5</span>) <span class="co"># A very tiny number</span></code></pre></div>
<pre><code>## [1] 1.490116e-08</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> xdis, <span class="dt">threshold =</span> <span class="dv">1</span> <span class="op">+</span><span class="st"> </span>e, <span class="dt">stats =</span> <span class="kw">c</span>(<span class="st">"mlg"</span>, <span class="st">"thresholds"</span>))</code></pre></div>
<pre><code>## $MLGS
## [1] 7 6 4 1 2 4 4
## 
## $THRESHOLDS
## [1] 0 1</code></pre>
<p>Now that we have set the threshold to <em>just above</em> 1, we have 5 unique MLLs since MLG 3 collapsed into MLG 4. Because of the way clustering happens, it’s important to point out what happens when you attempt to use a threshold of 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> xdis, <span class="dt">threshold =</span> <span class="dv">0</span>, <span class="dt">stats =</span> <span class="kw">c</span>(<span class="st">"mlg"</span>, <span class="st">"thresholds"</span>))</code></pre></div>
<pre><code>## $MLGS
## [1] 7 6 4 1 2 5 3
## 
## $THRESHOLDS
## numeric(0)</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x, <span class="st">"original"</span>)</code></pre></div>
<pre><code>## [1] 7 6 4 1 2 5 3</code></pre>
<p>Notice that all the MLGs are the same as the originally defined MLGs. Using a threshold of zero allows two MLGs separated by zero genetic distance to remain mutually unique.</p>
</div>
<div id="setting-the-genotypes" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-the-genotypes" class="anchor"></a>Setting the genotypes</h3>
<p>As demonstrated at the beginning of this section, if we wanted to reassign our sample genotypes to the collapsed version, all we would have to do is use the <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> commands above, but use the <code>&lt;-</code> operator to specify the threshold. Note that I will be using <code><a href="../reference/mlg.html">mlg.table()</a></code> to show the distribution of the multilocus genotypes before and after filtering.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    7 original multilocus genotypes 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.html">mlg.table</a></span>(x) <span class="co"># Before: 7 MLGs</span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;"></p>
<pre><code>##       MLG.1 MLG.2 MLG.3 MLG.4 MLG.5 MLG.6 MLG.7
## Total     1     1     1     1     1     1     1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> xdis) &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">+</span><span class="st"> </span>e
x</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    5 contracted multilocus genotypes
##      (1) [t], (xdis) [d], (farthest) [a] 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.html">mlg.table</a></span>(x) <span class="co"># After: 5 MLGs</span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-17-2.png" width="672" style="display: block; margin: auto;"></p>
<pre><code>##       MLG.1 MLG.2 MLG.4 MLG.6 MLG.7
## Total     1     1     3     1     1</code></pre>
<p>Notice how the information you see has changed. You can see that it’s printed out that you have 5 <strong>contracted</strong> multilocus genotypes, but you have some cryptic code underneath:</p>
<table class="table">
<thead><tr class="header">
<th align="center">symbol</th>
<th align="left">meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>[t]</strong></td>
<td align="left">threshold</td>
</tr>
<tr class="even">
<td align="center"><strong>[d]</strong></td>
<td align="left">distance (can be matrix or function)</td>
</tr>
<tr class="odd">
<td align="center"><strong>[a]</strong></td>
<td align="left">algorithm (see the next section)</td>
</tr>
</tbody>
</table>
<p>Genclone and snpclone objects will always remember what parameters were used for filtering multilocus genotypes, but the only catch is that, if you use your own supplied distance (matrix or function), you must be very careful not to delete it or change the object name.</p>
<p>This means that you don’t always have to specify the distance when assigning a threshold:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x) &lt;-<span class="st"> </span><span class="fl">4.51</span>
x</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    3 contracted multilocus genotypes
##      (4.51) [t], (xdis) [d], (farthest) [a] 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.html">mlg.table</a></span>(x)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;"></p>
<pre><code>##       MLG.1 MLG.4 MLG.6
## Total     2     3     2</code></pre>
<blockquote>
<p>DANGER! DANGER! While this is a convenient function, if you use a matrix or function that was created by you, you <strong>must not delete it or change its name</strong>. Only the name of the matrix/function is stored, so if you delete the matrix/function used to inform your filtering, you will get an error:</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rm</span>(xdis) <span class="co"># NOOOOOO!</span>
<span class="kw">try</span>(<span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x) &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">+</span><span class="st"> </span>e)</code></pre></div>
<pre><code>##  Error: cannot evaluate distance function, it might be missing.</code></pre>
<p>Basically, your object thinks that there should be a matrix called <strong>xdis</strong>, but it can’t find it anywhere. But don’t worry, we can restore it if we have the function available. We will use the <code>raw_dist()</code> function that we defined earlier.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> raw_dist) &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">+</span><span class="st"> </span>e
x</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    5 contracted multilocus genotypes
##      (1) [t], (raw_dist) [d], (farthest) [a] 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<p>The safest way, perhaps, is to use a function defined in <em>poppr</em>. For example, we’ll use Bruvo’s distance since it takes into account the real value of the alleles <span class="citation">(Bruvo et al. 2004)</span>.</p>
<blockquote>
<p>The arguments to your distance function will be stored in the object as well!</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bruvo.dist.html">bruvo.dist</a></span>(x, <span class="dt">replen =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</code></pre></div>
<pre><code>##             A         B         C         D         E       new
## B   0.4375000                                                  
## C   0.9062500 0.4687500                                        
## D   0.9355469 0.4980469 0.4687500                              
## E   0.4980469 0.9355469 0.9062500 0.4375000                    
## new 0.9375000 0.9375000 0.0000000 0.9375000 0.9375000          
## mut 0.8437500 0.7187500 0.2500000 0.7187500 0.8437500 0.0000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(x, <span class="dt">distance =</span> bruvo.dist, <span class="dt">replen =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)) &lt;-<span class="st"> </span><span class="fl">0.44</span>
x</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    3 contracted multilocus genotypes
##      (0.44) [t], (bruvo.dist) [d], (farthest) [a] 
##    7 haploid individuals
##    2 codominant loci
## 
## Population information:
## 
##    0 strata. 
##    0 populations defined.</code></pre>
<p>Of course, our multilocus genotypes are not changed forever, they are just stored in a different place. We can access the original, naïve multilocus genotypes by using the <code><a href="../reference/mll-method.html">mll()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x, <span class="st">"original"</span>)</code></pre></div>
<pre><code>## [1] 7 6 4 1 2 5 3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x) <span class="co"># contracted</span></code></pre></div>
<pre><code>## [1] 6 6 4 1 1 4 4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x) &lt;-<span class="st"> "original"</span>
<span class="kw"><a href="../reference/mll-method.html">mll</a></span>(x) <span class="co"># original</span></code></pre></div>
<pre><code>## [1] 7 6 4 1 2 5 3</code></pre>
</div>
<div id="tie-breakers-algorithms" class="section level3">
<h3 class="hasAnchor">
<a href="#tie-breakers-algorithms" class="anchor"></a>Tie breakers (algorithms)</h3>
<p>Underlying <code>mlg.filter</code> are three algorithms that decide what genotypes go together <span class="citation">(Kamvar, Brooks, and Grünwald 2015)</span>:</p>
<ul>
<li>nearest neighbor</li>
<li>farthest neighbor (default)</li>
<li>average neighbor (UPGMA)</li>
</ul>
<p>You can specify which algorithm you want to use in clustering with the <code>algorithm</code> argument. Each of these algorithms have different behaviors when it comes to collapsing multilocus genotypes. In short, farthest neighbor is the most conservative, nearest neighbor can have a chaining effect, and average neighbor is somewhere in between. Your choice of algorithm really depends on the biology of your organism.</p>
<p>To help visualize this, there is the function <code><a href="../reference/filter_stats.html">filter_stats()</a></code>, which will plot the output of the filtering algorithm. For simplicity, we will use the <code>Pinf</code> data set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(Pinf)
Pinf</code></pre></div>
<pre><code>## 
## This is a genclone object
## -------------------------
## Genotype information:
## 
##    72 multilocus genotypes 
##    86 tetraploid individuals
##    11 codominant loci
## 
## Population information:
## 
##     2 strata - Continent, Country
##     2 populations defined - South America, North America</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pinfreps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fix_replen.html">fix_replen</a></span>(Pinf, <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">6</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">2</span>))
pinf_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filter_stats.html">filter_stats</a></span>(Pinf, <span class="dt">distance =</span> bruvo.dist, <span class="dt">replen =</span> pinfreps, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can see that the different algorithms behave quite differently.</p>
</div>
</div>
<div id="custom-custom" class="section level2">
<h2 class="hasAnchor">
<a href="#custom-custom" class="anchor"></a>Custom (“custom”)</h2>
<p>Sometimes multilocus genotypes are defined by more than just genetic data and it would be useful to be able to analyze these genotype definitions. Luckily, there is a way to do this. <em>Poppr</em> has support for custom multilocus genotypes. For example, we have a simulated data set that has 50 samples with 26 multilocus genotypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(partial_clone)
pc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coercion-methods.html">as.genclone</a></span>(partial_clone)
<span class="kw"><a href="../reference/mll-method.html">mll</a></span>(pc)</code></pre></div>
<pre><code>##  [1]  8  9 23 24 22 21 10  3 22 11 24  9 25  4 12  2 14  1  9  9  9 26  7
## [24] 13 23  3 17 22  6 20 22 12  5 25 13 21 15 13 13 13  2 19 18 13 23 16
## [47]  1 11 25  4</code></pre>
<p>Let’s change the MLGs to letters instead of numbers. We will use <code><a href="../reference/mll.custom.html">mll.custom()</a></code> to do this.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">LETTERS[<span class="kw"><a href="../reference/mll-method.html">mll</a></span>(pc)]  <span class="co"># The new MLGs</span></code></pre></div>
<pre><code>##  [1] "H" "I" "W" "X" "V" "U" "J" "C" "V" "K" "X" "I" "Y" "D" "L" "B" "N"
## [18] "A" "I" "I" "I" "Z" "G" "M" "W" "C" "Q" "V" "F" "T" "V" "L" "E" "Y"
## [35] "M" "U" "O" "M" "M" "M" "B" "S" "R" "M" "W" "P" "A" "K" "Y" "D"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll.custom.html">mll.custom</a></span>(pc) &lt;-<span class="st"> </span>LETTERS[<span class="kw"><a href="../reference/mll-method.html">mll</a></span>(pc)]
<span class="kw"><a href="../reference/mlg.html">mlg.table</a></span>(pc)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;"></p>
<pre><code>##   A B C D E F G H I J K L M N O P Q R S T U V W X Y Z
## 1 0 1 0 0 1 1 0 1 1 0 0 0 0 1 1 0 0 0 0 0 0 2 2 0 2 0
## 2 1 0 1 2 0 0 0 0 1 0 1 0 1 0 0 1 0 0 1 1 1 0 0 0 1 1
## 3 1 0 0 0 0 0 1 0 1 1 0 1 2 0 0 0 1 1 0 0 0 1 1 1 0 0
## 4 0 1 1 0 0 0 0 0 2 0 1 1 3 0 0 0 0 0 0 0 1 1 0 1 0 0</code></pre>
<p>This even works with minimum spanning networks:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pcpal &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">"blue"</span>, <span class="st">"gold"</span>))
<span class="kw">set.seed</span>(<span class="dv">9001</span>)
pcmsn &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bruvo.msn.html">bruvo.msn</a></span>(pc, <span class="dt">replen =</span> <span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">nLoc</span>(pc)), <span class="dt">palette =</span> pcpal,
                   <span class="dt">vertex.label.color =</span> <span class="st">"firebrick"</span>, <span class="dt">vertex.label.font =</span> <span class="dv">2</span>,
                   <span class="dt">vertex.label.cex =</span> <span class="fl">1.5</span>)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Let’s say that we realized that we have strong evidence for MLG “Q” being the same as MLG “M”. If we wanted to change those, we could simply change the factor levels:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll.custom.html">mll.levels</a></span>(pc)[<span class="kw"><a href="../reference/mll.custom.html">mll.levels</a></span>(pc) <span class="op">==</span><span class="st"> "Q"</span>] &lt;-<span class="st"> "M"</span></code></pre></div>
<p>And we can plot again</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">9001</span>)
pcmsn &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bruvo.msn.html">bruvo.msn</a></span>(pc, <span class="dt">replen =</span> <span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">nLoc</span>(pc)), <span class="dt">palette =</span> pcpal,
                   <span class="dt">vertex.label.color =</span> <span class="st">"firebrick"</span>, <span class="dt">vertex.label.font =</span> <span class="dv">2</span>,
                   <span class="dt">vertex.label.cex =</span> <span class="fl">1.5</span>)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Notice how the minimum spanning network architecture stayed the same, but the labels had changed.</p>
</div>
</div>
<div id="diversity-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#diversity-analysis" class="anchor"></a>Diversity Analysis</h1>
<p>In the <strong>Data Import and Manipulation</strong> vignette, the first command demonstrated was the <code><a href="../reference/poppr.html">poppr()</a></code>, command, which produced a table of diversity statistics, most of which were calculated on counts of multilocus genotypes. These statistics can be replicated by using the function <code><a href="../reference/diversity_stats.html">diversity_stats()</a></code>. This function calculates diversity statistics based off of a table of multilocus genotypes. Let’s analyze the diversity of the peach brown rot pathogen, <em>Monilinia fructicola</em>, structured by Symptom (Fruit Rot:FR and Blossom Blight:BB) and Year <span class="citation">(Everhart and Scherm 2015)</span>. Let’s first look at the distribution of multilocus genotypes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(monpop)
<span class="kw">splitStrata</span>(monpop) &lt;-<span class="st"> </span><span class="er">~</span>Tree<span class="op">/</span>Year<span class="op">/</span>Symptom
montab &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mlg.html">mlg.table</a></span>(monpop, <span class="dt">strata =</span> <span class="op">~</span>Symptom<span class="op">/</span>Year)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-32-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can see from these plots that the Fruit Rot (FR) have a lot more clones in the canopies than Blossom Blight (BB). This makes biological sense since Fruit Rot infections are clonally propagated, whereas Blossom Blight infections are from sexual propagules <span class="citation">(Everhart and Scherm 2015)</span>. Let’s look at the diversity metrics.</p>
<div id="basic-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-statistics" class="anchor"></a>Basic Statistics</h2>
<p>The function <code><a href="../reference/diversity_stats.html">diversity_stats()</a></code> is used to get basic diversity statistics:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(monstat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/diversity_stats.html">diversity_stats</a></span>(montab))</code></pre></div>
<pre><code>##        Index
## Pop            H        G    lambda       E.5
##   BB_9  3.580258 25.92405 0.9614258 0.7145086
##   FR_9  3.103899 12.43711 0.9195955 0.5373402
##   BB_10 2.944439 19.00000 0.9473684 1.0000000
##   FR_10 4.066838 38.10132 0.9737542 0.6466786
##   BB_11 3.401197 30.00000 0.9666667 1.0000000
##   FR_11 3.094688 10.48941 0.9046657 0.4501543</code></pre>
</div>
<div id="confidence-intervals" class="section level2">
<h2 class="hasAnchor">
<a href="#confidence-intervals" class="anchor"></a>Confidence Intervals</h2>
<p>We can get confidence intervals for these statistics using <code><a href="../reference/diversity_ci.html">diversity_ci()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/diversity_ci.html">diversity_ci</a></span>(montab, <span class="dt">n =</span> 100L, <span class="dt">raw =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## 
## Confidence Intervals have been centered around observed statistic.
## Please see ?diversity_ci for details.</code></pre>
<p><img src="mlg_files/figure-html/unnamed-chunk-34-1.png" width="672" style="display: block; margin: auto;"></p>
<pre><code>##          H H.est           H.ci    G G.est             G.ci lambda
## BB_9  3.58  3.19 (3.341, 3.819) 25.9 19.26  (19.298, 32.55)  0.961
## FR_9  3.10  2.97 (2.912, 3.296) 12.4 11.98  (9.918, 14.956)  0.920
## BB_10 2.94  2.37 (2.664, 3.225) 19.0  9.81 (15.986, 22.014)  0.947
## FR_10 4.07  3.87 (3.926, 4.207) 38.1 33.63 (31.199, 45.004)  0.974
## BB_11 3.40  2.83 (3.173, 3.629) 30.0 15.28 (26.066, 33.934)  0.967
## FR_11 3.09  2.85 (2.802, 3.387) 10.5  9.81   (6.73, 14.249)  0.905
##       lambda.est      lambda.ci   E.5 E.5.est         E.5.ci
## BB_9       0.946 (0.942, 0.981) 0.715   0.776  (0.599, 0.83)
## FR_9       0.916 (0.902, 0.937) 0.537   0.593 (0.488, 0.587)
## BB_10      0.895 (0.913, 0.981) 1.000   0.894 (0.921, 1.079)
## FR_10      0.970  (0.968, 0.98) 0.647   0.691 (0.579, 0.714)
## BB_11      0.933 (0.948, 0.986) 1.000   0.889 (0.936, 1.064)
## FR_11      0.894 (0.861, 0.949) 0.450   0.533   (0.35, 0.55)</code></pre>
<p>You can see that there was a warning about centering the confidence interval. The boxplots you see are the distribution of the bootstrapped replicates, but they are known to be biased. We invite users to please read the documentation for <code><a href="../reference/diversity_ci.html">diversity_ci()</a></code> by typing <code><a href="../reference/diversity_ci.html">?diversity_ci</a></code>. It is very important to be careful when interpreting these results because sometimes confidence the confidence intervals exist outside of the possible range, as exemplified by BB_10 and BB_11.</p>
</div>
<div id="custom-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#custom-statistics" class="anchor"></a>Custom Statistics</h2>
<p>In all of the <code>diversity_*</code> functions, you can use your own custom diversity statistics. A common one we get requests for is the clonal fraction, <span class="math inline">\(\frac{MLG}{N}\)</span> or the number of multilocus genotypes over the number of samples. You can add it in by writing your own function. Keep in mind, that you should write it for both a matrix and a vector of counts if you want to be able to bootstrap it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myCF &lt;-<span class="st"> </span><span class="cf">function</span>(x){
 x &lt;-<span class="st"> </span><span class="kw">drop</span>(<span class="kw">as.matrix</span>(x))
 <span class="cf">if</span> (<span class="kw">length</span>(<span class="kw">dim</span>(x)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>){ <span class="co"># if it's a matrix</span>
   res &lt;-<span class="st"> </span><span class="kw">rowSums</span>(x <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)<span class="op">/</span><span class="kw">rowSums</span>(x)
 } <span class="cf">else</span> {                 <span class="co"># if it's a vector</span>
   res &lt;-<span class="st"> </span><span class="kw">sum</span>(x <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)<span class="op">/</span><span class="kw">sum</span>(x)
 }
 <span class="kw">return</span>(res)
}
(monstat2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/diversity_stats.html">diversity_stats</a></span>(montab, <span class="dt">CF =</span> myCF))</code></pre></div>
<pre><code>##        Index
## Pop            H        G    lambda       E.5        CF
##   BB_9  3.580258 25.92405 0.9614258 0.7145086 0.7031250
##   FR_9  3.103899 12.43711 0.9195955 0.5373402 0.2585366
##   BB_10 2.944439 19.00000 0.9473684 1.0000000 1.0000000
##   FR_10 4.066838 38.10132 0.9737542 0.6466786 0.3405018
##   BB_11 3.401197 30.00000 0.9666667 1.0000000 1.0000000
##   FR_11 3.094688 10.48941 0.9046657 0.4501543 0.4536082</code></pre>
<p>You can use filtered or custom MLGs to compare diversity. Here, I’m filtering genotypes in <code>monpop</code> that are different by only a single mutational step <span class="citation">(Bruvo et al. 2004)</span>.</p>
<blockquote>
<p>One mutational step for a single allele in Bruvo’s distance is equivalent to 0.5, so a single mutational step for a haploid organism with 13 loci is 0.5/13.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Repeat lengths are necessary</span>
reps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fix_replen.html">fix_replen</a></span>(monpop,
                   <span class="kw">c</span>(<span class="dt">CHMFc4 =</span> <span class="dv">7</span>, <span class="dt">CHMFc5 =</span> <span class="dv">2</span>, <span class="dt">CHMFc12 =</span> <span class="dv">4</span>,
                     <span class="dt">SEA =</span> <span class="dv">4</span>, <span class="dt">SED =</span> <span class="dv">4</span>, <span class="dt">SEE =</span> <span class="dv">2</span>, <span class="dt">SEG =</span> <span class="dv">6</span>,
                     <span class="dt">SEI =</span> <span class="dv">3</span>, <span class="dt">SEL =</span> <span class="dv">4</span>, <span class="dt">SEN =</span> <span class="dv">2</span>,
                     <span class="dt">SEP =</span> <span class="dv">4</span>, <span class="dt">SEQ =</span> <span class="dv">2</span>, <span class="dt">SER =</span> <span class="dv">4</span>))

<span class="co"># Adding a little bit, so the threshold is included.</span>
e &lt;-<span class="st"> </span>.Machine<span class="op">$</span>double.eps<span class="op">^</span><span class="fl">0.5</span>

<span class="co"># Using the default farthest neighbor algorithm to collapse genotypes</span>
<span class="kw"><a href="../reference/mlg.filter.html">mlg.filter</a></span>(monpop, <span class="dt">distance =</span> bruvo.dist, <span class="dt">replen =</span> reps) &lt;-<span class="st"> </span>(<span class="fl">0.5</span><span class="op">/</span><span class="dv">13</span>) <span class="op">+</span><span class="st"> </span>e
montabf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mlg.html">mlg.table</a></span>(monpop, <span class="dt">strata =</span> <span class="op">~</span>Symptom<span class="op">/</span>Year)</code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-37-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(monstatf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/diversity_stats.html">diversity_stats</a></span>(montabf, <span class="dt">CF =</span> myCF))</code></pre></div>
<pre><code>##        Index
## Pop            H        G    lambda       E.5        CF
##   BB_9  3.442117 23.81395 0.9580078 0.7541044 0.6093750
##   FR_9  3.003859 11.99001 0.9165973 0.5734961 0.2195122
##   BB_10 2.944439 19.00000 0.9473684 1.0000000 1.0000000
##   FR_10 3.861202 33.10974 0.9697974 0.6901987 0.2724014
##   BB_11 3.401197 30.00000 0.9666667 1.0000000 1.0000000
##   FR_11 2.754565  8.00766 0.8751196 0.4762511 0.3195876</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">monstat2 <span class="op">-</span><span class="st"> </span>monstatf <span class="co"># Take the difference from the unfiltered</span></code></pre></div>
<pre><code>##        Index
## Pop             H         G      lambda         E.5         CF
##   BB_9  0.1381408 2.1100971 0.003417969 -0.03959575 0.09375000
##   FR_9  0.1000406 0.4470973 0.002998215 -0.03615591 0.03902439
##   BB_10 0.0000000 0.0000000 0.000000000  0.00000000 0.00000000
##   FR_10 0.2056360 4.9915810 0.003956784 -0.04352013 0.06810036
##   BB_11 0.0000000 0.0000000 0.000000000  0.00000000 0.00000000
##   FR_11 0.3401226 2.4817496 0.029546179 -0.02609686 0.13402062</code></pre>
<p>We can see that filtered MLLs tend to be less diverse. This makes intuitive sense as it is creating larger classes of multilocus genotypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop) &lt;-<span class="st"> "original"</span></code></pre></div>
</div>
<div id="jack-knife-rarefaction" class="section level2">
<h2 class="hasAnchor">
<a href="#jack-knife-rarefaction" class="anchor"></a>Jack-knife rarefaction</h2>
<p>The function <code><a href="../reference/diversity_ci.html">diversity_ci()</a></code> and <code><a href="../reference/diversity_boot.html">diversity_boot()</a></code> have the option to perform jack knife rarefaction calculations. This means that your data will be randomly sub-sampled to either the smallest population size, or whatever is specified in the parameter <code>n.rare</code>, whichever is bigger. Here’s an example with the previous data set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(monrare &lt;-<span class="st"> </span><span class="kw"><a href="../reference/diversity_ci.html">diversity_ci</a></span>(montab, <span class="dt">n =</span> 100L, <span class="dt">rarefy =</span> <span class="ot">TRUE</span>, <span class="dt">raw =</span> <span class="ot">FALSE</span>))</code></pre></div>
<pre><code>## 
## Samples for rarefaction: 19</code></pre>
<p><img src="mlg_files/figure-html/unnamed-chunk-40-1.png" width="672" style="display: block; margin: auto;"></p>
<pre><code>##          H H.est           H.ci    G G.est            G.ci lambda
## BB_9  3.58  2.70 (2.465, 2.871) 25.9 13.70  (9.757, 17.19)  0.961
## FR_9  3.10  2.27 (1.934, 2.625) 12.4  8.35 (5.819, 12.448)  0.920
## BB_10 2.94  2.94           &lt;NA&gt; 19.0 19.00            &lt;NA&gt;  0.947
## FR_10 4.07  2.69 (2.406, 2.871) 38.1 13.91 (10.022, 17.19)  0.974
## BB_11 3.40  2.94           &lt;NA&gt; 30.0 19.00            &lt;NA&gt;  0.967
## FR_11 3.09  2.30 (1.826, 2.698) 10.5  8.12  (4.009, 13.37)  0.905
##       lambda.est      lambda.ci   E.5 E.5.est         E.5.ci
## BB_9       0.925 (0.898, 0.942) 0.715   0.906 (0.801, 0.972)
## FR_9       0.875  (0.828, 0.92) 0.537   0.825 (0.698, 0.941)
## BB_10      0.947           &lt;NA&gt; 1.000   1.000           &lt;NA&gt;
## FR_10      0.926   (0.9, 0.942) 0.647   0.922 (0.817, 0.972)
## BB_11      0.947           &lt;NA&gt; 1.000   1.000           &lt;NA&gt;
## FR_11      0.865  (0.75, 0.925) 0.450   0.759 (0.565, 0.896)</code></pre>
<p>This can give you comparable estimates of diversity when not all samples are of equal size.</p>
</div>
</div>
<div id="clone-correction" class="section level1">
<h1 class="hasAnchor">
<a href="#clone-correction" class="anchor"></a>Clone Correction</h1>
<p>Clone-correction works hierarchically and only uses the first MLG copy encountered in the data per population. This is straightforward for naïve MLGs, but for MLLs collapsed by genetic distance or custom MLLs, this might change the results very slightly. As an example, let’s look at the <code>monpop</code> filtered MLLs that we created earlier.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">nmll</a></span>(monpop, <span class="st">"original"</span>)</code></pre></div>
<pre><code>## [1] 264</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">nmll</a></span>(monpop, <span class="st">"contracted"</span>)</code></pre></div>
<pre><code>## [1] 214</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mll-method.html">mll</a></span>(monpop) &lt;-<span class="st"> "contracted"</span></code></pre></div>
<p>To show how the order of the samples can affect the sampling, we will take the sum of all pairwise distances between clone-corrected samples (corrected without respect to populations):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">monpop <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/clonecorrect.html">clonecorrect</a></span>(<span class="dt">strata =</span> <span class="ot">NA</span>) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># 1. clone correct whole data set</span>
<span class="st">  </span><span class="kw">dist</span>() <span class="op">%&gt;%</span><span class="st">                     </span><span class="co"># 2. calculate distance</span>
<span class="st">  </span><span class="kw">sum</span>()                          <span class="co"># 3. take the sum of the distance</span></code></pre></div>
<pre><code>## [1] 90268.73</code></pre>
<p>Now, what happens when we randomly sample individuals?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">999</span>)
monpop[<span class="kw">sample</span>(<span class="kw">nInd</span>(monpop))] <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># 1. shuffle samples</span>
<span class="st">  </span><span class="kw"><a href="../reference/clonecorrect.html">clonecorrect</a></span>(<span class="dt">strata =</span> <span class="ot">NA</span>) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># 2. clone correct whole data set</span>
<span class="st">  </span><span class="kw">dist</span>() <span class="op">%&gt;%</span><span class="st">                     </span><span class="co"># 3. calculate distance</span>
<span class="st">  </span><span class="kw">sum</span>()                          <span class="co"># 4. take the sum of the distance</span></code></pre></div>
<pre><code>## [1] 90359.52</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1000</span>)
monpop[<span class="kw">sample</span>(<span class="kw">nInd</span>(monpop))] <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># 1. shuffle samples</span>
<span class="st">  </span><span class="kw"><a href="../reference/clonecorrect.html">clonecorrect</a></span>(<span class="dt">strata =</span> <span class="ot">NA</span>) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># 2. clone correct whole data set</span>
<span class="st">  </span><span class="kw">dist</span>() <span class="op">%&gt;%</span><span class="st">                     </span><span class="co"># 3. calculate distance</span>
<span class="st">  </span><span class="kw">sum</span>()                          <span class="co"># 4. take the sum of the distance</span></code></pre></div>
<pre><code>## [1] 90324.07</code></pre>
<p>Notice how we are getting different results based on the order of samples. This does not mean that the procedure doesn’t work, it just means that we must be careful when clone-correcting modified multilocus genotypes.</p>
</div>
<div id="conclusions" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusions" class="anchor"></a>Conclusions</h1>
<p>We have demonstrated here new methods for treating multilocus genotypes on microsatellite data, but it is important to remember that all of these functions can work with any source of data stored in either a <strong>genclone</strong> or <strong>snpclone</strong> (derived from <strong>genlight</strong>) object. Especially with <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>, these new functions will allow for a more flexible analysis of WGS data of clonal organisms where true clones may differ by more than a few mutations/errors/missing data. As with all analyses, it is important to understand the algorithms used and take them into account when interpreting results.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Bruvo:2004">
<p>Bruvo, Ruzica, Nicolaas K. Michiels, Thomas G. D’Souza, and Hinrich Schulenburg. 2004. “A Simple Method for the Calculation of Microsatellite Genotype Distances Irrespective of Ploidy Level.” <em>Molecular Ecology</em> 13 (7). Blackwell Science Ltd: 2101–6. doi:<a href="https://doi.org/10.1111/j.1365-294X.2004.02209.x">10.1111/j.1365-294X.2004.02209.x</a>.</p>
</div>
<div id="ref-everhart2014finescale">
<p>Everhart, Sydney, and Harald Scherm. 2015. “Fine-Scale Genetic Structure of <em>Monilinia Fructicola</em> During Brown Rot Epidemics Within Individual Peach Tree Canopies.” <em>Phytopathology</em> 105 (4): 542–49. doi:<a href="https://doi.org/10.1094/PHYTO-03-14-0088-R">10.1094/PHYTO-03-14-0088-R</a>.</p>
</div>
<div id="ref-kamvar2015novel">
<p>Kamvar, Zhian N, Jonah C Brooks, and Niklaus J Grünwald. 2015. “Novel R tools for analysis of genome-wide population genetic data with emphasis on clonality.” <em>Frontiers in Genetics</em> 6: 208. doi:<a href="https://doi.org/10.3389/fgene.2015.00208">10.3389/fgene.2015.00208</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#multilocus-genotype-flavors">Multilocus Genotype Flavors</a><ul class="nav nav-pills nav-stacked">
<li><a href="#naive-original">Naïve (“original”)</a></li>
      <li><a href="#filtered-contracted">Filtered (“contracted”)</a></li>
      <li><a href="#custom-custom">Custom (“custom”)</a></li>
      </ul>
</li>
      <li>
<a href="#diversity-analysis">Diversity Analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#basic-statistics">Basic Statistics</a></li>
      <li><a href="#confidence-intervals">Confidence Intervals</a></li>
      <li><a href="#custom-statistics">Custom Statistics</a></li>
      <li><a href="#jack-knife-rarefaction">Jack-knife rarefaction</a></li>
      </ul>
</li>
      <li><a href="#clone-correction">Clone Correction</a></li>
      <li><a href="#conclusions">Conclusions</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
