<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate dissimilarity or Euclidean distance for genlight objects — bitwise.dist • poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Calculate dissimilarity or Euclidean distance for genlight objects — bitwise.dist" />

<meta property="og:description" content="This function calculates both dissimilarity and Euclidean distances for
genlight or snpclone objects." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate dissimilarity or Euclidean distance for genlight objects</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/bitwise.r'><code>R/bitwise.r</code></a></small>
    <div class="hidden name"><code>bitwise.dist.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function calculates both dissimilarity and Euclidean distances for
genlight or <a href='genclone-class.html'>snpclone</a> objects.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bitwise.dist</span>(<span class='no'>x</span>, <span class='kw'>percent</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>mat</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>missing_match</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>scale_missing</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>euclidean</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>differences_only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>0L</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a genlight or <a href='genclone-class.html'>snpclone</a> object.</p></td>
    </tr>
    <tr>
      <th>percent</th>
      <td><p><code>logical</code>. Should the distance be represented from 0 to
1? Default set to <code>TRUE</code>. <code>FALSE</code> will return the distance
represented as integers from 1 to n where n is the number of loci.
This option has no effect if <code>euclidean = TRUE</code></p></td>
    </tr>
    <tr>
      <th>mat</th>
      <td><p><code>logical</code>. Return a matrix object. Default set to
<code>FALSE</code>, returning a dist object. <code>TRUE</code> returns a matrix object.</p></td>
    </tr>
    <tr>
      <th>missing_match</th>
      <td><p><code>logical</code>. Determines whether two samples differing
by missing data in a location should be counted as matching at that
location. Default set to <code>TRUE</code>, which forces missing data to match
with anything. <code>FALSE</code> forces missing data to not match with any other
information, <strong>including other missing data</strong>.</p></td>
    </tr>
    <tr>
      <th>scale_missing</th>
      <td><p>A logical. If <code>TRUE</code>, comparisons with missing
data is scaled up proportionally to the number of columns used by
multiplying the value by <code>m / (m - x)</code> where m is the number of
loci and x is the number of missing sites. This option matches the behavior
of base R's <code><a href='https://www.rdocumentation.org/packages/stats/topics/dist'>dist()</a></code> function.
Defaults to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>euclidean</th>
      <td><p><code>logical</code>. if <code>TRUE</code>, the Euclidean distance will
be calculated.</p></td>
    </tr>
    <tr>
      <th>differences_only</th>
      <td><p><code>logical</code>. When <code>differences_only = TRUE</code>,
the output will reflect the number of different loci. The default setting,
<code>differences_only = FALSE</code>, reflects the number of different alleles.
Note: this has no effect on haploid organisms since 1 locus = 1 allele.
This option is NOT recommended.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>The maximum number of parallel threads to be used within this
function. A value of 0 (default) will attempt to use as many threads as
there are available cores/CPUs. In most cases this is ideal. A value of 1
will force the function to run serially, which may increase stability on
some systems. Other values may be specified, but should be used with
caution.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dist object containing pairwise distances between samples.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The default distance calculated here is quite simple and goes by
many names depending on its application. The most familiar name might be
the Hamming distance, or the number of differences between two strings.</p>
<p>As of poppr version 2.8.0, this function now also calculates Euclidean
distance and is considerably faster and more memory-efficient than the
standard <code><a href='https://www.rdocumentation.org/packages/stats/topics/dist'>dist()</a></code> function.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>This function is optimized for genlight and
<a href='genclone-class.html'>snpclone</a> objects. This does not mean that it is a
catch-all optimization for SNP data. Three assumptions must be met for this
function to work:</p><ol>
<li><p>SNPs are bi-allelic</p></li>
<li><p>Samples are haploid or diploid</p></li>
<li><p>All samples have the same ploidy</p></li>
</ol>
    <p>If the user supplies a genind or
<a href='genclone-class.html'>genclone</a> object, <code><a href='genetic_distance.html'>prevosti.dist()</a></code> will be used for
calculation.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='diss.dist.html'>diss.dist()</a></code>, <a href='genclone-class.html'>snpclone</a>,
genlight, <code><a href='win.ia.html'>win.ia()</a></code>,  <code><a href='samp.ia.html'>samp.ia()</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>999</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>glSim</span>(<span class='kw'>n.ind</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>n.snp.nonstruc</span> <span class='kw'>=</span> <span class='fl'>5e2</span>, <span class='kw'>n.snp.struc</span> <span class='kw'>=</span> <span class='fl'>5e2</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>)
<span class='no'>x</span></div><div class='output co'>#&gt;  /// GENLIGHT OBJECT /////////
#&gt; 
#&gt;  // 10 genotypes,  1,000 binary SNPs, size: 20.6 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  // Basic content
#&gt;    @gen: list of 10 SNPbin
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 3-7)
#&gt;    @other: a list containing: ancestral.pops 
#&gt; </div><div class='input'><span class='co'># Assess fraction of different alleles</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.time'>system.time</a></span>(<span class='no'>xd</span> <span class='kw'>&lt;-</span> <span class='fu'>bitwise.dist</span>(<span class='no'>x</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>))</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.002   0.000   0.002 </div><div class='input'><span class='no'>xd</span></div><div class='output co'>#&gt;         1      2      3      4      5      6      7      8      9
#&gt; 2  0.2105                                                        
#&gt; 3  0.2195 0.2180                                                 
#&gt; 4  0.3950 0.3825 0.3855                                          
#&gt; 5  0.4075 0.3910 0.3960 0.2175                                   
#&gt; 6  0.3920 0.3855 0.3985 0.2230 0.2185                            
#&gt; 7  0.3955 0.3920 0.3910 0.2265 0.2140 0.2385                     
#&gt; 8  0.3955 0.3830 0.3970 0.2245 0.1980 0.2255 0.2190              
#&gt; 9  0.3770 0.3805 0.3945 0.2200 0.2105 0.2210 0.2215 0.2155       
#&gt; 10 0.3980 0.3925 0.4005 0.2180 0.2035 0.2220 0.2185 0.2125 0.2140</div><div class='input'>
<span class='co'># Calculate Euclidean distance</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.time'>system.time</a></span>(<span class='no'>xdt</span> <span class='kw'>&lt;-</span> <span class='fu'>bitwise.dist</span>(<span class='no'>x</span>, <span class='kw'>euclidean</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>scale_missing</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>))</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.001   0.000   0.002 </div><div class='input'><span class='no'>xdt</span></div><div class='output co'>#&gt;           1        2        3        4        5        6        7        8
#&gt; 2  22.64950                                                               
#&gt; 3  23.00000 23.19483                                                      
#&gt; 4  35.24202 34.68429 34.91418                                             
#&gt; 5  35.98611 35.27038 35.88872 23.43075                                    
#&gt; 6  34.98571 35.14257 35.59494 23.87467 23.72762                           
#&gt; 7  35.31289 35.01428 35.27038 24.02082 23.49468 24.59675                  
#&gt; 8  35.05710 34.46738 35.55278 23.60085 22.27106 23.55844 23.49468         
#&gt; 9  34.32200 34.71311 35.22783 23.62202 22.82542 23.45208 23.55844 23.25941
#&gt; 10 35.49648 34.85685 35.90265 23.70654 22.95648 23.57965 23.51595 23.21637
#&gt;           9
#&gt; 2          
#&gt; 3          
#&gt; 4          
#&gt; 5          
#&gt; 6          
#&gt; 7          
#&gt; 8          
#&gt; 9          
#&gt; 10 23.10844</div><div class='input'>
</div><div class='input'>
<span class='co'># This function is more efficient in both memory and speed than [dist()] for</span>
<span class='co'># calculating Euclidean distance on genlight objects. For example, we can</span>
<span class='co'># observe a clear speed increase when we attempt a calculation on 100k SNPs</span>
<span class='co'># with 10% missing data:</span>

<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>999</span>)
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/sample'>sample</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>0</span>:<span class='fl'>2</span>, <span class='fl'>NA</span>),
                     <span class='fl'>100000</span> * <span class='fl'>50</span>,
                     <span class='kw'>replace</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
                     <span class='kw'>prob</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>0.3</span>, <span class='fl'>0.3</span>, <span class='fl'>0.3</span>, <span class='fl'>0.1</span>)),
              <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>50</span>)
<span class='no'>glite</span> <span class='kw'>&lt;-</span> <span class='fu'>new</span>(<span class='st'>"genlight"</span>, <span class='no'>mat</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>)

<span class='co'># Default Euclidean distance </span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.time'>system.time</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/dist'>dist</a></span>(<span class='no'>glite</span>))</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   2.426   0.080   2.551 </div><div class='input'>
<span class='co'># Bitwise dist</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.time'>system.time</a></span>(<span class='fu'>bitwise.dist</span>(<span class='no'>glite</span>, <span class='kw'>euclidean</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>scale_missing</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>))</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.904   0.004   0.966 </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Zhian N. Kamvar, Jonah C. Brooks</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

