<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate a dissimilarity distance matrix for SNP data. — bitwise.dist • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate a dissimilarity distance matrix for SNP data.</h1>
    </div>

    
    <p>This function performs the same task as <code><a href='diss.dist.html'>diss.dist</a></code>, calculating 
the fraction or number of different alleles between two genlight or snpclone
objects.</p>
    

    <pre class="usage"><span class='fu'>bitwise.dist</span>(<span class='no'>x</span>, <span class='kw'>percent</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>mat</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>missing_match</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>differences_only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>0</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a <code>genlight</code>, <code>genind</code>,
<code><a href='genclone-class.html'>genclone</a></code>, or <code><a href='genclone-class.html'>snpclone</a></code> object.</p></td>
    </tr>
    <tr>
      <th>percent</th>
      <td><p><code>logical</code>. Should the distance be represented from 0 to 
1? Default set to <code>TRUE</code>. <code>FALSE</code> will return the distance 
represented as integers from 1 to n where n is the number of loci.</p></td>
    </tr>
    <tr>
      <th>mat</th>
      <td><p><code>logical</code>. Return a matrix object. Default set to 
<code>FALSE</code>, returning a dist object. <code>TRUE</code> returns a matrix object.</p></td>
    </tr>
    <tr>
      <th>missing_match</th>
      <td><p><code>logical</code>. Determines whether two samples differing
by missing data in a location should be counted as matching at that 
location. Default set to <code>TRUE</code>, which forces missing data to match 
with anything. <code>FALSE</code> forces missing data to not match with any other
information, <strong>including other missing data</strong>.</p></td>
    </tr>
    <tr>
      <th>differences_only</th>
      <td><p><code>logical</code>. When <code>differences_only = TRUE</code>,
the output will reflect the number of different loci. The default setting,
<code>differences_only = FALSE</code>, reflects the number of different alleles.
Note: this has no effect on haploid organisms since 1 locus = 1 allele.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>The maximum number of parallel threads to be used within this 
function. A value of 0 (default) will attempt to use as many threads as 
there are available cores/CPUs. In most cases this is ideal. A value of 1 
will force the function to run serially, which may increase stability on 
some systems. Other values may be specified, but should be used with 
caution.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dist object containing pairwise distances between samples.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The distance calculated here is quite simple and goes by many names,
  depending on its application. The most familiar name might be the Hamming 
  distance, or the number of differences between two strings.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>If the user supplies a <code>genind</code> or <code>genclone</code> object,
  <code><a href='genetic_distance.html'>prevosti.dist</a></code> will be used for calculation.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='diss.dist.html'>diss.dist</a></code>,
   <code><a href='genclone-class.html'>snpclone</a></code>,
   <code><a href='http://www.rdocumentation.org/packages/adegenet/topics/genlight'>genlight</a></code>,
   <code><a href='win.ia.html'>win.ia</a></code>, 
   <code><a href='samp.ia.html'>samp.ia</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>set.seed</span>(<span class='fl'>999</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>glSim</span>(<span class='kw'>n.ind</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>n.snp.nonstruc</span> <span class='kw'>=</span> <span class='fl'>5e2</span>, <span class='kw'>n.snp.struc</span> <span class='kw'>=</span> <span class='fl'>5e2</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>)
<span class='no'>x</span></div><div class='output co'>#&gt;  /// GENLIGHT OBJECT /////////
#&gt; 
#&gt;  // 10 genotypes,  1,000 binary SNPs, size: 19.6 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  // Basic content
#&gt;    @gen: list of 10 SNPbin
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 3-7)
#&gt;    @other: a list containing: ancestral.pops 
#&gt; </div><div class='input'><span class='co'># Assess fraction of different alleles (finer measure, usually the most sensible)</span>
<span class='fu'>system.time</span>(<span class='no'>xd</span> <span class='kw'>&lt;-</span> <span class='fu'>bitwise.dist</span>(<span class='no'>x</span>))</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.016   0.014   0.030 </div><div class='input'><span class='no'>xd</span></div><div class='output co'>#&gt;         1      2      3      4      5      6      7      8      9
#&gt; 2  0.2105                                                        
#&gt; 3  0.2195 0.2180                                                 
#&gt; 4  0.3950 0.3825 0.3855                                          
#&gt; 5  0.4075 0.3910 0.3960 0.2175                                   
#&gt; 6  0.3920 0.3855 0.3985 0.2230 0.2185                            
#&gt; 7  0.3955 0.3920 0.3910 0.2265 0.2140 0.2385                     
#&gt; 8  0.3955 0.3830 0.3970 0.2245 0.1980 0.2255 0.2190              
#&gt; 9  0.3770 0.3805 0.3945 0.2200 0.2105 0.2210 0.2215 0.2155       
#&gt; 10 0.3980 0.3925 0.4005 0.2180 0.2035 0.2220 0.2185 0.2125 0.2140</div><div class='input'>
<span class='co'># Assess fraction of different loci (coarse measure)</span>
<span class='fu'>system.time</span>(<span class='no'>xdt</span> <span class='kw'>&lt;-</span> <span class='fu'>bitwise.dist</span>(<span class='no'>x</span>, <span class='kw'>differences_only</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>))</div><div class='output co'>#&gt;    user  system elapsed 
#&gt;   0.054   0.002   0.056 </div><div class='input'><span class='no'>xdt</span></div><div class='output co'>#&gt;        1     2     3     4     5     6     7     8     9
#&gt; 2  0.375                                                
#&gt; 3  0.394 0.385                                          
#&gt; 4  0.564 0.546 0.547                                    
#&gt; 5  0.575 0.551 0.544 0.378                              
#&gt; 6  0.564 0.539 0.562 0.384 0.374                        
#&gt; 7  0.563 0.563 0.551 0.391 0.366 0.413                  
#&gt; 8  0.572 0.555 0.559 0.395 0.346 0.399 0.381            
#&gt; 9  0.542 0.539 0.563 0.381 0.371 0.388 0.387 0.376      
#&gt; 10 0.564 0.570 0.557 0.373 0.347 0.388 0.379 0.368 0.375</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar, Jonah C. Brooks

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
