<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Bruvo's distance for microsatellites — bruvo.dist • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Bruvo's distance for microsatellites</h1>
    </div>

    
    <p>Calculate the average Bruvo's distance over all loci in a population.</p>
    

    <pre class="usage"><span class='fu'>bruvo.dist</span>(<span class='no'>pop</span>, <span class='kw'>replen</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>loss</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>by_locus</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pop</th>
      <td><p>a <code>genind</code> or <code><a href='genclone-class.html'>genclone</a></code> object</p></td>
    </tr>
    <tr>
      <th>replen</th>
      <td><p>a <code>vector</code> of <code>integers</code> indicating the length of the
nucleotide repeats for each microsatellite locus. E.g. a locus with a (CAT) 
repeat would have a replen value of 3. (Also see <code><a href='fix_replen.html'>fix_replen</a></code>)</p></td>
    </tr>
    <tr>
      <th>add</th>
      <td><p>if <code>TRUE</code>, genotypes with zero values will be treated under 
the genome addition model presented in Bruvo et al. 2004. See the
<strong>Note</strong> section for options.</p></td>
    </tr>
    <tr>
      <th>loss</th>
      <td><p>if <code>TRUE</code>, genotypes with zero values will be treated under 
the genome loss model presented in Bruvo et al. 2004. See the
<strong>Note</strong> section for options.</p></td>
    </tr>
    <tr>
      <th>by_locus</th>
      <td><p>indicator to get the results per locus. The default setting
is <code>by_locus = FALSE</code>, indicating that Bruvo's distance is to be
averaged over all loci. When <code>by_locus = TRUE</code>, a list of distance
matrices will be returned.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>an object of class <code>dist</code> or a list of these objects if
  <code>by_locus = TRUE</code></p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Bruvo's distance between two alleles is calculated as 
  $$d = 1 - 2^{-\mid x \mid}$$, where <strong>x</strong>
  is the number of repeat units between the two alleles (see the Algorithms 
  and Equations vignette for more details). These distances are calculated 
  over all combinations of alleles at a locus and then the minimum average
  distance between allele combinations is taken as the distance for that 
  locus. All loci are then averaged over to obtain the distance between two
  samples. Missing data is ignored (in the same fashion as 
  <code>mean(c(1:9, NA), na.rm = TRUE)</code>) if all alleles are missing. See the
  next section for other cases.</p><h3>Polyploids</h3>
    Ploidy is irrelevant with respect to calculation of Bruvo's 
  distance. However, since it makes a comparison between all alleles at a 
  locus, it only makes sense that the two loci need to have the same ploidy 
  level. Unfortunately for polyploids, it's often difficult to fully separate
  distinct alleles at each locus, so you end up with genotypes that appear to
  have a lower ploidy level than the organism.
    To help deal with these situations, Bruvo has suggested three methods for
  dealing with these differences in ploidy levels: <ul>
<li><p><strong>Infinite Model</strong> - The simplest way to deal with it is to count all
  missing alleles as infinitely large so that the distance between it and
  anything else is 1. Aside from this being computationally simple, it will
  tend to <strong>inflate distances between individuals</strong>.</p></li>
<li><p><strong>Genome Addition Model</strong> - If it is suspected that the organism has
  gone through a recent genome expansion, <strong>the missing alleles will be
  replace with all possible combinations of the observed alleles in the
  shorter genotype</strong>. For example, if there is a genotype of [69, 70, 0, 0]
  where 0 is a missing allele, the possible combinations are: [69, 70, 69,
  69], [69, 70, 69, 70], [69, 70, 70, 69], and [69, 70, 70, 70]. The
  resulting distances are then averaged over the number of comparisons.</p></li>
<li><p><strong>Genome Loss Model</strong> - This is similar to the genome addition model,
  except that it assumes that there was a recent genome reduction event and
  uses <strong>the observed values in the full genotype to fill the missing
  values in the short genotype</strong>. As with the Genome Addition Model, the
  resulting distances are averaged over the number of comparisons.</p></li>
<li><p><strong>Combination Model</strong> - Combine and average the genome addition and
  loss models.</p></li>
</ul>
    As mentioned above, the infinite model is biased, but it is not nearly as
  computationally intensive as either of the other models. The reason for
  this is that both of the addition and loss models requires replacement of
  alleles and recalculation of Bruvo's distance. The number of replacements
  required is equal to n^k where where <em>n</em> is the number of potential
  replacements and <em>k</em> is the number of alleles to be replaced.
  To reduce the number of calculations and assumptions otherwise, Bruvo's 
  distance will be calculated using the largest observed ploidy in pairwise 
  comparisons. This means that when comparing [69,70,71,0] and [59,60,0,0], 
  they will be treated as triploids.
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Do not use missingno with this function.</p><h3>Missing alleles and Bruvo's distance in <span class="pkg">poppr</span> versions &lt; 2.5</h3>
    In earlier versions of <span class="pkg">poppr</span>, the authors had assumed that, because
  the calculation of Bruvo's distance does not rely on orderd sets of
  alleles, the imputation methods in the genome addition and genome loss
  models would also assume unordered alleles for creating the hypothetical
  genotypes. This means that the results from this imputation did not
  consider all possible combinations of alleles, resulting in either an over-
  or under- estimation of Bruvo's distance between two samples with two or
  more missing alleles. This version of <span class="pkg">poppr</span> considers all possible
  combinations when calculating Bruvo's distance for incomplete genotype with
  a negligable gain in computation time.
    If you want to see the effect of this change on your data, you can use the
  global <span class="pkg">poppr</span> option <code>old.bruvo.model</code>. Currently, this option is
  <code>FALSE</code> and you can set it by using 
  <code>options(old.bruvo.model = TRUE)</code>, but make sure to reset it to 
  <code>FALSE</code> afterwards.
  <p></p><h3>Repeat Lengths (replen)</h3>
    The <code>replen</code> argument is crucial for proper analysis of Bruvo's
  distance since the calculation relies on the knowledge of the number of
  steps between alleles. To calculate Bruvo's distance, your raw allele calls
  are first divided by the repeat lengths and then rounded. This can create a
  problem with repeat lengths of even size due to the IEC 60559 standard that
  says rounding at 0.5 is to the nearest even number, meaning that it is
  possible for two alleles that are one step apart may appear to be exactly
  the same. This can be fixed by subtracting a tiny number from the repeat
  length with the function <code><a href='fix_replen.html'>fix_replen</a></code>. Please consider using
  this before running Bruvo's distance.
  <p></p><h3>Model Choice</h3>
 The <code>add</code> and <code>loss</code> arguments 
  modify the model choice accordingly: <ul>
<li><p><strong>Infinite 
  Model:</strong>  <code>add = FALSE, loss = FALSE</code></p></li>
<li><p><strong>Genome Addition 
  Model:</strong>  <code>add = TRUE, loss = FALSE</code></p></li>
<li><p><strong>Genome Loss Model:</strong> 
  <code>add = FALSE, loss = TRUE</code></p></li>
<li><p><strong>Combination Model</strong>
  <em>(DEFAULT):</em>  <code>add = TRUE, loss = TRUE</code></p></li>
</ul> Details of each model
  choice are described in the <strong>Details</strong> section, above. Additionally,
  genotypes containing all missing values at a locus will return a value of
  <code>NA</code> and not contribute to the average across loci. <p></p><h3>Repeat Lengths</h3>
 If the user does not provide a vector of 
  appropriate length for <code>replen</code> , it will be estimated by taking the 
  minimum difference among represented alleles at each locus. IT IS NOT 
  RECOMMENDED TO RELY ON THIS ESTIMATION.
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Ruzica Bruvo, Nicolaas K. Michiels, Thomas G. D'Souza, and 
  Hinrich Schulenburg. A simple method for the calculation of microsatellite 
  genotype distances irrespective of ploidy level. Molecular Ecology, 
  13(7):2101-2106, 2004.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='fix_replen.html'>fix_replen</a></code>, <code><a href='test_replen.html'>test_replen</a></code>,
  <code><a href='bruvo.boot.html'>bruvo.boot</a></code>, <code><a href='bruvo.msn.html'>bruvo.msn</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Please note that the data presented is assuming that the nancycat dataset </span>
<span class='co'># contains all dinucleotide repeats, it most likely is not an accurate</span>
<span class='co'># representation of the data.</span>

<span class='co'># Load the nancycats dataset and construct the repeat vector.</span>
<span class='fu'>data</span>(<span class='no'>nancycats</span>)
<span class='fu'>names</span>(<span class='fu'>alleles</span>(<span class='no'>nancycats</span>)) <span class='kw'>&lt;-</span> <span class='fu'>locNames</span>(<span class='no'>nancycats</span>) <span class='co'># small bug in this data set</span>
<span class='co'># Assume the alleles are all dinucleotide repeats.</span>
<span class='no'>ssr</span> <span class='kw'>&lt;-</span> <span class='fu'>rep</span>(<span class='fl'>2</span>, <span class='fu'>nLoc</span>(<span class='no'>nancycats</span>))
<span class='fu'><a href='test_replen.html'>test_replen</a></span>(<span class='no'>nancycats</span>, <span class='no'>ssr</span>)         <span class='co'># Are the repeat lengths consistent?</span></div><div class='output co'>#&gt;  fca8 fca23 fca43 fca45 fca77 fca78 fca90 fca96 fca37 
#&gt; FALSE  TRUE FALSE  TRUE  TRUE  TRUE FALSE FALSE  TRUE </div><div class='input'>(<span class='no'>ssr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='fix_replen.html'>fix_replen</a></span>(<span class='no'>nancycats</span>, <span class='no'>ssr</span>)) <span class='co'># Nope. We need to fix them.</span></div><div class='output co'>#&gt;    fca8   fca23   fca43   fca45   fca77   fca78   fca90   fca96   fca37 
#&gt; 1.99999 2.00000 1.99999 2.00000 2.00000 2.00000 1.99999 1.99999 2.00000 </div><div class='input'>
<span class='co'># Analyze the first population in nancycats</span>
<span class='fu'>bruvo.dist</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>1</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>)</div><div class='output co'>#&gt;           N215      N216      N217      N218      N219      N220      N221
#&gt; N216 0.2377930                                                            
#&gt; N217 0.3359375 0.5029297                                                  
#&gt; N218 0.4062195 0.5033875 0.5590007                                        
#&gt; N219 0.3964844 0.2929688 0.4600694 0.3940701                              
#&gt; N220 0.4491882 0.3984070 0.4947645 0.5104167 0.4114312                    
#&gt; N221 0.2617188 0.3217773 0.4288194 0.2603895 0.2517361 0.4110243          
#&gt; N222 0.4882202 0.4784546 0.5086263 0.5360243 0.4912652 0.2968479 0.4821777
#&gt; N223 0.3789062 0.5166016 0.3142361 0.4444173 0.5451389 0.4769965 0.3541667
#&gt; N224 0.5483093 0.4784851 0.5741916 0.3680556 0.4787055 0.3224826 0.3953722
#&gt;           N222      N223
#&gt; N216                    
#&gt; N217                    
#&gt; N218                    
#&gt; N219                    
#&gt; N220                    
#&gt; N221                    
#&gt; N222                    
#&gt; N223 0.4353027          
#&gt; N224 0.4956597 0.5307888</div><div class='input'>
<span class='fu'>not_run</span>({

  <span class='co'># get the per locus estimates:</span>
  <span class='fu'>bruvo.dist</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>1</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>, <span class='kw'>by_locus</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

  <span class='co'># View each population as a heatmap.</span>
  <span class='fu'>sapply</span>(<span class='fu'>popNames</span>(<span class='no'>nancycats</span>), <span class='kw'>function</span>(<span class='no'>x</span>)
  <span class='fu'>heatmap</span>(<span class='fu'>as.matrix</span>(<span class='fu'>bruvo.dist</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='no'>x</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>)), <span class='kw'>symm</span><span class='kw'>=</span><span class='fl'>TRUE</span>))
})</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
