<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a tree using Bruvo's Distance with non-parametric bootstrapping. — bruvo.boot • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Create a tree using Bruvo's Distance with non-parametric bootstrapping. — bruvo.boot" />

<meta property="og:description" content="Create a tree using Bruvo's Distance with non-parametric bootstrapping." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">2.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a tree using Bruvo's Distance with non-parametric bootstrapping.</h1>
    <small>Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/bruvo.r'><code>R/bruvo.r</code></a></small>
    </div>

    
    <p>Create a tree using Bruvo's Distance with non-parametric bootstrapping.</p>
    

    <pre class="usage"><span class='fu'>bruvo.boot</span>(<span class='no'>pop</span>, <span class='kw'>replen</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>loss</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>sample</span> <span class='kw'>=</span> <span class='fl'>100</span>,
  <span class='kw'>tree</span> <span class='kw'>=</span> <span class='st'>"upgma"</span>, <span class='kw'>showtree</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>root</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pop</th>
      <td><p>a <code>genind</code> or <code><a href='genclone-class.html'>genclone</a></code> object</p></td>
    </tr>
    <tr>
      <th>replen</th>
      <td><p>a <code>vector</code> of <code>integers</code> indicating the length of the
nucleotide repeats for each microsatellite locus.</p></td>
    </tr>
    <tr>
      <th>add</th>
      <td><p>if <code>TRUE</code>, genotypes with zero values will be treated under 
the genome addition model presented in Bruvo et al. 2004.</p></td>
    </tr>
    <tr>
      <th>loss</th>
      <td><p>if <code>TRUE</code>, genotypes with zero values will be treated under 
the genome loss model presented in Bruvo et al. 2004.</p></td>
    </tr>
    <tr>
      <th>sample</th>
      <td><p>an <code>integer</code> indicated the number of bootstrap replicates 
desired.</p></td>
    </tr>
    <tr>
      <th>tree</th>
      <td><p>any function that can generate a tree from a distance matrix.
Default is <code><a href='http://www.rdocumentation.org/packages/phangorn/topics/upgma'>upgma</a></code>.</p></td>
    </tr>
    <tr>
      <th>showtree</th>
      <td><p><code>logical</code> if <code>TRUE</code>, a tree will be plotted with 
nodelabels.</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p><code>integer</code> the cutoff value for bootstrap node label values
(between 0 and 100).</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p><code>logical</code> defaults to <code>FALSE</code>. If <code>TRUE</code>, a 
progress bar and messages will be suppressed.</p></td>
    </tr>
    <tr>
      <th>root</th>
      <td><p><code>logical</code> This is a parameter passed on to
<code><a href='http://www.rdocumentation.org/packages/ape/topics/boot.phylo'>boot.phylo</a></code>. If the <code>tree</code> argument produces a
rooted tree (e.g. "upgma"), then this value should be <code>TRUE</code>. If it
produces an unrooted tree (e.g. "nj"), then the value should be
<code>FALSE</code>. By default, it is set to <code>NULL</code>, which will assume an
unrooted phylogeny unless the function name contains "upgma".</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>any argument to be passed on to <code>boot.phylo</code>. eg. 
<code>quiet = TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a tree of class phylo with nodelables</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function will calculate a tree based off of Bruvo's distance
  and then utilize <code><a href='http://www.rdocumentation.org/packages/ape/topics/boot.phylo'>boot.phylo</a></code> to randomly sample loci with
  replacement, recalculate the tree, and tally up the bootstrap support
  (measured in percent success). While this function can take any tree
  function, it has native support for two algorithms: <code><a href='http://www.rdocumentation.org/packages/ape/topics/nj'>nj</a></code>
  and <code><a href='http://www.rdocumentation.org/packages/phangorn/topics/upgma'>upgma</a></code>. If you want to use any other functions,
  you must load the package before you use them (see examples).</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><strong>Please refer to the documentation for bruvo.dist for details on
  the algorithm.</strong> If the user does not provide a vector of appropriate length
  for <code>replen</code> , it will be estimated by taking the minimum difference
  among represented alleles at each locus. IT IS NOT RECOMMENDED TO RELY ON
  THIS ESTIMATION.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Ruzica Bruvo, Nicolaas K. Michiels, Thomas G. D'Souza, and
Hinrich Schulenburg. A simple method for the calculation of microsatellite
genotype distances irrespective of ploidy level. Molecular Ecology,
13(7):2101-2106, 2004.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='bruvo.dist.html'>bruvo.dist</a></code>, <code>nancycats</code>, 
  <code>upgma</code>, <code><a href='http://www.rdocumentation.org/packages/ape/topics/nj'>nj</a></code>, <code><a href='http://www.rdocumentation.org/packages/ape/topics/boot.phylo'>boot.phylo</a></code>, 
  <code><a href='http://www.rdocumentation.org/packages/ape/topics/nodelabels'>nodelabels</a></code>, <code>tab</code>, 
  <code><a href='missingno.html'>missingno</a></code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Please note that the data presented is assuming that the nancycat dataset </span>
<span class='co'># contains all dinucleotide repeats, it most likely is not an accurate</span>
<span class='co'># representation of the data.</span>

<span class='co'># Load the nancycats dataset and construct the repeat vector.</span>
<span class='fu'>data</span>(<span class='no'>nancycats</span>)
<span class='no'>ssr</span> <span class='kw'>&lt;-</span> <span class='fu'>rep</span>(<span class='fl'>2</span>, <span class='fl'>9</span>)

<span class='co'># Analyze the 1st population in nancycats</span>

<span class='fu'>bruvo.boot</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>1</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>)</div><div class='output co'>#&gt; 
#&gt; Bootstrapping...
#&gt; (note: calculation of node labels can take a while even after the progress bar is full)
#&gt; 
#&gt; Running bootstraps:       100 / 100
#&gt; Calculating bootstrap values... done.</div><div class='img'><img src='bruvo.boot-1.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; 
#&gt; Phylogenetic tree with 10 tips and 9 internal nodes.
#&gt; 
#&gt; Tip labels:
#&gt; 	N215, N216, N217, N218, N219, N220, ...
#&gt; Node labels:
#&gt; 	100, 21, 11, 69, 37, 49, ...
#&gt; 
#&gt; Rooted; includes branch lengths.</div><div class='input'>
</div><span class='co'># NOT RUN {</span>
<span class='co'># Always load the library before you specify the function.</span>
<span class='fu'>library</span>(<span class='st'>"ape"</span>)

<span class='co'># Estimate the tree based off of the BIONJ algorithm.</span>

<span class='fu'>bruvo.boot</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>9</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>, <span class='kw'>tree</span> <span class='kw'>=</span> <span class='no'>bionj</span>)

<span class='co'># Utilizing  balanced FastME</span>
<span class='fu'>bruvo.boot</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>9</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>, <span class='kw'>tree</span> <span class='kw'>=</span> <span class='no'>fastme.bal</span>)

<span class='co'># To change parameters for the tree, wrap it in a function.</span>
<span class='co'># For example, let's build the tree without utilizing subtree-prune-regraft</span>

<span class='no'>myFastME</span> <span class='kw'>&lt;-</span> <span class='kw'>function</span>(<span class='no'>x</span>) <span class='fu'><a href='http://www.rdocumentation.org/packages/ape/topics/fastme'>fastme.bal</a></span>(<span class='no'>x</span>, <span class='kw'>nni</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>spr</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>tbr</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='fu'>bruvo.boot</span>(<span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>9</span>), <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>ssr</span>, <span class='kw'>tree</span> <span class='kw'>=</span> <span class='no'>myFastME</span>)

<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar, Javier F. Tabima

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

