<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find and fix inconsistent repeat lengths — fix_replen • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find and fix inconsistent repeat lengths</h1>
    </div>

    
    <p>Attempts to fix inconsistent repeat lengths found by <code>test_replen</code></p>
    

    <pre class="usage"><span class='fu'>fix_replen</span>(<span class='no'>gid</span>, <span class='no'>replen</span>, <span class='kw'>e</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>, <span class='kw'>fix_some</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gid</th>
      <td><p>a <code>genind</code> or <code><a href='genclone-class.html'>genclone</a></code> object</p></td>
    </tr>
    <tr>
      <th>replen</th>
      <td><p>a numeric vector of repeat motif lengths.</p></td>
    </tr>
    <tr>
      <th>e</th>
      <td><p>a number to be subtracted or added to inconsistent repeat lengths to
allow for proper rounding.</p></td>
    </tr>
    <tr>
      <th>fix_some</th>
      <td><p>if <code>TRUE</code> (default), when there are inconsistent repeat 
lengths that cannot be fixed by subtracting or adding e, those than can be 
fixed will. If <code>FALSE</code>, the original repeat lengths will not be fixed.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a numeric vector of corrected repeat motif lengths.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function is modified from the version used in 
  <a href = 'http://dx.doi.org/10.5281/zenodo.13007'>http://dx.doi.org/10.5281/zenodo.13007</a>. Before being fed into the
  algorithm to calculate Bruvo's distance, the amplicon length is divided by
  the repeat unit length. Because of the amplified primer sequence attached
  to sequence repeat, this division does not always result in an integer and
  so the resulting numbers are rounded. The rounding also protects against
  slight mis-calls of alleles. Because we know that $$\frac{(A - e) - (B
  - e)}{r}$$ is equivalent to $$\frac{A - B}{r}$$, we know that the primer sequence will not alter the relationships
  between the alleles. Unfortunately for nucleotide repeats that have powers
  of 2, rounding in R is based off of the IEC 60559 standard (see
  <code>round</code>), that means that any number ending in 5 is rounded to
  the nearest <em>even</em> digit. This function will attempt to alleviate this
  problem by adding a very small amount to the repeat length so that division
  will not result in a 0.5. If this fails, the same amount will be
  subtracted. If neither of these work, a warning will be issued and it is up
  to the user to determine if the fault is in the allele calls or the repeat
  lengths.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Zhian N. Kamvar, Meg M. Larsen, Alan M. Kanaskie, Everett M. 
  Hansen, &amp; Niklaus J. Grünwald. Sudden_Oak_Death_in_Oregon_Forests: Spatial
  and temporal population dynamics of the sudden oak death epidemic in Oregon
  Forests. ZENODO, http://doi.org/10.5281/zenodo.13007, 2014.
    Kamvar, Z. N., Larsen, M. M., Kanaskie, A. M., Hansen, E. M., &amp; Grünwald,
  N. J. (2015). Spatial and temporal analysis of populations of the sudden
  oak death pathogen in Oregon forests. Phytopathology 105:982-989.
  doi: <a href = 'http://dx.doi.org/10.1094/PHYTO-12-14-0350-FI'>10.1094/PHYTO-12-14-0350-FI</a>
    Ruzica Bruvo, Nicolaas K. Michiels, Thomas G. D'Souza, and Hinrich 
  Schulenburg. A simple method for the calculation of microsatellite genotype
  distances irrespective of ploidy level. Molecular Ecology, 13(7):2101-2106,
  2004.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='test_replen.html'>test_replen</a></code> <code><a href='bruvo.dist.html'>bruvo.dist</a></code>
  <code><a href='bruvo.msn.html'>bruvo.msn</a></code> <code><a href='bruvo.boot.html'>bruvo.boot</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'>data</span>(<span class='no'>Pram</span>)
(<span class='no'>Pram_replen</span> <span class='kw'>&lt;-</span> <span class='fu'>setNames</span>(<span class='fu'>c</span>(<span class='fl'>3</span>, <span class='fl'>2</span>, <span class='fl'>4</span>, <span class='fl'>4</span>, <span class='fl'>4</span>), <span class='fu'>locNames</span>(<span class='no'>Pram</span>)))</div><div class='output co'>#&gt;  PrMS6A1  Pr9C3A1 PrMS39A1 PrMS45A1 PrMS43A1 
#&gt;        3        2        4        4        4 </div><div class='input'><span class='fu'>fix_replen</span>(<span class='no'>Pram</span>, <span class='no'>Pram_replen</span>)</div><div class='output co'>#&gt;  PrMS6A1  Pr9C3A1 PrMS39A1 PrMS45A1 PrMS43A1 
#&gt;  3.00000  2.00000  3.99999  4.00000  4.00000 </div><div class='input'><span class='co'># Let's start with an example of a tetranucleotide repeat motif and imagine</span>
<span class='co'># that there are twenty alleles all 1 step apart:</span>
(<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fl'>1</span>:<span class='fl'>20L</span> * <span class='fl'>4L</span>)</div><div class='output co'>#&gt;  [1]  4  8 12 16 20 24 28 32 36 40 44 48 52 56 60 64 68 72 76 80</div><div class='input'><span class='co'># These are the true lengths of the different alleles. Now, let's add the</span>
<span class='co'># primer sequence to them. </span>
(<span class='no'>PxP</span> <span class='kw'>&lt;-</span> <span class='no'>x</span> + <span class='fl'>21</span> + <span class='fl'>21</span>)</div><div class='output co'>#&gt;  [1]  46  50  54  58  62  66  70  74  78  82  86  90  94  98 102 106 110 114 118
#&gt; [20] 122</div><div class='input'><span class='co'># Now we make sure that x / 4 is equal to 1:20, which we know each have</span>
<span class='co'># 1 difference.</span>
<span class='no'>x</span>/<span class='fl'>4</span></div><div class='output co'>#&gt;  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20</div><div class='input'><span class='co'># Now, we divide the sequence with the primers by 4 and see what happens.</span>
(<span class='no'>PxPc</span> <span class='kw'>&lt;-</span> <span class='no'>PxP</span>/<span class='fl'>4</span>)</div><div class='output co'>#&gt;  [1] 11.5 12.5 13.5 14.5 15.5 16.5 17.5 18.5 19.5 20.5 21.5 22.5 23.5 24.5 25.5
#&gt; [16] 26.5 27.5 28.5 29.5 30.5</div><div class='input'>(<span class='no'>PxPcr</span> <span class='kw'>&lt;-</span> <span class='fu'>round</span>(<span class='no'>PxPc</span>))</div><div class='output co'>#&gt;  [1] 12 12 14 14 16 16 18 18 20 20 22 22 24 24 26 26 28 28 30 30</div><div class='input'><span class='fu'>diff</span>(<span class='no'>PxPcr</span>) <span class='co'># we expect all 1s</span></div><div class='output co'>#&gt;  [1] 0 2 0 2 0 2 0 2 0 2 0 2 0 2 0 2 0 2 0</div><div class='input'>
<span class='co'># Let's try that again by subtracting a tiny amount from 4</span>
(<span class='no'>PxPc</span> <span class='kw'>&lt;-</span> <span class='no'>PxP</span>/(<span class='fl'>4</span> - <span class='fl'>1e-5</span>))</div><div class='output co'>#&gt;  [1] 11.50003 12.50003 13.50003 14.50004 15.50004 16.50004 17.50004 18.50005
#&gt;  [9] 19.50005 20.50005 21.50005 22.50006 23.50006 24.50006 25.50006 26.50007
#&gt; [17] 27.50007 28.50007 29.50007 30.50008</div><div class='input'>(<span class='no'>PxPcr</span> <span class='kw'>&lt;-</span> <span class='fu'>round</span>(<span class='no'>PxPc</span>))</div><div class='output co'>#&gt;  [1] 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31</div><div class='input'><span class='fu'>diff</span>(<span class='no'>PxPcr</span>)</div><div class='output co'>#&gt;  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
