<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>GENclone and SNPclone classes — genclone-class • poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="GENclone and SNPclone classes — genclone-class" />

<meta property="og:description" content="GENclone is an S4 class that extends the 
  genind object. SNPclone is an S4 class
  that extends the genlight object. They will have
  all of the same attributes as their parent classes, but they will contain
  one extra slot to store extra information about multilocus genotypes." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>GENclone and SNPclone classes</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/classes.r'><code>R/classes.r</code></a></small>
    <div class="hidden name"><code>genclone-class.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><strong>GENclone</strong> is an S4 class that extends the 
  <code>genind</code> object.<br /> <strong>SNPclone</strong> is an S4 class
  that extends the <code>genlight</code> object.<br /><br /> They will have
  all of the same attributes as their parent classes, but they will contain
  one extra slot to store extra information about multilocus genotypes.</p>
    
    </div>

        
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The genclone and snpclone classes will allow for more optimized 
  methods of clone correction.</p>  
<p>Previously for genind and genlight objects, 
  multilocus genotypes were not retained after a data set was subset by 
  population. The new <strong><code>mlg</code></strong> slot allows us to assign the 
  multilocus genotypes and retain that information no matter how we subset 
  the data set. This new slot can either contain numeric values for 
  multilocus genotypes OR it can contain a special internal 
  <code><a href='MLG-class.html'>MLG</a></code> class that allows for custom multilocus genotype 
  definitions and filtering.</p>
    
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    
    <dl class='dl-horizontal'>
<dt><code>mlg</code></dt><dd><p>a vector representing multilocus genotypes for the data set OR an 
object of class <code><a href='MLG-class.html'>MLG</a></code>.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>When calculating multilocus genotypes for genclone objects, a rank 
  function is used, but calculation of multilocus genotypes for snpclone
  objects is distance-based (via <code><a href='bitwise.dist.html'>bitwise.dist</a></code> and 
  <code><a href='mlg.filter.html'>mlg.filter</a></code>). This means that genclone objects are sensitive
  to missing data, whereas snpclone objects are insensitive.</p>
    
    <h2 class="hasAnchor" id="extends"><a class="anchor" href="#extends"></a>Extends</h2>

    <p>The <code>genclone</code> class extends class 
  <code>"genind"</code>, directly. <br /> The <code>snpclone</code> class 
  extends class <code>"genlight"</code>, directly.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='coercion-methods.html'>as.genclone</a></code> <code><a href='snpclone-coercion-methods.html'>as.snpclone</a></code> 
  <code>genind</code> <code>genlight</code> 
  <code><a href='https://www.rdocumentation.org/packages/adegenet/topics/strata-methods'>strata</a></code> <code><a href='https://www.rdocumentation.org/packages/adegenet/topics/population-methods'>setPop</a></code> 
  <code><a href='MLG-class.html'>MLG</a></code> <code><a href='mll-method.html'>mll</a></code> <code><a href='mlg.filter.html'>mlg.filter</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># genclone objects can be created from genind objects</span>
<span class='co'>#</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>partial_clone</span>)
<span class='no'>partial_clone</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 50 individuals; 10 loci; 35 alleles; size: 22.6 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  50 x 35 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-5)
#&gt;    @loc.fac: locus factor for the 35 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt;    @type:  codom
#&gt;    @call: old2new_genind(object = x, donor = new(class(x)))
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 12-13)</div><div class='input'>(<span class='no'>pc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='coercion-methods.html'>as.genclone</a></span>(<span class='no'>partial_clone</span>))</div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    26 original multilocus genotypes 
#&gt;    50 diploid individuals
#&gt;    10 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     0 strata. 
#&gt;     4 populations defined - 1, 2, 3, 4</div><div class='input'>
<span class='co'># snpclone objects can be created from genlight objects</span>
<span class='co'>#</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>999</span>)
(<span class='no'>gl</span> <span class='kw'>&lt;-</span> <span class='fu'>glSim</span>(<span class='fl'>100</span>, <span class='fl'>0</span>, <span class='kw'>n.snp.struc</span> <span class='kw'>=</span> <span class='fl'>1e3</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>))</div><div class='output co'>#&gt;  /// GENLIGHT OBJECT /////////
#&gt; 
#&gt;  // 100 genotypes,  1,000 binary SNPs, size: 172 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  // Basic content
#&gt;    @gen: list of 100 SNPbin
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 44-56)
#&gt;    @other: a list containing: ancestral.pops 
#&gt; </div><div class='input'>(<span class='no'>sc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snpclone-coercion-methods.html'>as.snpclone</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/cbind'>rbind</a></span>(<span class='no'>gl</span>, <span class='no'>gl</span>, <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>), <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>))</div><div class='output co'>#&gt;  ||| SNPCLONE OBJECT |||||||||
#&gt; 
#&gt;  || 200 genotypes,  1,000 binary SNPs, size: 353.5 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  || Basic content
#&gt;    @gen: list of 200 SNPbin
#&gt;    @mlg: 100 original multilocus genotypes
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  || Optional content
#&gt;    @pop: population of each individual (group size range: 88-112)
#&gt;    @other: a list containing: elements without names 
#&gt; 
#&gt; NULL</div><div class='input'><span class='co'># </span>
<span class='co'># Use mlg.filter to create a distance threshold to define multilocus genotypes.</span>
<span class='fu'><a href='mlg.filter.html'>mlg.filter</a></span>(<span class='no'>sc</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>) <span class='kw'>&lt;-</span> <span class='fl'>0.25</span>
<span class='no'>sc</span> <span class='co'># 82 mlgs</span></div><div class='output co'>#&gt;  ||| SNPCLONE OBJECT |||||||||
#&gt; 
#&gt;  || 200 genotypes,  1,000 binary SNPs, size: 353.5 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  || Basic content
#&gt;    @gen: list of 200 SNPbin
#&gt;    @mlg: 82 contracted multilocus genotypes
#&gt;          (0.25) [t], (bitwise.dist) [d], (farthest) [a]
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  || Optional content
#&gt;    @pop: population of each individual (group size range: 88-112)
#&gt;    @other: a list containing: elements without names 
#&gt; 
#&gt; NULL</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#details">Details</a></li>

      <li><a href="#slots">Slots</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#extends">Extends</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Zhian N. Kamvar</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

