<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate Genetic Distance for a genind or genclone object. — nei.dist • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate Genetic Distance for a genind or genclone object.</h1>
    </div>

    
    <p>These functions are modified from the function <a href='http://www.rdocumentation.org/packages/adegenet/topics/dist.genpop'>dist.genpop</a> to
be applicable for distances between individuals.</p>
    

    <pre class="usage"><span class='fu'>nei.dist</span>(<span class='no'>x</span>, <span class='kw'>warning</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='fu'>edwards.dist</span>(<span class='no'>x</span>)

<span class='fu'>rogers.dist</span>(<span class='no'>x</span>)

<span class='fu'>reynolds.dist</span>(<span class='no'>x</span>)

<span class='fu'>provesti.dist</span>(<span class='no'>x</span>)

<span class='fu'>prevosti.dist</span>(<span class='no'>x</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a genind, <a href='genclone-class.html'>genclone</a>, or matrix object.</p></td>
    </tr>
    <tr>
      <th>warning</th>
      <td><p>If <code>TRUE</code>, a warning will be printed if any infinite 
values are detected and replaced. If <code>FALSE</code>, these values will be 
replaced without warning. See Details below.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>an object of class dist with the same number of observations as the 
  number of individuals in your data.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>It is important to be careful with the interpretation of these 
  distances as they were originally intended for calculation of 
  between-population distance. As Nei's distance is the negative log of 0:1, 
  this means that it is very possible to obtain distances of infinity. When 
  this happens, infinite values are corrected to be 10 * max(D) where D is
  the distance matrix without infinite values.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Prevosti's distance is identical to <code><a href='diss.dist.html'>diss.dist</a></code>, except 
  that <code><a href='diss.dist.html'>diss.dist</a></code> is optimized for a larger number of 
  individuals (n &gt; 125) at the cost of required memory. Both
  <code>prevosti.dist</code> and <code>provesti.dist</code> are the same function,
  <code>provesti.dist</code> is a spelling error and exists for backwards
  compatibility.
    These distances were adapted from the <span class="pkg">adegenet</span> function
  <code>dist.genpop</code> to work with <code>genind</code> objects.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Nei, M. (1972) Genetic distances between populations. American Naturalist,
106, 283-292.</p>
<p>Nei M. (1978) Estimation of average heterozygosity and genetic
distance from a small number of individuals. Genetics, 23, 341-369.</p>
<p>Avise, J. C. (1994) Molecular markers, natural history and evolution. Chapman &amp; Hall,
London.</p>
<p>Edwards, A.W.F. (1971) Distance between populations on the basis of gene
frequencies. Biometrics, 27, 873-881.</p>
<p>Cavalli-Sforza L.L. and Edwards A.W.F.
(1967) Phylogenetic analysis: models and estimation procedures. Evolution,
32, 550-570.</p>
<p>Hartl, D.L. and Clark, A.G. (1989) Principles of population
genetics. Sinauer Associates, Sunderland, Massachussetts (p. 303).</p>
<p>Reynolds, J. B., B. S. Weir, and C. C. Cockerham. (1983) Estimation of the
coancestry coefficient: basis for a short-term genetic distance. Genetics,
105, 767-779.</p>
<p>Rogers, J.S. (1972) Measures of genetic similarity and genetic distances.
Studies in Genetics, Univ. Texas Publ., 7213, 145-153.</p>
<p>Avise, J. C. (1994)
Molecular markers, natural history and evolution. Chapman &amp; Hall, London.</p>
<p>Prevosti A. (1974) La distancia genetica entre poblaciones. Miscellanea
Alcobe, 68, 109-118.</p>
<p>Prevosti A., Oca\~na J. and Alonso G. (1975) Distances
between populations of Drosophila subobscura, based on chromosome
arrangements frequencies. Theoretical and Applied Genetics, 45, 231-241.</p>
<p>For more information on dissimilarity indexes:</p>
<p>Gower J. and Legendre P. (1986)
Metric and Euclidean properties of dissimilarity coefficients. Journal of
Classification, 3, 5-48</p>
<p>Legendre P. and Legendre L. (1998) Numerical Ecology,
Elsevier Science B.V. 20, pp274-288.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='aboot.html'>aboot</a></code> <code><a href='diss.dist.html'>diss.dist</a></code> <code><a href='poppr.amova.html'>poppr.amova</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'>data</span>(<span class='no'>nancycats</span>)
(<span class='no'>nan9</span>   <span class='kw'>&lt;-</span> <span class='fu'><a href='popsub.html'>popsub</a></span>(<span class='no'>nancycats</span>, <span class='fl'>9</span>))</div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 9 individuals; 9 loci; 43 alleles; size: 15.3 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  9 x 43 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-8)
#&gt;    @loc.fac: locus factor for the 43 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt;    @type:  codom
#&gt;    @call: popsub(gid = nancycats, sublist = 9)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 9-9)
#&gt;    @other: a list containing: xy 
#&gt; </div><div class='input'>(<span class='no'>neinan</span> <span class='kw'>&lt;-</span> <span class='fu'>nei.dist</span>(<span class='no'>nan9</span>))</div><div class='output co'>#&gt;           N104      N105      N106      N107      N108      N109      N111
#&gt; N105 1.0579046                                                            
#&gt; N106 0.5219437 0.5990190                                                  
#&gt; N107 0.2009995 0.6263815 0.2350018                                        
#&gt; N108 0.3013348 0.9090384 0.2707986 0.2181184                              
#&gt; N109 0.6019864 0.2195294 0.5978370 0.5451567 0.7408023                    
#&gt; N111 0.8201799 2.6020033 0.7290191 0.8770096 0.5355122 1.5174765          
#&gt; N112 0.9373236 0.3784101 0.7408023 0.4257577 0.6061358 0.5584807 2.0759572
#&gt; N113 0.8754687 1.0579046 1.0074515 1.0883026 0.8319630 0.8896685 0.6531258
#&gt;           N112
#&gt; N105          
#&gt; N106          
#&gt; N107          
#&gt; N108          
#&gt; N109          
#&gt; N111          
#&gt; N112          
#&gt; N113 1.5251102</div><div class='input'>(<span class='no'>ednan</span>  <span class='kw'>&lt;-</span> <span class='fu'>edwards.dist</span>(<span class='no'>nan9</span>))</div><div class='output co'>#&gt;           N104      N105      N106      N107      N108      N109      N111
#&gt; N105 0.7517229                                                            
#&gt; N106 0.6127062 0.6387116                                                  
#&gt; N107 0.3913906 0.6808143 0.4463299                                        
#&gt; N108 0.5140990 0.7362578 0.5047434 0.4197600                              
#&gt; N109 0.6312058 0.4309644 0.6236096 0.6236096 0.6906433                    
#&gt; N111 0.7730660 0.9599128 0.7517229 0.7730660 0.6808143 0.8873047          
#&gt; N112 0.7517229 0.5739208 0.6906433 0.5855469 0.6312058 0.6564788 0.9305251
#&gt; N113 0.7878089 0.8022810 0.7878089 0.8223129 0.7580364 0.7730660 0.7297559
#&gt;           N112
#&gt; N105          
#&gt; N106          
#&gt; N107          
#&gt; N108          
#&gt; N109          
#&gt; N111          
#&gt; N112          
#&gt; N113 0.8687726</div><div class='input'>(<span class='no'>rodnan</span> <span class='kw'>&lt;-</span> <span class='fu'>rogers.dist</span>(<span class='no'>nan9</span>))</div><div class='output co'>#&gt;           N104      N105      N106      N107      N108      N109      N111
#&gt; N105 0.5813390                                                            
#&gt; N106 0.4295584 0.5000000                                                  
#&gt; N107 0.2222222 0.5081258 0.2222222                                        
#&gt; N108 0.3333333 0.5785674 0.2777778 0.2222222                              
#&gt; N109 0.4146723 0.2777778 0.3563452 0.3414591 0.4674563                    
#&gt; N111 0.5718072 0.8170412 0.5311377 0.5866932 0.4904682 0.6882725          
#&gt; N112 0.5664529 0.4119008 0.4702279 0.3888889 0.4146723 0.4674563 0.7465996
#&gt; N113 0.5718072 0.6192369 0.6043508 0.6097051 0.5866932 0.5866932 0.4606960
#&gt;           N112
#&gt; N105          
#&gt; N106          
#&gt; N107          
#&gt; N108          
#&gt; N109          
#&gt; N111          
#&gt; N112          
#&gt; N113 0.6829183</div><div class='input'>(<span class='no'>reynan</span> <span class='kw'>&lt;-</span> <span class='fu'>reynolds.dist</span>(<span class='no'>nan9</span>))</div><div class='output co'>#&gt;           N104      N105      N106      N107      N108      N109      N111
#&gt; N105 0.7934920                                                            
#&gt; N106 0.6255432 0.6915641                                                  
#&gt; N107 0.4588315 0.6770032 0.4364358                                        
#&gt; N108 0.5619515 0.7595545 0.5000000 0.4472136                              
#&gt; N109 0.6454972 0.5423261 0.6000000 0.5656854 0.6504436                    
#&gt; N111 0.7483315 0.8574929 0.6928203 0.6938887 0.6741999 0.7620008          
#&gt; N112 0.7200823 0.6488857 0.6504436 0.5516773 0.6454972 0.6123724 0.7977240
#&gt; N113 0.7337994 0.7934920 0.7071068 0.6948083 0.7071068 0.6938887 0.7223151
#&gt;           N112
#&gt; N105          
#&gt; N106          
#&gt; N107          
#&gt; N108          
#&gt; N109          
#&gt; N111          
#&gt; N112          
#&gt; N113 0.7620008</div><div class='input'>(<span class='no'>pronan</span> <span class='kw'>&lt;-</span> <span class='fu'>prevosti.dist</span>(<span class='no'>nan9</span>))</div><div class='output co'>#&gt;           N104      N105      N106      N107      N108      N109      N111
#&gt; N105 0.6111111                                                            
#&gt; N106 0.4444444 0.5000000                                                  
#&gt; N107 0.2222222 0.5555556 0.2222222                                        
#&gt; N108 0.3333333 0.6111111 0.2777778 0.2222222                              
#&gt; N109 0.4444444 0.2777778 0.3888889 0.3888889 0.5000000                    
#&gt; N111 0.6666667 0.9444444 0.6111111 0.6666667 0.5555556 0.8333333          
#&gt; N112 0.6111111 0.4444444 0.5000000 0.3888889 0.4444444 0.5000000 0.8888889
#&gt; N113 0.6666667 0.6666667 0.6666667 0.7222222 0.6666667 0.6666667 0.5555556
#&gt;           N112
#&gt; N105          
#&gt; N106          
#&gt; N107          
#&gt; N108          
#&gt; N109          
#&gt; N111          
#&gt; N112          
#&gt; N113 0.7777778</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar (poppr adaptation)
Thibaut Jombart (adegenet adaptation)
Daniel Chessel (ade4)

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
