<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="Aeut.html">Aeut</a></code> </p>
        </td>
        <td><p>Oomycete root rot pathogen *Aphanomyces euteiches* AFLP data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Pinf.html">Pinf</a></code> </p>
        </td>
        <td><p>Phytophthora infestans data from Mexico and South America.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Pram.html">Pram</a></code> </p>
        </td>
        <td><p>Phytophthora ramorum data from OR Forests and Nurseries (OR and CA)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aboot.html">aboot()</a></code> </p>
        </td>
        <td><p>Calculate a dendrogram with bootstrap support using any distance applicable
to genind or genclone objects.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bitwise.dist.html">bitwise.dist()</a></code> </p>
        </td>
        <td><p>Calculate dissimilarity or Euclidean distance for genlight objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="boot.ia.html">boot.ia()</a></code> </p>
        </td>
        <td><p>Bootstrap the index of association</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bruvo.boot.html">bruvo.boot()</a></code> </p>
        </td>
        <td><p>Create a tree using Bruvo's Distance with non-parametric bootstrapping.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bruvo.dist.html">bruvo.dist()</a></code> </p>
        </td>
        <td><p>Bruvo's distance for microsatellites</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bruvo.msn.html">bruvo.msn()</a></code> </p>
        </td>
        <td><p>Create minimum spanning network of selected populations using Bruvo's 
distance.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clonecorrect.html">clonecorrect()</a></code> </p>
        </td>
        <td><p>Remove potential bias caused by cloned genotypes in genind or genclone 
object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coercion-methods.html">bootgen2genind()</a></code> <code><a href="coercion-methods.html">as.genclone()</a></code> <code><a href="coercion-methods.html">genclone2genind()</a></code> <code><a href="coercion-methods.html">as.genambig()</a></code> </p>
        </td>
        <td><p>Switch between genind and genclone objects.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cutoff_predictor.html">cutoff_predictor()</a></code> </p>
        </td>
        <td><p>Predict cutoff thresholds for use with mlg.filter</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="diss.dist.html">diss.dist()</a></code> </p>
        </td>
        <td><p>Calculate a distance matrix based on relative dissimilarity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="diversity_boot.html">diversity_boot()</a></code> </p>
        </td>
        <td><p>Perform a bootstrap analysis on diversity statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="diversity_ci.html">diversity_ci()</a></code> </p>
        </td>
        <td><p>Perform bootstrap statistics, calculate, and plot confidence intervals.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="diversity_stats.html">diversity_stats()</a></code> </p>
        </td>
        <td><p>Produce a table of diversity statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_stats.html">filter_stats()</a></code> </p>
        </td>
        <td><p>Utilize all algorithms of mlg.filter</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fix_replen.html">fix_replen()</a></code> </p>
        </td>
        <td><p>Find and fix inconsistent repeat lengths</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genclone-class.html">genclone-class</a></code> </p>
        </td>
        <td><p>GENclone and SNPclone classes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genetic_distance.html">nei.dist()</a></code> <code><a href="genetic_distance.html">edwards.dist()</a></code> <code><a href="genetic_distance.html">rogers.dist()</a></code> <code><a href="genetic_distance.html">reynolds.dist()</a></code> <code><a href="genetic_distance.html">provesti.dist()</a></code> <code><a href="genetic_distance.html">prevosti.dist()</a></code> </p>
        </td>
        <td><p>Calculate Genetic Distance for a genind or genclone object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genind2genalex.html">genind2genalex()</a></code> </p>
        </td>
        <td><p>Export data from genind objects to genalex formatted *.csv files.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genotype_curve.html">genotype_curve()</a></code> </p>
        </td>
        <td><p>Produce a genotype accumulation curve</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getfile.html">getfile()</a></code> </p>
        </td>
        <td><p>Get a file name and path and store them in a list.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="greycurve.html">greycurve()</a></code> </p>
        </td>
        <td><p>Display a greyscale gradient adjusted to specific parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ia.html">ia()</a></code> <code><a href="ia.html">pair.ia()</a></code> <code><a href="ia.html">resample.ia()</a></code> <code><a href="ia.html">jack.ia()</a></code> </p>
        </td>
        <td><p>Index of Association</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="imsn.html">imsn()</a></code> </p>
        </td>
        <td><p>Create minimum spanning networks interactively</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="incomp.html">incomp()</a></code> </p>
        </td>
        <td><p>Check for samples that are incomparable due to missing data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="info_table.html">info_table()</a></code> </p>
        </td>
        <td><p>Create a table summarizing missing data or ploidy information of a genind or
genclone object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="informloci.html">informloci()</a></code> </p>
        </td>
        <td><p>Remove all non-phylogentically informative loci</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is.clone.html">is.snpclone()</a></code> <code><a href="is.clone.html">is.clone()</a></code> <code><a href="is.clone.html">is.genclone()</a></code> </p>
        </td>
        <td><p>Check for validity of a genclone or snpclone object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="locus_table.html">locus_table()</a></code> </p>
        </td>
        <td><p>Create a table of summary statistics per locus.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_haplotypes-method.html">make_haplotypes()</a></code> </p>
        </td>
        <td><p>Split samples from a genind object into pseudo-haplotypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="missingno.html">missingno()</a></code> </p>
        </td>
        <td><p>Treat missing data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mlg.html">mlg()</a></code> <code><a href="mlg.html">mlg.table()</a></code> <code><a href="mlg.html">mlg.vector()</a></code> <code><a href="mlg.html">mlg.crosspop()</a></code> <code><a href="mlg.html">mlg.id()</a></code> </p>
        </td>
        <td><p>Create counts, vectors, and matrices of multilocus genotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mlg.filter.html">mlg.filter()</a></code> <code><a href="mlg.filter.html">`mlg.filter<-`()</a></code> </p>
        </td>
        <td><p>MLG definitions based on genetic distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mll-method.html">mll()</a></code> <code><a href="mll-method.html">nmll()</a></code> <code><a href="mll-method.html">`mll<-`()</a></code> </p>
        </td>
        <td><p>Access and manipulate multilocus lineages.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mll.custom.html">mll.custom()</a></code> <code><a href="mll.custom.html">`mll.custom<-`()</a></code> <code><a href="mll.custom.html">mll.levels()</a></code> <code><a href="mll.custom.html">`mll.levels<-`()</a></code> </p>
        </td>
        <td><p>Define custom multilocus lineages</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mll.reset-method.html">mll.reset()</a></code> </p>
        </td>
        <td><p>Reset multilocus lineages</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="monpop.html">monpop</a></code> </p>
        </td>
        <td><p>Peach brown rot pathogen *Monilinia fructicola*</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="old2new_genclone.html">old2new_genclone()</a></code> </p>
        </td>
        <td><p>Convert an old genclone object to a new genclone object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="partial_clone.html">partial_clone</a></code> </p>
        </td>
        <td><p>Simulated data illustrating a Minimum Spanning Network based on Bruvo's
Distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pgen.html">pgen()</a></code> </p>
        </td>
        <td><p>Genotype Probability</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_poppr_msn.html">plot_poppr_msn()</a></code> </p>
        </td>
        <td><p>Plot minimum spanning networks produced in poppr.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="poppr-package.html">poppr-package</a></code> </p>
        </td>
        <td><p>The <span class="pkg">poppr</span> R package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="poppr.html">poppr()</a></code> </p>
        </td>
        <td><p>Produce a basic summary table for population genetic analyses.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="poppr.all.html">poppr.all()</a></code> </p>
        </td>
        <td><p>Process a list of files with poppr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="poppr.amova.html">poppr.amova()</a></code> </p>
        </td>
        <td><p>Perform Analysis of Molecular Variance (AMOVA) on genind or genclone objects.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="poppr.msn.html">poppr.msn()</a></code> </p>
        </td>
        <td><p>Create a minimum spanning network of selected populations using a distance 
matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="poppr_has_parallel.html">poppr_has_parallel()</a></code> </p>
        </td>
        <td><p>Determines whether openMP is support on this system.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="popsub.html">popsub()</a></code> </p>
        </td>
        <td><p>Subset data by population</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="private_alleles.html">private_alleles()</a></code> </p>
        </td>
        <td><p>Tabulate alleles the occur in only one population.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="psex.html">psex()</a></code> </p>
        </td>
        <td><p>Probability of encountering a genotype more than once by chance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rare_allele_correction.html">rare_allele_correction</a></code> </p>
        </td>
        <td><p>Correcting rare allele frequencies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read.genalex.html">read.genalex()</a></code> </p>
        </td>
        <td><p>Importing data from genalex formatted *.csv files.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recode_polyploids.html">recode_polyploids()</a></code> </p>
        </td>
        <td><p>Recode polyploid microsatellite data for use in frequency based statistics.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rraf.html">rraf()</a></code> </p>
        </td>
        <td><p>Round Robin Allele Frequencies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rrmlg.html">rrmlg()</a></code> </p>
        </td>
        <td><p>Round Robin Multilocus Genotypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="samp.ia.html">samp.ia()</a></code> </p>
        </td>
        <td><p>Calculate random samples of the index of association for genlight objects.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="shufflepop.html">shufflepop()</a></code> </p>
        </td>
        <td><p>Shuffle individuals in a <code>genclone</code> or
<code>genind</code> object independently over each locus.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="snpclone-coercion-methods.html">as.snpclone()</a></code> </p>
        </td>
        <td><p>Create a snpclone object from a genlight object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_replen.html">test_replen()</a></code> </p>
        </td>
        <td><p>Test repeat length consistency.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="win.ia.html">win.ia()</a></code> </p>
        </td>
        <td><p>Calculate windows of the index of association for genlight objects.</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

