<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Remove all non-phylogentically informative loci — informloci • poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Remove all non-phylogentically informative loci — informloci" />

<meta property="og:description" content="This function will facilitate in removing phylogenetically uninformative loci
from a genclone or genind object. 
The user has the ability to define what uninformative means by setting a 
cutoff value for either percentage of differentiating genotypes or minor 
allele frequency." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Remove all non-phylogentically informative loci</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/data_subset.r'><code>R/data_subset.r</code></a></small>
    <div class="hidden name"><code>informloci.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function will facilitate in removing phylogenetically uninformative loci
from a <code><a href='genclone-class.html'>genclone</a></code> or <code>genind</code> object. 
The user has the ability to define what uninformative means by setting a 
cutoff value for either percentage of differentiating genotypes or minor 
allele frequency.</p>
    
    </div>

    <pre class="usage"><span class='fu'>informloci</span>(<span class='no'>pop</span>, <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fl'>2</span>/<span class='fu'>nInd</span>(<span class='no'>pop</span>), <span class='kw'>MAF</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pop</th>
      <td><p>a <code><a href='genclone-class.html'>genclone</a></code> or <code>genind</code> 
object.</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p><code>numeric</code>. A number from 0 to 1 defining the minimum 
number of differentiating samples.</p></td>
    </tr>
    <tr>
      <th>MAF</th>
      <td><p><code>numeric</code>. A number from 0 to 1 defining the minimum minor
allele frequency. This is passed as the <code>thresh</code> parameter of
<code><a href='https://www.rdocumentation.org/packages/adegenet/topics/isPoly'>isPoly</a></code>.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p><code>logical</code>. When <code>quiet = TRUE</code> (default), messages 
indicating the loci removed will be printed to screen. When <code>quiet = 
FALSE</code>, nothing will be printed to screen.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>genind</code> object with user-defined informative loci.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function will remove uninformative loci using a traditional MAF
  cutoff (using <code><a href='https://www.rdocumentation.org/packages/adegenet/topics/isPoly'>isPoly</a></code> from <span class="pkg">adegenet</span>) as well
  as analyzing the number of observed genotypes in a locus. This is important
  for clonal organisms that can have fixed heterozygous sites not detected by
  MAF methods.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>This will have a few side effects that affect certain analyses. First, 
  the number of multilocus genotypes might be reduced due to the reduced 
  number of markers (if you are only using a genind object). Second, if you 
  plan on using this data for analysis of the index of association, be sure 
  to use the standardized version (rbarD) that corrects for the number of 
  observed loci.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># We will use a dummy data set to demonstrate how this detects uninformative</span>
<span class='co'># loci using both MAF and a cutoff.</span>

<span class='no'>genos</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"A/A"</span>, <span class='st'>"A/B"</span>, <span class='st'>"A/C"</span>, <span class='st'>"B/B"</span>, <span class='st'>"B/C"</span>, <span class='st'>"C/C"</span>)

<span class='no'>v</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/sample'>sample</a></span>(<span class='no'>genos</span>, <span class='fl'>100</span>, <span class='kw'>replace</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>w</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='no'>genos</span>[<span class='fl'>2</span>], <span class='fl'>99</span>), <span class='no'>genos</span>[<span class='fl'>3</span>])           <span class='co'># found by cutoff</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='no'>genos</span>[<span class='fl'>1</span>], <span class='fl'>98</span>), <span class='no'>genos</span>[<span class='fl'>3</span>], <span class='no'>genos</span>[<span class='fl'>2</span>]) <span class='co'># found by MAF</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='no'>genos</span>[<span class='fl'>1</span>], <span class='fl'>99</span>), <span class='no'>genos</span>[<span class='fl'>2</span>])           <span class='co'># found by both</span>
<span class='no'>z</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/sample'>sample</a></span>(<span class='no'>genos</span>, <span class='fl'>100</span>, <span class='kw'>replace</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>dat</span> <span class='kw'>&lt;-</span> <span class='fu'>df2genind</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/data.frame'>data.frame</a></span>(<span class='kw'>v</span> <span class='kw'>=</span> <span class='no'>v</span>, <span class='kw'>w</span> <span class='kw'>=</span> <span class='no'>w</span>, <span class='kw'>x</span> <span class='kw'>=</span> <span class='no'>x</span>, <span class='kw'>y</span> <span class='kw'>=</span> <span class='no'>y</span>, <span class='kw'>z</span> <span class='kw'>=</span> <span class='no'>z</span>), <span class='kw'>sep</span> <span class='kw'>=</span> <span class='st'>"/"</span>)

<span class='fu'>informloci</span>(<span class='no'>dat</span>)</div><div class='output co'>#&gt; <span class='message'>cutoff value: 2 % ( 2 samples ).</span></div><div class='output co'>#&gt; <span class='message'>MAF         : 0.01</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'> Found 3 uninformative loci </span>
#&gt; <span class='message'> ============================ </span>
#&gt; <span class='message'> 2 loci found with a cutoff of 2 samples :</span>
#&gt; <span class='message'> w, y </span>
#&gt; <span class='message'> 2 loci found with MAF &lt; 0.01 :</span>
#&gt; <span class='message'> x, y</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 100 individuals; 2 loci; 6 alleles; size: 14.6 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  100 x 6 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-3)
#&gt;    @loc.fac: locus factor for the 6 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt;    @type:  codom
#&gt;    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)
#&gt; 
#&gt;  // Optional content
#&gt;    - empty -</div><div class='input'>
</div><div class='input'><span class='co'># Ignore MAF</span>
<span class='fu'>informloci</span>(<span class='no'>dat</span>, <span class='kw'>MAF</span> <span class='kw'>=</span> <span class='fl'>0</span>)</div><div class='output co'>#&gt; <span class='message'>cutoff value: 2 % ( 2 samples ).</span></div><div class='output co'>#&gt; <span class='message'>MAF         : 0</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'> Found 2 uninformative loci </span>
#&gt; <span class='message'> ============================ </span>
#&gt; <span class='message'> 2 loci found with a cutoff of 2 samples :</span>
#&gt; <span class='message'> w, y </span>
#&gt; <span class='message'> 0 loci found with MAF &lt; 0  </span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 100 individuals; 3 loci; 9 alleles; size: 16.5 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  100 x 9 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-3)
#&gt;    @loc.fac: locus factor for the 9 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt;    @type:  codom
#&gt;    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)
#&gt; 
#&gt;  // Optional content
#&gt;    - empty -</div><div class='input'>
<span class='co'># Ignore cutoff</span>
<span class='fu'>informloci</span>(<span class='no'>dat</span>, <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fl'>0</span>)</div><div class='output co'>#&gt; <span class='message'>cutoff value: 0 % ( 0 samples ).</span></div><div class='output co'>#&gt; <span class='message'>MAF         : 0.01</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'> Found 2 uninformative loci </span>
#&gt; <span class='message'> ============================ </span>
#&gt; <span class='message'> 0 loci found with a cutoff of 0 samples   </span>
#&gt; <span class='message'> 2 loci found with MAF &lt; 0.01 :</span>
#&gt; <span class='message'> x, y</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 100 individuals; 3 loci; 9 alleles; size: 16.5 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  100 x 9 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-3)
#&gt;    @loc.fac: locus factor for the 9 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt;    @type:  codom
#&gt;    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)
#&gt; 
#&gt;  // Optional content
#&gt;    - empty -</div><div class='input'>
<span class='co'># Real data</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>H3N2</span>)
<span class='fu'>informloci</span>(<span class='no'>H3N2</span>)</div><div class='output co'>#&gt; <span class='message'>cutoff value: 0.105097214923805 % ( 2 samples ).</span></div><div class='output co'>#&gt; <span class='message'>MAF         : 0.01</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'> Found 5 uninformative loci </span>
#&gt; <span class='message'> ============================ </span>
#&gt; <span class='message'> 1 locus found with a cutoff of 2 samples :</span>
#&gt; <span class='message'> 597 </span>
#&gt; <span class='message'> 5 loci found with MAF &lt; 0.01 :</span>
#&gt; <span class='message'> 42, 313, 433, 597, 915</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 1,903 individuals; 120 loci; 322 alleles; size: 3.3 Mb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  1903 x 322 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 2-4)
#&gt;    @loc.fac: locus factor for the 322 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 1-1)
#&gt;    @type:  codom
#&gt;    @call: .local(x = x, i = i, j = j, loc = ..1, drop = drop)
#&gt; 
#&gt;  // Optional content
#&gt;    @other: a list containing: x  xy  epid 
#&gt; </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Zhian N. Kamvar</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

