<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a table of summary statistics per locus. — locus_table • poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Create a table of summary statistics per locus. — locus_table" />

<meta property="og:description" content="Create a table of summary statistics per locus." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a table of summary statistics per locus.</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/Index_calculations.r'><code>R/Index_calculations.r</code></a></small>
    <div class="hidden name"><code>locus_table.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Create a table of summary statistics per locus.</p>
    
    </div>

    <pre class="usage"><span class='fu'>locus_table</span>(<span class='no'>x</span>, <span class='kw'>index</span> <span class='kw'>=</span> <span class='st'>"simpson"</span>, <span class='kw'>lev</span> <span class='kw'>=</span> <span class='st'>"allele"</span>, <span class='kw'>population</span> <span class='kw'>=</span> <span class='st'>"ALL"</span>,
  <span class='kw'>information</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>a genind or <a href='genclone-class.html'>genclone</a>
object.</p></td>
    </tr>
    <tr>
      <th>index</th>
      <td><p>Which diversity index to use. Choices are</p><ul>
<li><p><code>"simpson"</code> (Default) to give Simpson's index</p></li>
<li><p><code>"shannon"</code>
to give the Shannon-Wiener index</p></li>
<li><p><code>"invsimpson"</code> to give the
Inverse Simpson's index aka the Stoddard and Tayor index.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>lev</th>
      <td><p>At what level do you want to analyze diversity? Choices are
<code>"allele"</code> (Default) or <code>"genotype"</code>.</p></td>
    </tr>
    <tr>
      <th>population</th>
      <td><p>Select the populations to be analyzed. This is the
parameter <code>sublist</code> passed on to the function <code><a href='popsub.html'>popsub()</a></code>.
Defaults to <code>"ALL"</code>.</p></td>
    </tr>
    <tr>
      <th>information</th>
      <td><p>When <code>TRUE</code> (Default), this will print out a header
of information to the R console.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a table with 4 columns indicating the Number of alleles/genotypes
observed, Diversity index chosen, Nei's 1978 gene diversity (expected
heterozygosity), and Evenness.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>The calculation of <code>Hexp</code> is \((\frac{n}{n-1}) 1 - \sum_{i = 
  1}^k{p^{2}_{i}}\) where p is the allele
frequencies at a given locus and n is the number of observed alleles (Nei,
1978) in each locus and then returning the average. Caution should be
exercised in interpreting the results of Hexp with polyploid organisms with
ambiguous ploidy. The lack of allelic dosage information will cause rare
alleles to be over-represented and artificially inflate the index. This is
especially true with small sample sizes.</p>
<p>If <code>lev = "genotype"</code>, then all statistics reflect <strong>genotypic</strong> diversity
within each locus. This includes the calculation for <code>Hexp</code>, which turns
into the unbiased Simpson's index.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Jari Oksanen, F. Guillaume Blanchet, Roeland Kindt, Pierre Legendre, Peter
R. Minchin, R. B. O'Hara, Gavin L. Simpson, Peter Solymos, M. Henry H.
Stevens, and Helene Wagner. vegan: Community Ecology Package, 2012. R
package version 2.0-5.</p>
<p>Niklaus J. Gr"unwald, Stephen B. Goodwin, Michael G. Milgroom, and William
E. Fry. Analysis of genotypic diversity data for populations of
microorganisms. Phytopathology, 93(6):738-46, 2003</p>
<p>J.A. Ludwig and J.F. Reynolds. Statistical Ecology. A Primer on Methods and
Computing. New York USA: John Wiley and Sons, 1988.</p>
<p>E.C. Pielou. Ecological Diversity. Wiley, 1975.</p>
<p>J.A. Stoddart and J.F. Taylor. Genotypic diversity: estimation and
prediction in samples. Genetics, 118(4):705-11, 1988.</p>
<p>Masatoshi Nei. Estimation of average heterozygosity and genetic distance
from a small number of individuals. Genetics, 89(3):583-590, 1978.</p>
<p>Claude Elwood Shannon. A mathematical theory of communication. Bell Systems
Technical Journal, 27:379-423,623-656, 1948</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://www.rdocumentation.org/packages/vegan/topics/diversity'>vegan::diversity()</a></code>, <code><a href='poppr.html'>poppr()</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>nancycats</span>)
<span class='fu'>locus_table</span>(<span class='no'>nancycats</span>[<span class='kw'>pop</span> <span class='kw'>=</span> <span class='fl'>5</span>])</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>allele = Number of observed alleles</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>1-D = Simpson index</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Hexp = Nei's 1978 gene diversity</span></div><div class='output co'>#&gt; <span class='message'>------------------------------------------</span></div><div class='output co'>#&gt;        summary
#&gt; locus   allele  1-D Hexp Evenness
#&gt;   fca8    8.00 0.78 0.81     0.78
#&gt;   fca23   6.00 0.66 0.69     0.65
#&gt;   fca43   5.00 0.79 0.82     0.97
#&gt;   fca45   5.00 0.76 0.79     0.90
#&gt;   fca77   7.00 0.74 0.77     0.74
#&gt;   fca78   3.00 0.38 0.39     0.63
#&gt;   fca90   5.00 0.64 0.66     0.76
#&gt;   fca96   5.00 0.63 0.65     0.76
#&gt;   fca37   4.00 0.19 0.19     0.42
#&gt;   mean    5.33 0.62 0.64     0.73</div><div class='input'><span class='co'># Analyze locus statistics for the North American population of P. infestans.</span>
<span class='co'># Note that due to the unknown dosage of alleles, many of these statistics</span>
<span class='co'># will be artificially inflated for polyploids.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>Pinf</span>)
<span class='fu'>locus_table</span>(<span class='no'>Pinf</span>, <span class='kw'>population</span> <span class='kw'>=</span> <span class='st'>"North America"</span>)</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>allele = Number of observed alleles</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>1-D = Simpson index</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Hexp = Nei's 1978 gene diversity</span></div><div class='output co'>#&gt; <span class='message'>------------------------------------------</span></div><div class='output co'>#&gt;       summary
#&gt; locus  allele    1-D   Hexp Evenness
#&gt;   Pi02  9.000  0.690  0.697    0.653
#&gt;   D13  21.000  0.895  0.906    0.684
#&gt;   Pi33  2.000  0.021  0.021    0.353
#&gt;   Pi04  4.000  0.545  0.551    0.764
#&gt;   Pi4B  5.000  0.596  0.603    0.736
#&gt;   Pi16  6.000  0.425  0.430    0.498
#&gt;   G11  15.000  0.824  0.833    0.625
#&gt;   Pi56  3.000  0.335  0.338    0.647
#&gt;   Pi63  3.000  0.310  0.313    0.568
#&gt;   Pi70  2.000  0.203  0.205    0.595
#&gt;   Pi89 11.000  0.627  0.634    0.549
#&gt;   mean  7.364  0.497  0.503    0.607</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Zhian N. Kamvar</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

