<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Split samples from a genind object into pseudo-haplotypes — make_haplotypes • poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Split samples from a genind object into pseudo-haplotypes — make_haplotypes" />

<meta property="og:description" content="Split samples from a genind object into pseudo-haplotypes" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Split samples from a genind object into pseudo-haplotypes</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/methods.r'><code>R/methods.r</code></a></small>
    <div class="hidden name"><code>make_haplotypes-method.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Split samples from a genind object into pseudo-haplotypes</p>
    
    </div>

    <pre class="usage"><span class='fu'>make_haplotypes</span>(<span class='no'>gid</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gid</th>
      <td><p>a <a href='https://www.rdocumentation.org/packages/adegenet/topics/new.genind'>genind</a> or <a href='https://www.rdocumentation.org/packages/adegenet/topics/genlight'>genlight</a> object.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a haploid genind object with an extra <a href='https://www.rdocumentation.org/packages/adegenet/topics/strata-methods'>strata</a>
column called "Individual".</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Certain analyses, such as <a href='poppr.amova.html'>amova</a> work best if within-sample
variance (error) can be estimated. Practically, this is performed by
splitting the genotypes across all loci to create multiple haplotypes. This
way, the within-sample distance can be calculated and incorporated into the
model. Please note that the haplotypes generated are based on the order of
the unphased alleles in the genind object and do not represent true
haplotypes.</p>
<p>Haploid data will be returned un-touched.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>The <a href='https://www.rdocumentation.org/packages/adegenet/topics/accessors'>other slot</a> will not be copied over to the new
genind object.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='poppr.amova.html'>poppr.amova()</a></code> <code><a href='https://www.rdocumentation.org/packages/pegas/topics/amova'>pegas::amova()</a></code> <code><a href='coercion-methods.html'>as.genambig()</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Diploid data is doubled -------------------------------------------------</span>

<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>nancycats</span>)
<span class='no'>nan9</span> <span class='kw'>&lt;-</span> <span class='no'>nancycats</span>[<span class='kw'>pop</span> <span class='kw'>=</span> <span class='fl'>9</span>]
<span class='no'>nan9hap</span> <span class='kw'>&lt;-</span> <span class='fu'>make_haplotypes</span>(<span class='no'>nan9</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: No strata found... creating one from the population.</span></div><div class='input'><span class='no'>nan9</span>              <span class='co'># 9 individuals from population 9</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 9 individuals; 9 loci; 108 alleles; size: 28.6 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  9 x 108 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-8)
#&gt;    @loc.fac: locus factor for the 108 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt;    @type:  codom
#&gt;    @call: .local(x = x, i = i, j = j, pop = 9, drop = drop)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 9-9)
#&gt;    @other: a list containing: xy 
#&gt; </div><div class='input'><span class='no'>nan9hap</span>           <span class='co'># 18 haplotypes</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 18 individuals; 9 loci; 43 alleles; size: 21.3 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  18 x 43 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 3-8)
#&gt;    @loc.fac: locus factor for the 43 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 1-1)
#&gt;    @type:  codom
#&gt;    @call: df2genind(X = newdf, ploidy = 1, strata = df)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 2-2)
#&gt;    @strata: a data frame with 2 columns ( pop, Individual )</div><div class='input'><span class='fu'>strata</span>(<span class='no'>nan9hap</span>)   <span class='co'># strata gains a new column: Individual</span></div><div class='output co'>#&gt;    pop Individual
#&gt; 01 P09       N104
#&gt; 02 P09       N104
#&gt; 03 P09       N105
#&gt; 04 P09       N105
#&gt; 05 P09       N106
#&gt; 06 P09       N106
#&gt; 07 P09       N107
#&gt; 08 P09       N107
#&gt; 09 P09       N108
#&gt; 10 P09       N108
#&gt; 11 P09       N109
#&gt; 12 P09       N109
#&gt; 13 P09       N111
#&gt; 14 P09       N111
#&gt; 15 P09       N112
#&gt; 16 P09       N112
#&gt; 17 P09       N113
#&gt; 18 P09       N113</div><div class='input'><span class='fu'>indNames</span>(<span class='no'>nan9hap</span>) <span class='co'># individuals are renamed sequentially</span></div><div class='output co'>#&gt;  [1] "01" "02" "03" "04" "05" "06" "07" "08" "09" "10" "11" "12" "13" "14" "15"
#&gt; [16] "16" "17" "18"</div><div class='input'>

<span class='co'># Mix ploidy data can be split, but should be treated with caution --------</span>
<span class='co'># </span>
<span class='co'># For example, the Pinf data set contains 86 tetraploid individuals, </span>
<span class='co'># but there appear to only be diploids and triploid genotypes. When </span>
<span class='co'># we convert to haplotypes, those with all missing data are dropped.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>Pinf</span>)
<span class='no'>Pinf</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    72 multilocus genotypes 
#&gt;    86 tetraploid individuals
#&gt;    11 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     2 strata - Continent, Country
#&gt;     2 populations defined - South America, North America</div><div class='input'><span class='no'>pmiss</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='info_table.html'>info_table</a></span>(<span class='no'>Pinf</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"ploidy"</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='img'><img src='make_haplotypes-method-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># No samples appear to be triploid across all loci. This will cause</span>
<span class='co'># several haplotypes to have a lot of missing data.</span>
<span class='no'>p_haps</span> <span class='kw'>&lt;-</span> <span class='fu'>make_haplotypes</span>(<span class='no'>Pinf</span>)
<span class='no'>p_haps</span></div><div class='output co'>#&gt; /// GENIND OBJECT /////////
#&gt; 
#&gt;  // 203 individuals; 11 loci; 95 alleles; size: 137.9 Kb
#&gt; 
#&gt;  // Basic content
#&gt;    @tab:  203 x 95 matrix of allele counts
#&gt;    @loc.n.all: number of alleles per locus (range: 2-25)
#&gt;    @loc.fac: locus factor for the 95 columns of @tab
#&gt;    @all.names: list of allele names for each locus
#&gt;    @ploidy: ploidy of each individual  (range: 1-1)
#&gt;    @type:  codom
#&gt;    @call: df2genind(X = newdf, ploidy = 1, strata = df)
#&gt; 
#&gt;  // Optional content
#&gt;    @pop: population of each individual (group size range: 2-3)
#&gt;    @strata: a data frame with 3 columns ( Continent, Country, Individual )</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/head'>head</a></span>(<span class='fu'>genind2df</span>(<span class='no'>p_haps</span>), <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>20</span>)</div><div class='output co'>#&gt;       pop Pi02  D13 Pi33 Pi04 Pi4B Pi16  G11 Pi56 Pi63 Pi70 Pi89
#&gt; 2  PiCO01 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;
#&gt; 3  PiCO01  162  136  203  166  213  178  156  174  157  192  179
#&gt; 4  PiCO01  162  136  203  170  217  178  156  176  157  192  181
#&gt; 6  PiCO02 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;
#&gt; 7  PiCO02  162  132  203  166  213  178  156  174  157  192  179
#&gt; 8  PiCO02  162  136  203  170  217  178  156  176  157  192  181
#&gt; 10 PiCO03 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;
#&gt; 11 PiCO03  162  136  203  166  213  178  156  174  157  192  179
#&gt; 12 PiCO03  162  136  203  170  217  178  156  176  157  192  181
#&gt; 14 PiCO04 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;
#&gt; 15 PiCO04  162  136  203  166  213  178  156  174  157  192  179
#&gt; 16 PiCO04  162  136  203  170  217  178  156  176  157  192  181
#&gt; 19 PiCO05  160  108  203  166  213  174  156  176  157  192  181
#&gt; 20 PiCO05  162  112  203  170  225  178  156  176  157  192  181
#&gt; 23 PiEC01  162  134  203  166  213  178  156  174  157  192  179
#&gt; 24 PiEC01  162  136  203  170  217  178  156  176  157  192  181
#&gt; 26 PiEC02 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;  205 &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;
#&gt; 27 PiEC02  162  136  203  166  213  178  156  174  157  192  179
#&gt; 28 PiEC02  162  144  203  170  217  178  156  176  157  192  181
#&gt; 31 PiEC03  162  136  203  166  205  178  156  174  157  192  179</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

