<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create counts, vectors, and matrices of multilocus genotypes. — mlg • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create counts, vectors, and matrices of multilocus genotypes.</h1>
    </div>

    
    <p>Create counts, vectors, and matrices of multilocus genotypes.</p>
    

    <pre class="usage"><span class='fu'>mlg</span>(<span class='no'>gid</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>mlg.table</span>(<span class='no'>gid</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sublist</span> <span class='kw'>=</span> <span class='st'>"ALL"</span>, <span class='kw'>blacklist</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>mlgsub</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>bar</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>total</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>color</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>background</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>mlg.vector</span>(<span class='no'>gid</span>, <span class='kw'>reset</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>mlg.crosspop</span>(<span class='no'>gid</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sublist</span> <span class='kw'>=</span> <span class='st'>"ALL"</span>, <span class='kw'>blacklist</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>mlgsub</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>indexreturn</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>df</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>mlg.id</span>(<span class='no'>gid</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gid</th>
      <td><p>a <code>genind</code>, <code><a href='genclone-class.html'>genclone</a></code>,
<code>genlight</code>, or <code><a href='genclone-class.html'>snpclone</a></code> object.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p><code>Logical</code>. If FALSE, progress of functions will be printed 
to the screen.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>a formula specifying the strata at which computation is to be
performed.</p></td>
    </tr>
    <tr>
      <th>sublist</th>
      <td><p>a <code>vector</code> of population names or indices that the user 
wishes to keep. Default to "ALL".</p></td>
    </tr>
    <tr>
      <th>blacklist</th>
      <td><p>a <code>vector</code> of population names or indices that the user
wishes to discard. Default to <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>mlgsub</th>
      <td><p>a <code>vector</code> of multilocus genotype indices with which to 
subset <code>mlg.table</code> and <code>mlg.crosspop</code>. NOTE: The resulting table 
from <code>mlg.table</code> will only contain countries with those MLGs</p></td>
    </tr>
    <tr>
      <th>bar</th>
      <td><p>deprecated. Same as <code>plot</code>. Retained for compatibility.</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p><code>logical</code> If <code>TRUE</code>, a bar graph for each population 
will be displayed showing the relative abundance of each MLG within the 
population.</p></td>
    </tr>
    <tr>
      <th>total</th>
      <td><p><code>logical</code> If <code>TRUE</code>, a row containing the sum of all 
represented MLGs is appended to the matrix produced by mlg.table.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p>an option to display a single barchart for mlg.table, colored by
population (note, this becomes facetted if <code>background = TRUE</code>).</p></td>
    </tr>
    <tr>
      <th>background</th>
      <td><p>an option to display the the total number of MLGs across
populations per facet in the background of the plot.</p></td>
    </tr>
    <tr>
      <th>reset</th>
      <td><p>logical. For genclone objects, the MLGs are defined by the input
data, but they do not change if more or less information is added (i.e.
loci are dropped). Setting <code>reset = TRUE</code> will recalculate MLGs.
Default is <code>FALSE</code>, returning the MLGs defined in the @mlg slot.</p></td>
    </tr>
    <tr>
      <th>indexreturn</th>
      <td><p><code>logical</code> If <code>TRUE</code>, a vector will be returned 
to index the columns of <code>mlg.table</code>.</p></td>
    </tr>
    <tr>
      <th>df</th>
      <td><p><code>logical</code> If <code>TRUE</code>, return a data frame containing the 
counts of the MLGs and what countries they are in. Useful for making graphs
with <code>ggplot</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<h3>mlg</h3>
 an integer describing the number of multilocus
genotypes observed. <p></p><h3>mlg.table</h3>
 a matrix with columns
indicating unique multilocus genotypes and rows indicating populations. 
This table can be used with the funciton <code><a href='diversity_stats.html'>diversity_stats</a></code> to calculate
the Shannon-Weaver index (H), Stoddart and Taylor's 
index (aka inverse Simpson's index; G), Simpson's index (lambda), and evenness (E5). 
<h3>mlg.vector</h3>
 a numeric vector naming the multilocus genotype of
each individual in the dataset. <p></p><h3>mlg.crosspop</h3>
 <ul>
<li><p>default a <code>list</code> where each element contains a named integer
vector representing the number of individuals represented from each
population in that MLG</p></li>
<li><p><code>indexreturn = TRUE</code> a <code>vector</code> of
integers defining the multilocus genotypes that have individuals crossing
populations</p></li>
<li><p><code>df = TRUE</code> A long form data frame with the
columns: MLG, Population, Count. Useful for graphing with ggplot2</p></li>
</ul>  
<h3>mlg.id</h3>
 a list of multilocus genotypes with the associated
individual names per MLG. 

<p>a <code>list</code> containing vectors of population names for each MLG.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Multilocus genotypes are the unique combination of alleles across
  all loci. For details of how these are calculated see <code><a href='https://cran.rstudio.com/web/packages/poppr/vignettes/mlg.html'>vignette("mlg",
  package = "poppr")</a></code>. In short, for genind and genclone objects, they are
  calculated by using a rank function on strings of alleles, which is
  sensitive to missing data. For genlight and snpclone objects, they are
  calculated with distance methods via <code><a href='bitwise.dist.html'>bitwise.dist</a></code> and
  <code><a href='mlg.filter.html'>mlg.filter</a></code>, which means that these are insensitive to missing
  data. Three different types of MLGs can be defined in <span class="pkg">poppr</span>:</p><ul>
<li><p>original - the default definition of multilocus genotypes as
  detailed above</p></li>
<li><p>contracted - these are multilocus genotypes
  collapsed into multilocus lineages (<code><a href='mll-method.html'>mll</a></code>) with genetic
  distance via <code><a href='mlg.filter.html'>mlg.filter</a></code></p></li>
<li><p>custom - user-defined
  multilocus genotypes. These are useful for information such as mycelial
  compatibility groups</p></li>
</ul><p><strong>All of the functions documented here will work on any of the MLG
  types defined in <span class="pkg">poppr</span></strong></p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>The resulting matrix of <code>mlg.table</code> can be used for analysis with 
the <code>vegan</code> package.</p>
<p>mlg.vector will recalculate the mlg vector for
  <code>genind</code> objects and will return the contents of the mlg
  slot in <code><a href='genclone-class.html'>genclone</a></code> objects. This means that MLGs will be
  different for subsetted <code>genind</code> objects.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='http://www.rdocumentation.org/packages/vegan/topics/diversity'>diversity</a></code> 
 <code><a href='diversity_stats.html'>diversity_stats</a></code>
 <code><a href='popsub.html'>popsub</a></code>
 <code><a href='mll-method.html'>mll</a></code>
 <code><a href='mlg.filter.html'>mlg.filter</a></code>
 <code><a href='mll.custom.html'>mll.custom</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># Load the data set</span>
<span class='fu'>data</span>(<span class='no'>Aeut</span>)

<span class='co'># Investigate the number of multilocus genotypes.</span>
<span class='no'>amlg</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg</span>(<span class='no'>Aeut</span>)</div><div class='output co'>#&gt; #############################
#&gt; # Number of Individuals:  187 
#&gt; # Number of MLG:  119 
#&gt; #############################</div><div class='input'><span class='no'>amlg</span> <span class='co'># 119</span></div><div class='output co'>#&gt; [1] 119</div><div class='input'>
<span class='co'># show the multilocus genotype vector </span>
<span class='no'>avec</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.vector</span>(<span class='no'>Aeut</span>)
<span class='no'>avec</span></div><div class='output co'>#&gt;   [1]  63  53  17  16  26  42   8  17  16  93  94 112  44 118  51 113  45 110
#&gt;  [19] 114   7 111  20  20  20  20  30  20  20  20  20  20  28  66  66  68  66
#&gt;  [37]  67  23  66  65  66  57  22 119  36  35  36  35  62  35  34  52  33  63
#&gt;  [55]   1  39  59  61  23  21  47  48  25  50  32  49  60   9  10  60  27  13
#&gt;  [73]  11  14  12  14  14  29  15  13  32  41  43  24  46  52  38  37  31  24
#&gt;  [91]  40  31 109  19  18  64 108  86  82  83   2   6   2  91 117 116  80  80
#&gt; [109]  58   5   3   4  56  84  54  73  80  89 106  80  80  80  88  70  76  70
#&gt; [127]  70  90 104  81  70  92  70  75  75  99  75  96  74  74  74  75  98  96
#&gt; [145]  69  69  75  69  69 100  55  72 105  72 115  71  85  87  59  77 107  76
#&gt; [163]  90  80  99  88  80  77  78  77  87  80  91 101  79 102  78 103 103  95
#&gt; [181]  97  81  80  97  97  97 101</div><div class='input'>
<span class='co'># Get a table</span>
<span class='no'>atab</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.table</span>(<span class='no'>Aeut</span>, <span class='kw'>color</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='img'><img src='mlg-3.png' alt='' width='700' height='432.632880098887' /></div><div class='input'><span class='no'>atab</span></div><div class='output co'>#&gt;            MLG.1 MLG.2 MLG.3 MLG.4 MLG.5 MLG.6 MLG.7 MLG.8 MLG.9 MLG.10 MLG.11
#&gt; Athena         1     0     0     0     0     0     1     1     1      1      1
#&gt; Mt. Vernon     0     2     1     1     1     1     0     0     0      0      0
#&gt;            MLG.12 MLG.13 MLG.14 MLG.15 MLG.16 MLG.17 MLG.18 MLG.19 MLG.20
#&gt; Athena          1      2      3      1      2      2      1      1      9
#&gt; Mt. Vernon      0      0      0      0      0      0      0      0      0
#&gt;            MLG.21 MLG.22 MLG.23 MLG.24 MLG.25 MLG.26 MLG.27 MLG.28 MLG.29
#&gt; Athena          1      1      2      2      1      1      1      1      1
#&gt; Mt. Vernon      0      0      0      0      0      0      0      0      0
#&gt;            MLG.30 MLG.31 MLG.32 MLG.33 MLG.34 MLG.35 MLG.36 MLG.37 MLG.38
#&gt; Athena          1      2      2      1      1      3      2      1      1
#&gt; Mt. Vernon      0      0      0      0      0      0      0      0      0
#&gt;            MLG.39 MLG.40 MLG.41 MLG.42 MLG.43 MLG.44 MLG.45 MLG.46 MLG.47
#&gt; Athena          1      1      1      1      1      1      1      1      1
#&gt; Mt. Vernon      0      0      0      0      0      0      0      0      0
#&gt;            MLG.48 MLG.49 MLG.50 MLG.51 MLG.52 MLG.53 MLG.54 MLG.55 MLG.56
#&gt; Athena          1      1      1      1      2      1      0      0      0
#&gt; Mt. Vernon      0      0      0      0      0      0      1      1      1
#&gt;            MLG.57 MLG.58 MLG.59 MLG.60 MLG.61 MLG.62 MLG.63 MLG.64 MLG.65
#&gt; Athena          1      0      1      2      1      1      2      1      1
#&gt; Mt. Vernon      0      1      1      0      0      0      0      0      0
#&gt;            MLG.66 MLG.67 MLG.68 MLG.69 MLG.70 MLG.71 MLG.72 MLG.73 MLG.74
#&gt; Athena          5      1      1      0      0      0      0      0      0
#&gt; Mt. Vernon      0      0      0      4      5      1      2      1      3
#&gt;            MLG.75 MLG.76 MLG.77 MLG.78 MLG.79 MLG.80 MLG.81 MLG.82 MLG.83
#&gt; Athena          0      0      0      0      0      0      0      0      0
#&gt; Mt. Vernon      5      2      3      2      1     10      2      1      1
#&gt;            MLG.84 MLG.85 MLG.86 MLG.87 MLG.88 MLG.89 MLG.90 MLG.91 MLG.92
#&gt; Athena          0      0      0      0      0      0      0      0      0
#&gt; Mt. Vernon      1      1      1      2      2      1      2      2      1
#&gt;            MLG.93 MLG.94 MLG.95 MLG.96 MLG.97 MLG.98 MLG.99 MLG.100 MLG.101
#&gt; Athena          1      1      0      0      0      0      0       0       0
#&gt; Mt. Vernon      0      0      1      2      4      1      2       1       2
#&gt;            MLG.102 MLG.103 MLG.104 MLG.105 MLG.106 MLG.107 MLG.108 MLG.109
#&gt; Athena           0       0       0       0       0       0       1       1
#&gt; Mt. Vernon       1       2       1       1       1       1       0       0
#&gt;            MLG.110 MLG.111 MLG.112 MLG.113 MLG.114 MLG.115 MLG.116 MLG.117
#&gt; Athena           1       1       1       1       1       0       0       0
#&gt; Mt. Vernon       0       0       0       0       0       1       1       1
#&gt;            MLG.118 MLG.119
#&gt; Athena           1       1
#&gt; Mt. Vernon       0       0</div><div class='input'>
<span class='co'># See where multilocus genotypes cross populations</span>
<span class='no'>acrs</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.crosspop</span>(<span class='no'>Aeut</span>) <span class='co'># MLG.59: (2 inds) Athena Mt. Vernon</span></div><div class='output co'>#&gt; MLG.59: (2 inds) Athena Mt. Vernon</div><div class='input'>
<span class='co'># See which individuals belong to each MLG</span>
<span class='no'>aid</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.id</span>(<span class='no'>Aeut</span>)
<span class='no'>aid</span>[<span class='st'>"59"</span>] <span class='co'># individuals 159 and 57</span></div><div class='output co'>#&gt; $`59`
#&gt; [1] "057" "159"
#&gt; </div><div class='input'>
</div><span class='co'># NOT RUN {</span>
<span class='co'># For the mlg.table, you can also choose to display the number of MLGs across</span>
<span class='co'># populations in the background</span>

<span class='fu'>mlg.table</span>(<span class='no'>Aeut</span>, <span class='kw'>background</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='fu'>mlg.table</span>(<span class='no'>Aeut</span>, <span class='kw'>background</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>color</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># A simple example. 10 individuals, 5 genotypes.</span>
<span class='no'>mat1</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>5</span>, <span class='fl'>25</span>:<span class='fl'>1</span>)
<span class='no'>mat1</span> <span class='kw'>&lt;-</span> <span class='fu'>rbind</span>(<span class='no'>mat1</span>, <span class='no'>mat1</span>)
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='kw'>nrow</span><span class='kw'>=</span><span class='fl'>10</span>, <span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>5</span>, <span class='fu'>paste</span>(<span class='no'>mat1</span>,<span class='no'>mat1</span>,<span class='kw'>sep</span><span class='kw'>=</span><span class='st'>"/"</span>))
<span class='no'>mat.gid</span> <span class='kw'>&lt;-</span> <span class='fu'>df2genind</span>(<span class='no'>mat</span>, <span class='kw'>sep</span><span class='kw'>=</span><span class='st'>"/"</span>)
<span class='fu'>mlg</span>(<span class='no'>mat.gid</span>)
<span class='fu'>mlg.vector</span>(<span class='no'>mat.gid</span>)
<span class='fu'>mlg.table</span>(<span class='no'>mat.gid</span>)

<span class='co'># Now for a more complicated example.</span>
<span class='co'># Data set of 1903 samples of the H3N2 flu virus genotyped at 125 SNP loci.</span>
<span class='fu'>data</span>(<span class='no'>H3N2</span>)
<span class='fu'>mlg</span>(<span class='no'>H3N2</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='no'>H.vec</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.vector</span>(<span class='no'>H3N2</span>)

<span class='co'># Changing the population vector to indicate the years of each epidemic.</span>
<span class='fu'>pop</span>(<span class='no'>H3N2</span>) <span class='kw'>&lt;-</span> <span class='fu'>other</span>(<span class='no'>H3N2</span>)$<span class='no'>x</span>$<span class='no'>country</span>
<span class='no'>H.tab</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.table</span>(<span class='no'>H3N2</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>total</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># Show which genotypes exist accross populations in the entire dataset.</span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.crosspop</span>(<span class='no'>H3N2</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># Let's say we want to visualize the multilocus genotype distribution for the</span>
<span class='co'># USA and Russia</span>
<span class='fu'>mlg.table</span>(<span class='no'>H3N2</span>, <span class='kw'>sublist</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"USA"</span>, <span class='st'>"Russia"</span>), <span class='kw'>bar</span><span class='kw'>=</span><span class='fl'>TRUE</span>)

<span class='co'># An exercise in subsetting the output of mlg.table and mlg.vector.</span>
<span class='co'># First, get the indices of each MLG duplicated across populations.</span>
<span class='no'>inds</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.crosspop</span>(<span class='no'>H3N2</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>indexreturn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># Since the columns of the table from mlg.table are equal to the number of</span>
<span class='co'># MLGs, we can subset with just the columns.</span>
<span class='no'>H.sub</span> <span class='kw'>&lt;-</span> <span class='no'>H.tab</span>[, <span class='no'>inds</span>]

<span class='co'># We can also do the same by using the mlgsub flag.</span>
<span class='no'>H.sub</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.table</span>(<span class='no'>H3N2</span>, <span class='kw'>mlgsub</span> <span class='kw'>=</span> <span class='no'>inds</span>)

<span class='co'># We can subset the original data set using the output of mlg.vector to</span>
<span class='co'># analyze only the MLGs that are duplicated across populations. </span>
<span class='no'>new.H</span> <span class='kw'>&lt;-</span> <span class='no'>H3N2</span>[<span class='no'>H.vec</span> <span class='kw'>%in%</span> <span class='no'>inds</span>, ]

<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
