<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MLG definitions based on genetic distance — mlg.filter • poppr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="MLG definitions based on genetic distance — mlg.filter" />

<meta property="og:description" content="Multilocus genotypes are initially defined by naive string matching, but this
definition does not take into account missing data or genotyping error,
casting these as unique genotypes. Defining multilocus genotypes by genetic
distance allows you to incorporate genotypes that have missing data o
genotyping error into their parent clusters." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.8.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MLG definitions based on genetic distance</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/methods.r'><code>R/methods.r</code></a></small>
    <div class="hidden name"><code>mlg.filter.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Multilocus genotypes are initially defined by naive string matching, but this
definition does not take into account missing data or genotyping error,
casting these as unique genotypes. Defining multilocus genotypes by genetic
distance allows you to incorporate genotypes that have missing data o
genotyping error into their parent clusters.</p>
    
    </div>

    <pre class="usage"><span class='fu'>mlg.filter</span>(<span class='no'>pop</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>missing</span> <span class='kw'>=</span> <span class='st'>"asis"</span>, <span class='kw'>memory</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>algorithm</span> <span class='kw'>=</span> <span class='st'>"farthest_neighbor"</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='st'>"diss.dist"</span>,
  <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='kw'>stats</span> <span class='kw'>=</span> <span class='st'>"MLGs"</span>, <span class='no'>...</span>)

<span class='fu'>mlg.filter</span>(<span class='no'>pop</span>, <span class='kw'>missing</span> <span class='kw'>=</span> <span class='st'>"asis"</span>, <span class='kw'>memory</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>algorithm</span> <span class='kw'>=</span> <span class='st'>"farthest_neighbor"</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='st'>"diss.dist"</span>,
  <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='no'>...</span>) <span class='kw'>&lt;-</span> <span class='no'>value</span></pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pop</th>
      <td><p>a <code><a href='genclone-class.html'>genclone</a></code>, <code><a href='genclone-class.html'>snpclone</a></code>, or
<code>genind</code> object.</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>a number indicating the minimum distance two MLGs must be
separated by to be considered different. Defaults to 0, which will reflect
the original (naive) MLG definition.</p></td>
    </tr>
    <tr>
      <th>missing</th>
      <td><p>any method to be used by <code><a href='missingno.html'>missingno</a></code>: "mean", 
"zero", "loci", "genotype", or "asis" (default).</p></td>
    </tr>
    <tr>
      <th>memory</th>
      <td><p>whether this function should remember the last distance matrix 
it generated. TRUE will attempt to reuse the last distance matrix if the 
other parameters are the same. (default) FALSE will ignore any stored 
matrices and not store any it generates.</p></td>
    </tr>
    <tr>
      <th>algorithm</th>
      <td><p>determines the type of clustering to be done.</p><dl class='dl-horizontal'>
  <dt>"farthest_neighbor"</dt><dd><p><em> (default) </em>merges clusters based on the 
  maximum distance between points in either cluster. This is the strictest of
  the three.</p></dd>
  <dt>"nearest_neighbor"</dt><dd><p>merges clusters based on the minimum distance
  between points in either cluster. This is the loosest of the three.</p></dd>
  <dt>"average_neighbor"</dt><dd><p>merges clusters based on the average distance
  between every pair of points between clusters.</p></dd>
</dl></td>
    </tr>
    <tr>
      <th>distance</th>
      <td><p>a character or function defining the distance to be applied 
to pop. Defaults to <code><a href='diss.dist.html'>diss.dist</a></code> for genclone objects and
<code><a href='bitwise.dist.html'>bitwise.dist</a></code> for snpclone objects. A matrix or table
containing distances between individuals (such as the output of 
<code><a href='genetic_distance.html'>rogers.dist</a></code>) is also accepted for this parameter.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>(unused) Previously, this was the maximum number of parallel 
threads to be used within this function. Default is 1 indicating that this
function will run serially. Any other number will result in a warning.</p></td>
    </tr>
    <tr>
      <th>stats</th>
      <td><p>a character vector specifying which statistics should be
returned (details below). Choices are "MLG", "THRESHOLDS", "DISTANCES",
"SIZES", or "ALL". If choosing "ALL" or more than one, a named list will be
returned.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>any parameters to be passed off to the distance method.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>the threshold at which genotypes should be collapsed.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Default, a vector of collapsed multilocus genotypes. Otherwise, any
  combination of the following:</p><h3>MLGs</h3>

  a numeric vector defining the multilocus genotype cluster of each
  individual in the dataset. Each genotype cluster is separated from every
  other genotype cluster by at least the defined threshold value, as 
  calculated by the selected algorithm.

<h3>THRESHOLDS</h3>

  A numeric vector representing the thresholds <strong>beyond</strong> which clusters
  of multilocus genotypes were collapsed. 

<h3>DISTANCES</h3>

  A square matrix representing the distances between each cluster.

<h3>SIZES</h3>

 The sizes of the multilocus genotype clusters in order. 


    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function will take in any distance matrix or function and
collapse multilocus genotypes below a given threshold. If you use this
function as the assignment method (mlg.filter(myData, distance = myDist) &lt;-
0.5), the distance function or matrix will be remembered by the object. This
means that if you define your own distance matrix or function, you must keep
it in memory to further utilize mlg.filter.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><code>mlg.vector</code> makes use of <code>mlg.vector</code> grouping prior to 
  applying the given threshold. Genotype numbers returned by
  <code>mlg.vector</code> represent the lowest numbered genotype (as returned by
  <code>mlg.vector</code>) in in each new multilocus genotype. Therefore
  <strong><code>mlg.filter</code> and <code>mlg.vector</code> return the same vector when
  threshold is set to 0 or less</strong>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='filter_stats.html'>filter_stats</a></code>, 
  <code><a href='cutoff_predictor.html'>cutoff_predictor</a></code>, 
  <code><a href='mll-method.html'>mll</a></code>, 
  <code><a href='genclone-class.html'>genclone</a></code>, 
  <code><a href='genclone-class.html'>snpclone</a></code>, 
  <code><a href='diss.dist.html'>diss.dist</a></code>, 
  <code><a href='bruvo.dist.html'>bruvo.dist</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>partial_clone</span>)
<span class='no'>pc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='coercion-methods.html'>as.genclone</a></span>(<span class='no'>partial_clone</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>) <span class='co'># convert to genclone object</span>

<span class='co'># Basic Use ---------------------------------------------------------------</span>


<span class='co'># Show MLGs at threshold 0.05</span>
<span class='fu'>mlg.filter</span>(<span class='no'>pc</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='st'>"nei.dist"</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>)</div><div class='output co'>#&gt;  [1]  8  7 23 24 22 21 10  3 22 11 24  7 25  4 12  2 14  1  7  7  7 26  7 13 23
#&gt; [26]  3 17 22  6 20 22 12  5 25 13 21 15 13 13 13  2 19 18 13 23 16  1 11 25  4</div><div class='input'><span class='no'>pc</span> <span class='co'># 26 mlgs</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    26 original multilocus genotypes 
#&gt;    50 diploid individuals
#&gt;    10 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     0 strata. 
#&gt;     4 populations defined - 1, 2, 3, 4</div><div class='input'>
<span class='co'># Set MLGs at threshold 0.05</span>
<span class='fu'>mlg.filter</span>(<span class='no'>pc</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='st'>"nei.dist"</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1L</span>) <span class='kw'>&lt;-</span> <span class='fl'>0.05</span>
<span class='no'>pc</span> <span class='co'># 25 mlgs</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    25 contracted multilocus genotypes
#&gt;       (0.05) [t], (nei.dist) [d], (farthest) [a] 
#&gt;    50 diploid individuals
#&gt;    10 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     0 strata. 
#&gt;     4 populations defined - 1, 2, 3, 4</div><div class='input'>
</div><div class='input'>
<span class='co'># The distance definition is persistant</span>
<span class='fu'>mlg.filter</span>(<span class='no'>pc</span>) <span class='kw'>&lt;-</span> <span class='fl'>0.1</span>
<span class='no'>pc</span> <span class='co'># 24 mlgs</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    24 contracted multilocus genotypes
#&gt;       (0.1) [t], (nei.dist) [d], (farthest) [a] 
#&gt;    50 diploid individuals
#&gt;    10 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     0 strata. 
#&gt;     4 populations defined - 1, 2, 3, 4</div><div class='input'>
<span class='co'># But you can still change the definition</span>
<span class='fu'>mlg.filter</span>(<span class='no'>pc</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='st'>"diss.dist"</span>, <span class='kw'>percent</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='kw'>&lt;-</span> <span class='fl'>0.1</span>
<span class='no'>pc</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    25 contracted multilocus genotypes
#&gt;       (0.1) [t], (diss.dist) [d], (farthest) [a] 
#&gt;    50 diploid individuals
#&gt;    10 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     0 strata. 
#&gt;     4 populations defined - 1, 2, 3, 4</div><div class='input'>
<span class='co'># Choosing a threshold ----------------------------------------------------</span>


<span class='co'># Thresholds for collapsing multilocus genotypes should not be arbitrary. It</span>
<span class='co'># is important to consider what threshold is suitable. One method of choosing</span>
<span class='co'># a threshold is to find a gap in the distance distribution that represents</span>
<span class='co'># clonal groups. You can look at this by analyzing the distribution of all</span>
<span class='co'># possible thresholds with the function "cutoff_predictor".</span>

<span class='co'># For this example, we'll use Bruvo's distance to predict the cutoff for</span>
<span class='co'># P. infestans.</span>

<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>Pinf</span>)
<span class='no'>Pinf</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    72 multilocus genotypes 
#&gt;    86 tetraploid individuals
#&gt;    11 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     2 strata - Continent, Country
#&gt;     2 populations defined - South America, North America</div><div class='input'><span class='co'># Repeat lengths are necessary for Bruvo's distance</span>
(<span class='no'>pinfreps</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='fix_replen.html'>fix_replen</a></span>(<span class='no'>Pinf</span>, <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>2</span>, <span class='fl'>2</span>, <span class='fl'>6</span>, <span class='fl'>2</span>, <span class='fl'>2</span>, <span class='fl'>2</span>, <span class='fl'>2</span>, <span class='fl'>2</span>, <span class='fl'>3</span>, <span class='fl'>3</span>, <span class='fl'>2</span>)))</div><div class='output co'>#&gt;    Pi02     D13    Pi33    Pi04    Pi4B    Pi16     G11    Pi56    Pi63    Pi70 
#&gt; 2.00000 2.00000 6.00000 2.00000 1.99999 2.00000 2.00000 2.00000 3.00000 3.00000 
#&gt;    Pi89 
#&gt; 1.99999 </div><div class='input'>
<span class='co'># Now we can collect information of the thresholds. We can set threshold = 1</span>
<span class='co'># because we know that this will capture the maximum possible distance:</span>
(<span class='no'>thresholds</span> <span class='kw'>&lt;-</span> <span class='fu'>mlg.filter</span>(<span class='no'>Pinf</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='no'>bruvo.dist</span>, <span class='kw'>stats</span> <span class='kw'>=</span> <span class='st'>"THRESHOLDS"</span>,
                          <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>pinfreps</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>1</span>))</div><div class='output co'>#&gt;  [1] 0.01262626 0.02189867 0.02272727 0.02272727 0.03535354 0.04166667
#&gt;  [7] 0.04261364 0.04545454 0.04699337 0.04758520 0.05681818 0.05681818
#&gt; [13] 0.05835701 0.06451231 0.06534091 0.06818182 0.07871686 0.07954545
#&gt; [19] 0.08772748 0.08877841 0.09375000 0.09469697 0.09722222 0.10144413
#&gt; [25] 0.12500000 0.13099747 0.13593750 0.13740234 0.14488636 0.15000000
#&gt; [31] 0.15656566 0.16688366 0.18115234 0.19317072 0.21022726 0.21590909
#&gt; [37] 0.21874983 0.22561553 0.23115234 0.23295450 0.23437499 0.23532197
#&gt; [43] 0.24147723 0.25850032 0.27201702 0.27500000 0.28338066 0.29208984
#&gt; [49] 0.30000000 0.30042336 0.30326990 0.30549961 0.30632250 0.31325684
#&gt; [55] 0.33639034 0.33709162 0.34089797 0.35170815 0.35619658 0.36079233
#&gt; [61] 0.36487924 0.39646464 0.40280346 0.40332030 0.42325106 0.43731965
#&gt; [67] 0.45436164 0.47047212 0.50827603 0.51496688 0.57474140</div><div class='input'><span class='co'># We can use these thresholds to find an appropriate cutoff</span>
(<span class='no'>pcut</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='cutoff_predictor.html'>cutoff_predictor</a></span>(<span class='no'>thresholds</span>))</div><div class='output co'>#&gt; [1] 0.1132221</div><div class='input'><span class='fu'>mlg.filter</span>(<span class='no'>Pinf</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='no'>bruvo.dist</span>, <span class='kw'>replen</span> <span class='kw'>=</span> <span class='no'>pinfreps</span>) <span class='kw'>&lt;-</span> <span class='no'>pcut</span>
<span class='no'>Pinf</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    48 contracted multilocus genotypes
#&gt;       (0.113) [t], (bruvo.dist) [d], (farthest) [a] 
#&gt;    86 tetraploid individuals
#&gt;    11 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     2 strata - Continent, Country
#&gt;     2 populations defined - South America, North America</div><div class='input'>
<span class='co'># This can also be visualized with the "filter_stats" function.</span>

<span class='co'># Special case: threshold = 0 ---------------------------------------------</span>


<span class='co'># It's important to remember that a threshold of 0 is equal to the original</span>
<span class='co'># MLG definition. This example will show a data set that contains genotypes</span>
<span class='co'># with missing data that share all alleles with other genotypes except for </span>
<span class='co'># the missing one.</span>

<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>monpop</span>)
<span class='no'>monpop</span> <span class='co'># 264 mlg</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    264 multilocus genotypes 
#&gt;    694 haploid individuals
#&gt;     13 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;      1 stratum - Pop
#&gt;     12 populations defined - 
#&gt; 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</div><div class='input'><span class='fu'>mlg.filter</span>(<span class='no'>monpop</span>) <span class='kw'>&lt;-</span> <span class='fl'>0</span>
<span class='fu'><a href='mll-method.html'>nmll</a></span>(<span class='no'>monpop</span>) <span class='co'># 264 mlg</span></div><div class='output co'>#&gt; [1] 264</div><div class='input'>
<span class='co'># In order to merge these genotypes with missing data, we should set the </span>
<span class='co'># threshold to be slightly higher than 0. We will use the smallest fraction </span>
<span class='co'># the computer can store.</span>

<span class='fu'>mlg.filter</span>(<span class='no'>monpop</span>) <span class='kw'>&lt;-</span> <span class='no'>.Machine</span>$<span class='no'>double.eps</span> ^ <span class='fl'>0.5</span>
<span class='fu'><a href='mll-method.html'>nmll</a></span>(<span class='no'>monpop</span>) <span class='co'># 236 mlg</span></div><div class='output co'>#&gt; [1] 236</div><div class='input'>
<span class='co'># Custom distance ---------------------------------------------------------</span>

<span class='co'># Custom genetic distances can be used either in functions from other</span>
<span class='co'># packages or user-defined functions</span>

<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>Pinf</span>)
<span class='no'>Pinf</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    48 contracted multilocus genotypes
#&gt;       (0.113) [t], (bruvo.dist) [d], (farthest) [a] 
#&gt;    86 tetraploid individuals
#&gt;    11 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     2 strata - Continent, Country
#&gt;     2 populations defined - South America, North America</div><div class='input'><span class='fu'>mlg.filter</span>(<span class='no'>Pinf</span>, <span class='kw'>distance</span> <span class='kw'>=</span> <span class='kw'>function</span>(<span class='no'>x</span>) <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/dist'>dist</a></span>(<span class='fu'>tab</span>(<span class='no'>x</span>))) <span class='kw'>&lt;-</span> <span class='fl'>3</span>
<span class='no'>Pinf</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    48 contracted multilocus genotypes
#&gt;       (3) [t], (function(x) dist(tab(x))) [d], (farthest) [a] 
#&gt;    86 tetraploid individuals
#&gt;    11 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     2 strata - Continent, Country
#&gt;     2 populations defined - South America, North America</div><div class='input'><span class='fu'>mlg.filter</span>(<span class='no'>Pinf</span>) <span class='kw'>&lt;-</span> <span class='fl'>4</span>
<span class='no'>Pinf</span></div><div class='output co'>#&gt; 
#&gt; This is a genclone object
#&gt; -------------------------
#&gt; Genotype information:
#&gt; 
#&gt;    33 contracted multilocus genotypes
#&gt;       (4) [t], (function(x) dist(tab(x))) [d], (farthest) [a] 
#&gt;    86 tetraploid individuals
#&gt;    11 codominant loci
#&gt; 
#&gt; Population information:
#&gt; 
#&gt;     2 strata - Continent, Country
#&gt;     2 populations defined - South America, North America</div><div class='input'>
<span class='co'># genlight / snpclone objects ---------------------------------------------</span>


<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/Random'>set.seed</a></span>(<span class='fl'>999</span>)
<span class='no'>gc</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snpclone-coercion-methods.html'>as.snpclone</a></span>(<span class='fu'>glSim</span>(<span class='fl'>100</span>, <span class='fl'>0</span>, <span class='kw'>n.snp.struc</span> <span class='kw'>=</span> <span class='fl'>1e3</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>))
<span class='no'>gc</span> <span class='co'># 100 mlgs</span></div><div class='output co'>#&gt;  ||| SNPCLONE OBJECT |||||||||
#&gt; 
#&gt;  || 100 genotypes,  1,000 binary SNPs, size: 183.3 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  || Basic content
#&gt;    @gen: list of 100 SNPbin
#&gt;    @mlg: 100 original multilocus genotypes
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  || Optional content
#&gt;    @pop: population of each individual (group size range: 50-50)
#&gt;    @other: a list containing: ancestral.pops 
#&gt; 
#&gt; NULL</div><div class='input'><span class='fu'>mlg.filter</span>(<span class='no'>gc</span>) <span class='kw'>&lt;-</span> <span class='fl'>0.25</span>
<span class='no'>gc</span> <span class='co'># 82 mlgs</span></div><div class='output co'>#&gt;  ||| SNPCLONE OBJECT |||||||||
#&gt; 
#&gt;  || 100 genotypes,  1,000 binary SNPs, size: 183.3 Kb
#&gt;  0 (0 %) missing data
#&gt; 
#&gt;  || Basic content
#&gt;    @gen: list of 100 SNPbin
#&gt;    @mlg: 82 contracted multilocus genotypes
#&gt;          (0.25) [t], (bitwise.dist) [d], (farthest) [a]
#&gt;    @ploidy: ploidy of each individual  (range: 2-2)
#&gt; 
#&gt;  || Optional content
#&gt;    @pop: population of each individual (group size range: 50-50)
#&gt;    @other: a list containing: ancestral.pops 
#&gt; 
#&gt; NULL</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

