<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genotype Probability — pgen • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genotype Probability</h1>
    </div>

    
    <p>Calculate the probability of genotypes based on the product of allele
frequencies over all loci.</p>
    

    <pre class="usage"><span class='fu'>pgen</span>(<span class='no'>gid</span>, <span class='kw'>pop</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>by_pop</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>freq</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gid</th>
      <td><p>a genind or genclone object.</p></td>
    </tr>
    <tr>
      <th>pop</th>
      <td><p>either a formula to set the population factor from the 
<code>strata</code> slot or a vector specifying the population factor for 
each sample. Defaults to <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>by_pop</th>
      <td><p>When this is <code>TRUE</code> (default), the calculation will be
done by population.</p></td>
    </tr>
    <tr>
      <th>log</th>
      <td><p>a <code>logical</code> if <code>log =TRUE</code> (default), the values
returned will be log(Pgen). If <code>log = FALSE</code>, the values returned will
be Pgen.</p></td>
    </tr>
    <tr>
      <th>freq</th>
      <td><p>a vector or matrix of allele frequencies. This defaults to 
<code>NULL</code>, indicating that the frequencies will be determined via 
round-robin approach in <code><a href='rraf.html'>rraf</a></code>. <strong>If this matrix or 
vector is not provided, zero-value allele frequencies will automatically be
corrected.</strong> For details, please see the documentation on
<a href='rare_allele_correction.html'>correcting rare alleles</a>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>options from <a href='rare_allele_correction.html'>correcting rare
alleles</a>. The default is to correct allele frequencies to 1/n</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A vector containing Pgen values per locus for each genotype in the 
  object.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Pgen is the probability of a given genotype occuring in a population
  assuming HWE. Thus, the value for diploids is
    $$P_{gen} = \left(\prod_{i=1}^m p_i\right)2^h$$
    where \(p_i\) are the allele frequencies and <em>h</em> is the count of the
  number of heterozygous sites in the sample (Arnaud-Haond et al. 2007; Parks
  and Werth, 1993). The allele frequencies, by default, are calculated using
  a round-robin approach where allele frequencies at a particular locus are 
  calculated on the clone-censored genotypes without that locus.
    To avoid issues with numerical precision of small numbers, this function 
  calculates pgen per locus by adding up log-transformed values of allele 
  frequencies. These can easily be transformed to return the true value (see
  examples).</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>For haploids, Pgen at a particular locus is the allele frequency. This 
  function cannot handle polyploids. Additionally, when the argument 
  <code>pop</code> is not <code>NULL</code>, <code>by_pop</code> is automatically <code>TRUE</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Arnaud-Haond, S., Duarte, C. M., Alberto, F., &amp; Serrão, E. A. 2007.
Standardizing methods to address clonality in population studies.
<em>Molecular Ecology</em>, 16(24), 5115-5139.</p>
<p>Parks, J. C., &amp; Werth, C. R. 1993. A study of spatial features of clones in a
population of bracken fern, <em>Pteridium aquilinum</em> (Dennstaedtiaceae).
<em>American Journal of Botany</em>, 537-544.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='psex.html'>psex</a></code>, <code><a href='rraf.html'>rraf</a></code>, <code><a href='rrmlg.html'>rrmlg</a></code>, 
<code><a href='rare_allele_correction.html'>rare_allele_correction</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>Pram</span>)
<span class='fu'>head</span>(<span class='fu'>pgen</span>(<span class='no'>Pram</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>))</div><div class='output co'>#&gt;             PrMS6A1 Pr9C3A1 PrMS39A1 PrMS45A1    PrMS43A1
#&gt; 1411152-10B     0.5     0.5     0.15      0.5 0.111111111
#&gt; 82              0.5     0.5     0.35      0.5 0.111111111
#&gt; 83              0.5     0.5     0.15      0.5 0.002298851
#&gt; 84              0.5     0.5     0.35      0.5 0.004597701
#&gt; 85              0.5     0.5     0.35      0.5 0.222222222
#&gt; 81-NM-1         0.5     0.5     0.35      0.5 0.111111111</div><div class='input'>
<span class='fu'>not_run</span>({
  <span class='co'># You can also supply the observed allele frequencies</span>
  <span class='no'>pramfreq</span> <span class='kw'>&lt;-</span> <span class='no'>Pram</span> <span class='kw'>%&gt;%</span> <span class='fu'>genind2genpop</span>() <span class='kw'>%&gt;%</span> <span class='fu'>tab</span>(<span class='kw'>freq</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
  <span class='fu'>head</span>(<span class='fu'>pgen</span>(<span class='no'>Pram</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>freq</span> <span class='kw'>=</span> <span class='no'>pramfreq</span>))

  <span class='co'># You can get the Pgen values over all loci by summing over the logged results:</span>
  <span class='fu'>pgen</span>(<span class='no'>Pram</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='kw'>%&gt;%</span>  <span class='co'># calculate pgen matrix</span>
    <span class='fu'>rowSums</span>(<span class='kw'>na.rm</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='kw'>%&gt;%</span> <span class='co'># take the sum of each row</span>
    <span class='fu'>exp</span>()                     <span class='co'># take the exponent of the results</span>

  <span class='co'># You can also take the product of the non-logged results:</span>
  <span class='fu'>apply</span>(<span class='fu'>pgen</span>(<span class='no'>Pram</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>), <span class='fl'>1</span>, <span class='no'>prod</span>, <span class='kw'>na.rm</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

  <span class='co'>## Rare Allele Correction ---------------------------------------------------</span>
  <span class='co'>##</span>
  <span class='co'># If you don't supply a table of frequencies, they are calculated with rraf </span>
  <span class='co'># with correction = TRUE. This is normally benign when analyzing large </span>
  <span class='co'># populations, but it can have a great effect on small populations. To help </span>
  <span class='co'># control this, you can supply arguments described in </span>
  <span class='co'># help("rare_allele_correction"). </span>


  <span class='co'># Default is to correct by 1/n per population. Since the calculation is </span>
  <span class='co'># performed on a smaller sample size due to round robin clone correction, it</span>
  <span class='co'># would be more appropriate to correct by 1/rrmlg at each locus. This is </span>
  <span class='co'># acheived by setting d = "rrmlg". Since this is a diploid, we would want to</span>
  <span class='co'># account for the number of chromosomes, and so we set mul = 1/2</span>
  <span class='fu'>head</span>(<span class='fu'>pgen</span>(<span class='no'>Pram</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>d</span> <span class='kw'>=</span> <span class='st'>"rrmlg"</span>, <span class='kw'>mul</span> <span class='kw'>=</span> <span class='fl'>1</span>/<span class='fl'>2</span>)) <span class='co'># compare with the output above</span>

  <span class='co'># If you wanted to treat all alleles as equally rare, then you would set a</span>
  <span class='co'># specific value (let's say the rare alleles are 1/100):</span>
  <span class='fu'>head</span>(<span class='fu'>pgen</span>(<span class='no'>Pram</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>e</span> <span class='kw'>=</span> <span class='fl'>1</span>/<span class='fl'>100</span>))
})</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar, Jonah Brooks, Stacy A. Krueger-Hadfield, Erik Sotka

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
