<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>The <span class="pkg">poppr</span> R package. — poppr-package • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">poppr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The <span class="pkg">poppr</span> R package.</h1>
    </div>

    
    <p><span class="pkg">Poppr</span> provides tools for population genetic analysis that
include genotypic diversity measures, genetic distances with bootstrap
support, native organization and handling of population hierarchies, and
clone correction.</p>
<p>To cite <span class="pkg">poppr</span>, please use <code>citation("poppr")</code>. When referring to
<span class="pkg">poppr</span> in your manuscript, please use lower case unless it occurs at the
beginning of a sentence.</p>
    

        
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This package relies on the <span class="pkg"><a href='http://www.rdocumentation.org/packages/adegenet/topics/adegenet.package'>adegenet</a></span> package. 
  It is built around the <code>genind</code> and 
  <code>genlight</code> object. Genind objects store genetic
  information in a table of allele frequencies while genlight objects store
  SNP data efficiently by packing binary allele calls into single bits.
  <span class="pkg">Poppr</span> has extended these object into new objects called 
  <code><a href='genclone-class.html'>genclone</a></code> and <code><a href='genclone-class.html'>snpclone</a></code>,
  respectively. These objects are designed for analysis of clonal organisms
  as they add the <strong>@mlg</strong> slot for keeping track of multilocus
  genotypes and multilocus lineages.</p><h3>Documentation</h3>
 Documentation is available for any function by 
  typing <code>?function_name</code> in the R console. Detailed topic explanations
  live in the package vignettes:
    <table><tr><td>
    <strong>Vignette</strong>              </td>
</tr><tr><td> <strong>command</strong></td>
</tr><tr><td>
    Data import and manipulation   </td>
</tr><tr><td> <code>NA</code></td>
</tr><tr><td>
    Algorithms and Equations      </td>
</tr><tr><td> <code>NA</code></td>
</tr></table>
    Essential functions for importing and manipulating data are detailed within
  the <em>Data import and manipulation</em> vignette, details on algorithms
  used in <span class="pkg">poppr</span> are within the <em>Algorithms and equations</em>
  vignette, and details for working with multilocus genotypes are in
  <em>Multilocus Genotype Analysis</em>. Examples of analyses are available in
  a primer written by Niklaus J. Grünwald, Zhian N. Kamvar, and Sydney E.
  Everhart at <a href = 'http://grunwaldlab.github.io/Population_Genetics_in_R'>http://grunwaldlab.github.io/Population_Genetics_in_R</a>.<p></p><h3>Getting help</h3>
 If you have a specific question or issue with 
  <span class="pkg">poppr</span>, feel free to contribute to the google group at 
  <a href = 'https://groups.google.com/group/poppr'>https://groups.google.com/group/poppr</a>. If you find a bug and 
  are a github user, you can submit bug reports at 
  <a href = 'https://github.com/grunwaldlab/poppr/issues'>https://github.com/grunwaldlab/poppr/issues</a>. Otherwise, leave a
  message on the groups. Personal emails are highly discouraged as they do 
  not allow others to learn.
    
    <h2 class="hasAnchor" id="functions-in-span-class-pkg-poppr-span-"><a class="anchor" href="#functions-in-span-class-pkg-poppr-span-"></a>Functions in &lt;span class=&quot;pkg&quot;&gt;poppr&lt;/span&gt;</h2>

    
    <p>Below are descriptions and links to functions found in <span class="pkg">poppr</span>. Be
  aware that all functions in <span class="pkg"><a href='http://www.rdocumentation.org/packages/adegenet/topics/adegenet.package'>adegenet</a></span> are also
  available. The functions are documented as:</p><ul>
<li><p><code>function</code> (data type) - Description</p></li>
</ul><p>Where &#8216;data type&#8217; refers to the type of data that can be used:</p><table><tr><td>
    <strong>m</strong> </td>
</tr><tr><td> a genclone or genind object </td>
</tr><tr><td>
    <strong>s</strong> </td>
</tr><tr><td> a snpclone or genlight object </td>
</tr></table>
    
    <h2 class="hasAnchor" id="data-import-export"><a class="anchor" href="#data-import-export"></a>Data import/export</h2>

    
    <ul>
<li><p><code><a href='getfile.html'>getfile</a></code> (x) - Provides a quick GUI to grab files for import</p></li>
<li><p><code><a href='read.genalex.html'>read.genalex</a></code> (x) - Reads GenAlEx formatted csv files to a genind object</p></li>
<li><p><code><a href='genind2genalex.html'>genind2genalex</a></code> (m) - Converts genind objects to GenAlEx formatted csv files</p></li>
<li><p><code><a href='coercion-methods.html'>genclone2genind</a></code> (m) - Removes the @mlg slot from genclone objects</p></li>
<li><p><code><a href='coercion-methods.html'>genclone2genind</a></code> (m) - Removes the @mlg slot from genclone objects</p></li>
<li><p><code><a href='coercion-methods.html'>bootgen2genind</a></code> (x) - see <code><a href='aboot.html'>aboot</a></code> for details)</p></li>
</ul>
    
    <h2 class="hasAnchor" id="data-manipulation"><a class="anchor" href="#data-manipulation"></a>Data manipulation</h2>

    
    <ul>
<li><p><code><a href='coercion-methods.html'>as.genclone</a></code> (m) - Converts genind objects to genclone objects</p></li>
<li><p><code><a href='missingno.html'>missingno</a></code> (m) - Handles missing data</p></li>
<li><p><code><a href='clonecorrect.html'>clonecorrect</a></code> (m | s) - Clone-censors at a specified population hierarchy</p></li>
<li><p><code><a href='informloci.html'>informloci</a></code> (m) - Detects and removes phylogenetically uninformative loci</p></li>
<li><p><code><a href='popsub.html'>popsub</a></code> (m | s) - Subsets genind objects by population</p></li>
<li><p><code><a href='shufflepop.html'>shufflepop</a></code> (m) - Shuffles genotypes at each locus using four different shuffling algorithms</p></li>
<li><p><code><a href='recode_polyploids.html'>recode_polyploids</a></code> (m | x) - Recodes polyploid data sets with missing alleles imported as "0"</p></li>
</ul>
    
    <h2 class="hasAnchor" id="genetic-distances"><a class="anchor" href="#genetic-distances"></a>Genetic distances</h2>

    
    <ul>
<li><p><code><a href='bruvo.dist.html'>bruvo.dist</a></code> (m) - Bruvo's distance (see also: <code><a href='fix_replen.html'>fix_replen</a></code>)</p></li>
<li><p><code><a href='diss.dist.html'>diss.dist</a></code> (m) - Absolute genetic distance (see prevosti.dist)</p></li>
<li><p><code><a href='genetic_distance.html'>nei.dist</a></code> (m | x) - Nei's 1978 genetic distance</p></li>
<li><p><code><a href='genetic_distance.html'>rogers.dist</a></code> (m | x) - Rogers' euclidean distance</p></li>
<li><p><code><a href='genetic_distance.html'>reynolds.dist</a></code> (m | x) - Reynolds' coancestry distance</p></li>
<li><p><code><a href='genetic_distance.html'>edwards.dist</a></code> (m | x) - Edwards' angular distance</p></li>
<li><p><code><a href='genetic_distance.html'>prevosti.dist</a></code> (m | x) - Prevosti's absolute genetic distance</p></li>
<li><p><code><a href='bitwise.dist.html'>bitwise.dist</a></code> (s) - Calculates fast pairwise distances for genlight objects.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="bootstrapping"><a class="anchor" href="#bootstrapping"></a>Bootstrapping</h2>

    
    <ul>
<li><p><code><a href='aboot.html'>aboot</a></code> (m | s | x) - Creates a bootstrapped dendrogram for any distance measure</p></li>
<li><p><code><a href='bruvo.boot.html'>bruvo.boot</a></code> (m) - Produces dendrograms with bootstrap support based on Bruvo's distance</p></li>
<li><p><code><a href='diversity_boot.html'>diversity_boot</a></code> (x) - Generates boostrap distributions of diversity statistics for multilocus genotypes</p></li>
<li><p><code><a href='diversity_ci.html'>diversity_ci</a></code> (m | s | x) - Generates confidence intervals for multilocus genotype diversity.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="analysis"><a class="anchor" href="#analysis"></a>Analysis</h2>

    
    <h3>Multilocus Genotypes</h3>
    <ul>
<li><p><code><a href='mlg.html'>mlg</a></code> (m | s) - Calculates the number of multilocus genotypes</p></li>
<li><p><code><a href='mll-method.html'>mll</a></code> (m | s) - Displays the current multilocus lineages (genotypes) defined.</p></li>
<li><p><code><a href='mll-method.html'>nmll</a></code> (m | s) - Same as <code><a href='mlg.html'>mlg</a></code>.</p></li>
<li><p><code><a href='mlg.html'>mlg.crosspop</a></code> (m | s) - Finds all multilocus genotypes that cross populations</p></li>
<li><p><code><a href='mlg.html'>mlg.table</a></code> (m | s) - Returns a table of populations by multilocus genotypes</p></li>
<li><p><code><a href='mlg.html'>mlg.vector</a></code> (m | s) - Returns a vector of a numeric multilocus genotype assignment for each individual</p></li>
<li><p><code><a href='mlg.html'>mlg.id</a></code> (m | s) - Finds all individuals associated with a single multilocus genotype</p></li>
<li><p><code><a href='mlg.filter.html'>mlg.filter</a></code> (m | s) - Collapses MLGs by genetic distance</p></li>
<li><p><code><a href='filter_stats.html'>filter_stats</a></code> (m | s) - Calculates mlg.filter for all algorithms and plots</p></li>
<li><p><code><a href='cutoff_predictor.html'>cutoff_predictor</a></code> (x) - Predicts cutoff threshold from mlg.filter.</p></li>
<li><p><code><a href='mll.custom.html'>mll.custom</a></code> (m | s) - Allows for the custom definition of multilocus lineages</p></li>
<li><p><code><a href='mll.custom.html'>mll.levels</a></code> (m | s) - Allows the user to change levels of custom MLLs.</p></li>
<li><p><code><a href='mll.reset-method.html'>mll.reset</a></code> (m | s) - Reset multilocus lineages.</p></li>
<li><p><code><a href='diversity_stats.html'>diversity_stats</a></code> (x) - Creates a table of diversity indices for multilocus genotypes.</p></li>
</ul>
    <h3>Other</h3>
    <ul>
<li><p><code><a href='poppr.amova.html'>poppr.amova</a></code> (m) - Analysis of Molecular Variance (as implemented in ade4)</p></li>
<li><p><code><a href='ia.html'>ia</a></code> (m) - Calculates the index of association</p></li>
<li><p><code><a href='ia.html'>pair.ia</a></code> (m) - Calculates the index of association for all loci pairs.</p></li>
<li><p><code><a href='ia.html'>jack.ia</a></code> (m) - Calculates the index of association over subsets of data.</p></li>
<li><p><code><a href='win.ia.html'>win.ia</a></code> (s) - Index of association windows for genlight objects.</p></li>
<li><p><code><a href='samp.ia.html'>samp.ia</a></code> (s) - Index of association on random subsets of loci for genlight objects.</p></li>
<li><p><code><a href='poppr.html'>poppr</a></code> (m | x) - Returns a diversity table by population</p></li>
<li><p><code><a href='poppr.all.html'>poppr.all</a></code> (m | x) - Returns a diversity table by population for all compatible files specified</p></li>
<li><p><code><a href='private_alleles.html'>private_alleles</a></code> (m) - Tabulates the occurrences of alleles that only occur in one population.</p></li>
<li><p><code><a href='locus_table.html'>locus_table</a></code> (m) - Creates a table of summary statistics per locus.</p></li>
<li><p><code><a href='rrmlg.html'>rrmlg</a></code> (m | x) - Round-robin multilocus genotype estimates.</p></li>
<li><p><code><a href='rraf.html'>rraf</a></code> (m) - Round-robin allele frequency estimates.</p></li>
<li><p><code><a href='pgen.html'>pgen</a></code> (m) - Probability of genotypes.</p></li>
<li><p><code><a href='psex.html'>psex</a></code> (m) - Probability of observing a genotype more than once.</p></li>
<li><p><code><a href='incomp.html'>incomp</a></code> (m) - Check data for incomparable samples.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="visualization"><a class="anchor" href="#visualization"></a>Visualization</h2>

    
    <ul>
<li><p><code><a href='imsn.html'>imsn</a></code> (m | s) - Interactive construction and visualization of minimum spanning networks</p></li>
<li><p><code><a href='plot_poppr_msn.html'>plot_poppr_msn</a></code> (m | s | x) - Plots minimum spanning networks produced in poppr with scale bar and legend</p></li>
<li><p><code><a href='greycurve.html'>greycurve</a></code> (x) - Helper to determine the appropriate parameters for adjusting the grey level for msn functions</p></li>
<li><p><code><a href='bruvo.msn.html'>bruvo.msn</a></code> (m) - Produces minimum spanning networks based off Bruvo's distance colored by population</p></li>
<li><p><code><a href='poppr.msn.html'>poppr.msn</a></code> (m | s | x) - Produces a minimum spanning network for any pairwise distance matrix related to the data</p></li>
<li><p><code><a href='info_table.html'>info_table</a></code> (m) - Creates a heatmap representing missing data or observed ploidy</p></li>
<li><p><code><a href='genotype_curve.html'>genotype_curve</a></code> (m | x) - Creates a series of boxplots to demonstrate how many markers are needed to represent the diversity of your data.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="datasets"><a class="anchor" href="#datasets"></a>Datasets</h2>

    
    <ul>
<li><p><code><a href='Aeut.html'>Aeut</a></code> - (AFLP) Oomycete root rot pathogen
<em>Aphanomyces euteiches</em> (Grünwald and Hoheisel, 2006)</p></li>
<li><p><code><a href='monpop.html'>monpop</a></code> - (SSR) Peach brown rot pathogen <em>Monilinia
fructicola</em> (Everhart and Scherm, 2015)</p></li>
<li><p><code><a href='partial_clone.html'>partial_clone</a></code> - (SSR) partially-clonal data simulated via
simuPOP (Peng and Amos, 2008)</p></li>
<li><p><code><a href='Pinf.html'>Pinf</a></code> - (SSR) Potato late blight pathogen
<em>Phytophthora infestans</em> (Goss et. al., 2014)</p></li>
<li><p><code><a href='Pram.html'>Pram</a></code> - (SSR) Sudden Oak Death pathogen <em>Phytophthora
ramorum</em> (Kamvar et. al., 2015; Goss et. al., 2009)</p></li>
</ul>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>--------- Papers announcing poppr ---------</p>
<p>Kamvar ZN, Tabima JF, Grünwald NJ. (2014) Poppr: an R package for genetic
analysis of populations with clonal, partially clonal, and/or sexual
reproduction. PeerJ 2:e281 <a href = 'http://dx.doi.org/10.7717/peerj.281'>http://dx.doi.org/10.7717/peerj.281</a></p>
<p>Kamvar ZN, Brooks JC and Grünwald NJ (2015) Novel R tools for analysis of 
genome-wide population genetic data with emphasis on clonality. Front. Genet.
6:208. doi: 10.3389/fgene.2015.00208 
<a href = 'http://dx.doi.org/10.3389/fgene.2015.00208'>http://dx.doi.org/10.3389/fgene.2015.00208</a></p>
<p>--------- Papers referencing data sets ---------</p>
<p>Grunwald, NJ and Hoheisel, G.A. 2006. Hierarchical Analysis of Diversity, 
Selfing, and Genetic Differentiation in Populations of the Oomycete 
<em>Aphanomyces euteiches</em>. Phytopathology 96:1134-1141 doi: 
<a href = 'http://dx.doi.org/10.1094/PHYTO-96-1134'>10.1094/PHYTO-96-1134</a></p>
<p>SE Everhart, H Scherm, (2015) Fine-scale genetic structure of <em>Monilinia
fructicola</em> during brown rot epidemics within individual peach tree canopies.
Phytopathology 105:542-549 doi: 
<a href = 'http://dx.doi.org/10.1094/PHYTO-03-14-0088-R'>10.1094/PHYTO-03-14-0088-R</a></p>
<p>Bo Peng and Christopher Amos (2008) Forward-time simulations of nonrandom 
mating populations using simuPOP. <em>bioinformatics</em>, 24 (11): 1408-1409.</p>
<p>Goss, Erica M., Javier F. Tabima, David EL Cooke, Silvia Restrepo, William E.
Fry, Gregory A. Forbes, Valerie J. Fieland, Martha Cardenas, and Niklaus J.
Grünwald. (2014) "The Irish potato famine pathogen <em>Phytophthora 
infestans</em> originated in central Mexico rather than the Andes." Proceedings 
of the National Academy of Sciences 111:8791-8796. doi: 
<a href = 'http://dx.doi.org/10.1073/pnas.1401884111'>10.1073/pnas.1401884111</a></p>
<p>Kamvar, Z. N., Larsen, M. M., Kanaskie, A. M., Hansen, E. M., &amp; Grünwald, N.
J. (2015). Spatial and temporal analysis of populations of the sudden oak
death pathogen in Oregon forests. Phytopathology 105:982-989. doi: 
<a href = 'http://dx.doi.org/10.1094/PHYTO-12-14-0350-FI'>10.1094/PHYTO-12-14-0350-FI</a></p>
    <p>Goss, E. M., Larsen, M., Chastagner, G. A., Givens, D. R., and Grünwald, N. 
J. 2009. Population genetic analysis infers migration pathways of 
<em>Phytophthora ramorum</em> in US nurseries. PLoS Pathog. 5:e1000583. doi: 
<a href = 'http://dx.doi.org/10.1371/journal.ppat.1000583'>10.1371/journal.ppat.1000583</a></p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#details">Details</a></li>

      <li><a href="#functions-in-span-class-pkg-poppr-span-">Functions in &lt;span class=&quot;pkg&quot;&gt;poppr&lt;/span&gt;</a></li>

      <li><a href="#data-import-export">Data import/export</a></li>

      <li><a href="#data-manipulation">Data manipulation</a></li>

      <li><a href="#genetic-distances">Genetic distances</a></li>

      <li><a href="#bootstrapping">Bootstrapping</a></li>

      <li><a href="#analysis">Analysis</a></li>

      <li><a href="#visualization">Visualization</a></li>

      <li><a href="#datasets">Datasets</a></li>

      <li><a href="#references">References</a></li>
          </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar, Jonah C. Brooks, Sydney E. Everhart, Javier F. 
  Tabima, Stacy Krueger-Hadfield, Erik Sotka, Niklaus J. Grünwald

Maintainer: Zhian N. Kamvar

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
