<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Produce a basic summary table for population genetic analyses. — poppr • poppr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Produce a basic summary table for population genetic analyses. — poppr" />

<meta property="og:description" content="For the poppr package description, please see
package?poppr
This function allows the user to quickly view indices of heterozygosity, 
evenness, and linkage to aid in the decision of a path to further analyze 
a specified dataset. It natively takes genind and 
genclone objects, but can convert any raw data formats 
that adegenet can take (fstat, structure, genetix, and genpop) as well as 
genalex files exported into a csv format (see read.genalex for 
details)." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">poppr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">2.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    </li>
    <li>
      <a href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Produce a basic summary table for population genetic analyses.</h1>
    <small class="dont-index">Source: <a href='https://github.com/grunwaldlab/poppr/blob/master/R/Index_calculations.r'><code>R/Index_calculations.r</code></a></small>
    <div class="hidden name"><code>poppr.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>For the <span class="pkg">poppr</span> package description, please see
<code><a href='poppr-package.html'>package?poppr</a></code></p>
<p>This function allows the user to quickly view indices of heterozygosity, 
evenness, and linkage to aid in the decision of a path to further analyze 
a specified dataset. It natively takes <code>genind</code> and 
<code><a href='genclone-class.html'>genclone</a></code> objects, but can convert any raw data formats 
that adegenet can take (fstat, structure, genetix, and genpop) as well as 
genalex files exported into a csv format (see <code><a href='read.genalex.html'>read.genalex</a></code> for 
details).</p>
    
    </div>

    <pre class="usage"><span class='fu'>poppr</span>(<span class='no'>dat</span>, <span class='kw'>total</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>sublist</span> <span class='kw'>=</span> <span class='st'>"ALL"</span>, <span class='kw'>blacklist</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sample</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>missing</span> <span class='kw'>=</span> <span class='st'>"ignore"</span>, <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>clonecorrect</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>strata</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>keep</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>hist</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>index</span> <span class='kw'>=</span> <span class='st'>"rbarD"</span>, <span class='kw'>minsamp</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>legend</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dat</th>
      <td><p>a <code>genind</code> object OR a 
<code><a href='genclone-class.html'>genclone</a></code> object OR any fstat, structure, genetix, 
genpop, or genalex formatted file.</p></td>
    </tr>
    <tr>
      <th>total</th>
      <td><p>When <code>TRUE</code> (default), indices will be calculated for the 
pooled populations.</p></td>
    </tr>
    <tr>
      <th>sublist</th>
      <td><p>a list of character strings or integers to indicate specific 
population names (accessed via <code><a href='http://www.rdocumentation.org/packages/adegenet/topics/accessors'>popNames</a></code>). 
Defaults to "ALL".</p></td>
    </tr>
    <tr>
      <th>blacklist</th>
      <td><p>a list of character strings or integers to indicate specific
populations to be removed from analysis. Defaults to NULL.</p></td>
    </tr>
    <tr>
      <th>sample</th>
      <td><p>an integer indicating the number of permutations desired to 
obtain p-values. Sampling will shuffle genotypes at each locus to simulate 
a panmictic population using the observed genotypes. Calculating the 
p-value includes the observed statistics, so set your sample number to one 
off for a round p-value (eg. <code>sample = 999</code> will give you p = 0.001 
and <code>sample = 1000</code> will give you p = 0.000999001).</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>an integer from 1 to 4 indicating the method of sampling 
desired. see <code><a href='shufflepop.html'>shufflepop</a></code> for details.</p></td>
    </tr>
    <tr>
      <th>missing</th>
      <td><p>how should missing data be treated? <code>"zero"</code> and 
<code>"mean"</code> will set the missing values to those documented in 
<code>tab</code>. <code>"loci"</code> and <code>"geno"</code> will remove any loci or
genotypes with missing data, respectively (see <code><a href='missingno.html'>missingno</a></code> for
more information.</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p><code>numeric</code> a number from 0 to 1 indicating the percent 
missing data allowed for analysis. This is to be used in conjunction with 
the flag <code>missing</code> (see <code><a href='missingno.html'>missingno</a></code> for details)</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p><code>FALSE</code> (default) will display a progress bar for each 
population analyzed.</p></td>
    </tr>
    <tr>
      <th>clonecorrect</th>
      <td><p>default <code>FALSE</code>. must be used with the <code>strata</code>
parameter, or the user will potentially get undesired results. see
<code><a href='clonecorrect.html'>clonecorrect</a></code> for details.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>a <code>formula</code> indicating the hierarchical levels to be used.
The hierarchies should be present in the <code>strata</code> slot. See
<code>strata</code> for details.</p></td>
    </tr>
    <tr>
      <th>keep</th>
      <td><p>an <code>integer</code>. This indicates which strata you wish to keep 
after clone correcting your data sets. To combine strata, just set keep 
from 1 to the number of straifications set in strata. see 
<code><a href='clonecorrect.html'>clonecorrect</a></code> for details.</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p><code>logical</code> if <code>TRUE</code> (default) and <code>sampling &gt; 0</code>, 
a histogram will be produced for each population.</p></td>
    </tr>
    <tr>
      <th>hist</th>
      <td><p><code>logical</code> Deprecated. Use plot.</p></td>
    </tr>
    <tr>
      <th>index</th>
      <td><p><code>character</code> Either "Ia" or "rbarD". If <code>hist = TRUE</code>, 
this will determine the index used for the visualization.</p></td>
    </tr>
    <tr>
      <th>minsamp</th>
      <td><p>an <code>integer</code> indicating the minimum number of individuals
to resample for rarefaction analysis. See <code><a href='http://www.rdocumentation.org/packages/vegan/topics/rarefy'>rarefy</a></code> for 
details.</p></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p><code>logical</code>. When this is set to <code>TRUE</code>, a legend 
describing the resulting table columns will be printed. Defaults to 
<code>FALSE</code></p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>arguments to be passed on to <code><a href='diversity_stats.html'>diversity_stats</a></code></p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame with populations in rows and the following columns:</p>
<dt>Pop</dt><dd><p>A vector indicating the population factor</p></dd> 
  <dt>N</dt><dd><p>An integer vector indicating the number of individuals/isolates in
  the specified population.</p></dd>
  <dt>MLG</dt><dd><p>An integer vector indicating the number of multilocus genotypes 
  found in the specified population, (see: <code><a href='mlg.html'>mlg</a></code>)</p></dd>
  <dt>eMLG</dt><dd><p>The expected number of MLG at the lowest common sample size
  (set by the parameter <code>minsamp</code>).</p></dd>
  <dt>SE</dt><dd><p>The standard error for the rarefaction analysis</p></dd>
  <dt>H</dt><dd><p>Shannon-Weiner Diversity index</p></dd>
  <dt>G</dt><dd><p>Stoddard and Taylor's Index</p></dd> 
  <dt>lambda</dt><dd><p>Simpson's index</p></dd> 
  <dt>E.5</dt><dd><p>Evenness</p></dd> 
  <dt>Hexp</dt><dd><p>Nei's gene diversity (expected heterozygosity)</p></dd>
  <dt>Ia</dt><dd><p>A numeric vector giving the value of the Index of Association for
  each population factor, (see <code><a href='ia.html'>ia</a></code>).</p></dd>
  <dt>p.Ia</dt><dd><p>A numeric vector indicating the p-value for Ia from the number
  of reshufflings indicated in <code>sample</code>. Lowest value is 1/n where n is
  the number of observed values.</p></dd>
  <dt>rbarD</dt><dd><p>A numeric vector giving the value of the Standardized Index of
  Association for each population factor, (see <code><a href='ia.html'>ia</a></code>).</p></dd>
  <dt>p.rD</dt><dd><p>A numeric vector indicating the p-value for rbarD from the
  number of reshuffles indicated in <code>sample</code>. Lowest value is 1/n where
  n is the number of observed values.</p></dd>
  <dt>File</dt><dd><p>A vector indicating the name of the original data file.</p></dd>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This table is intended to be a first look into the dynamics of 
  mutlilocus genotype diversity. Many of the statistics (except for the the 
  index of association) are simply based on counts of multilocus genotypes 
  and do not take into account the actual allelic states.
  <strong>Descriptions of the statistics can be found in the Algorithms and
  Equations vignette</strong>: <code>vignette("algo", package = "poppr")</code>.</p><h3>sampling</h3>
The sampling procedure is explicitly for testing the
  index of association. None of the other diversity statistics (H, G, lambda,
  E.5) are tested with this sampling due to the differing data types. To
  obtain confidence intervals for these statistics, please see 
  <code><a href='diversity_ci.html'>diversity_ci</a></code>.<p></p><h3>rarefaction</h3>
Rarefaction analysis is performed on the number of
  multilocus genotypes because it is relatively easy to estimate (Grünwald et
  al., 2003). To obtain rarefied estimates of diversity, it is possible to
  use <code><a href='diversity_ci.html'>diversity_ci</a></code> with the argument <code>rarefy = TRUE</code><p></p><h3>graphic</h3>
This function outputs a <span class="pkg">ggplot2</span> graphic of
  histograms. These can be manipulated to be visualized in another manner by
  retrieving the plot with the <code><a href='http://www.rdocumentation.org/packages/ggplot2/topics/last_plot'>last_plot</a></code> command from
  <span class="pkg">ggplot2</span>. A useful manipulation would be to arrange the graphs into a
  single column so that the values of the statistic line up: <br /> <code>p &lt;-
  last_plot(); p + facet_wrap(~population, ncol = 1, scales = "free_y")</code><br />
  The name for the groupings is "population" and the name for the x axis is
  "value".
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>The calculation of <code>Hexp</code> has changed from <span class="pkg">poppr</span> 1.x. It was
  previously calculated based on the diversity of multilocus genotypes, 
  resulting in a value of 1 for sexual populations. This was obviously not 
  Nei's 1978 expected heterozygosity. We have thus changed the statistic to 
  be the true value of Hexp by calculating \((\frac{n}{n-1}) 1 - \sum_{i =
  1}^k{p^{2}_{i}}\) where p is the allele
  frequencies at a given locus and n is the number of observed alleles (Nei,
  1978) in each locus and then returning the average. Caution should be 
  exercised in interpreting the results of Hexp with polyploid organisms with
  ambiguous ploidy. The lack of allelic dosage information will cause rare 
  alleles to be over-represented and artificially inflate the index. This is 
  especially true with small sample sizes.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Paul-Michael Agapow and Austin Burt. Indices of multilocus 
  linkage disequilibrium. <em>Molecular Ecology Notes</em>, 1(1-2):101-102, 
  2001</p>  
<p>A.H.D. Brown, M.W. Feldman, and E. Nevo. Multilocus structure of natural 
  populations of <em>Hordeum spontaneum</em>. <em>Genetics</em>, 96(2):523-536,
  1980.</p>  
<p>Niklaus J. Gr\"unwald, Stephen B. Goodwin, Michael G. Milgroom, and William
  E. Fry. Analysis of genotypic diversity data for populations of 
  microorganisms. Phytopathology, 93(6):738-46, 2003</p>  
<p>Bernhard Haubold and Richard R. Hudson. Lian 3.0: detecting linkage 
  disequilibrium in multilocus data. Bioinformatics, 16(9):847-849, 2000.</p>  
<p>Kenneth L.Jr. Heck, Gerald van Belle, and Daniel Simberloff. Explicit 
  calculation of the rarefaction diversity measurement and the determination 
  of sufficient sample size. Ecology, 56(6):pp. 1459-1461, 1975</p>  
<p>Masatoshi Nei. Estimation of average heterozygosity and genetic distance 
  from a small number of individuals. Genetics, 89(3):583-590, 1978.</p>  
<p>S H Hurlbert. The nonconcept of species diversity: a critique and 
  alternative parameters. Ecology, 52(4):577-586, 1971.</p>  
<p>J.A. Ludwig and J.F. Reynolds. Statistical Ecology. A Primer on Methods and
  Computing. New York USA: John Wiley and Sons, 1988.</p>  
<p>Simpson, E. H. Measurement of diversity. Nature 163: 688, 1949 
  doi:10.1038/163688a0</p>  
<p>Good, I. J. (1953). On the Population Frequency of Species and the 
  Estimation of Population Parameters. <em>Biometrika</em> 40(3/4): 237-264.</p>  
<p>Lande, R. (1996). Statistics and partitioning of species diversity, and 
  similarity among multiple communities. <em>Oikos</em> 76: 5-13.</p>  
<p>Jari Oksanen, F. Guillaume Blanchet, Roeland Kindt, Pierre Legendre, Peter 
  R. Minchin, R. B. O'Hara, Gavin L. Simpson, Peter Solymos, M. Henry H. 
  Stevens, and Helene Wagner. vegan: Community Ecology Package, 2012. R 
  package version 2.0-5.</p>  
<p>E.C. Pielou. Ecological Diversity. Wiley, 1975.</p>  
<p>Claude Elwood Shannon. A mathematical theory of communication. Bell Systems
  Technical Journal, 27:379-423,623-656, 1948</p>  
<p>J M Smith, N H Smith, M O'Rourke, and B G Spratt. How clonal are bacteria? 
  Proceedings of the National Academy of Sciences, 90(10):4384-4388, 1993.</p>  
<p>J.A. Stoddart and J.F. Taylor. Genotypic diversity: estimation and 
  prediction in samples. Genetics, 118(4):705-11, 1988.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='clonecorrect.html'>clonecorrect</a></code>, 
  <code><a href='poppr.all.html'>poppr.all</a></code>, 
  <code><a href='ia.html'>ia</a></code>, 
  <code><a href='missingno.html'>missingno</a></code>, 
  <code><a href='mlg.html'>mlg</a></code>, 
  <code><a href='diversity_stats.html'>diversity_stats</a></code>,
  <code><a href='diversity_ci.html'>diversity_ci</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>nancycats</span>)
<span class='fu'>poppr</span>(<span class='no'>nancycats</span>)</div><div class='output co'>#&gt; | P01 
#&gt; | P02 
#&gt; | P03 
#&gt; | P04 
#&gt; | P05 
#&gt; | P06 
#&gt; | P07 
#&gt; | P08 
#&gt; | P09 
#&gt; | P10 
#&gt; | P11 
#&gt; | P12 
#&gt; | P13 
#&gt; | P14 
#&gt; | P15 
#&gt; | P16 
#&gt; | P17 
#&gt; | Total </div><div class='output co'>#&gt;      Pop   N MLG eMLG       SE    H   G lambda E.5  Hexp      Ia   rbarD
#&gt; 1    P01  10  10   10 0.00e+00 2.30  10  0.900   1 0.649  0.1656  0.0211
#&gt; 2    P02  22  22   10 0.00e+00 3.09  22  0.955   1 0.701  0.1818  0.0230
#&gt; 3    P03  12  12   10 0.00e+00 2.48  12  0.917   1 0.719  0.3546  0.0452
#&gt; 4    P04  23  23   10 5.03e-07 3.14  23  0.957   1 0.750  0.4494  0.0563
#&gt; 5    P05  15  15   10 2.77e-07 2.71  15  0.933   1 0.640 -0.0475 -0.0060
#&gt; 6    P06  11  11   10 0.00e+00 2.40  11  0.909   1 0.745  0.3337  0.0426
#&gt; 7    P07  14  14   10 0.00e+00 2.64  14  0.929   1 0.667  0.2569  0.0326
#&gt; 8    P08  10  10   10 0.00e+00 2.30  10  0.900   1 0.752  0.2388  0.0301
#&gt; 9    P09   9   9    9 0.00e+00 2.20   9  0.889   1 0.694  2.0845  0.2636
#&gt; 10   P10  11  11   10 0.00e+00 2.40  11  0.909   1 0.698  0.5955  0.0763
#&gt; 11   P11  20  20   10 0.00e+00 3.00  20  0.950   1 0.783  0.2847  0.0363
#&gt; 12   P12  14  14   10 0.00e+00 2.64  14  0.929   1 0.667  0.4899  0.0643
#&gt; 13   P13  13  13   10 7.30e-08 2.56  13  0.923   1 0.688  0.1855  0.0237
#&gt; 14   P14  17  17   10 2.31e-07 2.83  17  0.941   1 0.789  0.2210  0.0282
#&gt; 15   P15  11  11   10 0.00e+00 2.40  11  0.909   1 0.723  0.6933  0.0873
#&gt; 16   P16  12  12   10 0.00e+00 2.48  12  0.917   1 0.700  0.2345  0.0295
#&gt; 17   P17  13  13   10 7.30e-08 2.56  13  0.923   1 0.605 -0.0906 -0.0138
#&gt; 18 Total 237 237   10 0.00e+00 5.47 237  0.996   1 0.774  0.1721  0.0218
#&gt;         File
#&gt; 1  nancycats
#&gt; 2  nancycats
#&gt; 3  nancycats
#&gt; 4  nancycats
#&gt; 5  nancycats
#&gt; 6  nancycats
#&gt; 7  nancycats
#&gt; 8  nancycats
#&gt; 9  nancycats
#&gt; 10 nancycats
#&gt; 11 nancycats
#&gt; 12 nancycats
#&gt; 13 nancycats
#&gt; 14 nancycats
#&gt; 15 nancycats
#&gt; 16 nancycats
#&gt; 17 nancycats
#&gt; 18 nancycats</div><div class='input'>
</div><span class='co'># NOT RUN {</span>
<span class='co'># Sampling</span>
<span class='fu'>poppr</span>(<span class='no'>nancycats</span>, <span class='kw'>sample</span> <span class='kw'>=</span> <span class='fl'>999</span>, <span class='kw'>total</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># Customizing the plot</span>
<span class='fu'>library</span>(<span class='st'>"ggplot2"</span>)
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/last_plot'>last_plot</a></span>()
<span class='no'>p</span> + <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/facet_wrap'>facet_wrap</a></span>(~<span class='no'>population</span>, <span class='kw'>scales</span> <span class='kw'>=</span> <span class='st'>"free_y"</span>, <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>)

<span class='co'># Turning off diversity statistics (see get_stats)</span>
<span class='fu'>poppr</span>(<span class='no'>nancycats</span>, <span class='kw'>total</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>H</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>G</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>E5</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># The previous version of poppr contained a definition of Hexp, which</span>
<span class='co'># was calculated as (N/(N - 1))*lambda. It basically looks like an unbiased </span>
<span class='co'># Simpson's index. This statistic was originally included in poppr because it</span>
<span class='co'># was originally included in the program multilocus. It was finally figured</span>
<span class='co'># to be an unbiased Simpson's diversity metric (Lande, 1996; Good, 1953).</span>

<span class='fu'>data</span>(<span class='no'>Aeut</span>)

<span class='no'>uSimp</span> <span class='kw'>&lt;-</span> <span class='kw'>function</span>(<span class='no'>x</span>){
  <span class='no'>lambda</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>vegan</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/vegan/topics/diversity'>diversity</a></span>(<span class='no'>x</span>, <span class='st'>"simpson"</span>)
  <span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>drop</span>(<span class='fu'>as.matrix</span>(<span class='no'>x</span>))
  <span class='kw'>if</span> (<span class='fu'>length</span>(<span class='fu'>dim</span>(<span class='no'>x</span>)) <span class='kw'>&gt;</span> <span class='fl'>1</span>){
    <span class='no'>N</span> <span class='kw'>&lt;-</span> <span class='fu'>rowSums</span>(<span class='no'>x</span>)
  } <span class='kw'>else</span> {
    <span class='no'>N</span> <span class='kw'>&lt;-</span> <span class='fu'>sum</span>(<span class='no'>x</span>)
  }
  <span class='fu'>return</span>((<span class='no'>N</span>/(<span class='no'>N</span>-<span class='fl'>1</span>))*<span class='no'>lambda</span>)
}
<span class='fu'>poppr</span>(<span class='no'>Aeut</span>, <span class='kw'>uSimp</span> <span class='kw'>=</span> <span class='no'>uSimp</span>)


<span class='co'># Demonstration with viral data</span>
<span class='co'># Note: this is a larger data set that could take a couple of minutes to run</span>
<span class='co'># on slower computers. </span>
<span class='fu'>data</span>(<span class='no'>H3N2</span>)
<span class='fu'>strata</span>(<span class='no'>H3N2</span>) <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='fu'>other</span>(<span class='no'>H3N2</span>)$<span class='no'>x</span>)
<span class='fu'>setPop</span>(<span class='no'>H3N2</span>) <span class='kw'>&lt;-</span> ~<span class='no'>country</span>
<span class='fu'>poppr</span>(<span class='no'>H3N2</span>, <span class='kw'>total</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>sublist</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='st'>"Austria"</span>, <span class='st'>"China"</span>, <span class='st'>"USA"</span>),
                                <span class='kw'>clonecorrect</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>strata</span> <span class='kw'>=</span> ~<span class='no'>country</span>/<span class='no'>year</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Zhian N. Kamvar

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

