<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="poppr">
<title>Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')` • poppr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`">
<meta property="og:description" content="poppr">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</h1>
                        <h4 data-toc-skip class="author">Zhian N. Kamvar
and Niklaus J. Grünwald</h4>
            
            <h4 data-toc-skip class="date">2024-03-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/vignettes/mlg.Rmd" class="external-link"><code>vignettes/mlg.Rmd</code></a></small>
      <div class="d-none name"><code>mlg.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>A multilocus genotype is a unique combination of alleles across two
or more loci. For organisms such as plant pathogens that often reproduce
clonally, multilocus genotypes are invaluable for identifying the mode
and spread of an organism. This document will describe in detail how you
can define multilocus genotypes/lineages and how you can use them for
your analyses of both genclone and snpclone objects.</p>
<p><br><br></p>
<center>
<img src="jalapeno_logo.png" width="300">
</center>
<hr>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
</div>
<div class="section level2">
<h2 id="multilocus-genotype-flavors">Multilocus Genotype Flavors<a class="anchor" aria-label="anchor" href="#multilocus-genotype-flavors"></a>
</h2>
<p>There are three different ways for constructing multilocus genotypes
in <em>poppr</em>.</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Method</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>"original"</code></td>
<td>Default MLG definition; strict matching</td>
</tr>
<tr class="even">
<td><code>"contracted"</code></td>
<td>Collapsing MLGs by genetic distance with
<code>mlg.filter</code>
</td>
</tr>
<tr class="odd">
<td><code>"custom"</code></td>
<td>User-defined multilocus genotypes</td>
</tr>
</tbody>
</table>
<p>The first is a simply naïve construction where all alleles must match
to make a unique multilocus genotype. New in version 2.0 is the ability
to use genetic distance to collapse multilocus genotypes or define
custom multilocus lineages based on other information. You can use the
function <code><a href="../reference/mll-method.html">mll()</a></code> to display and switch between different
multilocus genotypes/lineages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span></span>
<span><span class="va">monpop</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    264 multilocus genotypes </span></span>
<span><span class="co">##    694 haploid individuals</span></span>
<span><span class="co">##     13 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      1 stratum - Pop</span></span>
<span><span class="co">##     12 populations defined - </span></span>
<span><span class="co">## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span>, <span class="st">"original"</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span> <span class="co"># Showing the definitions for the first 20 samples</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: </span></span>
<span><span class="co">##  The @mlg slot does not contain an MLG class object.</span></span>
<span><span class="co">##  Returning the original mlgs. Please use:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  mll(monpop) &lt;- "original"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  to convert your object.</span></span></code></pre>
<pre><code><span><span class="co">##  [1] 260 179 169 168 167 221 152 133 144  78  78  79  81  44  41  40  40  38 119</span></span>
<span><span class="co">## [20] 118</span></span></code></pre>
<p>Notice how we got a warning. This is because the <code>monpop</code>
data set was created before <code><a href="../reference/mll-method.html">mll()</a></code> was implemented. Luckily,
this warning gives us information on how to fix it.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"original"</span></span>
<span><span class="va">monpop</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    264 original multilocus genotypes </span></span>
<span><span class="co">##    694 haploid individuals</span></span>
<span><span class="co">##     13 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      1 stratum - Pop</span></span>
<span><span class="co">##     12 populations defined - </span></span>
<span><span class="co">## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</span></span></code></pre>
<p>We can see that the description says 264 <strong>original</strong>
multilocus genotypes . This is how we know what the current MLG
definition is for our object. Let’ see what happens when we change
it.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"custom"</span></span>
<span><span class="va">monpop</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    264 custom multilocus genotypes </span></span>
<span><span class="co">##    694 haploid individuals</span></span>
<span><span class="co">##     13 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      1 stratum - Pop</span></span>
<span><span class="co">##     12 populations defined - </span></span>
<span><span class="co">## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</span></span></code></pre>
<p>Now it says 264 <strong>custom</strong> multilocus genotypes . Of
course, if we look at the MLGs, we will see that they appear to be the
same as our original definition:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span>, <span class="st">"custom"</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span> <span class="co"># Showing the definitions for the first 20 samples</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 260 179 169 168 167 221 152 133 144 78  78  79  81  44  41  40  40  38  119</span></span>
<span><span class="co">## [20] 118</span></span>
<span><span class="co">## 264 Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 ... 264</span></span></code></pre>
<p>All of these functions will work for both <strong>genclone</strong>
and <strong>snpclone</strong> objects. In this section, we will
demonstrate how to construct each of these three multilocus genotype
definitions in different data sets.</p>
<div class="section level3">
<h3 id="naïve-original">Naïve (“original”)<a class="anchor" aria-label="anchor" href="#na%C3%AFve-original"></a>
</h3>
<p>This is the default way <em>poppr</em> calculates multilocus
genotypes. You see it immediately when you load a genclone object. Let’s
use the example <code>monpop</code> from <span class="citation">(Everhart and Scherm 2015)</span>. First, we need to
reset the data to our original MLGs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"original"</span></span>
<span><span class="va">monpop</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    264 original multilocus genotypes </span></span>
<span><span class="co">##    694 haploid individuals</span></span>
<span><span class="co">##     13 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      1 stratum - Pop</span></span>
<span><span class="co">##     12 populations defined - </span></span>
<span><span class="co">## 7_09_BB, 26_09_BB, 26_09_FR, ..., 45_10_FR, 26_11_BB, 26_11_FR</span></span></code></pre>
<p>You notice that there are 694 samples, yet only 264 multilocus
genotypes in the data set over 13 loci. In this sub-section, I will
demonstrate how these MLGs are constructed.</p>
<p>The naïve definition simply takes strings of alleles and compares
them for equality. This method is quick and easily interpretable, but
means that things like genotyping error, hyper-variable loci, and
missing data all contribute to a unique genotype that might not be truly
unique <span class="citation">(Kamvar, Brooks, and Grünwald
2015)</span>.</p>
<p>To give an example, let’s imagine that we have five samples with
alleles typed at two loci.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grid_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>,</span>
<span>                         <span class="fl">1</span>, <span class="fl">1</span>,</span>
<span>                         <span class="fl">5</span>, <span class="fl">1</span>,</span>
<span>                         <span class="fl">9</span>, <span class="fl">1</span>,</span>
<span>                         <span class="fl">9</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>                       ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                       byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">grid_example</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">grid_example</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span><span class="va">grid_example</span></span></code></pre></div>
<pre><code><span><span class="co">##   x y</span></span>
<span><span class="co">## A 1 4</span></span>
<span><span class="co">## B 1 1</span></span>
<span><span class="co">## C 5 1</span></span>
<span><span class="co">## D 9 1</span></span>
<span><span class="co">## E 9 4</span></span></code></pre>
<p>We notice how they all have different multilocus genotypes. Indeed,
if we analyze them, we see that there are five multilocus genotypes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html" class="external-link">df2genind</a></span><span class="op">(</span><span class="va">grid_example</span>, ploidy <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>  <span class="co"># Look at the multilocus genotype table</span></span></code></pre></div>
<pre><code><span><span class="co">##   x.1 x.5 x.9 y.4 y.1</span></span>
<span><span class="co">## A   1   0   0   1   0</span></span>
<span><span class="co">## B   1   0   0   0   1</span></span>
<span><span class="co">## C   0   1   0   0   1</span></span>
<span><span class="co">## D   0   0   1   0   1</span></span>
<span><span class="co">## E   0   0   1   1   0</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">nmll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># count the number of multilocus genotypes</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>  <span class="co"># show the multilocus genotype definitions</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5 4 3 1 2</span></span></code></pre>
<p>What we did above was to analyze the <strong>tab</strong> slot of the
object showing the counts of alleles per sample across the two loci. We
can clearly see by eye that no two samples are alike. The
<code><a href="../reference/mll-method.html">nmll()</a></code> function counts the <strong>n</strong>umber of
<strong>m</strong>ulti<strong>l</strong>ocus <strong>l</strong>ineages
in the sample. the <code><a href="../reference/mll-method.html">mll()</a></code> function displays the assignment
of the multilocus lineages in the sample.</p>
<blockquote>
<p>Note: we used <code><a href="../reference/mll-method.html">mll()</a></code> to show us our multilocus lineages.
Later on, we will use this same function to switch between different
definitions. See ?mll for details.</p>
</blockquote>
<p>Now let’s say we included another sample with partial missing data.
Let’s say it was typed as allele “5” at the <strong>x</strong> locus,
but missing the allele at the <strong>y</strong> locus.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html" class="external-link">df2genind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">grid_example</span>, new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>, ploidy <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>  <span class="co"># Note the missing data at locus 2.</span></span></code></pre></div>
<pre><code><span><span class="co">##     x.1 x.5 x.9 y.4 y.1</span></span>
<span><span class="co">## A     1   0   0   1   0</span></span>
<span><span class="co">## B     1   0   0   0   1</span></span>
<span><span class="co">## C     0   1   0   0   1</span></span>
<span><span class="co">## D     0   0   1   0   1</span></span>
<span><span class="co">## E     0   0   1   1   0</span></span>
<span><span class="co">## new   0   1   0  NA  NA</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">nmll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6 5 3 1 2 4</span></span></code></pre>
<p>Now we see that we have 6 multilocus genotypes even though one of
them contains missing data.</p>
<p>Let’s also imagine that we included yet another sample that had a
low-frequency allele at locus <strong>y</strong>, “2”, but had the
allele “5” at locus <strong>x</strong>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grid_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">grid_example</span>,</span>
<span>                  new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>                  mut <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>                  <span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html" class="external-link">df2genind</a></span><span class="op">(</span><span class="va">grid_new</span>, ploidy <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     x.1 x.5 x.9 y.4 y.1 y.2</span></span>
<span><span class="co">## A     1   0   0   1   0   0</span></span>
<span><span class="co">## B     1   0   0   0   1   0</span></span>
<span><span class="co">## C     0   1   0   0   1   0</span></span>
<span><span class="co">## D     0   0   1   0   1   0</span></span>
<span><span class="co">## E     0   0   1   1   0   0</span></span>
<span><span class="co">## new   0   1   0  NA  NA  NA</span></span>
<span><span class="co">## mut   0   1   0   0   0   1</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">nmll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7 6 4 1 2 5 3</span></span></code></pre>
<p>Again, we get completely different genotypes, but notice how the
genotypes we added are closer to the genotype that shares the 5 allele.
The reason this happens is because the naïve algorithm reports the rank
of the concatenated alleles like so:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">xt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">paste</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           A           B           C           D           E         new </span></span>
<span><span class="co">##    "100100"    "100010"    "010010"    "001010"    "001100" "010NANANA" </span></span>
<span><span class="co">##         mut </span></span>
<span><span class="co">##    "010001"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rank.html" class="external-link">rank</a></span><span class="op">(</span><span class="va">xt</span>, ties.method <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   A   B   C   D   E new mut </span></span>
<span><span class="co">##   7   6   4   1   2   5   3</span></span></code></pre>
<p>Even though we have reason to believe that the samples “new” and
“mut” might actually have the genotype x.5 and y.1 (the MLG of sample
C), the algorithm sees them as different. If we calculate the pairwise
euclidean distances between samples, we see that “new”, “mut” and, “C”
are very similar to each other:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://ape-package.ird.fr/" class="external-link">"ape"</a></span><span class="op">)</span></span>
<span><span class="va">raw_dist</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/genind2df.html" class="external-link">genind2df</a></span><span class="op">(</span><span class="va">x</span>, usepop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="op">(</span><span class="va">xdis</span> <span class="op">&lt;-</span> <span class="fu">raw_dist</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            A        B        C        D        E      new</span></span>
<span><span class="co">## B   3.000000                                             </span></span>
<span><span class="co">## C   5.000000 4.000000                                    </span></span>
<span><span class="co">## D   8.544004 8.000000 4.000000                           </span></span>
<span><span class="co">## E   8.000000 8.544004 5.000000 3.000000                  </span></span>
<span><span class="co">## new 5.656854 5.656854 0.000000 5.656854 5.656854         </span></span>
<span><span class="co">## mut 4.472136 4.123106 1.000000 4.123106 4.472136 0.000000</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="fu"><a href="../reference/upgma.html">upgma</a></span><span class="op">(</span><span class="va">xdis</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-12-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Therefore, we might want to consider “new”, “mut” and, “C” to be the
same Multilocus Lineage (MLL). In the next section, you will see how to
collapse multilocus genotypes by genetic distance.</p>
</div>
<div class="section level3">
<h3 id="filtered-contracted">Filtered (“contracted”)<a class="anchor" aria-label="anchor" href="#filtered-contracted"></a>
</h3>
<p>To remedy the issues with a naïve definition of multilocus genotypes,
we can utilize genetic distance, which will allow us to collapse
multilocus genotypes that are under a specific distance threshold. The
main function we will use in this section is <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>.
It will create a dissimiliarity distance matrix from the data and then
filter based off of that matrix. You can also utilize your own supplied
distance matrix or distance function.</p>
<div class="section level4">
<h4 id="main-usage">Main usage<a class="anchor" aria-label="anchor" href="#main-usage"></a>
</h4>
<p>To use <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>, you need to define a thresold. This
threshold defines the minimum distance to consider two genotypes unique,
so anything below that will be clustered into the same MLG. In practice,
it’s used like this (using the data and the euclidean distance described
above):</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="co"># We have 7 MLGs before filtering</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    7 original multilocus genotypes </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">xdis</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span></span>
<span><span class="va">x</span> <span class="co"># Now we have 5 MLGs</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    5 contracted multilocus genotypes</span></span>
<span><span class="co">##      (1) [t], (xdis) [d], (farthest) [a] </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"original"</span> <span class="co"># We'll reset to the naive definition</span></span></code></pre></div>
<p>Here, we told <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> to set the threshold to 1
plus a very tiny number (this will be explained in the next section)
based off of the distance matrix <strong>xdis</strong>. This is a way to
manipulate the object in place. In this section, there will be examples
of retrieving the MLGs and other statistics from the function as well as
how to use different distance functions and reset the MLGs.</p>
</div>
<div class="section level4">
<h4 id="thresholds">Thresholds<a class="anchor" aria-label="anchor" href="#thresholds"></a>
</h4>
<p>As we saw above, in order to collapse all three new samples into the
original 5 MLGs, we had to set a threshold of <em>just above</em> 1. To
show you why, we will use the function to return to us the MLGs it
created and the thresholds that were passed to cluster MLLs (by using
the <code>stats</code> argument).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"original"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7 6 4 1 2 5 3</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">xdis</span>, threshold <span class="op">=</span> <span class="fl">1</span>, stats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mlg"</span>, <span class="st">"thresholds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $MLGS</span></span>
<span><span class="co">## [1] 7 6 4 1 2 3 3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $THRESHOLDS</span></span>
<span><span class="co">## [1] 0</span></span></code></pre>
<p>In our output, we can see that MLG 5 has collapsed into MLG 3. This
occured as soon as our threshold passed 0. Take a look at what happens
when we add a very small number to 1.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">e</span> <span class="op">&lt;-</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span><span class="op">)</span> <span class="co"># A very tiny number</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.490116e-08</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">xdis</span>, threshold <span class="op">=</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">e</span>, stats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mlg"</span>, <span class="st">"thresholds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $MLGS</span></span>
<span><span class="co">## [1] 7 6 4 1 2 4 4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $THRESHOLDS</span></span>
<span><span class="co">## [1] 0 1</span></span></code></pre>
<p>Now that we have set the threshold to <em>just above</em> 1, we have
5 unique MLLs since MLG 3 collapsed into MLG 4. Because of the way
clustering happens, it’s important to point out what happens when you
attempt to use a threshold of 0.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">xdis</span>, threshold <span class="op">=</span> <span class="fl">0</span>, stats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mlg"</span>, <span class="st">"thresholds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $MLGS</span></span>
<span><span class="co">## [1] 7 6 4 1 2 5 3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $THRESHOLDS</span></span>
<span><span class="co">## numeric(0)</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"original"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7 6 4 1 2 5 3</span></span></code></pre>
<p>Notice that all the MLGs are the same as the originally defined MLGs.
Using a threshold of zero allows two MLGs separated by zero genetic
distance to remain mutually unique.</p>
</div>
<div class="section level4">
<h4 id="setting-the-genotypes">Setting the genotypes<a class="anchor" aria-label="anchor" href="#setting-the-genotypes"></a>
</h4>
<p>As demonstrated at the beginning of this section, if we wanted to
reassign our sample genotypes to the collapsed version, all we would
have to do is use the <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code> commands above, but use
the <code>&lt;-</code> operator to specify the threshold. Note that I
will be using <code><a href="../reference/mlg.html">mlg.table()</a></code> to show the distribution of the
multilocus genotypes before and after filtering.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    7 original multilocus genotypes </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># Before: 7 MLGs</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##       MLG.1 MLG.2 MLG.3 MLG.4 MLG.5 MLG.6 MLG.7</span></span>
<span><span class="co">## Total     1     1     1     1     1     1     1</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">xdis</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">e</span></span>
<span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    5 contracted multilocus genotypes</span></span>
<span><span class="co">##      (1) [t], (xdis) [d], (farthest) [a] </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># After: 5 MLGs</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-17-2.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##       MLG.1 MLG.2 MLG.4 MLG.6 MLG.7</span></span>
<span><span class="co">## Total     1     1     3     1     1</span></span></code></pre>
<p>Notice how the information you see has changed. You can see that it’s
printed out that you have 5 <strong>contracted</strong> multilocus
genotypes, but you have some cryptic code underneath:</p>
<table class="table">
<thead><tr class="header">
<th align="center">symbol</th>
<th align="left">meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>[t]</strong></td>
<td align="left">threshold</td>
</tr>
<tr class="even">
<td align="center"><strong>[d]</strong></td>
<td align="left">distance (can be matrix or function)</td>
</tr>
<tr class="odd">
<td align="center"><strong>[a]</strong></td>
<td align="left">algorithm (see the next section)</td>
</tr>
</tbody>
</table>
<p>Genclone and snpclone objects will always remember what parameters
were used for filtering multilocus genotypes, but the only catch is
that, if you use your own supplied distance (matrix or function), you
must be very careful not to delete it or change the object name.</p>
<p>This means that you don’t always have to specify the distance when
assigning a threshold:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">4.51</span></span>
<span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    3 contracted multilocus genotypes</span></span>
<span><span class="co">##      (4.51) [t], (xdis) [d], (farthest) [a] </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##       MLG.1 MLG.4 MLG.6</span></span>
<span><span class="co">## Total     2     3     2</span></span></code></pre>
<blockquote>
<p>DANGER! DANGER! While this is a convenient function, if you use a
matrix or function that was created by you, you <strong>must not delete
it or change its name</strong>. Only the name of the matrix/function is
stored, so if you delete the matrix/function used to inform your
filtering, you will get an error:</p>
</blockquote>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">xdis</span><span class="op">)</span> <span class="co"># NOOOOOO!</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">e</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error : cannot evaluate distance function, it might be missing.</span></span></code></pre>
<pre><code><span><span class="co">##  Error: cannot evaluate distance function, it might be missing.</span></span></code></pre>
<p>Basically, your object thinks that there should be a matrix called
<strong>xdis</strong>, but it can’t find it anywhere. But don’t worry,
we can restore it if we have the function available. We will use the
<code>raw_dist()</code> function that we defined earlier.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">raw_dist</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">e</span></span>
<span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    5 contracted multilocus genotypes</span></span>
<span><span class="co">##      (1) [t], (raw_dist) [d], (farthest) [a] </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<p>The safest way, perhaps, is to use a function defined in
<em>poppr</em>. For example, we’ll use Bruvo’s distance since it takes
into account the real value of the alleles <span class="citation">(Bruvo
et al. 2004)</span>.</p>
<blockquote>
<p>The arguments to your distance function will be stored in the object
as well!</p>
</blockquote>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bruvo.dist.html">bruvo.dist</a></span><span class="op">(</span><span class="va">x</span>, replen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             A         B         C         D         E       new</span></span>
<span><span class="co">## B   0.4375000                                                  </span></span>
<span><span class="co">## C   0.9062500 0.4687500                                        </span></span>
<span><span class="co">## D   0.9355469 0.4980469 0.4687500                              </span></span>
<span><span class="co">## E   0.4980469 0.9355469 0.9062500 0.4375000                    </span></span>
<span><span class="co">## new 0.9375000 0.9375000 0.0000000 0.9375000 0.9375000          </span></span>
<span><span class="co">## mut 0.8437500 0.7187500 0.2500000 0.7187500 0.8437500 0.0000000</span></span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">x</span>, distance <span class="op">=</span> <span class="va">bruvo.dist</span>, replen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.44</span></span>
<span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    3 contracted multilocus genotypes</span></span>
<span><span class="co">##      (0.44) [t], (bruvo.dist) [d], (farthest) [a] </span></span>
<span><span class="co">##    7 haploid individuals</span></span>
<span><span class="co">##    2 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    0 strata. </span></span>
<span><span class="co">##    0 populations defined.</span></span></code></pre>
<p>Of course, our multilocus genotypes are not changed forever, they are
just stored in a different place. We can access the original, naïve
multilocus genotypes by using the <code><a href="../reference/mll-method.html">mll()</a></code> function:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"original"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7 6 4 1 2 5 3</span></span></code></pre>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># contracted</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6 6 4 1 1 4 4</span></span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"original"</span></span>
<span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># original</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7 6 4 1 2 5 3</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="tie-breakers-algorithms">Tie breakers (algorithms)<a class="anchor" aria-label="anchor" href="#tie-breakers-algorithms"></a>
</h4>
<p>Underlying <code>mlg.filter</code> are three algorithms that decide
what genotypes go together <span class="citation">(Kamvar, Brooks, and
Grünwald 2015)</span>:</p>
<ul>
<li>nearest neighbor</li>
<li>farthest neighbor (default)</li>
<li>average neighbor (UPGMA)</li>
</ul>
<p>You can specify which algorithm you want to use in clustering with
the <code>algorithm</code> argument. Each of these algorithms have
different behaviors when it comes to collapsing multilocus genotypes. In
short, farthest neighbor is the most conservative, nearest neighbor can
have a chaining effect, and average neighbor is somewhere in between.
Your choice of algorithm really depends on the biology of your
organism.</p>
<p>To help visualize this, there is the function
<code><a href="../reference/filter_stats.html">filter_stats()</a></code>, which will plot the output of the filtering
algorithm. For simplicity, we will use the <code>Pinf</code> data set
representing <em>Phytophthora infestans</em> samples from Mexico and
South America <span class="citation">(Goss et al. 2009)</span>.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span></span>
<span><span class="va">Pinf</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    72 multilocus genotypes </span></span>
<span><span class="co">##    86 tetraploid individuals</span></span>
<span><span class="co">##    11 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     2 strata - Continent, Country</span></span>
<span><span class="co">##     2 populations defined - South America, North America</span></span></code></pre>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pinfreps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fix_replen.html">fix_replen</a></span><span class="op">(</span><span class="va">Pinf</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">6</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pinf_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_stats.html">filter_stats</a></span><span class="op">(</span><span class="va">Pinf</span>, distance <span class="op">=</span> <span class="va">bruvo.dist</span>, replen <span class="op">=</span> <span class="va">pinfreps</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/plot_replen-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can see that the different algorithms behave similarly for small
thresholds, but begin to differ over larger thresholds.</p>
<p>This function is useful for finding all thresholds at which MLLs are
collapsed, which can help with choosing a threshold that collapses
putative clones in our sample into lineages.</p>
</div>
<div class="section level4">
<h4 id="choosing-a-threshold">Choosing a threshold<a class="anchor" aria-label="anchor" href="#choosing-a-threshold"></a>
</h4>
<p>After you have chosen a genetic distance and a filtering algorithm,
you must then decide on the threshold to represent the <strong>minimum
genetic distance at which two individuals would be considered from
different clonal lineages</strong>.</p>
<p>One method described in the literature of choosing a threshold is to
look for an initial, small peak in the histogram of pairwise genetic
distances and set the threshold to be between that peak and the larger
peak `<span class="citation">Bailleul, Stoeckel, and Arnaud-Haond
(2016)</span>. This initial peak likely represents clones differentiated
by a small set of random mutations. You can see this in the figure above
at a threshold of ~0.11 for the “farthest neighbor” algorithm.</p>
<p>However, if this peak is not obvious, then another method is to look
for the largest gap between all putative thresholds. For this, you can
use the <code><a href="../reference/cutoff_predictor.html">cutoff_predictor()</a></code> function with the output of
<code><a href="../reference/filter_stats.html">filter_stats()</a></code>. It should be noted that this method is not
a perfect solution. If we take the results from above, we can find the
threshold for each algorithm:</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">farthest_thresh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cutoff_predictor.html">cutoff_predictor</a></span><span class="op">(</span><span class="va">pinf_filtered</span><span class="op">$</span><span class="va">farthest</span><span class="op">$</span><span class="va">THRESHOLDS</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.1132221</span></span></code></pre>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">average_thresh</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cutoff_predictor.html">cutoff_predictor</a></span><span class="op">(</span><span class="va">pinf_filtered</span><span class="op">$</span><span class="va">average</span><span class="op">$</span><span class="va">THRESHOLDS</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.1084407</span></span></code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">nearest_thresh</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cutoff_predictor.html">cutoff_predictor</a></span><span class="op">(</span><span class="va">pinf_filtered</span><span class="op">$</span><span class="va">nearest</span><span class="op">$</span><span class="va">THRESHOLDS</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.1092773</span></span></code></pre>
<p>Now we can define multilocus lineages for <em>P. infestans</em> with
the following criteria:</p>
<table class="table"><tbody>
<tr class="odd">
<td align="center"><strong>[t]</strong></td>
<td align="left">threshold</td>
<td align="left">0.113</td>
</tr>
<tr class="even">
<td align="center"><strong>[d]</strong></td>
<td align="left">distance</td>
<td align="left">Bruvo’s Distance</td>
</tr>
<tr class="odd">
<td align="center"><strong>[a]</strong></td>
<td align="left">algorithm</td>
<td align="left">Farthest neighbor</td>
</tr>
</tbody></table>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">Pinf</span>, distance <span class="op">=</span> <span class="va">bruvo.dist</span>, replen <span class="op">=</span> <span class="va">pinfreps</span>, algorithm <span class="op">=</span> <span class="st">"f"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">farthest_thresh</span></span>
<span><span class="va">Pinf</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    48 contracted multilocus genotypes</span></span>
<span><span class="co">##       (0.113) [t], (bruvo.dist) [d], (farthest) [a] </span></span>
<span><span class="co">##    86 tetraploid individuals</span></span>
<span><span class="co">##    11 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     2 strata - Continent, Country</span></span>
<span><span class="co">##     2 populations defined - South America, North America</span></span></code></pre>
<blockquote>
<p>Note: Please take care to critically evaluate the results and make
sure it applies to your data. This function WILL give you an answer
whether or not one truely exists. Additionally, For diploid organisms,
another method of choosing a threshold is available in the
<strong>RClone</strong> package that involves simulating outcrossing and
inbreeding events <span class="citation">(Bailleul, Stoeckel, and
Arnaud-Haond 2016)</span>.</p>
</blockquote>
</div>
</div>
<div class="section level3">
<h3 id="custom-custom">Custom (“custom”)<a class="anchor" aria-label="anchor" href="#custom-custom"></a>
</h3>
<p>Sometimes multilocus genotypes are defined by more than just genetic
data and it would be useful to be able to analyze these genotype
definitions. Luckily, there is a way to do this. <em>Poppr</em> has
support for custom multilocus genotypes. For example, we have a
simulated data set that has 50 samples with 26 multilocus genotypes.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">partial_clone</span><span class="op">)</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">partial_clone</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  8  9 23 24 22 21 10  3 22 11 24  9 25  4 12  2 14  1  9  9  9 26  7 13 23</span></span>
<span><span class="co">## [26]  3 17 22  6 20 22 12  5 25 13 21 15 13 13 13  2 19 18 13 23 16  1 11 25  4</span></span></code></pre>
<p>Let’s change the MLGs to letters instead of numbers. We will use
<code><a href="../reference/mll.custom.html">mll.custom()</a></code> to do this.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LETTERS</span><span class="op">[</span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">]</span>  <span class="co"># The new MLGs</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "H" "I" "W" "X" "V" "U" "J" "C" "V" "K" "X" "I" "Y" "D" "L" "B" "N" "A" "I"</span></span>
<span><span class="co">## [20] "I" "I" "Z" "G" "M" "W" "C" "Q" "V" "F" "T" "V" "L" "E" "Y" "M" "U" "O" "M"</span></span>
<span><span class="co">## [39] "M" "M" "B" "S" "R" "M" "W" "P" "A" "K" "Y" "D"</span></span></code></pre>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll.custom.html">mll.custom</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-25-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##   A B C D E F G H I J K L M N O P Q R S T U V W X Y Z</span></span>
<span><span class="co">## 1 0 1 0 0 1 1 0 1 1 0 0 0 0 1 1 0 0 0 0 0 0 2 2 0 2 0</span></span>
<span><span class="co">## 2 1 0 1 2 0 0 0 0 1 0 1 0 1 0 0 1 0 0 1 1 1 0 0 0 1 1</span></span>
<span><span class="co">## 3 1 0 0 0 0 0 1 0 1 1 0 1 2 0 0 0 1 1 0 0 0 1 1 1 0 0</span></span>
<span><span class="co">## 4 0 1 1 0 0 0 0 0 2 0 1 1 3 0 0 0 0 0 0 0 1 1 0 1 0 0</span></span></code></pre>
<p>This even works with minimum spanning networks:</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcpal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"gold"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">9001</span><span class="op">)</span></span>
<span><span class="va">pcmsn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bruvo.msn.html">bruvo.msn</a></span><span class="op">(</span><span class="va">pc</span>, replen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nLoc</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">)</span>, palette <span class="op">=</span> <span class="va">pcpal</span>,</span>
<span>                   vertex.label.color <span class="op">=</span> <span class="st">"firebrick"</span>, vertex.label.font <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   vertex.label.cex <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-26-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Let’s say that we realized that we have strong evidence for MLG “Q”
being the same as MLG “M”. If we wanted to change those, we could simply
change the factor levels:</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll.custom.html">mll.levels</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="../reference/mll.custom.html">mll.levels</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span> <span class="op">==</span> <span class="st">"Q"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"M"</span></span></code></pre></div>
<p>And we can plot again</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">9001</span><span class="op">)</span></span>
<span><span class="va">pcmsn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bruvo.msn.html">bruvo.msn</a></span><span class="op">(</span><span class="va">pc</span>, replen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nLoc</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">)</span>, palette <span class="op">=</span> <span class="va">pcpal</span>,</span>
<span>                   vertex.label.color <span class="op">=</span> <span class="st">"firebrick"</span>, vertex.label.font <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   vertex.label.cex <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-28-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Notice how the minimum spanning network architecture stayed the same,
but the labels had changed.</p>
</div>
</div>
<div class="section level2">
<h2 id="diversity-analysis">Diversity Analysis<a class="anchor" aria-label="anchor" href="#diversity-analysis"></a>
</h2>
<p>In the <strong>Data Import and Manipulation</strong> vignette, the
first command demonstrated was the <code><a href="../reference/poppr.html">poppr()</a></code>, command, which
produced a table of diversity statistics, most of which were calculated
on counts of multilocus genotypes. These statistics can be replicated by
using the function <code><a href="../reference/diversity_stats.html">diversity_stats()</a></code>. This function
calculates diversity statistics based off of a table of multilocus
genotypes. Let’s analyze the diversity of the peach brown rot pathogen,
<em>Monilinia fructicola</em>, structured by Symptom (Fruit Rot:FR and
Blossom Blight:BB) and Year <span class="citation">(Everhart and Scherm
2015)</span>. Let’s first look at the distribution of multilocus
genotypes:</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">splitStrata</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">Tree</span><span class="op">/</span><span class="va">Year</span><span class="op">/</span><span class="va">Symptom</span></span>
<span><span class="va">montab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">monpop</span>, strata <span class="op">=</span> <span class="op">~</span><span class="va">Symptom</span><span class="op">/</span><span class="va">Year</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-29-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can see from these plots that the Fruit Rot (FR) have a lot more
clones in the canopies than Blossom Blight (BB). This makes biological
sense since Fruit Rot infections are clonally propagated, whereas
Blossom Blight infections are from sexual propagules <span class="citation">(Everhart and Scherm 2015)</span>. Let’s look at the
diversity metrics.</p>
<div class="section level3">
<h3 id="basic-statistics">Basic Statistics<a class="anchor" aria-label="anchor" href="#basic-statistics"></a>
</h3>
<p>The function <code><a href="../reference/diversity_stats.html">diversity_stats()</a></code> is used to get basic
diversity statistics:</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">monstat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diversity_stats.html">diversity_stats</a></span><span class="op">(</span><span class="va">montab</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Index</span></span>
<span><span class="co">## Pop            H        G    lambda       E.5</span></span>
<span><span class="co">##   BB_9  3.580258 25.92405 0.9614258 0.7145086</span></span>
<span><span class="co">##   FR_9  3.103899 12.43711 0.9195955 0.5373402</span></span>
<span><span class="co">##   BB_10 2.944439 19.00000 0.9473684 1.0000000</span></span>
<span><span class="co">##   FR_10 4.066838 38.10132 0.9737542 0.6466786</span></span>
<span><span class="co">##   BB_11 3.401197 30.00000 0.9666667 1.0000000</span></span>
<span><span class="co">##   FR_11 3.094688 10.48941 0.9046657 0.4501543</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="confidence-intervals">Confidence Intervals<a class="anchor" aria-label="anchor" href="#confidence-intervals"></a>
</h3>
<p>We can get confidence intervals for these statistics using
<code><a href="../reference/diversity_ci.html">diversity_ci()</a></code>:</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/diversity_ci.html">diversity_ci</a></span><span class="op">(</span><span class="va">montab</span>, n <span class="op">=</span> <span class="fl">100L</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Confidence Intervals have been centered around observed statistic.</span></span>
<span><span class="co">## Please see ?diversity_ci for details.</span></span></code></pre>
<p><img src="mlg_files/figure-html/unnamed-chunk-31-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##          H H.est           H.ci    G G.est             G.ci lambda lambda.est</span></span>
<span><span class="co">## BB_9  3.58  3.19 (3.341, 3.819) 25.9 19.26  (19.298, 32.55)  0.961      0.946</span></span>
<span><span class="co">## FR_9  3.10  2.97 (2.912, 3.296) 12.4 11.98  (9.918, 14.956)  0.920      0.916</span></span>
<span><span class="co">## BB_10 2.94  2.37 (2.664, 3.225) 19.0  9.81 (15.986, 22.014)  0.947      0.895</span></span>
<span><span class="co">## FR_10 4.07  3.87 (3.926, 4.207) 38.1 33.63 (31.199, 45.004)  0.974      0.970</span></span>
<span><span class="co">## BB_11 3.40  2.83 (3.173, 3.629) 30.0 15.28 (26.066, 33.934)  0.967      0.933</span></span>
<span><span class="co">## FR_11 3.09  2.85 (2.802, 3.387) 10.5  9.81   (6.73, 14.249)  0.905      0.894</span></span>
<span><span class="co">##            lambda.ci   E.5 E.5.est         E.5.ci</span></span>
<span><span class="co">## BB_9  (0.942, 0.981) 0.715   0.776  (0.599, 0.83)</span></span>
<span><span class="co">## FR_9  (0.902, 0.937) 0.537   0.593 (0.488, 0.587)</span></span>
<span><span class="co">## BB_10 (0.913, 0.981) 1.000   0.894 (0.921, 1.079)</span></span>
<span><span class="co">## FR_10  (0.968, 0.98) 0.647   0.691 (0.579, 0.714)</span></span>
<span><span class="co">## BB_11 (0.948, 0.986) 1.000   0.889 (0.936, 1.064)</span></span>
<span><span class="co">## FR_11 (0.861, 0.949) 0.450   0.533   (0.35, 0.55)</span></span></code></pre>
<p>You can see that there was a warning about centering the confidence
interval. The boxplots you see are the distribution of the bootstrapped
replicates, but they are known to be biased. We invite users to please
read the documentation for <code><a href="../reference/diversity_ci.html">diversity_ci()</a></code> by typing
<code><a href="../reference/diversity_ci.html">?diversity_ci</a></code>. It is very important to be careful when
interpreting these results because sometimes confidence the confidence
intervals exist outside of the possible range, as exemplified by BB_10
and BB_11.</p>
</div>
<div class="section level3">
<h3 id="custom-statistics">Custom Statistics<a class="anchor" aria-label="anchor" href="#custom-statistics"></a>
</h3>
<p>In all of the <code>diversity_*</code> functions, you can use your
own custom diversity statistics. A common one we get requests for is the
clonal fraction, <span class="math inline">\(\frac{MLG}{N}\)</span> or
the number of multilocus genotypes over the number of samples. You can
add it in by writing your own function. Keep in mind, that you should
write it for both a matrix and a vector of counts if you want to be able
to bootstrap it.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myCF</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span> <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/drop.html" class="external-link">drop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span> <span class="co"># if it's a matrix</span></span>
<span>   <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span> <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>                 <span class="co"># if it's a vector</span></span>
<span>   <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span> <span class="op">}</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="op">(</span><span class="va">monstat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diversity_stats.html">diversity_stats</a></span><span class="op">(</span><span class="va">montab</span>, CF <span class="op">=</span> <span class="va">myCF</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Index</span></span>
<span><span class="co">## Pop            H        G    lambda       E.5        CF</span></span>
<span><span class="co">##   BB_9  3.580258 25.92405 0.9614258 0.7145086 0.7031250</span></span>
<span><span class="co">##   FR_9  3.103899 12.43711 0.9195955 0.5373402 0.2585366</span></span>
<span><span class="co">##   BB_10 2.944439 19.00000 0.9473684 1.0000000 1.0000000</span></span>
<span><span class="co">##   FR_10 4.066838 38.10132 0.9737542 0.6466786 0.3405018</span></span>
<span><span class="co">##   BB_11 3.401197 30.00000 0.9666667 1.0000000 1.0000000</span></span>
<span><span class="co">##   FR_11 3.094688 10.48941 0.9046657 0.4501543 0.4536082</span></span></code></pre>
<p>You can use filtered or custom MLGs to compare diversity. Here, I’m
filtering genotypes in <code>monpop</code> that are different by only a
single mutational step <span class="citation">(Bruvo et al.
2004)</span>.</p>
<blockquote>
<p>One mutational step for a single allele in Bruvo’s distance is
equivalent to 0.5, so a single mutational step for a haploid organism
with 13 loci is 0.5/13.</p>
</blockquote>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Repeat lengths are necessary</span></span>
<span><span class="va">reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fix_replen.html">fix_replen</a></span><span class="op">(</span><span class="va">monpop</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>CHMFc4 <span class="op">=</span> <span class="fl">7</span>, CHMFc5 <span class="op">=</span> <span class="fl">2</span>, CHMFc12 <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                     SEA <span class="op">=</span> <span class="fl">4</span>, SED <span class="op">=</span> <span class="fl">4</span>, SEE <span class="op">=</span> <span class="fl">2</span>, SEG <span class="op">=</span> <span class="fl">6</span>,</span>
<span>                     SEI <span class="op">=</span> <span class="fl">3</span>, SEL <span class="op">=</span> <span class="fl">4</span>, SEN <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                     SEP <span class="op">=</span> <span class="fl">4</span>, SEQ <span class="op">=</span> <span class="fl">2</span>, SER <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Adding a little bit, so the threshold is included.</span></span>
<span><span class="va">e</span> <span class="op">&lt;-</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span></span>
<span></span>
<span><span class="co"># Using the default farthest neighbor algorithm to collapse genotypes</span></span>
<span><span class="fu"><a href="../reference/mlg.filter.html">mlg.filter</a></span><span class="op">(</span><span class="va">monpop</span>, distance <span class="op">=</span> <span class="va">bruvo.dist</span>, replen <span class="op">=</span> <span class="va">reps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">0.5</span><span class="op">/</span><span class="fl">13</span><span class="op">)</span> <span class="op">+</span> <span class="va">e</span></span>
<span><span class="va">montabf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">monpop</span>, strata <span class="op">=</span> <span class="op">~</span><span class="va">Symptom</span><span class="op">/</span><span class="va">Year</span><span class="op">)</span></span></code></pre></div>
<p><img src="mlg_files/figure-html/unnamed-chunk-34-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">monstatf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diversity_stats.html">diversity_stats</a></span><span class="op">(</span><span class="va">montabf</span>, CF <span class="op">=</span> <span class="va">myCF</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Index</span></span>
<span><span class="co">## Pop            H        G    lambda       E.5        CF</span></span>
<span><span class="co">##   BB_9  3.442117 23.81395 0.9580078 0.7541044 0.6093750</span></span>
<span><span class="co">##   FR_9  3.003859 11.99001 0.9165973 0.5734961 0.2195122</span></span>
<span><span class="co">##   BB_10 2.944439 19.00000 0.9473684 1.0000000 1.0000000</span></span>
<span><span class="co">##   FR_10 3.861202 33.10974 0.9697974 0.6901987 0.2724014</span></span>
<span><span class="co">##   BB_11 3.401197 30.00000 0.9666667 1.0000000 1.0000000</span></span>
<span><span class="co">##   FR_11 2.754565  8.00766 0.8751196 0.4762511 0.3195876</span></span></code></pre>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">monstat2</span> <span class="op">-</span> <span class="va">monstatf</span> <span class="co"># Take the difference from the unfiltered</span></span></code></pre></div>
<pre><code><span><span class="co">##        Index</span></span>
<span><span class="co">## Pop             H         G      lambda         E.5         CF</span></span>
<span><span class="co">##   BB_9  0.1381408 2.1100971 0.003417969 -0.03959575 0.09375000</span></span>
<span><span class="co">##   FR_9  0.1000406 0.4470973 0.002998215 -0.03615591 0.03902439</span></span>
<span><span class="co">##   BB_10 0.0000000 0.0000000 0.000000000  0.00000000 0.00000000</span></span>
<span><span class="co">##   FR_10 0.2056360 4.9915810 0.003956784 -0.04352013 0.06810036</span></span>
<span><span class="co">##   BB_11 0.0000000 0.0000000 0.000000000  0.00000000 0.00000000</span></span>
<span><span class="co">##   FR_11 0.3401226 2.4817496 0.029546179 -0.02609686 0.13402062</span></span></code></pre>
<p>We can see that filtered MLLs tend to be less diverse. This makes
intuitive sense as it is creating larger classes of multilocus
genotypes.</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"original"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="jack-knife-rarefaction">Jack-knife rarefaction<a class="anchor" aria-label="anchor" href="#jack-knife-rarefaction"></a>
</h3>
<p>The function <code><a href="../reference/diversity_ci.html">diversity_ci()</a></code> and
<code><a href="../reference/diversity_boot.html">diversity_boot()</a></code> have the option to perform jack knife
rarefaction calculations. This means that your data will be randomly
sub-sampled to either the smallest population size, or whatever is
specified in the parameter <code>n.rare</code>, whichever is bigger.
Here’s an example with the previous data set:</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">monrare</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diversity_ci.html">diversity_ci</a></span><span class="op">(</span><span class="va">montab</span>, n <span class="op">=</span> <span class="fl">100L</span>, rarefy <span class="op">=</span> <span class="cn">TRUE</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Samples for rarefaction: 19</span></span></code></pre>
<p><img src="mlg_files/figure-html/unnamed-chunk-37-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##          H H.est           H.ci    G G.est            G.ci lambda lambda.est</span></span>
<span><span class="co">## BB_9  3.58  2.71 (2.492, 2.944) 25.9 13.93     (9.494, 19)  0.961      0.926</span></span>
<span><span class="co">## FR_9  3.10  2.31 (1.922, 2.625) 12.4  8.75 (5.306, 12.448)  0.920      0.880</span></span>
<span><span class="co">## BB_10 2.94  2.94           &lt;NA&gt; 19.0 19.00            &lt;NA&gt;  0.947      0.947</span></span>
<span><span class="co">## FR_10 4.07  2.70 (2.362, 2.871) 38.1 13.98  (8.804, 17.19)  0.974      0.926</span></span>
<span><span class="co">## BB_11 3.40  2.94           &lt;NA&gt; 30.0 19.00            &lt;NA&gt;  0.967      0.947</span></span>
<span><span class="co">## FR_11 3.09  2.32  (1.831, 2.64) 10.5  8.42   (4.4, 12.448)  0.905      0.871</span></span>
<span><span class="co">##            lambda.ci   E.5 E.5.est         E.5.ci</span></span>
<span><span class="co">## BB_9  (0.895, 0.947) 0.715   0.907     (0.797, 1)</span></span>
<span><span class="co">## FR_9   (0.811, 0.92) 0.537   0.831   (0.7, 0.912)</span></span>
<span><span class="co">## BB_10           &lt;NA&gt; 1.000   1.000           &lt;NA&gt;</span></span>
<span><span class="co">## FR_10 (0.886, 0.942) 0.647   0.919 (0.775, 0.972)</span></span>
<span><span class="co">## BB_11           &lt;NA&gt; 1.000   1.000           &lt;NA&gt;</span></span>
<span><span class="co">## FR_11  (0.773, 0.92) 0.450   0.772   (0.627, 0.9)</span></span></code></pre>
<p>This can give you comparable estimates of diversity when not all
samples are of equal size.</p>
</div>
</div>
<div class="section level2">
<h2 id="clone-correction">Clone Correction<a class="anchor" aria-label="anchor" href="#clone-correction"></a>
</h2>
<p>Clone-correction works hierarchically and only uses the first MLG
copy encountered in the data per population. This is straightforward for
naïve MLGs, but for MLLs collapsed by genetic distance or custom MLLs,
this might change the results very slightly. As an example, let’s look
at the <code>monpop</code> filtered MLLs that we created earlier.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">nmll</a></span><span class="op">(</span><span class="va">monpop</span>, <span class="st">"original"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 264</span></span></code></pre>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">nmll</a></span><span class="op">(</span><span class="va">monpop</span>, <span class="st">"contracted"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 214</span></span></code></pre>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mll-method.html">mll</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"contracted"</span></span></code></pre></div>
<p>To show how the order of the samples can affect the sampling, we will
take the sum of all pairwise distances between clone-corrected samples
(corrected without respect to populations):</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">monpop</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span>strata <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span>  <span class="co"># 1. clone correct whole data set</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span>                     <span class="co"># 2. calculate distance</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span>                          <span class="co"># 3. take the sum of the distance</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 90268.73</span></span></code></pre>
<p>Now, what happens when we randomly sample individuals?</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">999</span><span class="op">)</span></span>
<span><span class="va">monpop</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span> <span class="co"># 1. shuffle samples</span></span>
<span>  <span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span>strata <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span>  <span class="co"># 2. clone correct whole data set</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span>                     <span class="co"># 3. calculate distance</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span>                          <span class="co"># 4. take the sum of the distance</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 90257.92</span></span></code></pre>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">monpop</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">monpop</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span> <span class="co"># 1. shuffle samples</span></span>
<span>  <span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span>strata <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span>  <span class="co"># 2. clone correct whole data set</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span>                     <span class="co"># 3. calculate distance</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span>                          <span class="co"># 4. take the sum of the distance</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 90311.96</span></span></code></pre>
<p>Notice how we are getting different results based on the order of
samples. This does not mean that the procedure doesn’t work, it just
means that we must be careful when clone-correcting modified multilocus
genotypes.</p>
</div>
<div class="section level2">
<h2 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>We have demonstrated here new methods for treating multilocus
genotypes on microsatellite data, but it is important to remember that
all of these functions can work with any source of data stored in either
a <strong>genclone</strong> or <strong>snpclone</strong> (derived from
<strong>genlight</strong>) object. Especially with
<code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>, these new functions will allow for a more
flexible analysis of WGS data of clonal organisms where true clones may
differ by more than a few mutations/errors/missing data. As with all
analyses, it is important to understand the algorithms used and take
them into account when interpreting results.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-arnaud2007standardizing" class="csl-entry">
Arnaud-Haond, Sophie, Carlos M Duarte, Filipe Alberto, and Ester A
Serrão. 2007. <span>“Standardizing Methods to Address Clonality in
Population Studies.”</span> <em>Molecular Ecology</em> 16 (24): 5115–39.
<a href="https://doi.org/10.1111/j.1365-294X.2007.03535.x" class="external-link">https://doi.org/10.1111/j.1365-294X.2007.03535.x</a>.
</div>
<div id="ref-bailleul2016rclone" class="csl-entry">
Bailleul, Diane, Solenn Stoeckel, and Sophie Arnaud-Haond. 2016.
<span>“<span>RClone</span>: A Package to Identify
<span>MultiLocus</span> Clonal Lineages and Handle Clonal Data Sets in
<span>R</span>.”</span> Edited by Timothée Poisot. <em>Methods in
Ecology and Evolution</em> 7 (8): 966–70. <a href="https://doi.org/10.1111/2041-210x.12550" class="external-link">https://doi.org/10.1111/2041-210x.12550</a>.
</div>
<div id="ref-Bruvo:2004" class="csl-entry">
Bruvo, Ruzica, Nicolaas K. Michiels, Thomas G. D’Souza, and Hinrich
Schulenburg. 2004. <span>“A Simple Method for the Calculation of
Microsatellite Genotype Distances Irrespective of Ploidy Level.”</span>
<em>Molecular Ecology</em> 13 (7): 2101–6. <a href="https://doi.org/10.1111/j.1365-294X.2004.02209.x" class="external-link">https://doi.org/10.1111/j.1365-294X.2004.02209.x</a>.
</div>
<div id="ref-everhart2014finescale" class="csl-entry">
Everhart, Sydney, and Harald Scherm. 2015. <span>“Fine-Scale Genetic
Structure of <span><em>Monilinia Fructicola</em></span> During Brown Rot
Epidemics Within Individual Peach Tree Canopies.”</span>
<em>Phytopathology</em> 105 (4): 542–49. <a href="https://doi.org/10.1094/PHYTO-03-14-0088-R" class="external-link">https://doi.org/10.1094/PHYTO-03-14-0088-R</a>.
</div>
<div id="ref-goss2009population" class="csl-entry">
Goss, Erica M, Meg Larsen, Gary A Chastagner, Donald R Givens, and
Niklaus J Grünwald. 2009. <span>“Population Genetic Analysis Infers
Migration Pathways of Phytophthora Ramorum in US Nurseries.”</span>
<em>PLoS Pathog</em> 5 (9): e1000583.
</div>
<div id="ref-kamvar2015novel" class="csl-entry">
Kamvar, Zhian N, Jonah C Brooks, and Niklaus J Grünwald. 2015.
<span>“<span class="nocase">Novel R tools for analysis of genome-wide
population genetic data with emphasis on clonality</span>.”</span>
<em>Frontiers in Genetics</em> 6: 208. <a href="https://doi.org/10.3389/fgene.2015.00208" class="external-link">https://doi.org/10.3389/fgene.2015.00208</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
