<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="poppr">
<title>Data import and manipulation in poppr version `r packageVersion('poppr')` • poppr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data import and manipulation in poppr version `r packageVersion('poppr')`">
<meta property="og:description" content="poppr">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-default navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">poppr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/poppr-package.html#functions-in-poppr">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mlg.html">Analysis of Multilocus Genotypes and Lineages in poppr `r packageVersion('poppr')`</a>
    <a class="dropdown-item" href="../articles/poppr_manual.html">Data import and manipulation in poppr version `r packageVersion('poppr')`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/grunwaldlab/poppr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Data import and manipulation in poppr version `r packageVersion('poppr')`</h1>
                        <h4 data-toc-skip class="author">Zhian N. Kamvar
and Niklaus J. Grünwald</h4>
            
            <h4 data-toc-skip class="date">2024-03-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/grunwaldlab/poppr/blob/HEAD/vignettes/poppr_manual.Rmd" class="external-link"><code>vignettes/poppr_manual.Rmd</code></a></small>
      <div class="d-none name"><code>poppr_manual.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p><em>Poppr</em> provides open-source, cross-platform tools for quick
analysis of population genetic data enabling focus on data analysis and
interpretation. While there are a plethora of packages for population
genetic analysis, few are able to offer quick and easy analysis of
populations with mixed reproductive modes. <em>Poppr</em>’s main
advantage is the ease of use and integration with other packages such as
<em>adegenet</em> and <em>vegan</em>, including support for novel
methods such as clone correction, multilocus genotype analysis,
calculation of Bruvo’s distance, and the index of association. New
features in version 2.0 include generation of minimum spanning networks
with reticulation, calculation of the index of association for genomic
data, and filtering multilocus genotypes based on genetic distance.</p>
<p><br><br></p>
<center>
<img src="jalapeno_logo.png" width="300">
</center>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="intro">Introduction<a class="anchor" aria-label="anchor" href="#intro"></a>
</h2>
<div class="section level3">
<h3 id="intro:purpose">Purpose<a class="anchor" aria-label="anchor" href="#intro:purpose"></a>
</h3>
<p><em>Poppr</em> is an R package with convenient functions for analysis
of genetic data with mixed modes of reproduction including sexual and
clonal reproduction. While there are many R packages in CRAN and other
repositories with tools for population genetic analyses, few are
appropriate for populations with mixed modes of reproduction. There are
several stand-alone programs that can handle these types of data sets,
but they are often platform specific and often only accept specific data
types. Furthermore, a typical analysis often involves switching between
many programs, and converting data to each specific format.</p>
<p><em>Poppr</em> is designed to make analysis of populations with mixed
reproductive modes more streamlined and user friendly so that the
researcher using it can focus on data analysis and interpretation.
<em>Poppr</em> allows analysis of haploid and diploid
dominant/co-dominant marker data including microsatellites, Single
Nucleotide Polymorphisms (SNP), and Amplified Fragment Length
Polymorphisms (AFLP). To avoid creating yet another file format that is
specific to a program, <em>poppr</em> was created on the backbone of the
popular R package <em>adegenet</em> and can take all the file formats
that <em>adegenet</em> can take (Genpop, Genetix, Fstat, and Structure)
and newly introduces compatibility with GenAlEx formatted files
(exported to CSV). This means that anything you can analyze in
<em>adegenet</em> can be further analyzed with <em>poppr</em>.</p>
<p>The real power of <em>poppr</em> is in the data manipulation and
analytic tools. <em>Poppr</em> has the ability to bootstrap,
clone-censor, and subset data sets. With <em>poppr</em> you can also
quickly calculate Bruvo’s distance, the index of association, and easily
determine which multilocus genotypes are shared across populations.</p>
<p>In version 2.0, tools for genomic data were introduced <span class="citation">(Zhian N. Kamvar, Brooks, and Grünwald 2015)</span>.
These tools allow researchers to define what it means to be a clone lost
in a sea of genomic data, generate bootstrapped dendrograms with any
genetic distance, and calculate minimum spanning networks with
reticulations to reveal the underlying population structure of your
clonal data.</p>
</div>
<div class="section level3">
<h3 id="resources">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h3>
<p>This vignette will cover all of the material you need to know to
efficiently analyze data in <em>poppr</em>. For information on methods
of analysis (eg. index of association, distance measures, AMOVA, …),
please read the manual pages provided for each function.</p>
<p>As <em>poppr</em> expanded from version 1.0, the vignette also
expanded to be 80+ pages. As a result, it became clear that over 22,000
words was less of a manual and more of a novella with a terrible plot.
To remedy this, this vignette will focus only on data manipulation and a
separate vignette, “algo”, has been written to give algorithmic details
of analyses introduced with <em>poppr</em>.</p>
<p>As of spring 2014, Drs. Niklaus J. Grünwald, Sydney E. Everhart, and
I have co- authored a primer on using R for population genetic analysis.
It is located at <a href="https://grunwaldlab.github.io/Population_Genetics_in_R/" class="external-link uri">https://grunwaldlab.github.io/Population_Genetics_in_R/</a>
and the source code can be found <a href="https://github.com/grunwaldlab/Population_Genetics_in_R/" class="external-link">on our
github site</a>.</p>
</div>
<div class="section level3">
<h3 id="getting-help">Getting Help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h3>
<p>If you have any questions or feedback, feel free to send a message to
the <em>poppr</em> forum at <a href="https://groups.google.com/d/forum/poppr" class="external-link uri">https://groups.google.com/d/forum/poppr</a>. You can submit
bug reports there or on our github site: <a href="https://github.com/grunwaldlab/poppr/issues" class="external-link uri">https://github.com/grunwaldlab/poppr/issues</a></p>
</div>
<div class="section level3">
<h3 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h3>
<p>Much thanks goes to Sydney E. Everhart for alpha testing, beta
testing, feature requests, proofreading, data contribution, and moral
support throughout the writing of this package and manual. Thanks also
to Brian Knaus, Ignazio Carbone, David Cooke, Corine Schoebel, Jane
Stewart, and Zaid Abdo for beta testing and feedback.</p>
<p>The following data sets are included in <em>poppr</em>:</p>
<ul>
<li><p><code>Pinf</code> (SSR, <em>Phytophthora infestans</em>)<span class="citation">(Erica M. Goss et al. 2014)</span></p></li>
<li><p><code>monpop</code> (SSR, <em>Monilinia fructicola</em>)<span class="citation">(Everhart and Scherm 2015)</span></p></li>
<li><p><code>Aeut</code> (AFLP, <em>Aphanomyces eutieches</em>)<span class="citation">(N. J. Grünwald and Hoheisel 2006)</span></p></li>
<li><p><code>Pram</code> (SSR, <em>Phytophthora ramorum</em>)<span class="citation">(Z. Kamvar et al. 2015; Erica M. Goss et al.
2009)</span></p></li>
</ul>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>To cite <em>poppr</em>, please type in your R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"poppr"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## To cite poppr in publications or presentations, please specify poppr</span></span>
<span><span class="co">## version 2.9.6 and with the following citation:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Kamvar ZN, Tabima JF, Grünwald NJ. (2014) Poppr: an R package for</span></span>
<span><span class="co">##   genetic analysis of populations with clonal, partially clonal, and/or</span></span>
<span><span class="co">##   sexual reproduction. PeerJ 2:e281. doi: 10.7717/peerj.281</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Kamvar ZN, Brooks JC and Grünwald NJ (2015) Novel R tools for</span></span>
<span><span class="co">##   analysis of genome-wide population genetic data with emphasis on</span></span>
<span><span class="co">##   clonality. Front. Genet. 6:208. doi: 10.3389/fgene.2015.00208</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## To see these entries in BibTeX format, use 'print(&lt;citation&gt;,</span></span>
<span><span class="co">## bibtex=TRUE)', 'toBibtex(.)', or set</span></span>
<span><span class="co">## 'options(citation.bibtex.max=999)'.</span></span></code></pre>
<p>The formal publication for the first version of <em>poppr</em> was
published in the journal <strong>PeerJ</strong>: <a href="https://peerj.com/articles/281/" class="external-link uri">https://peerj.com/articles/281/</a> <span class="citation">(Zhian N. Kamvar, Tabima, and Grünwald 2014)</span>.
The new features in version 2.0 were published in
<strong>Frontiers</strong> <a href="https://doi.org/10.3389/fgene.2015.00208" class="external-link uri">https://doi.org/10.3389/fgene.2015.00208</a> <span class="citation">(Zhian N. Kamvar, Brooks, and Grünwald
2015)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation-and-quick-start">Installation and Quick Start<a class="anchor" aria-label="anchor" href="#installation-and-quick-start"></a>
</h2>
<div class="section level3">
<h3 id="intro:install">Installation<a class="anchor" aria-label="anchor" href="#intro:install"></a>
</h3>
<p>This manual assumes you have installed R. If you have not, please
refer to The CRAN home page at <a href="https://cran.r-project.org/" class="external-link uri">https://cran.r-project.org/</a>. We also recommend the
Rstudio IDE (<a href="https://rstudio.com/" class="external-link uri">https://rstudio.com/</a>), which allows the user to view the
R console, environment, scripts, and plots in a single window.</p>
<div class="section level4">
<h4 id="from-cran">From CRAN<a class="anchor" aria-label="anchor" href="#from-cran"></a>
</h4>
<p>To install <em>poppr</em> from CRAN, select “Package Installer” from
the menu “Packages &amp; Data” in the GUI or type:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"poppr"</span>, repos <span class="op">=</span> <span class="st">"https://cloud.r-project.org/"</span><span class="op">)</span></span></code></pre></div>
<p>All dependencies will also be installed. In the unfortunate case this
does not work, please consult <a href="https://cran.r-project.org/doc/manuals/R-admin.html#Installing-packages" class="external-link uri">https://cran.r-project.org/doc/manuals/R-admin.html#Installing-packages</a>.</p>
</div>
<div class="section level4">
<h4 id="from-github">From GitHub<a class="anchor" aria-label="anchor" href="#from-github"></a>
</h4>
<p>GitHub is a repository where you can find all stable and development
versions of <em>poppr</em>.</p>
<p>Since <em>poppr</em> contains C code, it needs to be compiled, which
means that you need a working C compiler. If you are on Linux, you
should have that, but if you are on Windows or OSX, you might need to
download some special tools:</p>
<dl>
<dt>Windows</dt>
<dd>
<p>Download Rtools: <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link uri">https://cran.r-project.org/bin/windows/Rtools/</a></p>
</dd>
<dt>OSX</dt>
<dd>
<p>Download Xcode: <a href="https://developer.apple.com/xcode/" class="external-link uri">https://developer.apple.com/xcode/</a></p>
</dd>
</dl>
<p>To install from GitHub, you do not need to download the tarball since
there is a package called <em>devtools</em> that will download and
install the package for you directly from GitHub. After you have
installed all dependencies (see above section), you should download
<em>devtools</em>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span></code></pre></div>
<p>Now you can execute the command <code>install_github()</code> with
the user and repository name:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"grunwaldlab/poppr"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="intro:qstart">Quick start<a class="anchor" aria-label="anchor" href="#intro:qstart"></a>
</h3>
<p>The author assumes that if you have reached this point in the manual,
then you have successfully installed R and <em>poppr</em>. Before
proceeding, you should be aware that R is case sensitive. This means
that the words “Case” and “case” are different. You should also know
where your R package library is located. In this section, you will learn
how to locate a file, import it to R, and make a first analysis using
the <code><a href="../reference/poppr.html">poppr()</a></code> function.</p>
<blockquote>
<p><strong>What or where is my R package library?</strong></p>
<p>R is as powerful as it is through a community of people who submit
extra code called “packages” to help it do specific things. These
packages live in a certain place on your computer called an R library.
You can find out where this library is by typing
<code><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths()</a></code></p>
</blockquote>
<p>Importing a file into R involves you knowing the path to your file
and then typing that into R’s console. <code><a href="../reference/getfile.html">getfile()</a></code> will help
provide a point and click interface for selecting a file. This is simply
a tool to help you get started. As you get better with R, you might feel
that you don”t need it at all.</p>
<p>First, tell your computer to search R’s library, load the
<em>poppr</em> package, and use <code><a href="../reference/getfile.html">getfile()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: adegenet</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ade4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    /// adegenet 2.1.10 is loaded ////////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    &gt; overview: '?adegenet'</span></span>
<span><span class="co">##    &gt; tutorials/doc/questions: 'adegenetWeb()' </span></span>
<span><span class="co">##    &gt; bug reports/feature requests: adegenetIssues()</span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'pegas':</span></span>
<span><span class="co">##   method      from</span></span>
<span><span class="co">##   print.amova ade4</span></span></code></pre>
<pre><code><span><span class="co">## This is poppr version 2.9.6. To get started, type package?poppr</span></span>
<span><span class="co">## OMP parallel support: unavailable</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getfile.html">getfile</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>A pop up window will appear like this<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;This window sometimes appears behind your current
session of R, depending on the GUI and you will have to toggle to this
window&lt;/p&gt;"><sup>1</sup></a>:</p>
<div class="figure" style="text-align: center">
<img src="getfile.png" alt="A popup window as it appears in OSX (Mountain Lion)" width="500px"><p class="caption">
A popup window as it appears in OSX (Mountain Lion)
</p>
</div>
<p>You should navigate to your R library and select the file called
“rootrot.csv”. If you don”t know where your the poppr package lives, you
can find it by typing <code>find.package("poppr")</code> into the R
command line. Once we select a file, the file name and its path will be
stored in the variable, <code>x</code>. We can confirm that by typing
<code>x</code> into R”s command line.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## $files</span></span>
<span><span class="co">## [1] "/path/to/R/poppr/files/rootrot.csv"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $path</span></span>
<span><span class="co">## [1] "/path/to/R/poppr/files"</span></span></code></pre>
<p>Here we can see that <code>x</code> is a list with two entries:
<code>$files</code> shows the files you selected and <code>$path</code>
shows the path to those files.</p>
<p>We will use <code>x$files</code> to access the file. The file is in
the GenAlEx format, so we will import it using
<code><a href="../reference/read.genalex.html">read.genalex()</a></code> and then analyze it with the function
<code><a href="../reference/poppr.html">poppr()</a></code> to get a table of diversity indices per
population.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myData</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.genalex.html">read.genalex</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">files</span><span class="op">)</span></span>
<span><span class="va">myData</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">popdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/poppr.html">poppr</a></span><span class="op">(</span><span class="va">myData</span><span class="op">)</span></span></code></pre></div>
<p>The output of <code><a href="../reference/poppr.html">poppr()</a></code> was assigned to the variable
<code>popdata</code>, so let”s look at the data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">popdata</span></span></code></pre></div>
<pre><code><span><span class="co">##             Pop   N MLG  eMLG    SE     H     G lambda   E.5   Hexp    Ia rbarD   File</span></span>
<span><span class="co">## 1      Athena_1   9   7  7.00 0.000 1.889  6.23  0.840 0.932 0.1081  2.92 0.210 myData</span></span>
<span><span class="co">## 2      Athena_2  12  12 10.00 0.000 2.485 12.00  0.917 1.000 0.1999  4.16 0.128 myData</span></span>
<span><span class="co">## 3      Athena_3  10   2  2.00 0.000 0.325  1.22  0.180 0.571 0.0107  2.00 1.000 myData</span></span>
<span><span class="co">## 4      Athena_4  13   9  7.15 0.769 1.946  5.12  0.805 0.687 0.1012  5.49 0.372 myData</span></span>
<span><span class="co">## 5      Athena_5  10   7  7.00 0.000 1.834  5.56  0.820 0.866 0.0802  4.53 0.353 myData</span></span>
<span><span class="co">## 6      Athena_6   5   5  5.00 0.000 1.609  5.00  0.800 1.000 0.1143  2.46 0.190 myData</span></span>
<span><span class="co">## 7      Athena_7  11  10  9.18 0.386 2.272  9.31  0.893 0.955 0.1584  2.13 0.086 myData</span></span>
<span><span class="co">## 8      Athena_8   8   6  6.00 0.000 1.667  4.57  0.781 0.831 0.0791  3.86 0.323 myData</span></span>
<span><span class="co">## 9      Athena_9  10  10 10.00 0.000 2.303 10.00  0.900 1.000 0.1532  2.82 0.118 myData</span></span>
<span><span class="co">## 10    Athena_10   9   8  8.00 0.000 2.043  7.36  0.864 0.948 0.1230  2.85 0.137 myData</span></span>
<span><span class="co">## 11 Mt. Vernon_1  10   9  9.00 0.000 2.164  8.33  0.880 0.952 0.1766  7.13 0.276 myData</span></span>
<span><span class="co">## 12 Mt. Vernon_2   6   6  6.00 0.000 1.792  6.00  0.833 1.000 0.3940 20.65 0.492 myData</span></span>
<span><span class="co">## 13 Mt. Vernon_3   8   6  6.00 0.000 1.667  4.57  0.781 0.831 0.1033  2.12 0.106 myData</span></span>
<span><span class="co">## 14 Mt. Vernon_4  12   8  6.83 0.665 1.814  4.50  0.778 0.681 0.0823  3.01 0.255 myData</span></span>
<span><span class="co">## 15 Mt. Vernon_5  17   7  5.54 0.828 1.758  5.07  0.803 0.848 0.0544  2.68 0.340 myData</span></span>
<span><span class="co">## 16 Mt. Vernon_6  12  11  9.32 0.466 2.369 10.29  0.903 0.958 0.3001 19.50 0.467 myData</span></span>
<span><span class="co">## 17 Mt. Vernon_7  12   9  7.82 0.649 2.095  7.20  0.861 0.870 0.0601  1.21 0.153 myData</span></span>
<span><span class="co">## 18 Mt. Vernon_8  13   9  7.35 0.764 2.032  6.26  0.840 0.794 0.0430  1.15 0.169 myData</span></span>
<span><span class="co">## 19        Total 187 119  9.61 0.612 4.558 68.97  0.986 0.720 0.3651 14.37 0.271 myData</span></span></code></pre>
<p>The fields you see in the output include:</p>
<ul>
<li><p><code>Pop -</code> Population name (Note that “Total” also means
“Pooled”).</p></li>
<li><p><code>N -</code> Number of individuals observed.</p></li>
<li><p><code>MLG -</code> Number of multilocus genotypes (MLG)
observed.</p></li>
<li><p><code>eMLG -</code> The number of expected MLG at the smallest
sample size <span class="math inline">\(\geq 10\)</span> based on
rarefaction. <span class="citation">(Hurlbert 1971)</span></p></li>
<li><p><code>SE -</code> Standard error based on <code>eMLG</code> <span class="citation">(Heck, Belle, and Simberloff 1975)</span></p></li>
<li><p><code>H -</code> Shannon-Wiener Index of MLG diversity. <span class="citation">(Shannon 1948)</span></p></li>
<li><p><code>G -</code> Stoddart and Taylor”s Index of MLG diversity.
<span class="citation">(Stoddart and Taylor 1988)</span></p></li>
<li><p><code>lambda -</code> Simpson”s index, <span class="math inline">\(\lambda\)</span>.</p></li>
<li><p><code>E.5 -</code> Evenness, <span class="math inline">\(E_5\)</span>. [<span class="citation">Pielou
(1975)</span>]<span class="citation">(Ludwig and Reynolds
1988)</span><span class="citation">(Niklaus J. Grünwald et al.
2003)</span></p></li>
<li><p><code>Hexp -</code> Nei”s 1978 Expected Heterozygosity. <span class="citation">(Nei 1978)</span></p></li>
<li><p><code>Ia -</code> The index of association, <span class="math inline">\(I_A\)</span>. <span class="citation">(Brown,
Feldman, and Nevo 1980)</span> <span class="citation">(Smith et al.
1993)</span> <span class="citation">(Agapow and Burt
2001)</span></p></li>
<li><p><code>rbarD -</code> The standardized index of association, <span class="math inline">\(\bar{r}_d\)</span>. <span class="citation">(Agapow
and Burt 2001)</span></p></li>
</ul>
<p>These fields are further described in the function
<code>poppr</code>. You can access the help page for <code>poppr</code>
by typing <code><a href="../reference/poppr.html">?poppr</a></code> in your R console.</p>
<p>One thing to note about this output is the <code>NaN</code> in the
column labeled <code>SE</code>. In R, <code>NaN</code> means “Not a
number”. This is produced from calculation of a standard error based on
rarefaction analysis. Occasionally, this calculation will encounter a
situation in which it must attempt to take a square root of a negative
number. Since the root of any negative number is not defined in the set
of real numbers, it must therefore have an imaginary component, <span class="math inline">\(i\)</span>. Unfortunately, R will not represent
the imaginary components of numbers unless you specifically tell it to
do so. By default, R represents these as <code>NaN</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="import-and-data-types">Import and Data types<a class="anchor" aria-label="anchor" href="#import-and-data-types"></a>
</h2>
<div class="section level3">
<h3 id="intro:import">Importing data into poppr {Get out of my dreams and into my R}<a class="anchor" aria-label="anchor" href="#intro:import"></a>
</h3>
<div class="section level4">
<h4 id="how-does-r-represent-data-objective-data">How does R represent data? {Objective: data}<a class="anchor" aria-label="anchor" href="#how-does-r-represent-data-objective-data"></a>
</h4>
<p>Working with data in R means that these data have to be stored inside
an “object”, which is stored in the computer”s memory. Because of this,
it”s important to know the difference between a <strong>file</strong>
and an <strong>object</strong>. When anyone talking about importing a
<strong>file</strong> into R, they are referring to a spreadsheet or
text file that lives in a folder on your hard drive. Spreadsheet files
(saved as *.csv files) are normally imported through the R function
<code><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table()</a></code>. The output of <code><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table()</a></code> is a
<code>data.frame</code>. A <code>data.frame</code> is an
<strong>object</strong> represented in your computer’s memory. This
means that it only exists for as long as R is running.</p>
<p>The good thing about having objects stored in memory is that you can
manipulate them in any way and not affect the source of those data.
Since R is a scripted language (instead of point-and-click), any of your
manipulations can be saved in a separate R file that can be easily
adapted to new data. Of course, most data are not going to be entered
into R manually. Usually they will be formatted in a manner that can be
read by popular population genetics programs.</p>
<p>As previously mentioned, since <em>poppr</em> is based on
<em>adegenet</em>, it’s possible to read in the following file formats
into a <code>genind</code> object with the function
<code><a href="https://rdrr.io/pkg/adegenet/man/import2genind.html" class="external-link">import2genind()</a></code>:</p>
<ul>
<li><p><strong>Fstat</strong></p></li>
<li><p><strong>Genepop</strong></p></li>
<li><p><strong>Genetix</strong></p></li>
<li><p><strong>Structure</strong></p></li>
</ul>
<p>Here, we introduce a new way of importing data into a
<code>genind</code> or <code>genclone</code> object from a GenAlEx
formatted file.</p>
</div>
<div class="section level4">
<h4 id="intro:import:read-genalex">Function: read.genalex<a class="anchor" aria-label="anchor" href="#intro:import:read-genalex"></a>
</h4>
<p>A very popular program for population genetics is GenAlEx (<a href="https://biology-assets.anu.edu.au/GenAlEx/Welcome.html" class="external-link uri">https://biology-assets.anu.edu.au/GenAlEx/Welcome.html</a>)
<span class="citation">(Rod Peakall and Smouse 2012; R. Peakall and
Smouse 2006)</span>. GenAlEx runs within the Excel environment and can
be very powerful in its analyses. <em>Poppr</em> has added the ability
to read *.CSV files<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;*.CSV files are comma separated files that are easily
machine readable.&lt;/p&gt;"><sup>2</sup></a> produced in the GenAlEx format. It can
handle data types containing regions and geographic coordinates, but
currently cannot import allelic frequency data from GenAlEx. Using the
<em>poppr</em> function <code><a href="../reference/read.genalex.html">read.genalex()</a></code> will import your
data into <em>adegenet</em>”s <code>genind</code> object or
<em>poppr</em>”s <code>genclone</code> object (more information on that
below). For ways of formatting a GenAlEx file, see the manual here: <a href="https://biology-assets.anu.edu.au/GenAlEx/Download_files/GenAlEx%206.5%20Guide.pdf" class="external-link uri">https://biology-assets.anu.edu.au/GenAlEx/Download_files/GenAlEx%206.5%20Guide.pdf</a></p>
<p>Below is an example of the GenAlEx format. We will use the data set
called <code>microbov</code> from the <em>adegenet</em> package to
generate it. The data contains three demographic factors: Country,
Species and Breed contained within the <code>@other</code> slot
(detailed in <a href="#intro:genind:other">the ‘other’ slot</a>). We
will first set these as the population strata, define the population as
the combination of the strata, and then save a file to the desktop.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">microbov</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">microbov</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">microbov</span><span class="op">)</span><span class="op">)</span> <span class="co"># set the strata</span></span>
<span><span class="va">microbov</span></span>
<span><span class="fu"><a href="../reference/genind2genalex.html">genind2genalex</a></span><span class="op">(</span><span class="va">microbov</span>, file <span class="op">=</span> <span class="st">"~/Desktop/microbov.csv"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Extracting the table ... Writing the table to ~/Desktop/microbov.csv ... Done.</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="unmod_dat.png" alt="The first 15 individuals and 4 loci of the microbov data set. The first column contains the individual names, the second column contains the population names, and each subsequent column represents microsatellite genetic data. Highlighted in red is a list of populations and their relative sizes." width="700px"><p class="caption">
The first 15 individuals and 4 loci of the microbov data set. The first
column contains the individual names, the second column contains the
population names, and each subsequent column represents microsatellite
genetic data. Highlighted in red is a list of populations and their
relative sizes.
</p>
</div>
<p>The GenAlEx format contains individuals in rows and loci in columns.
Individual data begins at row 4. Column A always contains individual
names and column B defines the population of each individual. Notice
here that the three demographic factors from the data have been
concatenated with a “_”. This allows us to import more than one
population factor to use as hierarchical levels in a
<code>genclone</code> object.</p>
<p>The First three rows contain information pertaining to the global
data set. The only important information for <em>poppr</em> is the
information contained in row 3 and the first three columns of row 1.</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="left"><strong>A</strong></th>
<th align="left"><strong>B</strong></th>
<th align="left"><strong>C</strong></th>
<th align="left"><strong>D</strong></th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>1</strong></td>
<td align="left"># of Loci</td>
<td align="left"># of Individuals</td>
<td align="left"># of Populations</td>
<td align="left">Pop1 Size</td>
<td>…</td>
</tr>
<tr class="even">
<td><strong>2</strong></td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">Pop1 Name</td>
<td>…</td>
</tr>
<tr class="odd">
<td><strong>3</strong></td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">Locus 1</td>
<td align="left">…</td>
<td></td>
</tr>
</tbody>
</table>
<p>Highlighted in red in figure above are definitions of the number of
populations and their respective sizes. As this is redundant
information, we can remove it. Below is an example of a valid data set
that can be imported into <em>poppr</em>.</p>
<div class="figure" style="text-align: center">
<img src="mod_dat.png" alt="The first 15 individuals and 4 loci of the microbov data set. This is the same figure as above, however the populations and counts have been removed from the header row and the third number in the header has been replaced by 1" width="700px"><p class="caption">
The first 15 individuals and 4 loci of the microbov data set. This is
the same figure as above, however the populations and counts have been
removed from the header row and the third number in the header has been
replaced by 1
</p>
</div>
<p>All GenAlEx formatted data can be imported with the command
<code><a href="../reference/read.genalex.html">read.genalex()</a></code>, detailed below:</p>
<hr>
<pre><code><span><span class="fu"><a href="../reference/read.genalex.html">read.genalex</a></span><span class="op">(</span><span class="va">genalex</span>, ploidy <span class="op">=</span> <span class="fl">2</span>, geo <span class="op">=</span> <span class="cn">FALSE</span>, region <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    genclone <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span>, recode <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>genalex -</code> a *.CSV file exported from GenAlEx on your
disk (For example: <code>my_genalex_file.csv</code>).</p></li>
<li><p><code>ploidy -</code> a number indicating the ploidy for the data
set (eg 2 for diploids, 1 for haploids).</p></li>
<li><p><code>geo -</code> GenAlEx allows you to have geographic data
within your file. To do this for <em>poppr</em>, you will need to follow
the first format outlined in the GenAlEx manual and place the geographic
data AFTER all genetic and demographic data with one blank column
separating it (See the GenAlEx Manual for details). If you have
geographic information in your file, set this flag to <code>TRUE</code>
and it will be included within the resulting genind object in the
<code>@other</code> slot. (If you don”t know what that is, don”t worry.
It will be explained later in <a href="#intro:genind:other">the ‘other’
slot</a>)</p></li>
<li><p><code>region -</code> To format your GenAlEx file to include
regions, you can choose to include a separate column for regional data,
or, since regional data must be in contiguous blocks, you can simply
format it in the same way you would any other data (see the GenAlEx
manual for details). If you have your file organized in this manner,
select this option and the regional information will be stored in the
<code>@other</code> slot of the resulting <code>genind</code> object or
be incorporated into the hierarchy of the <code>genclone</code>
object.</p></li>
<li><p><code>genclone -</code> This flag will convert your data into a
<code>genclone</code> object (see <a href="#intro:genclone">Send in the
clones</a> for more info).</p></li>
<li><p><code>sep -</code> The separator argument for columns in your
data. It defaults to “,”.</p></li>
<li><p><code>recode -</code> If your data is polyploid data, this gives
you the option to recode it. (See <a href="#intro:import:polyploid">About Polyploid Data</a> for
details)</p></li>
</ul>
<hr>
<p>Note that regional data and geographic data are not mutually
exclusive. You can have both in one file, just make sure that they are
on the same sheet and that the geographic data is always placed after
all genetic and demographic data.</p>
<p>We have a short example of GenAlEx formatted data with no geographic
or regional formatting. We will first see where the data is using the
command <code><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file()</a></code></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"files/rootrot.csv"</span>, package<span class="op">=</span><span class="st">"poppr"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "/path/to/R/library/poppr/files/rootrot.csv"</span></span></code></pre>
<p>Now import the data into <em>poppr</em> like so:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rootrot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.genalex.html">read.genalex</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"files/rootrot.csv"</span>, package<span class="op">=</span><span class="st">"poppr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Executing rootrot shows that this file is now in
<code>genclone</code> format and can be used with any function in
<em>poppr</em> and <em>adegenet</em></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rootrot</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    119 original multilocus genotypes </span></span>
<span><span class="co">##    187 diploid individuals</span></span>
<span><span class="co">##     56 dominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      1 stratum - Pop</span></span>
<span><span class="co">##     18 populations defined - </span></span>
<span><span class="co">## Athena_1, Athena_2, Athena_3, ..., Mt. Vernon_6, Mt. Vernon_7, Mt. Vernon_8</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="intro:import:other">Other ways of importing data<a class="anchor" aria-label="anchor" href="#intro:import:other"></a>
</h4>
<p><em>Adegenet</em> already supports the import of
<strong>FSTAT</strong>, <strong>Structure</strong>,
<strong>Genpop</strong>, and <strong>Genetix</strong> formatted files,
so if you have data in those formats, you can import them using the
function <code>import2genind</code>. For sequence data, check if you can
use <code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.dna()</a></code> from the <em>ape</em> package to import your
data. If you can, then you can use the <em>adegenet</em> function
<code><a href="https://rdrr.io/pkg/adegenet/man/sequences.html" class="external-link">DNAbin2genind()</a></code>. If you don”t have any of these formats
handy, you can still import your data using R”s <code>read.table</code>
along with <code><a href="https://rdrr.io/pkg/adegenet/man/df2genind.html" class="external-link">df2genind()</a></code> from <em>adegenet</em>. For more
information, see <em>adegenet</em>’s “Getting Started” vignette.</p>
</div>
<div class="section level4">
<h4 id="intro:import:genind2genalex">Function: genind2genalex (exporting data)<a class="anchor" aria-label="anchor" href="#intro:import:genind2genalex"></a>
</h4>
<p>Of course, being able to export data is just as useful as being able
to import it, so we have this handy little function that will write a
GenAlEx formatted file to wherever you desire.<br><strong>WARNING: This will overwrite any file that exists with the same
name.</strong></p>
<hr>
<pre><code><span><span class="fu"><a href="../reference/genind2genalex.html">genind2genalex</a></span><span class="op">(</span><span class="va">gid</span>, filename <span class="op">=</span> <span class="st">""</span>, overwrite <span class="op">=</span> <span class="cn">FALSE</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    pop <span class="op">=</span> <span class="cn">NULL</span>, allstrata <span class="op">=</span> <span class="cn">TRUE</span>, geo <span class="op">=</span> <span class="cn">FALSE</span>, geodf <span class="op">=</span> <span class="st">"xy"</span>, </span>
<span>    sep <span class="op">=</span> <span class="st">","</span>, sequence <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genind</code> object.</p></li>
<li><p><code>filename -</code> This is where you specify the path to the
new file you wish to create. If you specify only a filename with no
path, it will place the file in your current working directory.</p></li>
<li><p><code>quiet -</code> If this is set to <code>FALSE</code>, a
status message will be printed to the console as the extraction
progresses.</p></li>
<li><p><code>pop -</code> A vector specifying a custom population OR a
formula specifying the strata to be combined in the new file.</p></li>
<li><p><code>allstrata -</code> This is <code>TRUE</code> by default and
will combine all of the strata in your file unless you have specified a
new population factor.</p></li>
<li><p><code>geo -</code> Set to <code>TRUE</code>, if you have a data
frame or matrix in the <code>@other</code> slot of your
<code>genind</code> object that contains geographic coordinates for all
individuals or all populations. Setting this to <code>TRUE</code> means
the resulting file will have two extra columns at the end of your file
with geographic coordinates.</p></li>
<li><p><code>geodf -</code> The name of the data frame or matrix
containing the geographic coordinates.</p></li>
<li><p><code>sep -</code> A separator to separate columns in the
resulting file.</p></li>
</ul>
<hr>
<p>First, a simple example for the rootrot data we demonstrated in an
earlier section:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genind2genalex.html">genind2genalex</a></span><span class="op">(</span><span class="va">rootrot</span>, <span class="st">"~/Desktop/rootrot.csv"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Extracting the table ... Writing the table to ~/Desktop/rootrot.csv ... Done.</span></span></code></pre>
<p>Here’s an example of exporting the nancycats data set into GenAlEx
format with geographic information. If we look at the nancycats
geographic information, we can see it”s coordinates for each population,
but not each individual:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span><span class="op">$</span><span class="va">xy</span></span></code></pre></div>
<pre><code><span><span class="co">##            x         y</span></span>
<span><span class="co">## P01 263.3498 171.10939</span></span>
<span><span class="co">## P02 183.5028 122.40790</span></span>
<span><span class="co">## P03 391.1050 254.70148</span></span>
<span><span class="co">## P04 458.6121  41.72336</span></span>
<span><span class="co">## P05 182.7769 219.08398</span></span>
<span><span class="co">## P06 335.2121 344.83557</span></span>
<span><span class="co">## P07 359.1662 375.36486</span></span>
<span><span class="co">## P08 271.3345  67.89132</span></span>
<span><span class="co">## P09 256.8169 150.02964</span></span>
<span><span class="co">## P10 270.6086  17.00917</span></span>
<span><span class="co">## P11 493.4544 237.25618</span></span>
<span><span class="co">## P12 305.4510  85.33663</span></span>
<span><span class="co">## P13 462.9674  86.79040</span></span>
<span><span class="co">## P14 429.5768 291.04587</span></span>
<span><span class="co">## P15 531.2003 115.13903</span></span>
<span><span class="co">## P16 407.8003  99.87438</span></span>
<span><span class="co">## P17 345.3745 251.79393</span></span></code></pre>
<p>To export it:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genind2genalex.html">genind2genalex</a></span><span class="op">(</span><span class="va">nancycats</span>, <span class="st">"~/Desktop/nancycats_pop_xy.csv"</span>, geo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Extracting the table ... Writing the table to ~/Desktop/nancycats_pop_xy.csv ... Done.</span></span></code></pre>
<p>If we wanted to assign a geographic coordinate to each individual, we
can use this trick knowing that there are 17 populations in the data
set:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nan2</span>           <span class="op">&lt;-</span> <span class="va">nancycats</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nan2</span><span class="op">)</span><span class="op">$</span><span class="va">xy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nan2</span><span class="op">)</span><span class="op">$</span><span class="va">xy</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">pop</a></span><span class="op">(</span><span class="va">nan2</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nan2</span><span class="op">)</span><span class="op">$</span><span class="va">xy</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            x        y</span></span>
<span><span class="co">## P17 345.3745 251.7939</span></span>
<span><span class="co">## P17 345.3745 251.7939</span></span>
<span><span class="co">## P17 345.3745 251.7939</span></span>
<span><span class="co">## P17 345.3745 251.7939</span></span>
<span><span class="co">## P17 345.3745 251.7939</span></span>
<span><span class="co">## P17 345.3745 251.7939</span></span></code></pre>
<p>Now we can export it to a different file.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genind2genalex.html">genind2genalex</a></span><span class="op">(</span><span class="va">nan2</span>, <span class="st">"~/Desktop/nancycats_inds_xy.csv"</span>, geo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Extracting the table ... Writing the table to ~/Desktop/nancycats_inds_xy.csv ... Done.</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="intro:genind">Getting to know <em>adegenet</em>”s <code>genind</code> object<a class="anchor" aria-label="anchor" href="#intro:genind"></a>
</h3>
<p>Since <em>poppr</em> was built around <em>adegenet</em>”s framework,
it is important to know how <em>adegenet</em> stores data in the
<code>genind</code> object, as that is the object used by
<em>poppr</em>. To create a <code>genind</code> object,
<em>adegenet</em> takes a data frame of genotypes (rows) across multiple
loci (columns) and converts them into a matrix of individual allelic
counts at each locus <span class="citation">(Jombart 2008)</span>.</p>
<p>For example, Let’s say we had data with 3 diploid individuals each
with 3 loci that had 3, 4, and 5 allelic states respectively:</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="left">locus1</th>
<th align="left">locus2</th>
<th align="left">locus3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="left">101/101</td>
<td align="left">201/201</td>
<td align="left">301/302</td>
</tr>
<tr class="even">
<td>2</td>
<td align="left">102/103</td>
<td align="left">202/203</td>
<td align="left">301/303</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="left">102/102</td>
<td align="left">203/204</td>
<td align="left">304/305</td>
</tr>
</tbody>
</table>
<p>The resulting <code>genind</code> object would contain a matrix that
has 3 rows and 12 columns. Below is a schematic of what that would look
like. Each column represents a separate allele, each row represents an
individual and each color represents a different locus.</p>
<p><img src="poppr_manual_files/figure-html/ex_data_picture-1.png" width="700px" style="display: block; margin: auto;"></p>
<p>When we look at the data derived from table above, we see that we
have a matrix of individual allele counts at each locus.</p>
<pre><code><span><span class="co">##   locus1.101 locus1.102 locus1.103 locus2.201 locus2.202 locus2.203 locus2.204 locus3.301</span></span>
<span><span class="co">## 1          2          0          0          2          0          0          0          1</span></span>
<span><span class="co">## 2          0          1          1          0          1          1          0          1</span></span>
<span><span class="co">## 3          0          2          0          0          0          1          1          0</span></span>
<span><span class="co">##   locus3.302 locus3.303 locus3.304 locus3.305</span></span>
<span><span class="co">## 1          1          0          0          0</span></span>
<span><span class="co">## 2          0          1          0          0</span></span>
<span><span class="co">## 3          0          0          1          1</span></span></code></pre>
<p>At each locus, the allele counts for each individual sum to the
ploidy, <span class="math inline">\(p\)</span>. Homozygotes are denoted
as having an allele count of <span class="math inline">\(p\)</span> at a
single allele within a locus, while heterozygotes have their allele
counts represented as <span class="math inline">\(&lt;p\)</span> where
<span class="math inline">\(p\)</span> = ploidy across multiple alleles
in a single locus. Along with this matrix, are elements that define the
names of the individuals, loci, alleles, and populations. If you wish to
know more, see the <em>adegenet</em> “Getting Started” manual.</p>
<div class="section level4">
<h4 id="intro:genind:other">The other slot<a class="anchor" aria-label="anchor" href="#intro:genind:other"></a>
</h4>
<p>The other slot is a place in the <code>genind</code> object that can
be used to store useful information about the data. We saw earlier that
it could store demographic information, now let’s explore a different
example. Bruvo’s distance is based off of a stepwise mutation model for
microsatellites. This requires us to know the length of the repeat of
each locus. We could store the repeat lengths in a separate variable in
our R environment, but we are at risk of losing that. One way to prevent
it from being lost would be to place it in the “other” slot. For the
purpose of this example, we will use the “nancycats” data set from the
<em>adegenet</em> package and assume that it has di-nucleotide repeats
at all of its loci.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span> <span class="co"># Load the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span> <span class="co"># geographical coordinates</span></span></code></pre></div>
<pre><code><span><span class="co">## $xy</span></span>
<span><span class="co">##            x         y</span></span>
<span><span class="co">## P01 263.3498 171.10939</span></span>
<span><span class="co">## P02 183.5028 122.40790</span></span>
<span><span class="co">## P03 391.1050 254.70148</span></span>
<span><span class="co">## P04 458.6121  41.72336</span></span>
<span><span class="co">## P05 182.7769 219.08398</span></span>
<span><span class="co">## P06 335.2121 344.83557</span></span>
<span><span class="co">## P07 359.1662 375.36486</span></span>
<span><span class="co">## P08 271.3345  67.89132</span></span>
<span><span class="co">## P09 256.8169 150.02964</span></span>
<span><span class="co">## P10 270.6086  17.00917</span></span>
<span><span class="co">## P11 493.4544 237.25618</span></span>
<span><span class="co">## P12 305.4510  85.33663</span></span>
<span><span class="co">## P13 462.9674  86.79040</span></span>
<span><span class="co">## P14 429.5768 291.04587</span></span>
<span><span class="co">## P15 531.2003 115.13903</span></span>
<span><span class="co">## P16 407.8003  99.87438</span></span>
<span><span class="co">## P17 345.3745 251.79393</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">repeats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nLoc</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span><span class="op">)</span> <span class="co">#nLoc = number of loci</span></span>
<span><span class="va">repeats</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2 2 2 2 2 2 2 2 2</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span><span class="op">$</span><span class="va">repeat_lengths</span> <span class="op">&lt;-</span> <span class="va">repeats</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span> <span class="co"># two items named xy and repeat_lengths</span></span></code></pre></div>
<pre><code><span><span class="co">## $xy</span></span>
<span><span class="co">##            x         y</span></span>
<span><span class="co">## P01 263.3498 171.10939</span></span>
<span><span class="co">## P02 183.5028 122.40790</span></span>
<span><span class="co">## P03 391.1050 254.70148</span></span>
<span><span class="co">## P04 458.6121  41.72336</span></span>
<span><span class="co">## P05 182.7769 219.08398</span></span>
<span><span class="co">## P06 335.2121 344.83557</span></span>
<span><span class="co">## P07 359.1662 375.36486</span></span>
<span><span class="co">## P08 271.3345  67.89132</span></span>
<span><span class="co">## P09 256.8169 150.02964</span></span>
<span><span class="co">## P10 270.6086  17.00917</span></span>
<span><span class="co">## P11 493.4544 237.25618</span></span>
<span><span class="co">## P12 305.4510  85.33663</span></span>
<span><span class="co">## P13 462.9674  86.79040</span></span>
<span><span class="co">## P14 429.5768 291.04587</span></span>
<span><span class="co">## P15 531.2003 115.13903</span></span>
<span><span class="co">## P16 407.8003  99.87438</span></span>
<span><span class="co">## P17 345.3745 251.79393</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $repeat_lengths</span></span>
<span><span class="co">## [1] 2 2 2 2 2 2 2 2 2</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="intro:genclone">The <code>genclone</code> object {send in the clones}<a class="anchor" aria-label="anchor" href="#intro:genclone"></a>
</h3>
<p>The <code>genclone</code> class was defined in order to make working
with clonal organisms more intuitive. It is built off of the
<code>genind</code> object and has dedicated slots for defined
multilocus genotypes. The name <code>genclone</code> refers to the fact
that it has the ability to handle genotypes of clonal organisms (but it
is also used for sexual populations).</p>
<p>In previous versions of <em>poppr</em>, the <code>genclone</code>
object contained a hierarchy slot as well. This slot was moved to
<em>adegenet</em> and its name was changed to “strata”. This slot allows
you to carry around several definitions for populations in the same data
set.</p>
<p>The function <code>as.genclone</code> allows the user to convert a
<code>genind</code> object to a <code>genclone</code> object. The
following example will demonstrate that the <code>genclone</code> object
is an extension of the <code>genind</code> object as well as the
advantages of having populations pre-defined in your data set.</p>
<div class="section level4">
<h4 id="intro:genclone:as-genclone">Function: as.genclone<a class="anchor" aria-label="anchor" href="#intro:genclone:as-genclone"></a>
</h4>
<hr>
<pre><code><span><span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, <span class="va">mlg</span>, mlgclass <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><strong>x -</strong> a <code>genind</code> object to be
converted.</p></li>
<li><p><strong>… -</strong> any arguments to be passed to the
<code>genind</code> constructor.</p></li>
<li><p><strong>mlg -</strong> a vector representing the multilocus
genotype definitions of your data.</p></li>
<li><p><strong>mlgclass -</strong> if <code>TRUE</code>, the MLGs
represented in your object will be converted to an MLG class object,
which allows for custom MLG definitions.</p></li>
</ul>
<hr>
<p>Let”s show an example of a <code>genclone</code> object. First, we
will take an existing <code>genind</code> object and convert it using
the function <code>as.genclone</code> (We can also use the function <a href="#intro:import:read.genalex">read.genalex()</a>to import as
<code>genclone</code> or <code>genind</code> objects). We will use the
<code>Aeut</code> data set because it is a clonal data set that has a
simple population strata <span class="citation">(N. J. Grünwald and
Hoheisel 2006)</span>. The data set is here: <a href="https://doi.org/10.6084/m9.figshare.877104" class="external-link uri">https://doi.org/10.6084/m9.figshare.877104</a> and it is
AFLP data of the root rot pathogen <em>Aphanomyces euteiches</em>
collected from two different fields in NW Oregon and W Washington, USA.
These fields were divided up into subplots from which samples were
collected. The fields represent the population and the subplots
represent the subpopulation. Let”s take a look at what the
<code>genind</code> object looks like:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span>
<span><span class="va">Aeut</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 187 individuals; 56 loci; 56 alleles; size: 68.9 Kb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  187 x 56 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 56-56)</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 2-2)</span></span>
<span><span class="co">##    @type:  PA</span></span>
<span><span class="co">##    @call: old2new_genind(object = x, donor = new(class(x)))</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @pop: population of each individual (group size range: 90-97)</span></span>
<span><span class="co">##    @other: a list containing: population_hierarchy</span></span></code></pre>
<p>We can see that there is a data frame in the <code>@other</code> slot
called “population_hierarchy”. This is left over from <em>poppr</em>
version 1.x behavior. Since the <code>genind</code> object now has a
<code>@strata</code> slot, we can use it to set the stratification
(previously called “hierarchy”).</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span><span class="op">$</span><span class="va">population_hierarchy</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">Aeut</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 187 individuals; 56 loci; 56 alleles; size: 72.7 Kb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  187 x 56 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 56-56)</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 2-2)</span></span>
<span><span class="co">##    @type:  PA</span></span>
<span><span class="co">##    @call: old2new_genind(object = x, donor = new(class(x)))</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @pop: population of each individual (group size range: 90-97)</span></span>
<span><span class="co">##    @strata: a data frame with 2 columns ( Pop, Subpop )</span></span>
<span><span class="co">##    @other: a list containing: population_hierarchy</span></span></code></pre>
<p>Now we can convert this to a <code>genclone</code> object, which will
store information about our multilocus genotypes for us.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span>
<span><span class="va">agc</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    119 original multilocus genotypes </span></span>
<span><span class="co">##    187 diploid individuals</span></span>
<span><span class="co">##     56 dominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      2 strata - Pop, Subpop</span></span>
<span><span class="co">##      2 populations defined - Athena, Mt. Vernon</span></span></code></pre>
<p>If we wanted to, we could also convert it back to a
<code>genind</code> object.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/coercion-methods.html">genclone2genind</a></span><span class="op">(</span><span class="va">agc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 187 individuals; 56 loci; 56 alleles; size: 83.7 Kb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  187 x 56 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 56-56)</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 2-2)</span></span>
<span><span class="co">##    @type:  PA</span></span>
<span><span class="co">##    @call: genclone2genind(x = agc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @pop: population of each individual (group size range: 90-97)</span></span>
<span><span class="co">##    @strata: a data frame with 2 columns ( Pop, Subpop )</span></span>
<span><span class="co">##    @other: a list containing: population_hierarchy</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="intro:import:polyploid">About polyploid data<a class="anchor" aria-label="anchor" href="#intro:import:polyploid"></a>
</h3>
<blockquote>
<p><strong>WARNING</strong></p>
<p>Treat polyploid data with care. Please read this section carefully
and consult the help pages for all functions mentioned here.</p>
</blockquote>
<p>With diploid or haploid data, genotypes are unambiguous. It is often
clear when it is homozygous or heterozygous. With polyploid data,
genotypes can be ambiguous. For example, a tetraploid individual with
the apparent genotype of <strong>A/B</strong> could actually have one of
three genotypes: <strong>A/A/A/B</strong>, <strong>A/A/B/B</strong>, or
<strong>A/B/B/B</strong>. This ambiguity prevents a researcher from
accurately calling all alleles present. In <em>adegenet</em>, it was
previously difficult to import polyploid data because of this ambiguity
as data was required to be unambiguous or missing.</p>
<p>A solution to this problem is to code missing alleles as “0”. An
example of this is found within the <code>Pinf</code> data set in
<em>poppr</em> <span class="citation">(Erica M. Goss et al.
2014)</span>. First, we will look at where we have polyploid allele
calls.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span></span>
<span><span class="va">Pinf</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    72 multilocus genotypes </span></span>
<span><span class="co">##    86 tetraploid individuals</span></span>
<span><span class="co">##    11 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     2 strata - Continent, Country</span></span>
<span><span class="co">##     2 populations defined - South America, North America</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ptab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/info_table.html">info_table</a></span><span class="op">(</span><span class="va">Pinf</span>, type <span class="op">=</span> <span class="st">"ploidy"</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/pinf_table-1.png" width="480" style="display: block; margin: auto;"></p>
<p>We look at the last six samples over two loci, Pi63 (3 alleles,
triploid) and Pi70 (3 alleles, diploid) to examine how the data is
represented.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pinf</span><span class="op">[</span>loc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">locNames</a></span><span class="op">(</span><span class="va">Pinf</span><span class="op">)</span><span class="op">[</span><span class="fl">9</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Pi63.000 Pi63.148 Pi63.151 Pi63.157 Pi70.000 Pi70.189 Pi70.192 Pi70.195</span></span>
<span><span class="co">## PiPE22        1        1        1        1        2        1        1        0</span></span>
<span><span class="co">## PiPE23        1        1        1        1        2        1        1        0</span></span>
<span><span class="co">## PiPE24        1        1        1        1        2        1        1        0</span></span>
<span><span class="co">## PiPE25        1        1        1        1        2        1        1        0</span></span>
<span><span class="co">## PiPE26        2        0        0        2        2        0        1        1</span></span>
<span><span class="co">## PiPE27        1        1        1        1        2        1        1        0</span></span></code></pre>
<p>Each column in this data represents a different allele at a
particular locus. <code>Pi63.148</code> is the allele <code>148</code>
at locus <code>Pi63</code>. Each row is an individual. The numbers
represent the fraction of a given allele that makes up the individual
genotype at a particular locus. What we can see here is that the number
of columns is 8 when we expect only 6 (2 loci x 3 alleles). The first
allele at each locus is <code>000</code>. Let’s take a look at the data
in a human-readable format.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pinfdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/genind2df.html" class="external-link">genind2df</a></span><span class="op">(</span><span class="va">Pinf</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">Pinfdf</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">11</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   Pi63            Pi70</span></span>
<span><span class="co">## PiPE22 000/148/151/157 000/000/189/192</span></span>
<span><span class="co">## PiPE23 000/148/151/157 000/000/189/192</span></span>
<span><span class="co">## PiPE24 000/148/151/157 000/000/189/192</span></span>
<span><span class="co">## PiPE25 000/148/151/157 000/000/189/192</span></span>
<span><span class="co">## PiPE26 000/000/157/157 000/000/192/195</span></span>
<span><span class="co">## PiPE27 000/148/151/157 000/000/189/192</span></span></code></pre>
<p>It”s more clear now that we have a data set of tetraploid individuals
where some genotypes appear diploid (<code>000/000/157/157</code>) and
some appear triploid (<code>000/148/151/157</code>). The tetraploid
genotype is padded with zeroes to make up the difference in ploidy.</p>
<p>This method allows Bruvo’s Distance <span class="citation">(Bruvo et
al. 2004)</span> and the Index of Association <span class="citation">(Brown, Feldman, and Nevo 1980; Smith et al. 1993;
Agapow and Burt 2001)</span> to work with polyploids as they
specifically recognize the zeroes as being missing data. A side effect,
unfortunately is that the extra zeroes appear as extra alleles. As this
affects all frequency-based statistics (except for the ones noted
above), the user should reformat their data set with the function
<code>recode_polyploids</code>, which will remove the zeroes.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pinf_rc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/recode_polyploids.html">recode_polyploids</a></span><span class="op">(</span><span class="va">Pinf</span>, newploidy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">Pinf_rc</span> <span class="co"># Notice that the new ploidy is accounted for.</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    72 multilocus genotypes </span></span>
<span><span class="co">##    86 diploid (55) and triploid (31) individuals</span></span>
<span><span class="co">##    11 codominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     2 strata - Continent, Country</span></span>
<span><span class="co">##     2 populations defined - South America, North America</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pinf_rc</span><span class="op">[</span>loc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">locNames</a></span><span class="op">(</span><span class="va">Pinf_rc</span><span class="op">)</span><span class="op">[</span><span class="fl">9</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span> <span class="va">tail</span></span></code></pre></div>
<pre><code><span><span class="co">##        Pi63.148 Pi63.151 Pi63.157 Pi70.189 Pi70.192 Pi70.195</span></span>
<span><span class="co">## PiPE22        1        1        1        1        1        0</span></span>
<span><span class="co">## PiPE23        1        1        1        1        1        0</span></span>
<span><span class="co">## PiPE24        1        1        1        1        1        0</span></span>
<span><span class="co">## PiPE25        1        1        1        1        1        0</span></span>
<span><span class="co">## PiPE26        0        0        2        0        1        1</span></span>
<span><span class="co">## PiPE27        1        1        1        1        1        0</span></span></code></pre>
<p>Below, we show the observed genotypes:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pinfrcdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/genind2df.html" class="external-link">genind2df</a></span><span class="op">(</span><span class="va">Pinf_rc</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">Pinfrcdf</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">11</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               Pi63    Pi70</span></span>
<span><span class="co">## PiPE22 148/151/157 189/192</span></span>
<span><span class="co">## PiPE23 148/151/157 189/192</span></span>
<span><span class="co">## PiPE24 148/151/157 189/192</span></span>
<span><span class="co">## PiPE25 148/151/157 189/192</span></span>
<span><span class="co">## PiPE26     157/157 192/195</span></span>
<span><span class="co">## PiPE27 148/151/157 189/192</span></span></code></pre>
<p>If you have imported your data as recoded polyploid data, you can use
the argument “addzero” to fill out the ploidy:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Pinf_rc</span><span class="op">[</span>loc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">locNames</a></span><span class="op">(</span><span class="va">Pinf_rc</span><span class="op">)</span><span class="op">[</span><span class="fl">9</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/recode_polyploids.html">recode_polyploids</a></span><span class="op">(</span>addzero <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/grapes-greater-than-grapes.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Pi63.0 Pi63.157 Pi63.148 Pi63.151 Pi70.0 Pi70.192 Pi70.189 Pi70.195</span></span>
<span><span class="co">## PiPE22      0        1        1        1      1        1        1        0</span></span>
<span><span class="co">## PiPE23      0        1        1        1      1        1        1        0</span></span>
<span><span class="co">## PiPE24      0        1        1        1      1        1        1        0</span></span>
<span><span class="co">## PiPE25      0        1        1        1      1        1        1        0</span></span>
<span><span class="co">## PiPE26      1        2        0        0      1        1        0        1</span></span>
<span><span class="co">## PiPE27      0        1        1        1      1        1        1        0</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="data-manip">Data Manipulation<a class="anchor" aria-label="anchor" href="#data-manip"></a>
</h2>
<p>One tedious aspect of population genetic analysis is the need for
repeated data manipulation. <em>Poppr</em> includes novel functions for
clone- censoring your data sets, removing genotypes or loci with missing
data, removing uninformative loci, and shuffling populations.</p>
<div class="section level3">
<h3 id="data-manip:missing">Replace or remove missing data {Inside the golden days of missing
data}<a class="anchor" aria-label="anchor" href="#data-manip:missing"></a>
</h3>
<p>A data set without missing data is always ideal, but often not
achievable. The <em>poppr</em> function <code>missingno</code> exists to
handle missing data. Missing data can mean different things based on
your data type. For microsatellites, missing data might represent any
source of error that could cause a PCR product to not amplify in gel
electrophoresis, which may or may not be biologically relevant. For a
DNA alignment, missing data could mean something as simple as an
insertion or deletion, which is biologically relevant. The choice to
exclude or estimate data has very different implications for the type of
data you have.</p>
<blockquote>
<p><strong>Treatment of Missing data is a non-trivial task.</strong> You
should understand the nature of missing data in your data set before
treatment.</p>
</blockquote>
<hr>
<div class="section level4">
<h4 id="data-manip:missing:missingno">Function: missingno<a class="anchor" aria-label="anchor" href="#data-manip:missing:missingno"></a>
</h4>
<p><code>missingno</code> is a function that serves as a way to exclude
specific areas that contain systematic missing data. There are four
methods available,</p>
<pre><code><span><span class="fu"><a href="../reference/missingno.html">missingno</a></span><span class="op">(</span><span class="va">pop</span>, type <span class="op">=</span> <span class="st">"loci"</span>, cutoff <span class="op">=</span> <span class="fl">0.05</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span>, freq <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<p><code>pop -</code> a <code>genind</code> object.</p>
<p><code>type -</code> This could be one of five options:</p>
<ol class="example" style="list-style-type: decimal">
<li><p>“loci” This is to be used for a data set that has systematic
problems with certain loci that contain null alleles or simply failed to
amplify. This will remove loci with a defined threshold of missing data
from the data set.</p></li>
<li><p>“geno” This is to be used for genotypes (individuals) in your
data set where many null alleles are present. Individuals with a defined
threshold missing data will be removed.</p></li>
<li><p>“asis” This argument will retain missing data in your data set.
It”s useful for functions that utilize <code>missingno</code>
internally, such as <code><a href="../reference/mlg.filter.html">mlg.filter()</a></code>,
<code><a href="../reference/poppr.amova.html">poppr.amova()</a></code>, or <code><a href="../reference/aboot.html">aboot()</a></code>.</p></li>
<li><p>“mean” This replaces missing data with the mean allele
frequencies in the entire data set .</p></li>
<li><p>“zero” or “0” This replaces missing data with zero, signifying a
new allele .</p></li>
</ol>
<p><code>cutoff -</code> This is a numeric value from 0 to 1 indicating
the percent allowable missing data for either loci or genotypes. If you
have, for example, two loci containing missing 5% and 10% missing data,
respectively and you set <code>cutoff = 0.05</code>,
<code>missingno</code> will remove the second locus. Percent missing
data for genotypes is considered the percent missing loci over number of
total loci.</p>
<p><code>quiet -</code> When this is set to <code>FALSE</code>, the
number of missing values replaced will be printed to screen if the
method is “zero” or “mean”. It will print the number of loci or
individuals removed if the method is “loci” or “geno”.</p>
<p><code>freq -</code> This is used for compatibility with the
<code><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab()</a></code> method for <code>genind</code> object. It will
convert counts of alleles to frequencies of alleles, rendering an object
that will return warnings.</p>
<hr>
<p>Let’s take a look at what this does by focusing in on areas with
missing data. We”ll use the data set <code>nancycats</code> as an
example. Using the <em>poppr</em> function <code>info_table</code>, we
can assess missing data within populations.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/info_table.html">info_table</a></span><span class="op">(</span><span class="va">nancycats</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/initializing_poppr-1.png" width="700px" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">##           Locus</span></span>
<span><span class="co">## Population  fca8 fca23 fca43 fca45 fca77 fca78 fca90 fca96 fca37  Mean</span></span>
<span><span class="co">##      P01   0.200     .     .     .     .     .     .     .     . 0.022</span></span>
<span><span class="co">##      P02       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P03       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P04       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P05       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P06       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P07   0.357     .     .     .     .     .     .     .     . 0.040</span></span>
<span><span class="co">##      P08       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P09       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P10       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P11   0.150     .     . 0.400     .     .     . 0.050     . 0.067</span></span>
<span><span class="co">##      P12   0.214     .     .     .     .     .     .     .     . 0.024</span></span>
<span><span class="co">##      P13       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P14   0.412     .     .     .     .     .     .     .     . 0.046</span></span>
<span><span class="co">##      P15       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P16       .     .     .     .     .     .     .     .     .     .</span></span>
<span><span class="co">##      P17       .     .     . 1.000     .     .     . 0.615     . 0.179</span></span>
<span><span class="co">##      Total 0.084     .     . 0.089     .     .     . 0.038     . 0.023</span></span></code></pre>
<p>We can see that locus fca8 has a lot of missing data. To demonstrate
the function <code><a href="../reference/missingno.html">missingno()</a></code>, we will zoom into the first five
individuals at the first locus.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">8</span><span class="op">:</span><span class="fl">13</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      fca8.133 fca8.135 fca8.137 fca8.139 fca8.141 fca8.143</span></span>
<span><span class="co">## N215       NA       NA       NA       NA       NA       NA</span></span>
<span><span class="co">## N216       NA       NA       NA       NA       NA       NA</span></span>
<span><span class="co">## N217        0        1        0        0        0        1</span></span>
<span><span class="co">## N218        1        1        0        0        0        0</span></span>
<span><span class="co">## N219        1        1        0        0        0        0</span></span></code></pre>
<p>When looking at this data set, recall how a <code>genind</code>
object is formatted. You have a matrix representing counts of alleles.
For diploids, if you see 1, that means it is heterozygous at that
allele, and a 2 means it”s homozygous. Here we see three heterozygotes
and two individuals with missing data (indicated by NA). Let”s look at
what happens when we exclude loci and genotypes with &gt;5% missing
data).</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nanloci</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/missingno.html">missingno</a></span><span class="op">(</span><span class="va">nancycats</span>, <span class="st">"loci"</span><span class="op">)</span></span>
<span><span class="va">nangeno</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/missingno.html">missingno</a></span><span class="op">(</span><span class="va">nancycats</span>, <span class="st">"geno"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">nanloci</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">8</span><span class="op">:</span><span class="fl">13</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      fca23.144 fca23.146 fca23.148 fca23.150 fca43.133 fca43.135</span></span>
<span><span class="co">## N215         0         1         0         0         0         0</span></span>
<span><span class="co">## N216         0         2         0         0         0         0</span></span>
<span><span class="co">## N217         0         1         0         0         0         0</span></span>
<span><span class="co">## N218         0         0         0         0         0         0</span></span>
<span><span class="co">## N219         0         1         0         0         0         0</span></span></code></pre>
<p>Notice how we now see columns named fca23.128 and fca23.130. This is
showing us another locus because we have removed the first. Recall from
the summary table that the first locus had 16 alleles, and the second
had 11. Now that we’ve removed loci containing missing data, all others
have shifted over.<br>
Let’s look at the loci names and number of individuals.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">nanloci</span><span class="op">)</span>     <span class="co"># Individuals</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 237</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">locNames</a></span><span class="op">(</span><span class="va">nanloci</span><span class="op">)</span> <span class="co"># Names of the loci</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "fca23" "fca43" "fca77" "fca78" "fca90" "fca96" "fca37"</span></span></code></pre>
<p>You can see that the number of individuals stayed the same but the
loci “fca8”, “fca45”, and “fca96” were removed. Let’s look at what
happened when we removed individuals.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/tab.html" class="external-link">tab</a></span><span class="op">(</span><span class="va">nangeno</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">8</span><span class="op">:</span><span class="fl">13</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      fca8.133 fca8.135 fca8.137 fca8.139 fca8.141 fca8.143</span></span>
<span><span class="co">## N217        0        1        0        0        0        1</span></span>
<span><span class="co">## N218        1        1        0        0        0        0</span></span>
<span><span class="co">## N219        1        1        0        0        0        0</span></span>
<span><span class="co">## N220        0        1        0        0        0        1</span></span>
<span><span class="co">## N221        0        2        0        0        0        0</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">nangeno</span><span class="op">)</span>     <span class="co"># Individuals</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 199</span></span></code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">locNames</a></span><span class="op">(</span><span class="va">nangeno</span><span class="op">)</span> <span class="co"># Names of the loci</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "fca8"  "fca23" "fca43" "fca45" "fca77" "fca78" "fca90" "fca96" "fca37"</span></span></code></pre>
<p>We can see here that the number of individuals decreased, yet we have
the same number of loci. Notice how the frequency matrix changes in both
scenarios? In the scenario with “loci”, we removed several columns of
the data set, and so with our sub-setting, we see alleles from the
second locus. In the scenario with “geno”, we removed several rows of
the data set so we see other individuals in our sub-setting.</p>
</div>
</div>
<div class="section level3">
<h3 id="data-manip:divide">Extract populations {Divide (populations) and conquer (your
analysis)}<a class="anchor" aria-label="anchor" href="#data-manip:divide"></a>
</h3>
<p>This <em>poppr</em> function <code>popsub</code> makes subsetting
<code>genind</code> or genlight objects by population easier:</p>
<div class="section level4">
<h4 id="data-manip:divide:popsub">Function: popsub<a class="anchor" aria-label="anchor" href="#data-manip:divide:popsub"></a>
</h4>
<p>The command <code>popsub</code> is powerful in that it allows you to
choose exactly what populations you choose to include or exclude from
your analyses. As with many R functions, you can also use this within a
function to avoid creating a new variable to keep track of.</p>
<hr>
<pre><code><span><span class="fu"><a href="../reference/popsub.html">popsub</a></span><span class="op">(</span><span class="va">gid</span>, sublist <span class="op">=</span> <span class="st">"ALL"</span>, exclude <span class="op">=</span> <span class="cn">NULL</span>, blacklist <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>    mat <span class="op">=</span> <span class="cn">NULL</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genind</code> object.</p></li>
<li><p><code>sublist -</code> vector of populations or integers
representing the populations in your data set you wish to
<strong>retain</strong>. For example:
<code>sublist = c(pop_z, pop_y)</code> or
<code>sublist = 1:2</code>.</p></li>
<li><p><code>exclude -</code> vector of populations or integers
representing the populations in your data set you wish to
<strong>exclude</strong>. This can take the same type of arguments as
sublist, and can be used in conjunction with sublist for when you want a
range of populations, but know that there is one in there that you do
not want to analyze. For example:
<code>sublist = 1:15, exclude = pop_x</code>. One very useful thing
about the exclude is that it allows the user to be extremely paranoid
about the data. You can set the exclude to contain populations that are
not even in your data set and it will still work!</p></li>
<li><p><code>mat -</code> A matrix produced from the
<code>mlg.table</code> function. This overrides the <em>pop</em>
argument and subsets this table instead.</p></li>
</ul>
<hr>
<p>To demonstrate this tool, we’ll use the H3N2 virus data set provided
in <em>adegenet</em>. It contains a data frame in the “other” slot
called “x” that contains information about the year of epidemic,
country, etc.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"H3N2"</span>, package <span class="op">=</span> <span class="st">"adegenet"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">H3N2</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 1,903 individuals; 125 loci; 334 alleles; size: 4.1 Mb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  1903 x 334 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 2-4)</span></span>
<span><span class="co">##    @loc.fac: locus factor for the 334 columns of @tab</span></span>
<span><span class="co">##    @all.names: list of allele names for each locus</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 1-1)</span></span>
<span><span class="co">##    @type:  codom</span></span>
<span><span class="co">##    @call: .local(x = x, i = i, j = j, drop = drop)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @strata: a data frame with 17 columns ( accession, length, host, segment, subtype, country, ... )</span></span>
<span><span class="co">##    @other: a list containing: x  xy  epid</span></span></code></pre>
<p>We will demonstrate the <code>popsub</code> functionality by setting
the population factor to “country”. Note, in this section, I am naming
the variables staring with “v” indicating “virus”.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/population-methods.html" class="external-link">setPop</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">country</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">popNames</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="co"># Only two countries from North America.</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Japan"          "USA"            "Finland"        "China"          "South Korea"   </span></span>
<span><span class="co">##  [6] "Norway"         "Taiwan"         "France"         "Latvia"         "Netherlands"   </span></span>
<span><span class="co">## [11] "Bulgaria"       "Turkey"         "United Kingdom" "Denmark"        "Austria"       </span></span>
<span><span class="co">## [16] "Canada"         "Italy"          "Russia"         "Bangladesh"     "Egypt"         </span></span>
<span><span class="co">## [21] "Germany"        "Romania"        "Ukraine"        "Czech Republic" "Greece"        </span></span>
<span><span class="co">## [26] "Iceland"        "Ireland"        "Sweden"         "Nepal"          "Saudi Arabia"  </span></span>
<span><span class="co">## [31] "Switzerland"    "Iran"           "Mongolia"       "Spain"          "Slovenia"      </span></span>
<span><span class="co">## [36] "Croatia"        "Algeria"</span></span></code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v_na</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/popsub.html">popsub</a></span><span class="op">(</span><span class="va">H3N2</span>, sublist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"USA"</span>, <span class="st">"Canada"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">popNames</a></span><span class="op">(</span><span class="va">v_na</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "USA"    "Canada"</span></span></code></pre>
<p>If we want to see the population size, we can use the
<em>adegenet</em> function <code><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd()</a></code>:</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>NorthAmerica <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">v_na</span><span class="op">)</span>, Total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NorthAmerica        Total </span></span>
<span><span class="co">##          665         1903</span></span></code></pre>
<p>You can see that the population factors are correct and that the size
of the data set is considerably smaller. Let’s see the data set without
the North American countries.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v_na_minus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/popsub.html">popsub</a></span><span class="op">(</span><span class="va">H3N2</span>, exclude <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"USA"</span>, <span class="st">"Canada"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">popNames</a></span><span class="op">(</span><span class="va">v_na_minus</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Japan"          "Finland"        "China"          "South Korea"    "Norway"        </span></span>
<span><span class="co">##  [6] "Taiwan"         "France"         "Latvia"         "Netherlands"    "Bulgaria"      </span></span>
<span><span class="co">## [11] "Turkey"         "United Kingdom" "Denmark"        "Austria"        "Italy"         </span></span>
<span><span class="co">## [16] "Russia"         "Bangladesh"     "Egypt"          "Germany"        "Romania"       </span></span>
<span><span class="co">## [21] "Ukraine"        "Czech Republic" "Greece"         "Iceland"        "Ireland"       </span></span>
<span><span class="co">## [26] "Sweden"         "Nepal"          "Saudi Arabia"   "Switzerland"    "Iran"          </span></span>
<span><span class="co">## [31] "Mongolia"       "Spain"          "Slovenia"       "Croatia"        "Algeria"</span></span></code></pre>
<p>Let”s make sure that the number of individuals in both data sets is
equal to the number of individuals in our original data set:</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">v_na_minus</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">v_na</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Now we have data sets with and without North America. Let’s try
something a bit more challenging. Let’s say that we want the first 10
populations in alphabetical order, but we know that we still don”t want
any countries in North America. We can easily do this by using the
<em>base</em> function <code>sort</code>.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsort</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">popNames</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="va">vsort</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Algeria"        "Austria"        "Bangladesh"     "Bulgaria"       "Canada"        </span></span>
<span><span class="co">##  [6] "China"          "Croatia"        "Czech Republic" "Denmark"        "Egypt"</span></span></code></pre>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">valph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/popsub.html">popsub</a></span><span class="op">(</span><span class="va">H3N2</span>, sublist <span class="op">=</span> <span class="va">vsort</span>, exclude <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"USA"</span>, <span class="st">"Canada"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">popNames</a></span><span class="op">(</span><span class="va">valph</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "China"          "Bulgaria"       "Denmark"        "Austria"        "Bangladesh"    </span></span>
<span><span class="co">## [6] "Egypt"          "Czech Republic" "Croatia"        "Algeria"</span></span></code></pre>
<p>And that, is how you subset your data with <em>poppr</em>!</p>
</div>
</div>
<div class="section level3">
<h3 id="data-manip:cc">Clone-censor data sets {Attack of the clone correction}<a class="anchor" aria-label="anchor" href="#data-manip:cc"></a>
</h3>
<p>Clone correction refers to the ability of keeping one observation of
each MLG in a given population (or sub-population). Clone correcting can
be hazardous if its done by hand (even on small data sets) and it
requires a defined population hierarchy to get relevant results.
<em>Poppr</em> has a clone correcting function that that will correct
down to the lowest level of any defined population hierarchy. Note that
clone correction in <em>poppr</em> is sensitive to missing data, as it
treats all missing data as a single extra allele.</p>
<p>This function will create new data sets, but it is also utilized by
the functions <code><a href="../reference/poppr.html">poppr()</a></code> and <code><a href="../reference/poppr.amova.html">poppr.amova()</a></code>
natively.</p>
<div class="section level4">
<h4 id="data-manip:cc:clonecorrect">Function: clonecorrect<a class="anchor" aria-label="anchor" href="#data-manip:cc:clonecorrect"></a>
</h4>
<p>This function will return a clone corrected data set corrected for
the lowest population level. Population levels are specified with the
<code>hier</code> flag. You can choose to combine the population
hierarchy to analyze at the lowest population level by choosing
<code>combine = TRUE</code>.</p>
<hr>
<pre><code><span><span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span><span class="va">pop</span>, strata <span class="op">=</span> <span class="fl">1</span>, combine <span class="op">=</span> <span class="cn">FALSE</span>, keep <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genclone</code> object with a defined
hierarchy or a <code>genind</code> object that has a population
hierarchy data frame in the <code>@other</code> slot. Note, the
<code>genind</code> object does not necessarily require a population
factor to begin with.</p></li>
<li><p><code>strata -</code> A hierarchical formula (eg.
<code>\simPop/Subpop</code>), representing the hierarchical levels in
your data.</p></li>
<li><p><code>combine -</code> Do you want to combine the population
hierarchy? If it”s set to <code>FALSE</code> (default), you will be
returned an object with the top most hierarchical level as a population
factor unless the <strong>keep</strong> argument is defined. If set to
<code>TRUE</code>, the hierarchy will be returned combined.</p></li>
<li><p><code>keep -</code> This flag is to be used if you set
<code>combine = FALSE</code>. This will tell clone correct to return a
specific combination of your hierarchy defined as integers. For example,
imagine a hierarchy that needs to be clone corrected at three levels:
<em>Population</em> by <em>Year</em> by <em>Month</em>. If you wanted to
only run an analysis on the <em>Population</em> level, you would set
<code>keep = 1</code> since <em>Population</em> is the first level of
the hierarchy. On the other hand, if you wanted to run analysis on
<em>Year</em> by <em>Month</em>, you would set <code>keep = 2:3</code>
since those are the second and third levels of the hierarchy.</p></li>
</ul>
<hr>
<p>Let”s look at ways to clone-correct our data. We”ll look at our
<em>A. euteichies</em> data that we used in an earlier section since
that data set is known to include clonal populations <span class="citation">(N. J. Grünwald and Hoheisel 2006)</span>. Try playing
around with the data and see what different combinations of the
<code>strata</code>, and <code>keep</code> flags produce. Below, I will
give three examples of clone corrections at the sample level with
respect to field, at the field level, and finally, at the level of the
entire data set.</p>
<p>First, we will examine the original data set.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">strata</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">other</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span><span class="op">$</span><span class="va">population_hierarchy</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">Aeut</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 187 individuals; 56 loci; 56 alleles; size: 72.7 Kb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  187 x 56 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 56-56)</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 2-2)</span></span>
<span><span class="co">##    @type:  PA</span></span>
<span><span class="co">##    @call: old2new_genind(object = x, donor = new(class(x)))</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @pop: population of each individual (group size range: 90-97)</span></span>
<span><span class="co">##    @strata: a data frame with 2 columns ( Pop, Subpop )</span></span>
<span><span class="co">##    @other: a list containing: population_hierarchy</span></span></code></pre>
<p>When you read in data with <code><a href="../reference/read.genalex.html">read.genalex()</a></code>, the default
is to represent it in a <code>genclone</code> object. Since the
<code><a href="../reference/clonecorrect.html">clonecorrect()</a></code> function works on multilocus genotype
definitions, It’s more efficient to convert it to a
<code>genclone</code> object first. We will also rename the strata to
“field” and “sample” to make the biological relevance of the data
clearer.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aphan</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion-methods.html">as.genclone</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/strata-methods.html" class="external-link">nameStrata</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">field</span><span class="op">/</span><span class="va">sample</span></span></code></pre></div>
<p>Now we correct by sample with respect to field and keep the field as
the population.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span><span class="va">aphan</span>,  strata <span class="op">=</span> <span class="op">~</span><span class="va">Pop</span><span class="op">/</span><span class="va">Subpop</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    119 original multilocus genotypes </span></span>
<span><span class="co">##    141 diploid individuals</span></span>
<span><span class="co">##     56 dominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      2 strata - Pop, Subpop</span></span>
<span><span class="co">##      2 populations defined - Athena, Mt. Vernon</span></span></code></pre>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Your turn: Use the same stratification and use combine = TRUE and then</span></span>
<span><span class="co"># keep = 1:2. Is there any difference?</span></span></code></pre></div>
<p>Correcting by field. Notice how the number of MLG is much closer to
our census.</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span><span class="va">aphan</span>, strata <span class="op">=</span> <span class="op">~</span><span class="va">Pop</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    119 original multilocus genotypes </span></span>
<span><span class="co">##    120 diploid individuals</span></span>
<span><span class="co">##     56 dominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      2 strata - Pop, Subpop</span></span>
<span><span class="co">##      2 populations defined - Athena, Mt. Vernon</span></span></code></pre>
<p>Correcting over whole data set. Our MLG is equal to our census.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonecorrect.html">clonecorrect</a></span><span class="op">(</span><span class="va">aphan</span>, strata <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## This is a genclone object</span></span>
<span><span class="co">## -------------------------</span></span>
<span><span class="co">## Genotype information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    119 original multilocus genotypes </span></span>
<span><span class="co">##    119 diploid individuals</span></span>
<span><span class="co">##     56 dominant loci</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Population information:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##      2 strata - Pop, Subpop</span></span>
<span><span class="co">##      2 populations defined - Athena, Mt. Vernon</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="data-manip:shuffle">Permutations and bootstrap resampling {every day I’m shuffling (data
sets)}<a class="anchor" aria-label="anchor" href="#data-manip:shuffle"></a>
</h3>
<p>A common null hypothesis for populations with mixed reproductive
modes is panmixia, or to put it simply: lots of sex. <em>Poppr</em>
randomly shuffles data sets in order to calculate P-values for the index
of association (<span class="math inline">\(I_A\)</span> and <span class="math inline">\(\bar{r} _d\)</span>) <span class="citation">(Agapow and Burt 2001)</span> using 4 different
methods:</p>
<table class="table">
<thead><tr class="header">
<th align="right">method</th>
<th align="left">strategy</th>
<th align="left">units sampled</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">permutation</td>
<td align="left">alleles</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="left">simulation</td>
<td align="left">alleles</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="left">simulation</td>
<td align="left">alleles</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">permutation</td>
<td align="left">genotypes</td>
</tr>
</tbody>
</table>
<p>These methods are detailed below. We will create a dummy data set to
be shuffled by each example below. Let”s assume a single diploid locus
with four alleles (1, 2, 3, 4) with the frequencies of 0.1, 0.2, 0.3,
and 0.4, respectively:</p>
<table class="table">
<caption>Original<span data-label="tab:original"></span>
</caption>
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">4/4</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">4/1</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">4/3</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">2/2</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">3/3</td>
</tr>
</tbody>
</table>
<p>The 4 methods are detailed below.</p>
<hr>
<div class="section level4">
<h4 id="data-manip:shuffle:shufflepop">Function: shufflepop<a class="anchor" aria-label="anchor" href="#data-manip:shuffle:shufflepop"></a>
</h4>
<pre><code><span><span class="fu"><a href="../reference/shufflepop.html">shufflepop</a></span><span class="op">(</span><span class="va">pop</span>, method <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genind</code> object.</p></li>
<li>
<p><code>method -</code> a number indicating the method of sampling
you wish to use. The following methods are available for use:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Permute Alleles (default)</strong> This is a sampling
scheme that will <strong>permute alleles within the locus.</strong></p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">3/4</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">2/3</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">4/4</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">2/1</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">3/4</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">1/3</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">2/4</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">3/4</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">4/3</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">4/2</td>
</tr>
</tbody>
</table>
<p>As you can see, The heterozygosity has changed, yet the allelic
frequencies remain the same. Overall this would show you what would
happen if the sample you had underwent panmixis within this sample
itself.</p>
</li>
<li>
<p><strong>Parametric Bootstrap</strong> The previous scheme
reshuffled the observed sample, but the parametric bootstrap
<strong>draws samples from a multinomial distribution using the observed
allele frequencies as weights.</strong></p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">1/3</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">3/3</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">3/2</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">4/4</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">4/2</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">3/4</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">2/3</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">4/2</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">4/4</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">4/2</td>
</tr>
</tbody>
</table>
<p>Notice how the heterozygosity has changed along with the allelic
frequencies. The frequencies for alleles 3 and 4 have switched in the
first data set, and we”ve lost allele 1 in the second data set purely by
chance! This type of sampling scheme attempts to show you what the true
population would look like if it were panmictic and your original sample
gave you a basis for estimating expected allele frequencies. Since
estimates are made from the observed allele frequencies, small samples
will produce skewed results.</p>
</li>
<li>
<p><strong>Non-Parametric Bootstrap</strong> The third method is
sampling with replacement, again <strong>drawing from a multinomial
distribution, but with no assumption about the allele
frequencies</strong>.</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">1/3</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">3/3</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">3/1</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">2/2</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">3/1</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">1/3</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">3/1</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">2/3</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">2/1</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">4/3</td>
</tr>
</tbody>
</table>
<p>Again, heterozygosity and allele frequencies are not maintained, but
now all of the alleles have a 1 in 4 chance of being chosen.</p>
</li>
<li>
<p><strong>Multilocus permutation</strong> This is called Multilocus
permutation because it does the same thing as the permutation analysis
in the program <em>multilocus</em> by Paul Agapow and Austin Burt <span class="citation">(Agapow and Burt 2001)</span>. This will shuffle the
genotypes at each locus.</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">3/3</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">4/1</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">2/2</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">4/4</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">4/3</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="center">A1/A2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="center">4/4</td>
</tr>
<tr class="even">
<td>2</td>
<td align="center">2/2</td>
</tr>
<tr class="odd">
<td>3</td>
<td align="center">3/3</td>
</tr>
<tr class="even">
<td>4</td>
<td align="center">4/3</td>
</tr>
<tr class="odd">
<td>5</td>
<td align="center">4/1</td>
</tr>
</tbody>
</table>
<p>Note that you have the same genotypes after shuffling, so at each
locus, you will maintain the same allelic frequencies and
heterozygosity. So, in this sample, you will only see a homozygote with
allele 2. This also ensures that the P-values associated with <span class="math inline">\(I_A\)</span> and <span class="math inline">\(\bar{r} _d\)</span> are exactly the same. This
method <strong>assumes that alleles are not independently assorting
within individuals.</strong> This strategy is <strong>useful if you
suspect the population is inbreeding</strong> (Jerome Goudet, personal
communication).</p>
</li>
</ol>
</li>
</ul>
<hr>
<p>These shuffling schemes have been implemented for the index of
association, but there may be other summary statistics you can use
<code>shufflepop</code> for. All you have to do is use the function
<code>replicate</code>. Let”s use average Bruvo”s distance with the
first population of the data set <code>nancycats</code> as an
example:</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span>
<span><span class="va">nan1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/popsub.html">popsub</a></span><span class="op">(</span><span class="va">nancycats</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">9</span><span class="op">)</span> <span class="co"># Assuming dinucleotide repeats.</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="../reference/bruvo.dist.html">bruvo.dist</a></span><span class="op">(</span><span class="va">nan1</span>, replen <span class="op">=</span> <span class="va">reps</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">observed</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.4180619</span></span></code></pre>
<p>You could use this method to replicate the resampling 999 times and
then create a histogram to visualize a distribution of what would happen
under different assumptions of panmixia.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">9999</span><span class="op">)</span></span>
<span><span class="va">bd.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fl">999</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="../reference/bruvo.dist.html">bruvo.dist</a></span><span class="op">(</span><span class="fu"><a href="../reference/shufflepop.html">shufflepop</a></span><span class="op">(</span><span class="va">nan1</span>, method <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, replen <span class="op">=</span> <span class="va">reps</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">bd.test</span>, xlab <span class="op">=</span> <span class="st">"Bruvo's Distance"</span>, main <span class="op">=</span> <span class="st">"Average Bruvo's distance over 999 randomizations"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">observed</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">'topleft'</span>, legend<span class="op">=</span><span class="st">"observed"</span>, col<span class="op">=</span><span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/bd_histogram-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="data-manip:informloci">Removing uninformative loci {Cut It Out!}<a class="anchor" aria-label="anchor" href="#data-manip:informloci"></a>
</h3>
<p>Phylogenetically uninformative loci are those that have only one
sample differentiating from the rest. This can lead to biased results
when using multilocus analyses such as the index of association <span class="citation">(Brown, Feldman, and Nevo 1980; Smith et al.
1993)</span>. These nuisance loci can be removed with the following
function.</p>
<div class="section level4">
<h4 id="data-manip:informloci:informloci">Function: informloci<a class="anchor" aria-label="anchor" href="#data-manip:informloci:informloci"></a>
</h4>
<hr>
<pre><code><span><span class="fu"><a href="../reference/informloci.html">informloci</a></span><span class="op">(</span><span class="va">pop</span>, cutoff <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nInd</a></span><span class="op">(</span><span class="va">pop</span><span class="op">)</span>, MAF <span class="op">=</span> <span class="fl">0.01</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genind</code> object.</p></li>
<li><p><code>cutoff -</code> this represents the minimum fraction of
individuals needed for a locus to be considered informative. The default
is set to <span class="math inline">\(2/n\)</span> with <span class="math inline">\(n\)</span> being the number of individuals in the
data set (represented by the <em>adegenet</em> function
<code>nInd</code>). Essentially, this means that any locus with fewer
than 2 observations differing will be removed. The user can also specify
a fraction of observations for the cutoff (eg. 0.05).</p></li>
<li><p><code>MAF -</code> (Minor Allele Frequency), This defaults to
0.01 indicating that loci that contain one allele representing 1 - MAF
of the data will be removed.</p></li>
<li><p><code>quiet -</code> if <code>TRUE</code>, nothing will be
printed to the screen, if <code>FALSE</code> (default), the cutoff value
in percentage and number of individuals will be printed as well as the
names of the uninfomrative loci found.</p></li>
</ul>
<hr>
<p>Here’s a quick example using the H3N2 virus SNP data set from an
earlier section. We will only retain loci that have a minor allele
frequency of <span class="math inline">\(\geq 5\%\)</span></p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H.five</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/informloci.html">informloci</a></span><span class="op">(</span><span class="va">H3N2</span>, cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## cutoff value: 5 % ( 95 samples ).</span></span>
<span><span class="co">## MAF         : 0.01</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  Found 49 uninformative loci </span></span>
<span><span class="co">##  ============================ </span></span>
<span><span class="co">##  47 loci found with a cutoff of 95 samples :</span></span>
<span><span class="co">##  157, 177, 233, 243, 262, 267, 280, 303, 313, 327, 357, 382, 384, 399, 412, 418,</span></span>
<span><span class="co">## 424, 425, 429, 433, 451, 470, 529, 546, 555, 557, 564, 576, 592, 595, 597, 602,</span></span>
<span><span class="co">## 612, 627, 642, 647, 648, 654, 658, 663, 667, 681, 717, 806, 824, 837, 882 </span></span>
<span><span class="co">##  5 loci found with MAF &lt; 0.01 :</span></span>
<span><span class="co">##  42, 313, 433, 597, 915</span></span></code></pre>
<p>Now what happens when you have all informative loci. We”ll use the
nancycats data set, which has microsatellite loci. It is important to
note that this is searching for loci with a specified genotype frequency
as fixed heterozygous sites are also uninformative:</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span>
<span><span class="va">naninform</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/informloci.html">informloci</a></span><span class="op">(</span><span class="va">nancycats</span>, cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## cutoff value: 5 % ( 12 samples ).</span></span></code></pre>
<pre><code><span><span class="co">## MAF         : 0.01</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## All sites polymorphic</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mlg">Multilocus Genotype Analysis<a class="anchor" aria-label="anchor" href="#mlg"></a>
</h2>
<p>In populations with mixed sexual and clonal reproduction, it common
to have multiple samples from the same population that have the same set
of alleles at all loci. Here, we introduce tools for tracking MLGs
within and across populations in <code>genind</code>, and
<code>genlight</code> objects from the <em>adegenet</em> package. We
will be using the H3N2 data set containing SNP data from isolates of the
H3N2 virus from 2002 to 2006. Note that <code>genclone</code> and
<code>snpclone</code> objects are optimal for these analyses. For a more
in-depth document on methods for multilocus genotypes in <em>poppr</em>,
see the “Multilocus Genotype Analysis” vignette by typing</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"mlg"</span>, package <span class="op">=</span> <span class="st">"poppr"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="mlg:mlg">How many multilocus genotypes are in our data set? {Just a
peek}<a class="anchor" aria-label="anchor" href="#mlg:mlg"></a>
</h3>
<p>Counting the number of MLGs in a population is the first step for
these analyses as they allow us to see how many clones exist. With the
<code>genclone</code> object, This information is already displayed when
we view the object.</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H3N2</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 1,903 individuals; 125 loci; 334 alleles; size: 4.1 Mb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  1903 x 334 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 2-4)</span></span>
<span><span class="co">##    @loc.fac: locus factor for the 334 columns of @tab</span></span>
<span><span class="co">##    @all.names: list of allele names for each locus</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 1-1)</span></span>
<span><span class="co">##    @type:  codom</span></span>
<span><span class="co">##    @call: .local(x = x, i = i, j = j, drop = drop)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @pop: population of each individual (group size range: 1-646)</span></span>
<span><span class="co">##    @strata: a data frame with 17 columns ( accession, length, host, segment, subtype, country, ... )</span></span>
<span><span class="co">##    @other: a list containing: x  xy  epid</span></span></code></pre>
<p>If we need to store the number of MLGs as a variable, we can simply
run the <code><a href="../reference/mlg.html">mlg()</a></code> command.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H3N2_mlg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## #############################</span></span>
<span><span class="co">## # Number of Individuals:  1903 </span></span>
<span><span class="co">## # Number of MLG:  752 </span></span>
<span><span class="co">## #############################</span></span></code></pre>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H3N2_mlg</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 752</span></span></code></pre>
<p>Since the number of individuals exceeds the number of multilocus
genotypes, we conclude that this data set contains clones. Let”s examine
what populations these clones belong to.</p>
</div>
<div class="section level3">
<h3 id="mlg:cross">MLGs across populations {clone-ing around}<a class="anchor" aria-label="anchor" href="#mlg:cross"></a>
</h3>
<p>Since you have the ability to define hierarchical levels of your data
set freely, it is quite possible to see some of the same MLGs across
different populations. Tracking them by hand can be a nightmare with
large data sets. Luckily, <code>mlg.crosspop</code> has you covered in
that regard.</p>
<div class="section level4">
<h4 id="mlg:cross:mlg-crosspop">Function: mlg.crosspop<a class="anchor" aria-label="anchor" href="#mlg:cross:mlg-crosspop"></a>
</h4>
<p>Analyze the MLGs that cross populations within your data set. This
has three output modes. The default one gives a list of MLGs, and for
each MLG, it gives a named numeric vector indicating the abundance of
that MLG in each population. Alternate outputs are described with
<code>indexreturn</code> and <code>df</code>.</p>
<hr>
<pre><code><span><span class="fu"><a href="../reference/mlg.html">mlg.crosspop</a></span><span class="op">(</span><span class="va">gid</span>, strata <span class="op">=</span> <span class="cn">NULL</span>, sublist <span class="op">=</span> <span class="st">"ALL"</span>, exclude <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>    blacklist <span class="op">=</span> <span class="cn">NULL</span>, mlgsub <span class="op">=</span> <span class="cn">NULL</span>, indexreturn <span class="op">=</span> <span class="cn">FALSE</span>, df <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genind</code> object.</p></li>
<li><p><code>sublist -</code> Populations to include (Defaults to
“ALL”). see <code><a href="../reference/popsub.html">popsub()</a></code></p></li>
<li><p><code>exclude -</code> Populations to exclude. see
<code><a href="../reference/popsub.html">popsub()</a></code></p></li>
<li><p><code>mlgsub -</code> see <code><a href="../reference/mlg.html">mlg.table()</a></code> Only analyze
specified MLGs. The vector for this flag can be produced by this
function as you will see later in this vignette.</p></li>
<li><p><code>indexreturn -</code> return a vector of indices of MLGs.
(You can use these in the <code>mlgsub</code> flag, or you can use them
to subset the columns of an MLG table).</p></li>
<li><p><code>df -</code> return a data frame containing the MLGs, the
populations they cross, and the number of copies you find in each
population. This is useful for making graphs in
<em>ggplot2</em>.</p></li>
<li><p><code>quiet -</code> <code>TRUE</code> or <code>FALSE</code>.
Should the populations be printed to screen as they are processed? (will
print nothing if <code>indexreturn</code> is <code>TRUE</code>)</p></li>
</ul>
<hr>
<p>We can see what MLGs cross different populations and then give a
vector that shows how many populations each one of those MLGs
crosses.</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/population-methods.html" class="external-link">setPop</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">country</span></span>
<span><span class="va">v.dup</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.crosspop</a></span><span class="op">(</span><span class="va">H3N2</span>, quiet<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Here is a snippet of what the output looks like when
<code>quiet</code> is <code>FALSE</code>. It will print out the MLG
name, the total number of individuals that make up that MLG, and the
populations where that MLG can be found.</p>
<pre><code><span><span class="co">## MLG.3: (12 inds) USA Denmark </span></span>
<span><span class="co">## MLG.9: (16 inds) Japan USA Finland Denmark </span></span>
<span><span class="co">## MLG.31: (9 inds) Japan Canada </span></span>
<span><span class="co">## MLG.75: (23 inds) Japan USA Finland Norway Denmark Austria Russia Ireland </span></span>
<span><span class="co">## MLG.80: (2 inds) USA Denmark </span></span>
<span><span class="co">## MLG.86: (7 inds) Denmark Austria </span></span>
<span><span class="co">## MLG.95: (2 inds) USA Bangladesh </span></span>
<span><span class="co">## MLG.97: (8 inds) USA Austria Bangladesh Romania </span></span>
<span><span class="co">## MLG.104: (3 inds) USA France </span></span>
<span><span class="co">## MLG.110: (16 inds) Japan USA China</span></span></code></pre>
<p>The output of this function is a list of MLGs, each containing a
vector indicating the number of copies in each population. We”ll count
the number of populations each MLG crosses using the function
<code>sapply</code> to loop over the data with the function
<code>length</code>.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">v.dup</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $MLG.3</span></span>
<span><span class="co">##     USA Denmark </span></span>
<span><span class="co">##       4       8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $MLG.9</span></span>
<span><span class="co">##   Japan     USA Finland Denmark </span></span>
<span><span class="co">##       1      13       1       1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $MLG.31</span></span>
<span><span class="co">##  Japan Canada </span></span>
<span><span class="co">##      2      7 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $MLG.75</span></span>
<span><span class="co">##   Japan     USA Finland  Norway Denmark Austria  Russia Ireland </span></span>
<span><span class="co">##       2       8       2       1       6       2       1       1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $MLG.80</span></span>
<span><span class="co">##     USA Denmark </span></span>
<span><span class="co">##       1       1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $MLG.86</span></span>
<span><span class="co">## Denmark Austria </span></span>
<span><span class="co">##       3       4</span></span></code></pre>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v.num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">v.dup</span>, <span class="va">length</span><span class="op">)</span> <span class="co"># count the number of populations each MLG crosses.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">v.num</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  MLG.3  MLG.9 MLG.31 MLG.75 MLG.80 MLG.86 </span></span>
<span><span class="co">##      2      4      2      8      2      2</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="mlg:table">Producing MLG tables and graphs {bringing something to the
table}<a class="anchor" aria-label="anchor" href="#mlg:table"></a>
</h3>
<p>We can also create a table of MLGs per population as well as bar
graphs to give us a visual representation of the data. This is achieved
through the function <code>mlg.table</code></p>
<div class="section level4">
<h4 id="mlg:table:mlg-table">Function: mlg.table<a class="anchor" aria-label="anchor" href="#mlg:table:mlg-table"></a>
</h4>
<p>This function will produce a matrix containing counts of MLGs
(columns) per population (rows). If there are not populations defined in
your data set, a vector will be produced instead.</p>
<hr>
<pre><code><span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">gid</span>, strata <span class="op">=</span> <span class="cn">NULL</span>, sublist <span class="op">=</span> <span class="st">"ALL"</span>, exclude <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>    blacklist <span class="op">=</span> <span class="cn">NULL</span>, mlgsub <span class="op">=</span> <span class="cn">NULL</span>, bar <span class="op">=</span> <span class="cn">TRUE</span>, plot <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    total <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="cn">FALSE</span>, background <span class="op">=</span> <span class="cn">FALSE</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<ul>
<li><p><code>pop -</code> a <code>genind</code> object.</p></li>
<li><p><code>sublist -</code> Populations to include (Defaults to
“ALL”). see <code><a href="../reference/popsub.html">popsub()</a></code></p></li>
<li><p><code>exclude -</code> Populations to exclude. see
<code><a href="../reference/popsub.html">popsub()</a></code></p></li>
<li><p><code>mlgsub -</code> a vector containing the indices of MLGs you
wish to subset your table with.</p></li>
<li><p><code>plot -</code> <code>TRUE</code> or <code>FALSE</code>. If
<code>TRUE</code>, a bar plot will be printed for each population with
more than one individual.</p></li>
<li><p><code>total -</code> When set to <code>TRUE</code>, the pooled
data set will be added to the table. Defaults to
<code>FALSE</code>.</p></li>
<li><p><code>quiet -</code> Defaults to <code>FALSE</code>: population
names will be printed to the console as they are processed.</p></li>
</ul>
<hr>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v.tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">H3N2</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">v.tab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="co"># Showing the first 10 columns and rows of the table.</span></span></code></pre></div>
<pre><code><span><span class="co">##             MLG.1 MLG.2 MLG.3 MLG.4 MLG.5 MLG.6 MLG.7 MLG.8 MLG.9 MLG.10</span></span>
<span><span class="co">## Japan           0     0     0     0     0     0     1     2     1      0</span></span>
<span><span class="co">## USA             0     2     4     1     1     0     0     0    13      0</span></span>
<span><span class="co">## Finland         0     0     0     0     0     0     0     0     1      0</span></span>
<span><span class="co">## China           0     0     0     0     0     0     0     0     0      0</span></span>
<span><span class="co">## South Korea     0     0     0     0     0     1     0     0     0      0</span></span>
<span><span class="co">## Norway          1     0     0     0     0     0     0     0     0      0</span></span>
<span><span class="co">## Taiwan          0     0     0     0     0     0     0     0     0      0</span></span>
<span><span class="co">## France          0     0     0     0     0     0     0     0     0      0</span></span>
<span><span class="co">## Latvia          0     0     0     0     0     0     0     0     0      0</span></span>
<span><span class="co">## Netherlands     0     0     0     0     0     0     0     0     0      0</span></span></code></pre>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">H3N2</span>, sublist <span class="op">=</span> <span class="st">"Norway"</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="poppr_manual_files/figure-html/mlgbarplot-1.png" alt="An example of a bar-chart produced by `mlg.table`. Note that this data set would produce several such charts but only the chart for Norway is shown here." width="700"><p class="caption">
An example of a bar-chart produced by <code>mlg.table</code>. Note that
this data set would produce several such charts but only the chart for
Norway is shown here.
</p>
</div>
<p>The MLG table is not limited to use with <em>poppr</em>. In fact, one
of the main advantages of <code><a href="../reference/mlg.html">mlg.table()</a></code> is that it allows
easy access to diversity functions present in the package <em>vegan</em>
<span class="citation">(Oksanen et al. 2012)</span>. One example is to
create a rarefaction curve for each population in your data set giving
the number of expected MLGs for a given sample size. For more
information, type <code><a href="https://rdrr.io/pkg/vegan/man/diversity.html" class="external-link">help(diversity, package=vegan)</a></code> in your R
console.</p>
<p>For the sake of this example, instead of drawing a curve for each of
the 37 countries represented in this sample, let”s set the hierarchical
level to year.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/population-methods.html" class="external-link">setPop</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">year</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="co"># Check the data to make sure it's correct.</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## // Number of individuals: 1903</span></span>
<span><span class="co">## // Group sizes: 158 415 399 469 462</span></span>
<span><span class="co">## // Number of alleles per locus: 3 3 4 2 4 2 3 2 4 3 4 2 4 3 2 2 3 3 2 2 3 3 3 2 2 2 2 2 2 2 2 2 2 4 4 3 3 3 4 2 2 2 4 3 2 3 4 2 3 2 3 2 2 2 4 2 2 2 2 2 2 2 4 4 4 3 3 2 3 4 3 2 3 3 3 3 2 3 2 4 2 3 2 2 3 3 3 3 2 2 2 2 3 2 3 2 3 2 3 2 3 3 2 2 2 3 2 2 2 3 3 3 2 2 3 3 3 3 4 2 3 3 4 3 2</span></span>
<span><span class="co">## // Number of alleles per group: 203 255 232 262 240</span></span>
<span><span class="co">## // Percentage of missing data: 2.36 %</span></span>
<span><span class="co">## // Observed heterozygosity: 0</span></span></code></pre>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/vegandevs/vegan" class="external-link">"vegan"</a></span><span class="op">)</span></span>
<span><span class="va">H.year</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">H3N2</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/rarefy.html" class="external-link">rarecurve</a></span><span class="op">(</span><span class="va">H.year</span>, ylab<span class="op">=</span><span class="st">"Number of expected MLGs"</span>, sample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">H.year</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          border <span class="op">=</span> <span class="cn">NA</span>, fill <span class="op">=</span> <span class="cn">NA</span>, font <span class="op">=</span> <span class="fl">2</span>, cex <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="poppr_manual_files/figure-html/mlgrareplot-1.png" alt="An example of a rarefaction curve produced using a MLG table." width="700"><p class="caption">
An example of a rarefaction curve produced using a MLG table.
</p>
</div>
<p>The minimum value from the <em>base</em> function
<code><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums()</a></code> of the table represents the minimum common sample
size of all populations defined in the table. Setting the “sample” flag
draws the horizontal and vertical lines you see on the graph. The
intersections of these lines correspond to the numbers you would find if
you ran the function <code>poppr</code> on this data set (under the
column “<code>eMLG</code>”).</p>
</div>
</div>
<div class="section level3">
<h3 id="mlg:mix">Combining MLG functions {getting into the mix}<a class="anchor" aria-label="anchor" href="#mlg:mix"></a>
</h3>
<p>Alone, the different functionalities are neat. Combined, we can
create interesting data sets. Let”s say we wanted to know which MLGs
were duplicated across the regions of the United Kingdom, Germany,
Netherlands, and Norway. All we have to do is use the
<code>sublist</code> flag in the function:</p>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/population-methods.html" class="external-link">setPop</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">country</span></span>
<span><span class="va">UGNN.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"United Kingdom"</span>, <span class="st">"Germany"</span>, <span class="st">"Netherlands"</span>, <span class="st">"Norway"</span><span class="op">)</span></span>
<span><span class="va">UGNN</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.crosspop</a></span><span class="op">(</span><span class="va">H3N2</span>, sublist<span class="op">=</span><span class="va">UGNN.list</span>, indexreturn<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>OK, the output tells us that there are three MLGs that are crossing
between these populations, but we do not know how many are in each. We
can easily find that out if we subset our original table,
<code>v.tab</code>.</p>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">UGNN</span> <span class="co"># Note that we have three numbers here. This will index the columns for us.</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 315 317 620</span></span></code></pre>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">UGNN.list</span> <span class="co"># And let's not forget that we have the population names.</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "United Kingdom" "Germany"        "Netherlands"    "Norway"</span></span></code></pre>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v.tab</span><span class="op">[</span><span class="va">UGNN.list</span>, <span class="va">UGNN</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                MLG.315 MLG.317 MLG.620</span></span>
<span><span class="co">## United Kingdom       1       0       0</span></span>
<span><span class="co">## Germany              0       1       1</span></span>
<span><span class="co">## Netherlands          0       0       0</span></span>
<span><span class="co">## Norway               2       3       1</span></span></code></pre>
<p>Now we can see that Norway has a higher incidence of nearly all of
these MLGs. We can investigate the incidence of these MLGs throughout
our data set. One thing that the <code>genclone</code> object keeps
track of is a single vector defining the unique multilocus genotypes
within the data. These are represented as integers and can be accessed
with <code>mlg.vector</code>. This is useful for finding MLGs that
correspond to certain individuals or populations. Let”s use
<code>mlg.vector</code> to find individuals corresponding to the MLGs.
First we”ll investigate what the output of this function looks like.</p>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v.vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.vector</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">v.vec</span><span class="op">)</span> <span class="co"># Analyze the structure.</span></span></code></pre></div>
<pre><code><span><span class="co">##  int [1:1903] 605 605 672 675 674 673 670 671 670 678 ...</span></span></code></pre>
<p>The integers produced are the MLG assignment of each individual in
the same order as the data set. This means that the first two
individuals have the exact same set of alleles at each locus, so they
have the same MLG: 605. If we look at the number of unique integers in
the vector, it corresponds to the number of observed multilocus
genotypes:</p>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">v.vec</span><span class="op">)</span><span class="op">)</span> <span class="co"># count the number of MLGs</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 752</span></span></code></pre>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H3N2</span> <span class="co"># equal to the first number in this output.</span></span></code></pre></div>
<pre><code><span><span class="co">## /// GENIND OBJECT /////////</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // 1,903 individuals; 125 loci; 334 alleles; size: 4.1 Mb</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Basic content</span></span>
<span><span class="co">##    @tab:  1903 x 334 matrix of allele counts</span></span>
<span><span class="co">##    @loc.n.all: number of alleles per locus (range: 2-4)</span></span>
<span><span class="co">##    @loc.fac: locus factor for the 334 columns of @tab</span></span>
<span><span class="co">##    @all.names: list of allele names for each locus</span></span>
<span><span class="co">##    @ploidy: ploidy of each individual  (range: 1-1)</span></span>
<span><span class="co">##    @type:  codom</span></span>
<span><span class="co">##    @call: .local(x = x, i = i, j = j, drop = drop)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##  // Optional content</span></span>
<span><span class="co">##    @pop: population of each individual (group size range: 1-646)</span></span>
<span><span class="co">##    @strata: a data frame with 17 columns ( accession, length, host, segment, subtype, country, ... )</span></span>
<span><span class="co">##    @other: a list containing: x  xy  epid</span></span></code></pre>
<p>We will take <code>UGNN</code> (MLGs crossing UK, Germany,
Netherlands, and Norway) and compare its elements to the MLG vector
(<code>v.vec</code>) to see where else they occur.</p>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">UGNN</span> <span class="co"># Show what we are looking for</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 315 317 620</span></span></code></pre>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">UGNN_match</span> <span class="op">&lt;-</span> <span class="va">v.vec</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">UGNN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">UGNN_match</span><span class="op">)</span> <span class="co"># How many individuals matched to those three MLGs?</span></span></code></pre></div>
<pre><code><span><span class="co">## UGNN_match</span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##  1881    22</span></span></code></pre>
<p>22 individuals matched to those three MLGs. We can use this vector to
show us the 22 individuals.</p>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">indNames</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span><span class="op">[</span><span class="va">UGNN_match</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       0329       0330       0331       0332       0341       0342       0345       0556 </span></span>
<span><span class="co">## "CY026119" "CY026120" "CY026121" "CY026122" "CY026131" "CY026132" "CY026135" "EU502462" </span></span>
<span><span class="co">##       0557       0558       0870       0974       1112       1113       1114       1122 </span></span>
<span><span class="co">## "EU502463" "EU502464" "EU501513" "AB243868" "DQ883618" "DQ883619" "DQ883620" "DQ883628" </span></span>
<span><span class="co">##       1193       1209       1210       1281       1288       1426 </span></span>
<span><span class="co">## "EU501609" "EU501642" "EU501643" "EU501735" "EU501742" "EU502513"</span></span></code></pre>
<p>Note that there is an alternative way to list individuals matching
specific MLGs using the function <code>mlg.id</code>. This function will
return a list where each element represents a unique MLG. You can use
this data to find out which individuals correspond to specific MLGs.
Each element in the list is named with the MLG, but the index does not
necessarily match up, so it is important to convert your query MLGs to
strings:</p>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H3N2.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.id</a></span><span class="op">(</span><span class="va">H3N2</span><span class="op">)</span></span>
<span><span class="va">H3N2.id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">UGNN</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $`315`</span></span>
<span><span class="co">##       0974       1112       1122       1209       1210       1281       1288 </span></span>
<span><span class="co">## "AB243868" "DQ883618" "DQ883628" "EU501642" "EU501643" "EU501735" "EU501742" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`317`</span></span>
<span><span class="co">##       0870       1113       1114       1193       1426 </span></span>
<span><span class="co">## "EU501513" "DQ883619" "DQ883620" "EU501609" "EU502513" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`620`</span></span>
<span><span class="co">##       0329       0330       0331       0332       0341       0342       0345       0556 </span></span>
<span><span class="co">## "CY026119" "CY026120" "CY026121" "CY026122" "CY026131" "CY026132" "CY026135" "EU502462" </span></span>
<span><span class="co">##       0557       0558 </span></span>
<span><span class="co">## "EU502463" "EU502464"</span></span></code></pre>
<p>We can also use the vector of MLGs to subset <code><a href="../reference/mlg.html">mlg.table()</a></code>
with the <code>mlgsub</code> flag.</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">H3N2</span>, mlgsub <span class="op">=</span> <span class="va">UGNN</span><span class="op">)</span></span></code></pre></div>
<p>That showed us exactly which populations these three MLGs came from
in our data set.</p>
<div class="figure" style="text-align: center">
<img src="poppr_manual_files/figure-html/mlgsub_flagshow-1.png" alt="All populations that contain these MLGs" width="700"><p class="caption">
All populations that contain these MLGs
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="manipulating-graphics">Manipulating Graphics<a class="anchor" aria-label="anchor" href="#manipulating-graphics"></a>
</h3>
<p><em>Poppr</em> utilizes <em>ggplot2</em> to produce many of its
graphs. One advantage it gives the user is the ability to manipulate
these graphs. With base R graphs, the only manipulation that can be
performed is by adding elements to the graph. It is a static image. The
ggplot graphs are actually represented as objects in your R environment.
We can use the function <code><a href="https://ggplot2.tidyverse.org/reference/last_plot.html" class="external-link">last_plot()</a></code> from <em>ggplot2</em>
to be able to grab the plot that was plotted last in our window. Let’s
illustrate this using a MLG bar graph from the Aeut data set.</p>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://grunwaldlab.github.io/poppr/" class="external-link">"poppr"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span>
<span><span class="va">Aeut.tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mlg.html">mlg.table</a></span><span class="op">(</span><span class="va">Aeut</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/Aeut_MLG_Aeut-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/last_plot.html" class="external-link">last_plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We’ve captured our plot using <code><a href="https://ggplot2.tidyverse.org/reference/last_plot.html" class="external-link">last_plot()</a></code> and now we can
manipulate it. One common need is to change the title. We can easily do
that with the function <code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle()</a></code>. Let”s say we wanted to
label it “<em>Aphanomyces euteiches</em> multilocus genotype
distribution”. We would use
<code>ggtitle(Aphanomyces euteiches multilocus genotype distribution)</code>.
Unfortunately, we need italics for a latin binomial. One way to acheive
this is by using the <code><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression()</a></code> function and declaring
which text needs to be italicized.</p>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myTitle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu">italic</span><span class="op">(</span><span class="st">"Aphanomyces euteiches"</span><span class="op">)</span>, </span>
<span>                      <span class="st">" multilocus genotype distribution"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">pt</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">myTitle</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Multilocus genotype"</span><span class="op">)</span><span class="op">)</span> <span class="co"># We can label the x axis, too</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/Aeut_MLG_title-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Let”s say we wanted to remove the grey background. We could use the
<code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme()</a></code> function to do this, or we could use a theme
already implemented in ggplot2 called <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw()</a></code>.</p>
<div class="sourceCode" id="cb172"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ptt</span> <span class="op">&lt;-</span> <span class="va">pt</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/Aeut_MLG_theme-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Uh-oh. The x axis labels are now horizontal when they should be
vertical. Since it”s the overall distribution we’re interested in, we
don’t really need them anyways. We can remove them with
<code>axis.text.x</code> and <code>axis.ticks.x</code> (we’ll also
remove the x axis gridlines because they’re ugly).</p>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ptta</span> <span class="op">&lt;-</span> <span class="va">ptt</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.major.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/Aeut_MLG_theme_axis-1.png" width="672" style="display: block; margin: auto;"></p>
<p>And, if for some bizarre reason, you liked the color gradient in
<em>poppr</em> version 1, you can get that back by adding the
<em>fill</em> aesthetic:</p>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ptttaf</span> <span class="op">&lt;-</span> <span class="va">ptta</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="poppr_manual_files/figure-html/Aeut_silly-1.png" width="672" style="display: block; margin: auto;"></p>
<p>This allows you to produce publication quality graphs directly in R.
Please see Hadley Wickham”s <em>ggplot2</em> package for more details
<span class="citation">(Wickham 2009)</span>. Note that if you don”t
like using <em>ggplot2</em>, you can access the data in the ggplot2
object and plot the data yourself:</p>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">p</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   Population MLG    count order</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Athena     MLG.20     9 1    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Athena     MLG.66     5 2    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Athena     MLG.14     3 3    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Athena     MLG.35     3 4    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Athena     MLG.13     2 5    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Athena     MLG.16     2 6</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="appendix:graphics">Exporting Graphics<a class="anchor" aria-label="anchor" href="#appendix:graphics"></a>
</h3>
<p>R has the ability to produce nice graphics from most any type of
data, but to get these graphics into a report, presentation, or
manuscript can be a bit challenging. It”s no secret that the R
Documentation pages are a little difficult to interpret, so I will give
the reader here a short example on how to export graphics from R. Note
that any code here that will produce images will also be present in
other places in this vignette. The default installation of the R GUI is
quite minimal, and for an easy way to manage your plots and code, I
strongly encourage the user to use Rstudio <a href="https://rstudio.com/" class="external-link uri">https://rstudio.com/</a>.</p>
<div class="section level4">
<h4 id="appendix:graphics:basics">Basics<a class="anchor" aria-label="anchor" href="#appendix:graphics:basics"></a>
</h4>
<p>Before you export graphics, you have to ask yourself what they will
be used for. If you want to use the graphic for a website, you might
want to opt for a low-resolution image so that it can load quickly. With
printing, you”ll want to make sure that you have a scalable or at least
a very high resolution image. Here, I will give some general guidelines
for graphics (note that these are merely suggestions, not defined
rules).</p>
<ul>
<li><p><strong>What you see is not always what you get</strong> I have
often seen presentations where the colors were too light or posters with
painfully pixellated graphs. Think about what you are going to be using
a graphic for and how it will appear to the intended audience given the
media type.</p></li>
<li><p><strong><span class="math inline">\(\geq\)</span> 300 dpi unless
its for a web page</strong> For any sort of printed material that
requires a raster based image, 300dpi (dots per inch) is the absolute
minimum resolution you should use. For simple black and white line
images, 1200dpi is better. This will leave you with crisp, professional
looking images.</p></li>
<li><p><strong>If possible, save to SVG, then rasterize</strong> Raster
images (bmp, png, jpg, etc…) are based off of the number of pixels or
dots per inch it takes to render the image. This means that the raster
image is more or less a very fine mosaic. Vector images (SVG) are built
upon several interconnected polygons, arcs, and lines that scale
relative to one another to create your graphic. With vector graphics,
you can produce a plot and scale it to the size of a building if you
wanted to. When you save to an SVG file first, you can also manipulate
it in programs such as Adobe Illustrator or Inkscape.</p></li>
<li><p><strong>Before saving, make sure the units and dimensions are
correct</strong> Unless you really wanted to save a graph that”s over 6
feet wide.</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="appendix:graphics:editors">Image Editors<a class="anchor" aria-label="anchor" href="#appendix:graphics:editors"></a>
</h4>
<p>Often times, fine details such as labels on networks need to be
tweaked by hand. Luckily, there are a wide variety of programs that can
help you do that. Here is a short list of image editors (both free and
for a price) that you can use to edit your graphics.</p>
<ul>
<li>
<p>Bitmap based editors (for jpeg, bmp, png, etc…)</p>
<ul>
<li>
<strong>The GIMP</strong> Free, cross-platform. <a href="https://www.gimp.org" class="external-link uri">https://www.gimp.org</a>
</li>
<li>
<strong>Paint.net</strong> Free, Windows only. <a href="https://www.getpaint.net" class="external-link uri">https://www.getpaint.net</a>
</li>
<li>
<strong>Adobe Photoshop</strong> Pricey, Windows and Mac. <a href="https://www.adobe.com/products/photoshop.html" class="external-link uri">https://www.adobe.com/products/photoshop.html</a>
</li>
</ul>
</li>
<li>
<p>Scalable Vector Graphics based editors (for svg, pdf)</p>
<ul>
<li>
<strong>Inkscape</strong> Free, cross-platform <a href="https://inkscape.org" class="external-link uri">https://inkscape.org</a>
</li>
<li>
<strong>Adobe Illustrator</strong> Pricey, Windows and Mac. <a href="https://www.adobe.com/products/illustrator.html" class="external-link uri">https://www.adobe.com/products/illustrator.html</a>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="appendix:graphics:ggplot2">Exporting ggplot2 graphics<a class="anchor" aria-label="anchor" href="#appendix:graphics:ggplot2"></a>
</h4>
<p><em>ggplot2</em> is a fantastic package that <em>poppr</em> uses to
produce graphs for the <code><a href="../reference/mlg.html">mlg.table()</a></code>, <code><a href="../reference/poppr.html">poppr()</a></code>,
and <code><a href="../reference/ia.html">ia()</a></code> functions. Saving a plot with <em>ggplot2</em> is
performed with one command after your plot has rendered:</p>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span> <span class="co"># Load the data set.</span></span>
<span><span class="fu"><a href="../reference/poppr.html">poppr</a></span><span class="op">(</span><span class="va">nancycats</span>, sample <span class="op">=</span> <span class="fl">999</span><span class="op">)</span> <span class="co"># Produce a single plot.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"nancycats.pdf"</span><span class="op">)</span></span></code></pre></div>
<p>Note that you can name the file anything, and <code>ggsave</code>
will save it in that format for you. The details are in the
documentation and you can access it by typing <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">help(ggsave)</a></code>
in your R console. The important things to note are that you can set a
<code>width</code>, <code>height</code>, and <code>unit</code>. The only
downside to this function is that you can only save one plot at a time.
If you want to be able to save multiple plots, read on to the next
section.</p>
</div>
<div class="section level4">
<h4 id="appendix:graphics:export">Exporting any graphics<a class="anchor" aria-label="anchor" href="#appendix:graphics:export"></a>
</h4>
<p>Some of the functions that <em>poppr</em> offers will give you
multiple plots, and if you want to save them all, using
<code>ggsave</code> will require a lot of tedious typing and clicking.
Luckily, R has Functions that will save any plot you generate in nearly
any image format you want. You can save in raster images such as png,
bpm, and jpeg. You can also save in vector based images such as svg,
pdf, and postscript. The important thing to remember is that when you
are saving in a raster format, the default units of measurement are
“pixels”, but you can change that by specifying your unit of choice and
a resolution.</p>
<p>For raster images and svg files, you can only save your plots in
multiple files, but pdf and postscript plots can be saved in one file as
multiple pages. All of these functions have the same basic form. You
call the function to specify the file type you want (eg.
<code>pdf(myfile.pdf)</code>), create any graphs that you want to
create, and then make sure to close the session with the function
<code><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off()</a></code>. Let”s give an example saving to pdf and png
files.</p>
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"nancy_pair%02d.png"</span>, width <span class="op">=</span> <span class="fl">14</span>, height <span class="op">=</span> <span class="fl">14</span>, units <span class="op">=</span> <span class="st">"in"</span>, res <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="va">poppairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/seppop.html" class="external-link">seppop</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span>, <span class="va">pair.ia</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.25</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Since this data set is made up of 30 populations with more than 1
individual, this will save 30 files to your working directory named
“nancy_pair_barchart01.png…nancy_pair_barchart30.png”. The way R knows
how to number these files is because of the <code>%02d</code> part of
the command. That”s telling R to use a number that is two digits long in
place of that expression. All of these files will be 14x14” and will
have a resolution of 300 dots per inch. If you wanted to do the same
thing, but place them all in one file, you should use the pdf
option.</p>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"nancy_pair.png"</span>, width <span class="op">=</span> <span class="fl">14</span>, height <span class="op">=</span> <span class="fl">14</span>, compress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">poppairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/seppop.html" class="external-link">seppop</a></span><span class="op">(</span><span class="va">nancycats</span><span class="op">)</span>, <span class="va">pair.ia</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.25</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Remember, it is important not to forget to type
<code><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off()</a></code> when you are done making graphs. Note that I did
not have to specify a resolution for this image since it is based off of
vector graphics.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Agapow:2001" class="csl-entry">
Agapow, Paul-Michael, and Austin Burt. 2001. <span>“Indices of
Multilocus Linkage Disequilibrium.”</span> <em>Molecular Ecology
Notes</em> 1 (1-2): 101–2. <a href="https://doi.org/10.1046/j.1471-8278.2000.00014.x" class="external-link">https://doi.org/10.1046/j.1471-8278.2000.00014.x</a>.
</div>
<div id="ref-Brown:1980" class="csl-entry">
Brown, A.H. D., M.W. Feldman, and E. Nevo. 1980. <span>“MULTILOCUS
STRUCTURE OF NATURAL POPULATIONS OF <em>Hordeum Spontaneum</em>.”</span>
<em>Genetics</em> 96 (2): 523–36.
</div>
<div id="ref-Bruvo:2004" class="csl-entry">
Bruvo, Ruzica, Nicolaas K. Michiels, Thomas G. D’Souza, and Hinrich
Schulenburg. 2004. <span>“A Simple Method for the Calculation of
Microsatellite Genotype Distances Irrespective of Ploidy Level.”</span>
<em>Molecular Ecology</em> 13 (7): 2101–6. <a href="https://doi.org/10.1111/j.1365-294X.2004.02209.x" class="external-link">https://doi.org/10.1111/j.1365-294X.2004.02209.x</a>.
</div>
<div id="ref-everhart2014finescale" class="csl-entry">
Everhart, Sydney, and Harald Scherm. 2015. <span>“Fine-Scale Genetic
Structure of <span><em>Monilinia Fructicola</em></span> During Brown Rot
Epidemics Within Individual Peach Tree Canopies.”</span>
<em>Phytopathology</em> 105 (4): 542–49. <a href="https://doi.org/10.1094/PHYTO-03-14-0088-R" class="external-link">https://doi.org/10.1094/PHYTO-03-14-0088-R</a>.
</div>
<div id="ref-goss2009population" class="csl-entry">
Goss, Erica M, Meg Larsen, Gary A Chastagner, Donald R Givens, and
Niklaus J Grünwald. 2009. <span>“Population Genetic Analysis Infers
Migration Pathways of Phytophthora Ramorum in US Nurseries.”</span>
<em>PLoS Pathog</em> 5 (9): e1000583.
</div>
<div id="ref-goss2014irish" class="csl-entry">
Goss, Erica M., Javier F. Tabima, David E. L. Cooke, Silvia Restrepo,
William E. Fry, Gregory A. Forbes, Valerie J. Fieland, Martha Cardenas,
and Niklaus J. Grünwald. 2014. <span>“The Irish Potato Famine Pathogen
Phytophthora Infestans Originated in Central Mexico Rather Than the
Andes.”</span> <em>Proceedings of the National Academy of Sciences</em>
111 (24): 8791–96. <a href="https://doi.org/10.1073/pnas.1401884111" class="external-link">https://doi.org/10.1073/pnas.1401884111</a>.
</div>
<div id="ref-Grunwald:2006" class="csl-entry">
Grünwald, N. J., and G. Hoheisel. 2006. <span>“Hierarchical Analysis of
Diversity, Selfing, and Genetic Differentiation in Populations of the
Oomycete <em>Aphanomyces Euteiches</em>.”</span> <em>Phytopathology</em>
96 (10): 1134–41.
</div>
<div id="ref-Grunwald:2003" class="csl-entry">
Grünwald, Niklaus J., Stephen B. Goodwin, Michael G. Milgroom, and
William E. Fry. 2003. <span>“Analysis of Genotypic Diversity Data for
Populations of Microorganisms.”</span> <em>Phytopathology</em> 93 (6):
738–46. <a href="https://doi.org/10.1094/PHYTO.2003.93.6.738" class="external-link">https://doi.org/10.1094/PHYTO.2003.93.6.738</a>.
</div>
<div id="ref-Heck:1975" class="csl-entry">
Heck, Kenneth L.Jr., Gerald van Belle, and Daniel Simberloff. 1975.
<span>“Explicit Calculation of the Rarefaction Diversity Measurement and
the Determination of Sufficient Sample Size.”</span> <em>Ecology</em> 56
(6): pp. 1459–1461.
</div>
<div id="ref-Hurlbert:1971" class="csl-entry">
Hurlbert, S H. 1971. <span>“The Nonconcept of Species Diversity: A
Critique and Alternative Parameters.”</span> <em>Ecology</em> 52 (4):
577–86.
</div>
<div id="ref-Jombart:2008" class="csl-entry">
Jombart, Thibaut. 2008. <span>“Adegenet: A <span>R</span> Package for
the Multivariate Analysis of Genetic Markers.”</span>
<em>Bioinformatics</em> 24 (11): 1403–5. <a href="https://doi.org/10.1093/bioinformatics/btn129" class="external-link">https://doi.org/10.1093/bioinformatics/btn129</a>.
</div>
<div id="ref-kamvar2015novel" class="csl-entry">
Kamvar, Zhian N, Jonah C Brooks, and Niklaus J Grünwald. 2015.
<span>“<span class="nocase">Novel R tools for analysis of genome-wide
population genetic data with emphasis on clonality</span>.”</span>
<em>Frontiers in Genetics</em> 6: 208. <a href="https://doi.org/10.3389/fgene.2015.00208" class="external-link">https://doi.org/10.3389/fgene.2015.00208</a>.
</div>
<div id="ref-kamvar2014poppr" class="csl-entry">
Kamvar, Zhian N., Javier F. Tabima, and Niklaus J. Grünwald. 2014.
<span>“<em>Poppr</em>: An <span>R</span> Package for Genetic Analysis of
Populations with Clonal, Partially Clonal, and/or Sexual
Reproduction.”</span> <em>PeerJ</em> 2 (March): e281. <a href="https://doi.org/10.7717/peerj.281" class="external-link">https://doi.org/10.7717/peerj.281</a>.
</div>
<div id="ref-kamvar2015spatial" class="csl-entry">
Kamvar, ZN, MM Larsen, AM Kanaskie, EM Hansen, and NJ Grünwald. 2015.
<span>“Spatial and Temporal Analysis of Populations of the Sudden Oak
Death Pathogen in Oregon Forests.”</span> <em>Phytopathology</em>,
PHYTO–12.
</div>
<div id="ref-Ludwig:1988" class="csl-entry">
Ludwig, J. A., and J. F. Reynolds. 1988. <em>Statistical Ecology. A
Primer on Methods and Computing</em>. New York USA: John Wiley; Sons.
</div>
<div id="ref-Nei:1978" class="csl-entry">
Nei, Masatoshi. 1978. <span>“ESTIMATION OF AVERAGE HETEROZYGOSITY AND
GENETIC DISTANCE FROM a SMALL NUMBER OF INDIVIDUALS.”</span>
<em>Genetics</em> 89 (3): 583–90.
</div>
<div id="ref-vegan" class="csl-entry">
Oksanen, Jari, F. Guillaume Blanchet, Roeland Kindt, Pierre Legendre,
Peter R. Minchin, R. B. O’Hara, Gavin L. Simpson, Peter Solymos, M.
Henry H. Stevens, and Helene Wagner. 2012. <em>Vegan: Community Ecology
Package</em>. <a href="https://CRAN.R-project.org/package=vegan" class="external-link">https://CRAN.R-project.org/package=vegan</a>.
</div>
<div id="ref-Peakall:2012" class="csl-entry">
Peakall, Rod, and Peter E. Smouse. 2012. <span>“<span>GenAlEx</span>
6.5: Genetic Analysis in Excel. Population Genetic Software for Teaching
and Research—an Update.”</span> <em>Bioinformatics</em> 28 (19):
2537–39. <a href="https://doi.org/10.1093/bioinformatics/bts460" class="external-link">https://doi.org/10.1093/bioinformatics/bts460</a>.
</div>
<div id="ref-Peakall:2006" class="csl-entry">
Peakall, R., and P. E. Smouse. 2006. <span>“<span>GenAlEx</span> 6:
Genetic Analysis in Excel. Population Genetic Software for Teaching and
Research.”</span> <em>Molecular Ecology Notes</em> 6 (1): 288–295+.
</div>
<div id="ref-Pielou:1975" class="csl-entry">
Pielou, EC. 1975. <em>Ecological Diversity–John Wiley &amp; Sons</em>.
<em>New York</em>.
</div>
<div id="ref-Shannon:1948" class="csl-entry">
Shannon, Claude Elwood. 1948. <span>“A Mathematical Theory of
Communication.”</span> <em>Bell Systems Technical Journal</em> 27:
379–423, 623–56.
</div>
<div id="ref-Smith:1993" class="csl-entry">
Smith, J M, N H Smith, M O’Rourke, and B G Spratt. 1993. <span>“How
Clonal Are Bacteria?”</span> <em>Proceedings of the National Academy of
Sciences</em> 90 (10): 4384–88. <a href="https://doi.org/10.1073/pnas.90.10.4384" class="external-link">https://doi.org/10.1073/pnas.90.10.4384</a>.
</div>
<div id="ref-Stoddart:1988" class="csl-entry">
Stoddart, J. A., and J. F. Taylor. 1988. <span>“Genotypic Diversity:
Estimation and Prediction in Samples.”</span> <em>Genetics</em> 118 (4):
705–11.
</div>
<div id="ref-ggplot2" class="csl-entry">
Wickham, Hadley. 2009. <em>Ggplot2: Elegant Graphics for Data
Analysis</em>. Springer New York.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhian N. Kamvar, Javier F. Tabima, Jonah C. Brooks, David Folarin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
